{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "85bb3bc5-7a86-4ae5-a3a6-fd2a46306756",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[info] Loaded 119 unique IDs from ImmPort_URECA_Cross_Black.csv\n",
      "[done] Input variants read: 1186020\n",
      "[done] Output variants written: 1186020\n",
      "[done] Kept 119 sample(s). VCF saved to: chr22.filtered_by_subset.vcf\n",
      "[done] CSV saved to: chr22.filtered_by_subset.csv\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "      <th>FORMAT</th>\n",
       "      <th>8119_45f3</th>\n",
       "      <th>...</th>\n",
       "      <th>1ce6_4a8b</th>\n",
       "      <th>8da2_4ffa</th>\n",
       "      <th>16a0_4f5a</th>\n",
       "      <th>aacf_4181</th>\n",
       "      <th>4438_4db6</th>\n",
       "      <th>19c1_493c</th>\n",
       "      <th>1455_4f52</th>\n",
       "      <th>d084_4cd9</th>\n",
       "      <th>9e45_40b0</th>\n",
       "      <th>0820_4702</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr22</td>\n",
       "      <td>10510105</td>\n",
       "      <td>rs1325858619</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>620.55</td>\n",
       "      <td>VQSRTrancheSNP99.95to100.00;DP10.GQ20.nocalls_0.1</td>\n",
       "      <td>AC=12;AF=0.023;AN=284;DP=664;ExcessHet=0;FS=0;...</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>./.:0,0:0:.:0,0,0</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0:1,0:1:3:0,3,14</td>\n",
       "      <td>./.:0,0:0:.:0,0,0</td>\n",
       "      <td>./.:0,0:0:.:0,0,0</td>\n",
       "      <td>0/0:6,0:6:18:0,18,209</td>\n",
       "      <td>0/0:4,0:4:12:0,12,129</td>\n",
       "      <td>./.:0,0:0:.:0,0,0</td>\n",
       "      <td>./.:0,0:0:.:0,0,0</td>\n",
       "      <td>./.:0,0:0:.:0,0,0</td>\n",
       "      <td>./.:0,0:0:.:0,0,0</td>\n",
       "      <td>./.:0,0:0:.:0,0,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr22</td>\n",
       "      <td>10510161</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>44.49</td>\n",
       "      <td>VQSRTrancheSNP99.90to99.95;DP10.GQ20.nocalls_0.1</td>\n",
       "      <td>AC=2;AF=0.002985;AN=334;DP=972;ExcessHet=0.003...</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>./.:0,0:0:.:0,0,0</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0:1,0:1:3:0,3,14</td>\n",
       "      <td>./.:0,0:0:.:0,0,0</td>\n",
       "      <td>./.:0,0:0:.:0,0,0</td>\n",
       "      <td>0/0:5,0:5:15:0,15,164</td>\n",
       "      <td>0/0:7,0:7:21:0,21,223</td>\n",
       "      <td>./.:0,0:0:.:0,0,0</td>\n",
       "      <td>./.:0,0:0:.:0,0,0</td>\n",
       "      <td>./.:0,0:0:.:0,0,0</td>\n",
       "      <td>./.:0,0:0:.:0,0,0</td>\n",
       "      <td>./.:0,0:0:.:0,0,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr22</td>\n",
       "      <td>10510212</td>\n",
       "      <td>rs1452389754</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>353.84</td>\n",
       "      <td>VQSRTrancheSNP99.90to99.95;DP10.GQ20.nocalls_0.1</td>\n",
       "      <td>AC=6;AF=0.014;AN=428;BaseQRankSum=0.549;DP=170...</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>./.:0,0:0:.:0,0,0</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0:1,0:1:3:0,3,14</td>\n",
       "      <td>./.:0,0:0:.:0,0,0</td>\n",
       "      <td>./.:0,0:0:.:0,0,0</td>\n",
       "      <td>0/0:7,0:7:21:0,21,223</td>\n",
       "      <td>0/0:7,0:7:21:0,21,223</td>\n",
       "      <td>./.:1,0:1:.:0,0,0</td>\n",
       "      <td>./.:0,0:0:.:0,0,0</td>\n",
       "      <td>./.:0,0:0:.:0,0,0</td>\n",
       "      <td>./.:0,0:0:.:0,0,0</td>\n",
       "      <td>0/0:1,0:1:3:0,3,35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr22</td>\n",
       "      <td>10510292</td>\n",
       "      <td>rs1304841723</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>141.69</td>\n",
       "      <td>VQSRTrancheSNP99.90to99.95;DP10.GQ20.nocalls_0.1</td>\n",
       "      <td>AC=2;AF=0.00335;AN=592;BaseQRankSum=0;DP=2205;...</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>./.:0,0:0:.:0,0,0</td>\n",
       "      <td>...</td>\n",
       "      <td>./.:0,0:0:.:0,0,0</td>\n",
       "      <td>./.:0,0:0:.:0,0,0</td>\n",
       "      <td>./.:0,0:0:.:0,0,0</td>\n",
       "      <td>0/0:4,0:4:12:0,12,139</td>\n",
       "      <td>0/0:2,0:2:6:0,6,70</td>\n",
       "      <td>0/0:1,0:1:3:0,3,35</td>\n",
       "      <td>./.:0,0:0:.:0,0,0</td>\n",
       "      <td>0/0:1,0:1:3:0,3,14</td>\n",
       "      <td>./.:0,0:0:.:0,0,0</td>\n",
       "      <td>0/0:2,0:2:6:0,6,49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr22</td>\n",
       "      <td>10510338</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>457.42</td>\n",
       "      <td>VQSRTrancheSNP99.90to99.95;DP10.GQ20.nocalls_0.1</td>\n",
       "      <td>AC=0;AF=0.0009434;AN=528;BaseQRankSum=-0.811;D...</td>\n",
       "      <td>GT:AD:DP:GQ:PGT:PID:PL:PS</td>\n",
       "      <td>./.:0,0:0:.:.:.:0,0,0:.</td>\n",
       "      <td>...</td>\n",
       "      <td>./.:0,0:0:.:.:.:0,0,0:.</td>\n",
       "      <td>./.:0,0:0:.:.:.:0,0,0:.</td>\n",
       "      <td>./.:0,0:0:.:.:.:0,0,0:.</td>\n",
       "      <td>0/0:8,0:8:24:.:.:0,24,278:.</td>\n",
       "      <td>0/0:7,0:7:21:.:.:0,21,223:.</td>\n",
       "      <td>./.:0,0:0:.:.:.:0,0,0:.</td>\n",
       "      <td>./.:0,0:0:.:.:.:0,0,0:.</td>\n",
       "      <td>0/0:1,0:1:3:.:.:0,3,14:.</td>\n",
       "      <td>./.:0,0:0:.:.:.:0,0,0:.</td>\n",
       "      <td>0/0:1,0:1:3:.:.:0,3,35:.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 128 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  #CHROM       POS            ID REF ALT    QUAL  \\\n",
       "0  chr22  10510105  rs1325858619   T   A  620.55   \n",
       "1  chr22  10510161             .   T   A   44.49   \n",
       "2  chr22  10510212  rs1452389754   A   T  353.84   \n",
       "3  chr22  10510292  rs1304841723   T   C  141.69   \n",
       "4  chr22  10510338             .   C   T  457.42   \n",
       "\n",
       "                                              FILTER  \\\n",
       "0  VQSRTrancheSNP99.95to100.00;DP10.GQ20.nocalls_0.1   \n",
       "1   VQSRTrancheSNP99.90to99.95;DP10.GQ20.nocalls_0.1   \n",
       "2   VQSRTrancheSNP99.90to99.95;DP10.GQ20.nocalls_0.1   \n",
       "3   VQSRTrancheSNP99.90to99.95;DP10.GQ20.nocalls_0.1   \n",
       "4   VQSRTrancheSNP99.90to99.95;DP10.GQ20.nocalls_0.1   \n",
       "\n",
       "                                                INFO  \\\n",
       "0  AC=12;AF=0.023;AN=284;DP=664;ExcessHet=0;FS=0;...   \n",
       "1  AC=2;AF=0.002985;AN=334;DP=972;ExcessHet=0.003...   \n",
       "2  AC=6;AF=0.014;AN=428;BaseQRankSum=0.549;DP=170...   \n",
       "3  AC=2;AF=0.00335;AN=592;BaseQRankSum=0;DP=2205;...   \n",
       "4  AC=0;AF=0.0009434;AN=528;BaseQRankSum=-0.811;D...   \n",
       "\n",
       "                      FORMAT                8119_45f3  ...  \\\n",
       "0             GT:AD:DP:GQ:PL        ./.:0,0:0:.:0,0,0  ...   \n",
       "1             GT:AD:DP:GQ:PL        ./.:0,0:0:.:0,0,0  ...   \n",
       "2             GT:AD:DP:GQ:PL        ./.:0,0:0:.:0,0,0  ...   \n",
       "3             GT:AD:DP:GQ:PL        ./.:0,0:0:.:0,0,0  ...   \n",
       "4  GT:AD:DP:GQ:PGT:PID:PL:PS  ./.:0,0:0:.:.:.:0,0,0:.  ...   \n",
       "\n",
       "                 1ce6_4a8b                8da2_4ffa                16a0_4f5a  \\\n",
       "0       0/0:1,0:1:3:0,3,14        ./.:0,0:0:.:0,0,0        ./.:0,0:0:.:0,0,0   \n",
       "1       0/0:1,0:1:3:0,3,14        ./.:0,0:0:.:0,0,0        ./.:0,0:0:.:0,0,0   \n",
       "2       0/0:1,0:1:3:0,3,14        ./.:0,0:0:.:0,0,0        ./.:0,0:0:.:0,0,0   \n",
       "3        ./.:0,0:0:.:0,0,0        ./.:0,0:0:.:0,0,0        ./.:0,0:0:.:0,0,0   \n",
       "4  ./.:0,0:0:.:.:.:0,0,0:.  ./.:0,0:0:.:.:.:0,0,0:.  ./.:0,0:0:.:.:.:0,0,0:.   \n",
       "\n",
       "                     aacf_4181                    4438_4db6  \\\n",
       "0        0/0:6,0:6:18:0,18,209        0/0:4,0:4:12:0,12,129   \n",
       "1        0/0:5,0:5:15:0,15,164        0/0:7,0:7:21:0,21,223   \n",
       "2        0/0:7,0:7:21:0,21,223        0/0:7,0:7:21:0,21,223   \n",
       "3        0/0:4,0:4:12:0,12,139           0/0:2,0:2:6:0,6,70   \n",
       "4  0/0:8,0:8:24:.:.:0,24,278:.  0/0:7,0:7:21:.:.:0,21,223:.   \n",
       "\n",
       "                 19c1_493c                1455_4f52                 d084_4cd9  \\\n",
       "0        ./.:0,0:0:.:0,0,0        ./.:0,0:0:.:0,0,0         ./.:0,0:0:.:0,0,0   \n",
       "1        ./.:0,0:0:.:0,0,0        ./.:0,0:0:.:0,0,0         ./.:0,0:0:.:0,0,0   \n",
       "2        ./.:1,0:1:.:0,0,0        ./.:0,0:0:.:0,0,0         ./.:0,0:0:.:0,0,0   \n",
       "3       0/0:1,0:1:3:0,3,35        ./.:0,0:0:.:0,0,0        0/0:1,0:1:3:0,3,14   \n",
       "4  ./.:0,0:0:.:.:.:0,0,0:.  ./.:0,0:0:.:.:.:0,0,0:.  0/0:1,0:1:3:.:.:0,3,14:.   \n",
       "\n",
       "                 9e45_40b0                 0820_4702  \n",
       "0        ./.:0,0:0:.:0,0,0         ./.:0,0:0:.:0,0,0  \n",
       "1        ./.:0,0:0:.:0,0,0         ./.:0,0:0:.:0,0,0  \n",
       "2        ./.:0,0:0:.:0,0,0        0/0:1,0:1:3:0,3,35  \n",
       "3        ./.:0,0:0:.:0,0,0        0/0:2,0:2:6:0,6,49  \n",
       "4  ./.:0,0:0:.:.:.:0,0,0:.  0/0:1,0:1:3:.:.:0,3,35:.  \n",
       "\n",
       "[5 rows x 128 columns]"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Initially, Read in 1 Chromosome and Test Processing Pipeline\n",
    "\n",
    "import os\n",
    "import gzip\n",
    "import pandas as pd\n",
    "\n",
    "# === USER INPUTS ===\n",
    "PHENO_PATH = \"ImmPort_URECA_Cross_Black.csv\"  # your attached phenotype subset\n",
    "PHENO_ID_COL = \"id\"                           # column in phenotype file with sample IDs\n",
    "VCF_INPUT = \"URECA.WGS.annotated.recoded.chr22.vcf.gz\"  # chromosome 22, gzipped\n",
    "OUT_CSV = \"chr22.filtered_by_subset.csv\"\n",
    "OUT_VCF = \"chr22.filtered_by_subset.vcf\"\n",
    "DROP_ROWS_ALL_MISSING = False  # set True to drop variants where all kept samples are \"./.\"\n",
    "# ====================\n",
    "\n",
    "# 1) Load the phenotype subset IDs--Here we are subsetting to the black population\n",
    "pheno = pd.read_csv(PHENO_PATH)\n",
    "subset_ids = (\n",
    "    pheno[PHENO_ID_COL]\n",
    "    .astype(str)\n",
    "    .str.strip()\n",
    "    .tolist()\n",
    ")\n",
    "subset_ids_set = set(subset_ids)\n",
    "print(f\"[info] Loaded {len(subset_ids_set)} unique IDs from {PHENO_PATH}\")\n",
    "\n",
    "if not os.path.exists(VCF_INPUT):\n",
    "    raise FileNotFoundError(f\"Input VCF not found: {VCF_INPUT}\")\n",
    "\n",
    "meta_lines = []\n",
    "header_parts = None\n",
    "kept_samples = []\n",
    "keep_col_idxs = None\n",
    "base_col_count = 9  # VCF fixed columns\n",
    "\n",
    "# 2) First pass: read header/meta, decide which samples to keep\n",
    "with gzip.open(VCF_INPUT, \"rt\") as f:\n",
    "    for line in f:\n",
    "        if line.startswith(\"##\"):\n",
    "            meta_lines.append(line.rstrip(\"\\n\"))\n",
    "            continue\n",
    "        if line.startswith(\"#CHROM\"):\n",
    "            header_parts = line.rstrip(\"\\n\").split(\"\\t\")\n",
    "            all_samples = header_parts[base_col_count:]\n",
    "            kept_samples = [s for s in all_samples if s in subset_ids_set]\n",
    "\n",
    "            found_set = set(kept_samples)\n",
    "            missing = sorted(list(subset_ids_set - found_set))\n",
    "            if missing:\n",
    "                print(f\"[warn] {len(missing)} subset ID(s) not present in VCF header, e.g.: {missing[:10]}\")\n",
    "            if not kept_samples:\n",
    "                raise RuntimeError(\"No overlapping sample IDs between phenotype subset and VCF.\")\n",
    "\n",
    "            # indices to keep = fixed 9 + selected samples\n",
    "            sample_name_to_idx = {s: base_col_count + i for i, s in enumerate(all_samples)}\n",
    "            keep_col_idxs = list(range(base_col_count)) + [sample_name_to_idx[s] for s in kept_samples]\n",
    "            break\n",
    "\n",
    "if header_parts is None:\n",
    "    raise RuntimeError(\"VCF header (#CHROM line) not found in input.\")\n",
    "\n",
    "filtered_header = header_parts[:base_col_count] + kept_samples\n",
    "\n",
    "def is_missing(gt_field: str) -> bool:\n",
    "    # genotype (before ':') is missing if \"./.\" or \".|.\" or \".\"\n",
    "    if gt_field is None:\n",
    "        return True\n",
    "    gt = gt_field.split(\":\", 1)[0] if gt_field else \"\"\n",
    "    return gt in {\"./.\", \".|.\", \".\"}\n",
    "\n",
    "# 3) Second pass: stream data, collect filtered rows; also write VCF as we go\n",
    "filtered_rows = []\n",
    "n_rows_in = 0\n",
    "n_rows_out = 0\n",
    "\n",
    "with gzip.open(VCF_INPUT, \"rt\") as fin, open(OUT_VCF, \"w\") as vout:\n",
    "    # write meta lines to output VCF\n",
    "    for ml in meta_lines:\n",
    "        vout.write(ml + \"\\n\")\n",
    "    # write filtered header to output VCF\n",
    "    vout.write(\"\\t\".join(filtered_header) + \"\\n\")\n",
    "\n",
    "    # move file pointer to first data line (after #CHROM)\n",
    "    for line in fin:\n",
    "        if line.startswith(\"#CHROM\"):\n",
    "            break\n",
    "\n",
    "    for line in fin:\n",
    "        if not line or line.startswith(\"#\"):\n",
    "            continue\n",
    "        n_rows_in += 1\n",
    "        parts = line.rstrip(\"\\n\").split(\"\\t\")\n",
    "        row = [parts[i] for i in keep_col_idxs]\n",
    "\n",
    "        if DROP_ROWS_ALL_MISSING and len(row) > base_col_count:\n",
    "            sample_fields = row[base_col_count:]\n",
    "            if all(is_missing(sf) for sf in sample_fields):\n",
    "                continue  # skip variant\n",
    "\n",
    "        # keep for CSV\n",
    "        filtered_rows.append(row)\n",
    "        n_rows_out += 1\n",
    "\n",
    "        # write to VCF\n",
    "        vout.write(\"\\t\".join(row) + \"\\n\")\n",
    "\n",
    "print(f\"[done] Input variants read: {n_rows_in}\")\n",
    "print(f\"[done] Output variants written: {n_rows_out}\")\n",
    "print(f\"[done] Kept {len(kept_samples)} sample(s). VCF saved to: {OUT_VCF}\")\n",
    "\n",
    "#4) Save CSV\n",
    "vcf_df = pd.DataFrame(filtered_rows, columns=filtered_header)\n",
    "vcf_df.to_csv(OUT_CSV, index=False)\n",
    "print(f\"[done] CSV saved to: {OUT_CSV}\")\n",
    "\n",
    "# (Optional) quick peek\n",
    "vcf_df.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4d38b69f-9f3a-4bcf-aadc-ea6f77acbf36",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[drop_multiallelics] variants: 1186020 -> 1059892 (−126128); samples: 119 -> 119 (−0)\n",
      "[keep_SNPs_only] variants: 1059892 -> 973431 (−86461); samples: 119 -> 119 (−0)\n",
      "[drop_unannotated_IDs] variants: 973431 -> 928432 (−44999); samples: 119 -> 119 (−0)\n",
      "[QUAL>=70] variants: 928432 -> 904827 (−23605); samples: 119 -> 119 (−0)\n",
      "[keep_MAF>=0.05] variants: 904827 -> 155154 (−749673); samples: 119 -> 119 (−0)\n",
      "[save] CSV -> chr22.processed.filtered.csv (variants=155154, samples=119)\n",
      "[save] VCF -> chr22.processed.filtered.vcf (variants=155154, samples=119)\n",
      "\n",
      "=== QC Summary ===\n",
      "                step  variants_before  variants_after  variants_removed  samples_before  samples_after  samples_removed criterion\n",
      "  drop_multiallelics          1186020         1059892            126128             119            119                0       NaN\n",
      "      keep_SNPs_only          1059892          973431             86461             119            119                0       NaN\n",
      "drop_unannotated_IDs           973431          928432             44999             119            119                0 ID != '.'\n",
      "            QUAL>=70           928432          904827             23605             119            119                0       NaN\n",
      "      keep_MAF>=0.05           904827          155154            749673             119            119                0       NaN\n"
     ]
    }
   ],
   "source": [
    "# Subset data to only black subjects, and QC VCF data\n",
    "import os\n",
    "import pandas as pd\n",
    "\n",
    "# ========= User inputs =========\n",
    "IN_CSV = \"chr22.filtered_by_subset.csv\"       # input from your earlier step\n",
    "IN_VCF = \"chr22.filtered_by_subset.vcf\"       # for copying meta/header lines\n",
    "OUT_CSV = \"chr22.processed.filtered.csv\"\n",
    "OUT_VCF = \"chr22.processed.filtered.vcf\"\n",
    "\n",
    "QUAL_MIN = 70\n",
    "DP_MIN = 10                  # used only when DP is present and numeric\n",
    "MAF_MIN = 0.05               # keep variants with MAF >= 0.05\n",
    "# ===============================\n",
    "\n",
    "\n",
    "# ---------- helpers ----------\n",
    "def snp_and_biallelic(ref: str, alt: str) -> bool:\n",
    "    \"\"\"True if single-ALT biallelic SNP (REF len 1, ALT len 1, no commas).\"\"\"\n",
    "    if not isinstance(ref, str) or not isinstance(alt, str):\n",
    "        return False\n",
    "    if \",\" in alt:\n",
    "        return False\n",
    "    return len(ref) == 1 and len(alt) == 1\n",
    "\n",
    "def is_callable_geno(sample_field: str, dp_idx: int | None) -> bool:\n",
    "    \"\"\"\n",
    "    A genotype (per-sample field) is 'callable' if:\n",
    "      - GT is present (not ./., .|., or .), AND\n",
    "      - if DP exists and is numeric, DP > DP_MIN.\n",
    "    If DP is absent/missing/non-numeric, we DO NOT penalize; GT is enough.\n",
    "    \"\"\"\n",
    "    if not isinstance(sample_field, str) or sample_field in (\"\", \".\"):\n",
    "        return False\n",
    "\n",
    "    toks = sample_field.split(\":\")\n",
    "    gt = toks[0] if toks else \"\"\n",
    "    if gt in {\"./.\", \".|.\", \".\"}:\n",
    "        return False\n",
    "\n",
    "    if dp_idx is not None:\n",
    "        if dp_idx >= len(toks):\n",
    "            return True  # DP field missing for this sample -> don't penalize\n",
    "        dp_raw = toks[dp_idx]\n",
    "        if dp_raw in (\".\", \"\"):\n",
    "            return True  # absent DP -> don't penalize\n",
    "        try:\n",
    "            dp = int(dp_raw)\n",
    "        except Exception:\n",
    "            return True  # non-numeric DP -> don't penalize\n",
    "        return dp > DP_MIN\n",
    "\n",
    "    # No DP in FORMAT -> rely on GT only\n",
    "    return True\n",
    "\n",
    "def maf_from_row(row: pd.Series, sample_cols: list[str]) -> float | None:\n",
    "    \"\"\"\n",
    "    Compute MAF for a biallelic site using callable genotypes only.\n",
    "    Counts alleles from GT; skips alleles other than 0/1; respects DP when present.\n",
    "    Returns None if no callable alleles.\n",
    "    \"\"\"\n",
    "    fmt = str(row[\"FORMAT\"]) if \"FORMAT\" in row.index else \"\"\n",
    "    fmt_keys = fmt.split(\":\") if fmt else []\n",
    "    dp_idx = fmt_keys.index(\"DP\") if \"DP\" in fmt_keys else None\n",
    "\n",
    "    ac = 0  # ALT allele count\n",
    "    an = 0  # total allele count\n",
    "\n",
    "    for s in sample_cols:\n",
    "        field = row[s]\n",
    "        if not is_callable_geno(field, dp_idx):\n",
    "            continue\n",
    "        gt = str(field).split(\":\", 1)[0]\n",
    "        sep = \"/\" if \"/\" in gt else (\"|\" if \"|\" in gt else None)\n",
    "        if sep is None:\n",
    "            # haploid or malformed -> skip\n",
    "            continue\n",
    "        alleles = gt.split(sep)\n",
    "        for a in alleles:\n",
    "            if a in (\"0\", \"1\"):\n",
    "                an += 1\n",
    "                if a == \"1\":\n",
    "                    ac += 1\n",
    "            # ignore others\n",
    "\n",
    "    if an == 0:\n",
    "        return None\n",
    "    p_alt = ac / an\n",
    "    return min(p_alt, 1 - p_alt)\n",
    "\n",
    "def log_step(name: str, before_df: pd.DataFrame, after_df: pd.DataFrame, log_list: list, extra: dict | None = None):\n",
    "    info = {\n",
    "        \"step\": name,\n",
    "        \"variants_before\": before_df.shape[0],\n",
    "        \"variants_after\":  after_df.shape[0],\n",
    "        \"variants_removed\": before_df.shape[0] - after_df.shape[0],\n",
    "        \"samples_before\": max(len(before_df.columns) - 9, 0),\n",
    "        \"samples_after\":  max(len(after_df.columns) - 9, 0),\n",
    "        \"samples_removed\": max(len(before_df.columns) - 9, 0) - max(len(after_df.columns) - 9, 0),\n",
    "    }\n",
    "    if extra:\n",
    "        info.update(extra)\n",
    "    log_list.append(info)\n",
    "    print(f\"[{name}] variants: {info['variants_before']} -> {info['variants_after']} \"\n",
    "          f\"(−{info['variants_removed']}); samples: {info['samples_before']} -> {info['samples_after']} \"\n",
    "          f\"(−{info['samples_removed']})\")\n",
    "\n",
    "\n",
    "# ---------- 0) Load ----------\n",
    "vcf_df = pd.read_csv(IN_CSV, low_memory=False)\n",
    "if vcf_df.shape[1] < 10:\n",
    "    raise RuntimeError(\n",
    "        \"Parsed CSV has fewer than 10 columns. Expected fixed 9 VCF columns + FORMAT + sample columns.\\n\"\n",
    "        \"Check that the input CSV came from the subset step that kept FORMAT+samples.\"\n",
    "    )\n",
    "\n",
    "fixed_cols = list(vcf_df.columns[:9])   # #CHROM POS ID REF ALT QUAL FILTER INFO FORMAT\n",
    "sample_cols = list(vcf_df.columns[9:])\n",
    "log = []\n",
    "\n",
    "# ---------- 1) Drop multiallelic (max-alleles = 2) ----------\n",
    "step1 = vcf_df[~vcf_df[\"ALT\"].astype(str).str.contains(\",\", na=True)].copy()\n",
    "log_step(\"drop_multiallelics\", vcf_df, step1, log)\n",
    "\n",
    "# ---------- 2) Keep SNPs only ----------\n",
    "mask_snp = step1.apply(lambda r: snp_and_biallelic(str(r[\"REF\"]), str(r[\"ALT\"])), axis=1)\n",
    "step2 = step1[mask_snp].copy()\n",
    "log_step(\"keep_SNPs_only\", step1, step2, log)\n",
    "\n",
    "# ---------- 3) Drop unannotated variants (ID == '.' or missing) ----------\n",
    "mask_annot = step2[\"ID\"].notna() & step2[\"ID\"].astype(str).str.strip().ne(\".\")\n",
    "step2_annot = step2[mask_annot].copy()\n",
    "log_step(\"drop_unannotated_IDs\", step2, step2_annot, log, extra={\"criterion\": \"ID != '.'\"})\n",
    "\n",
    "# ---------- 4) QUAL >= 70 ----------\n",
    "step2_annot[\"QUAL\"] = pd.to_numeric(step2_annot[\"QUAL\"], errors=\"coerce\")\n",
    "step3 = step2_annot[step2_annot[\"QUAL\"] >= QUAL_MIN].copy()\n",
    "log_step(f\"QUAL>={QUAL_MIN}\", step2_annot, step3, log)\n",
    "\n",
    "# ---------- 5) MAF >= 0.05 (computed from callable GTs; DP>10 if DP exists) ----------\n",
    "cur = step3\n",
    "sample_cols = list(cur.columns[9:])\n",
    "if cur.shape[0] == 0 or len(sample_cols) == 0:\n",
    "    step4 = cur.copy()\n",
    "else:\n",
    "    mafs = cur.apply(lambda r: maf_from_row(r, sample_cols), axis=1)\n",
    "    step4 = cur[(mafs.notna()) & (mafs >= MAF_MIN)].copy()\n",
    "log_step(f\"keep_MAF>={MAF_MIN}\", cur, step4, log)\n",
    "\n",
    "final_df = step4\n",
    "\n",
    "# ---------- 6) Save CSV ----------\n",
    "final_df.to_csv(OUT_CSV, index=False)\n",
    "print(f\"[save] CSV -> {OUT_CSV} (variants={final_df.shape[0]}, samples={len(final_df.columns)-9})\")\n",
    "\n",
    "# ---------- 7) Save VCF ----------\n",
    "if not os.path.exists(IN_VCF):\n",
    "    raise FileNotFoundError(f\"Could not find {IN_VCF} to copy meta/header for VCF output.\")\n",
    "\n",
    "meta_lines = []\n",
    "with open(IN_VCF, \"r\") as f:\n",
    "    for line in f:\n",
    "        if line.startswith(\"##\"):\n",
    "            meta_lines.append(line.rstrip(\"\\n\"))\n",
    "        elif line.startswith(\"#CHROM\"):\n",
    "            break\n",
    "\n",
    "kept_header = list(final_df.columns[:9]) + list(final_df.columns[9:])\n",
    "with open(OUT_VCF, \"w\") as vout:\n",
    "    for ml in meta_lines:\n",
    "        vout.write(ml + \"\\n\")\n",
    "    vout.write(\"\\t\".join(kept_header) + \"\\n\")\n",
    "    for _, row in final_df.iterrows():\n",
    "        vout.write(\"\\t\".join(str(row[col]) for col in kept_header) + \"\\n\")\n",
    "\n",
    "print(f\"[save] VCF -> {OUT_VCF} (variants={final_df.shape[0]}, samples={len(final_df.columns)-9})\")\n",
    "\n",
    "# ---------- 8) Summary ----------\n",
    "summary = pd.DataFrame(log)\n",
    "print(\"\\n=== QC Summary ===\")\n",
    "print(summary.to_string(index=False))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "cf20550c-3579-424d-9044-933acf320cf2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[annotate] unique rsIDs to query: 155154\n",
      "  processed 200/155154\n",
      "  processed 400/155154\n",
      "  processed 600/155154\n",
      "  processed 800/155154\n",
      "  processed 1000/155154\n",
      "  processed 1200/155154\n",
      "  processed 1400/155154\n",
      "  processed 1600/155154\n",
      "  processed 1800/155154\n",
      "  processed 2000/155154\n",
      "  processed 2200/155154\n",
      "  processed 2400/155154\n",
      "  processed 2600/155154\n",
      "  processed 2800/155154\n",
      "  processed 3000/155154\n",
      "  processed 3200/155154\n",
      "  processed 3400/155154\n",
      "  processed 3600/155154\n",
      "  processed 3800/155154\n",
      "  processed 4000/155154\n",
      "  processed 4200/155154\n",
      "  processed 4400/155154\n",
      "  processed 4600/155154\n",
      "  processed 4800/155154\n",
      "  processed 5000/155154\n",
      "  processed 5200/155154\n",
      "  processed 5400/155154\n",
      "  processed 5600/155154\n",
      "  processed 5800/155154\n",
      "  processed 6000/155154\n",
      "  processed 6200/155154\n",
      "  processed 6400/155154\n",
      "  processed 6600/155154\n",
      "  processed 6800/155154\n",
      "  processed 7000/155154\n",
      "  processed 7200/155154\n",
      "  processed 7400/155154\n",
      "  processed 7600/155154\n",
      "  processed 7800/155154\n",
      "  processed 8000/155154\n",
      "  processed 8200/155154\n",
      "  processed 8400/155154\n",
      "  processed 8600/155154\n",
      "  processed 8800/155154\n",
      "  processed 9000/155154\n",
      "  processed 9200/155154\n",
      "  processed 9400/155154\n",
      "  processed 9600/155154\n",
      "  processed 9800/155154\n",
      "  processed 10000/155154\n",
      "  processed 10200/155154\n",
      "  processed 10400/155154\n",
      "  processed 10600/155154\n",
      "  processed 10800/155154\n",
      "  processed 11000/155154\n",
      "  processed 11200/155154\n",
      "  processed 11400/155154\n",
      "  processed 11600/155154\n",
      "  processed 11800/155154\n",
      "  processed 12000/155154\n",
      "  processed 12200/155154\n",
      "  processed 12400/155154\n",
      "  processed 12600/155154\n",
      "  processed 12800/155154\n",
      "  processed 13000/155154\n",
      "  processed 13200/155154\n",
      "  processed 13400/155154\n",
      "  processed 13600/155154\n",
      "  processed 13800/155154\n",
      "  processed 14000/155154\n",
      "  processed 14200/155154\n",
      "  processed 14400/155154\n",
      "  processed 14600/155154\n",
      "  processed 14800/155154\n",
      "  processed 15000/155154\n",
      "  processed 15200/155154\n",
      "  processed 15400/155154\n",
      "  processed 15600/155154\n",
      "  processed 15800/155154\n",
      "  processed 16000/155154\n",
      "  processed 16200/155154\n",
      "  processed 16400/155154\n",
      "  processed 16600/155154\n",
      "  processed 16800/155154\n",
      "  processed 17000/155154\n",
      "  processed 17200/155154\n",
      "  processed 17400/155154\n",
      "  processed 17600/155154\n",
      "  processed 17800/155154\n",
      "  processed 18000/155154\n",
      "  processed 18200/155154\n",
      "  processed 18400/155154\n",
      "  processed 18600/155154\n",
      "  processed 18800/155154\n",
      "  processed 19000/155154\n",
      "  processed 19200/155154\n",
      "  processed 19400/155154\n",
      "  processed 19600/155154\n",
      "  processed 19800/155154\n",
      "  processed 20000/155154\n",
      "  processed 20200/155154\n",
      "  processed 20400/155154\n",
      "  processed 20600/155154\n",
      "  processed 20800/155154\n",
      "  processed 21000/155154\n",
      "  processed 21200/155154\n",
      "  processed 21400/155154\n",
      "  processed 21600/155154\n",
      "  processed 21800/155154\n",
      "  processed 22000/155154\n",
      "  processed 22200/155154\n",
      "  processed 22400/155154\n",
      "  processed 22600/155154\n",
      "  processed 22800/155154\n",
      "  processed 23000/155154\n",
      "  processed 23200/155154\n",
      "  processed 23400/155154\n",
      "  processed 23600/155154\n",
      "  processed 23800/155154\n",
      "  processed 24000/155154\n",
      "  processed 24200/155154\n",
      "  processed 24400/155154\n",
      "  processed 24600/155154\n",
      "  processed 24800/155154\n",
      "  processed 25000/155154\n",
      "  processed 25200/155154\n",
      "  processed 25400/155154\n",
      "  processed 25600/155154\n",
      "  processed 25800/155154\n",
      "  processed 26000/155154\n",
      "  processed 26200/155154\n",
      "  processed 26400/155154\n",
      "  processed 26600/155154\n",
      "  processed 26800/155154\n",
      "  processed 27000/155154\n",
      "  processed 27200/155154\n",
      "  processed 27400/155154\n",
      "  processed 27600/155154\n",
      "  processed 27800/155154\n",
      "  processed 28000/155154\n",
      "  processed 28200/155154\n",
      "  processed 28400/155154\n",
      "  processed 28600/155154\n",
      "  processed 28800/155154\n",
      "  processed 29000/155154\n",
      "  processed 29200/155154\n",
      "  processed 29400/155154\n",
      "  processed 29600/155154\n",
      "  processed 29800/155154\n",
      "  processed 30000/155154\n",
      "  processed 30200/155154\n",
      "  processed 30400/155154\n",
      "  processed 30600/155154\n",
      "  processed 30800/155154\n",
      "  processed 31000/155154\n",
      "  processed 31200/155154\n",
      "  processed 31400/155154\n",
      "  processed 31600/155154\n",
      "  processed 31800/155154\n",
      "  processed 32000/155154\n",
      "  processed 32200/155154\n",
      "  processed 32400/155154\n",
      "  processed 32600/155154\n",
      "  processed 32800/155154\n",
      "  processed 33000/155154\n",
      "  processed 33200/155154\n",
      "  processed 33400/155154\n",
      "  processed 33600/155154\n",
      "  processed 33800/155154\n",
      "  processed 34000/155154\n",
      "  processed 34200/155154\n",
      "  processed 34400/155154\n",
      "  processed 34600/155154\n",
      "  processed 34800/155154\n",
      "  processed 35000/155154\n",
      "  processed 35200/155154\n",
      "  processed 35400/155154\n",
      "  processed 35600/155154\n",
      "  processed 35800/155154\n",
      "  processed 36000/155154\n",
      "  processed 36200/155154\n",
      "  processed 36400/155154\n",
      "  processed 36600/155154\n",
      "  processed 36800/155154\n",
      "  processed 37000/155154\n",
      "  processed 37200/155154\n",
      "  processed 37400/155154\n",
      "  processed 37600/155154\n",
      "  processed 37800/155154\n",
      "  processed 38000/155154\n",
      "  processed 38200/155154\n",
      "  processed 38400/155154\n",
      "  processed 38600/155154\n",
      "  processed 38800/155154\n",
      "  processed 39000/155154\n",
      "  processed 39200/155154\n",
      "  processed 39400/155154\n",
      "  processed 39600/155154\n",
      "  processed 39800/155154\n",
      "  processed 40000/155154\n",
      "  processed 40200/155154\n",
      "  processed 40400/155154\n",
      "  processed 40600/155154\n",
      "  processed 40800/155154\n",
      "  processed 41000/155154\n",
      "  processed 41200/155154\n",
      "  processed 41400/155154\n",
      "  processed 41600/155154\n",
      "  processed 41800/155154\n",
      "  processed 42000/155154\n",
      "  processed 42200/155154\n",
      "  processed 42400/155154\n",
      "  processed 42600/155154\n",
      "  processed 42800/155154\n",
      "  processed 43000/155154\n",
      "  processed 43200/155154\n",
      "  processed 43400/155154\n",
      "  processed 43600/155154\n",
      "  processed 43800/155154\n",
      "  processed 44000/155154\n",
      "  processed 44200/155154\n",
      "  processed 44400/155154\n",
      "  processed 44600/155154\n",
      "  processed 44800/155154\n",
      "  processed 45000/155154\n",
      "  processed 45200/155154\n",
      "  processed 45400/155154\n",
      "  processed 45600/155154\n",
      "  processed 45800/155154\n",
      "  processed 46000/155154\n",
      "  processed 46200/155154\n",
      "  processed 46400/155154\n",
      "  processed 46600/155154\n",
      "  processed 46800/155154\n",
      "  processed 47000/155154\n",
      "  processed 47200/155154\n",
      "  processed 47400/155154\n",
      "  processed 47600/155154\n",
      "  processed 47800/155154\n",
      "  processed 48000/155154\n",
      "  processed 48200/155154\n",
      "  processed 48400/155154\n",
      "  processed 48600/155154\n",
      "  processed 48800/155154\n",
      "  processed 49000/155154\n",
      "  processed 49200/155154\n",
      "  processed 49400/155154\n",
      "  processed 49600/155154\n",
      "  processed 49800/155154\n",
      "  processed 50000/155154\n",
      "  processed 50200/155154\n",
      "  processed 50400/155154\n",
      "  processed 50600/155154\n",
      "  processed 50800/155154\n",
      "  processed 51000/155154\n",
      "  processed 51200/155154\n",
      "  processed 51400/155154\n",
      "  processed 51600/155154\n",
      "  processed 51800/155154\n",
      "  processed 52000/155154\n",
      "  processed 52200/155154\n",
      "  processed 52400/155154\n",
      "  processed 52600/155154\n",
      "  processed 52800/155154\n",
      "  processed 53000/155154\n",
      "  processed 53200/155154\n",
      "  processed 53400/155154\n",
      "  processed 53600/155154\n",
      "  processed 53800/155154\n",
      "  processed 54000/155154\n",
      "  processed 54200/155154\n",
      "  processed 54400/155154\n",
      "  processed 54600/155154\n",
      "  processed 54800/155154\n",
      "  processed 55000/155154\n",
      "  processed 55200/155154\n",
      "  processed 55400/155154\n",
      "  processed 55600/155154\n",
      "  processed 55800/155154\n",
      "  processed 56000/155154\n",
      "  processed 56200/155154\n",
      "  processed 56400/155154\n",
      "  processed 56600/155154\n",
      "  processed 56800/155154\n",
      "  processed 57000/155154\n",
      "  processed 57200/155154\n",
      "  processed 57400/155154\n",
      "  processed 57600/155154\n",
      "  processed 57800/155154\n",
      "  processed 58000/155154\n",
      "  processed 58200/155154\n",
      "  processed 58400/155154\n",
      "  processed 58600/155154\n",
      "  processed 58800/155154\n",
      "  processed 59000/155154\n",
      "  processed 59200/155154\n",
      "  processed 59400/155154\n",
      "  processed 59600/155154\n",
      "  processed 59800/155154\n",
      "  processed 60000/155154\n",
      "  processed 60200/155154\n",
      "  processed 60400/155154\n",
      "  processed 60600/155154\n",
      "  processed 60800/155154\n",
      "  processed 61000/155154\n",
      "  processed 61200/155154\n",
      "  processed 61400/155154\n",
      "  processed 61600/155154\n",
      "  processed 61800/155154\n",
      "  processed 62000/155154\n",
      "  processed 62200/155154\n",
      "  processed 62400/155154\n",
      "  processed 62600/155154\n",
      "  processed 62800/155154\n",
      "  processed 63000/155154\n",
      "  processed 63200/155154\n",
      "  processed 63400/155154\n",
      "  processed 63600/155154\n",
      "  processed 63800/155154\n",
      "  processed 64000/155154\n",
      "  processed 64200/155154\n",
      "  processed 64400/155154\n",
      "  processed 64600/155154\n",
      "  processed 64800/155154\n",
      "  processed 65000/155154\n",
      "  processed 65200/155154\n",
      "  processed 65400/155154\n",
      "  processed 65600/155154\n",
      "  processed 65800/155154\n",
      "  processed 66000/155154\n",
      "  processed 66200/155154\n",
      "  processed 66400/155154\n",
      "  processed 66600/155154\n",
      "  processed 66800/155154\n",
      "  processed 67000/155154\n",
      "  processed 67200/155154\n",
      "  processed 67400/155154\n",
      "  processed 67600/155154\n",
      "  processed 67800/155154\n",
      "  processed 68000/155154\n",
      "  processed 68200/155154\n",
      "  processed 68400/155154\n",
      "  processed 68600/155154\n",
      "  processed 68800/155154\n",
      "  processed 69000/155154\n",
      "  processed 69200/155154\n",
      "  processed 69400/155154\n",
      "  processed 69600/155154\n",
      "  processed 69800/155154\n",
      "  processed 70000/155154\n",
      "  processed 70200/155154\n",
      "  processed 70400/155154\n",
      "  processed 70600/155154\n",
      "  processed 70800/155154\n",
      "  processed 71000/155154\n",
      "  processed 71200/155154\n",
      "  processed 71400/155154\n",
      "  processed 71600/155154\n",
      "  processed 71800/155154\n",
      "  processed 72000/155154\n",
      "  processed 72200/155154\n",
      "  processed 72400/155154\n",
      "  processed 72600/155154\n",
      "  processed 72800/155154\n",
      "  processed 73000/155154\n",
      "  processed 73200/155154\n",
      "  processed 73400/155154\n",
      "  processed 73600/155154\n",
      "  processed 73800/155154\n",
      "  processed 74000/155154\n",
      "  processed 74200/155154\n",
      "  processed 74400/155154\n",
      "  processed 74600/155154\n",
      "  processed 74800/155154\n",
      "  processed 75000/155154\n",
      "  processed 75200/155154\n",
      "  processed 75400/155154\n",
      "  processed 75600/155154\n",
      "  processed 75800/155154\n",
      "  processed 76000/155154\n",
      "  processed 76200/155154\n",
      "  processed 76400/155154\n",
      "  processed 76600/155154\n",
      "  processed 76800/155154\n",
      "  processed 77000/155154\n",
      "  processed 77200/155154\n",
      "  processed 77400/155154\n",
      "  processed 77600/155154\n",
      "  processed 77800/155154\n",
      "  processed 78000/155154\n",
      "  processed 78200/155154\n",
      "  processed 78400/155154\n",
      "  processed 78600/155154\n",
      "  processed 78800/155154\n",
      "  processed 79000/155154\n",
      "  processed 79200/155154\n",
      "  processed 79400/155154\n",
      "  processed 79600/155154\n",
      "  processed 79800/155154\n",
      "  processed 80000/155154\n",
      "  processed 80200/155154\n",
      "  processed 80400/155154\n",
      "  processed 80600/155154\n",
      "  processed 80800/155154\n",
      "  processed 81000/155154\n",
      "  processed 81200/155154\n",
      "  processed 81400/155154\n",
      "  processed 81600/155154\n",
      "  processed 81800/155154\n",
      "  processed 82000/155154\n",
      "  processed 82200/155154\n",
      "  processed 82400/155154\n",
      "  processed 82600/155154\n",
      "  processed 82800/155154\n",
      "  processed 83000/155154\n",
      "  processed 83200/155154\n",
      "  processed 83400/155154\n",
      "  processed 83600/155154\n",
      "  processed 83800/155154\n",
      "  processed 84000/155154\n",
      "  processed 84200/155154\n",
      "  processed 84400/155154\n",
      "  processed 84600/155154\n",
      "  processed 84800/155154\n",
      "  processed 85000/155154\n",
      "  processed 85200/155154\n",
      "  processed 85400/155154\n",
      "  processed 85600/155154\n",
      "  processed 85800/155154\n",
      "  processed 86000/155154\n",
      "  processed 86200/155154\n",
      "  processed 86400/155154\n",
      "  processed 86600/155154\n",
      "  processed 86800/155154\n",
      "  processed 87000/155154\n",
      "  processed 87200/155154\n",
      "  processed 87400/155154\n",
      "  processed 87600/155154\n",
      "  processed 87800/155154\n",
      "  processed 88000/155154\n",
      "  processed 88200/155154\n",
      "  processed 88400/155154\n",
      "  processed 88600/155154\n",
      "  processed 88800/155154\n",
      "  processed 89000/155154\n",
      "  processed 89200/155154\n",
      "  processed 89400/155154\n",
      "  processed 89600/155154\n",
      "  processed 89800/155154\n",
      "  processed 90000/155154\n",
      "  processed 90200/155154\n",
      "  processed 90400/155154\n",
      "  processed 90600/155154\n",
      "  processed 90800/155154\n",
      "  processed 91000/155154\n",
      "  processed 91200/155154\n",
      "  processed 91400/155154\n",
      "  processed 91600/155154\n",
      "  processed 91800/155154\n",
      "  processed 92000/155154\n",
      "  processed 92200/155154\n",
      "  processed 92400/155154\n",
      "  processed 92600/155154\n",
      "  processed 92800/155154\n",
      "  processed 93000/155154\n",
      "  processed 93200/155154\n",
      "  processed 93400/155154\n",
      "  processed 93600/155154\n",
      "  processed 93800/155154\n",
      "  processed 94000/155154\n",
      "  processed 94200/155154\n",
      "  processed 94400/155154\n",
      "  processed 94600/155154\n",
      "  processed 94800/155154\n",
      "  processed 95000/155154\n",
      "  processed 95200/155154\n",
      "  processed 95400/155154\n",
      "  processed 95600/155154\n",
      "  processed 95800/155154\n",
      "  processed 96000/155154\n",
      "  processed 96200/155154\n",
      "  processed 96400/155154\n",
      "  processed 96600/155154\n",
      "  processed 96800/155154\n",
      "  processed 97000/155154\n",
      "  processed 97200/155154\n",
      "  processed 97400/155154\n",
      "  processed 97600/155154\n",
      "  processed 97800/155154\n",
      "  processed 98000/155154\n",
      "  processed 98200/155154\n",
      "  processed 98400/155154\n",
      "  processed 98600/155154\n",
      "  processed 98800/155154\n",
      "  processed 99000/155154\n",
      "  processed 99200/155154\n",
      "  processed 99400/155154\n",
      "  processed 99600/155154\n",
      "  processed 99800/155154\n",
      "  processed 100000/155154\n",
      "  processed 100200/155154\n",
      "  processed 100400/155154\n",
      "  processed 100600/155154\n",
      "  processed 100800/155154\n",
      "  processed 101000/155154\n",
      "  processed 101200/155154\n",
      "  processed 101400/155154\n",
      "  processed 101600/155154\n",
      "  processed 101800/155154\n",
      "  processed 102000/155154\n",
      "  processed 102200/155154\n",
      "  processed 102400/155154\n",
      "  processed 102600/155154\n",
      "  processed 102800/155154\n",
      "  processed 103000/155154\n",
      "  processed 103200/155154\n",
      "  processed 103400/155154\n",
      "  processed 103600/155154\n",
      "  processed 103800/155154\n",
      "  processed 104000/155154\n",
      "  processed 104200/155154\n",
      "  processed 104400/155154\n",
      "  processed 104600/155154\n",
      "  processed 104800/155154\n",
      "  processed 105000/155154\n",
      "  processed 105200/155154\n",
      "  processed 105400/155154\n",
      "  processed 105600/155154\n",
      "  processed 105800/155154\n",
      "  processed 106000/155154\n",
      "  processed 106200/155154\n",
      "  processed 106400/155154\n",
      "  processed 106600/155154\n",
      "  processed 106800/155154\n",
      "  processed 107000/155154\n",
      "  processed 107200/155154\n",
      "  processed 107400/155154\n",
      "  processed 107600/155154\n",
      "  processed 107800/155154\n",
      "  processed 108000/155154\n",
      "  processed 108200/155154\n",
      "  processed 108400/155154\n",
      "  processed 108600/155154\n",
      "  processed 108800/155154\n",
      "  processed 109000/155154\n",
      "  processed 109200/155154\n",
      "  processed 109400/155154\n",
      "  processed 109600/155154\n",
      "  processed 109800/155154\n",
      "  processed 110000/155154\n",
      "  processed 110200/155154\n",
      "  processed 110400/155154\n",
      "  processed 110600/155154\n",
      "  processed 110800/155154\n",
      "  processed 111000/155154\n",
      "  processed 111200/155154\n",
      "  processed 111400/155154\n",
      "  processed 111600/155154\n",
      "  processed 111800/155154\n",
      "  processed 112000/155154\n",
      "  processed 112200/155154\n",
      "  processed 112400/155154\n",
      "  processed 112600/155154\n",
      "  processed 112800/155154\n",
      "  processed 113000/155154\n",
      "  processed 113200/155154\n",
      "  processed 113400/155154\n",
      "  processed 113600/155154\n",
      "  processed 113800/155154\n",
      "  processed 114000/155154\n",
      "  processed 114200/155154\n",
      "  processed 114400/155154\n",
      "  processed 114600/155154\n",
      "  processed 114800/155154\n",
      "  processed 115000/155154\n",
      "  processed 115200/155154\n",
      "  processed 115400/155154\n",
      "  processed 115600/155154\n",
      "  processed 115800/155154\n",
      "  processed 116000/155154\n",
      "  processed 116200/155154\n",
      "  processed 116400/155154\n",
      "  processed 116600/155154\n",
      "  processed 116800/155154\n",
      "  processed 117000/155154\n",
      "  processed 117200/155154\n",
      "  processed 117400/155154\n",
      "  processed 117600/155154\n",
      "  processed 117800/155154\n",
      "  processed 118000/155154\n",
      "  processed 118200/155154\n",
      "  processed 118400/155154\n",
      "  processed 118600/155154\n",
      "  processed 118800/155154\n",
      "  processed 119000/155154\n",
      "  processed 119200/155154\n",
      "  processed 119400/155154\n",
      "  processed 119600/155154\n",
      "  processed 119800/155154\n",
      "  processed 120000/155154\n",
      "  processed 120200/155154\n",
      "  processed 120400/155154\n",
      "  processed 120600/155154\n",
      "  processed 120800/155154\n",
      "  processed 121000/155154\n",
      "  processed 121200/155154\n",
      "  processed 121400/155154\n",
      "  processed 121600/155154\n",
      "  processed 121800/155154\n",
      "  processed 122000/155154\n",
      "  processed 122200/155154\n",
      "  processed 122400/155154\n",
      "  processed 122600/155154\n",
      "  processed 122800/155154\n",
      "  processed 123000/155154\n",
      "  processed 123200/155154\n",
      "  processed 123400/155154\n",
      "  processed 123600/155154\n",
      "  processed 123800/155154\n",
      "  processed 124000/155154\n",
      "  processed 124200/155154\n",
      "  processed 124400/155154\n",
      "  processed 124600/155154\n",
      "  processed 124800/155154\n",
      "  processed 125000/155154\n",
      "  processed 125200/155154\n",
      "  processed 125400/155154\n",
      "  processed 125600/155154\n",
      "  processed 125800/155154\n",
      "  processed 126000/155154\n",
      "  processed 126200/155154\n",
      "  processed 126400/155154\n",
      "  processed 126600/155154\n",
      "  processed 126800/155154\n",
      "  processed 127000/155154\n",
      "  processed 127200/155154\n",
      "  processed 127400/155154\n",
      "  processed 127600/155154\n",
      "  processed 127800/155154\n",
      "  processed 128000/155154\n",
      "  processed 128200/155154\n",
      "  processed 128400/155154\n",
      "  processed 128600/155154\n",
      "  processed 128800/155154\n",
      "  processed 129000/155154\n",
      "  processed 129200/155154\n",
      "  processed 129400/155154\n",
      "  processed 129600/155154\n",
      "  processed 129800/155154\n",
      "  processed 130000/155154\n",
      "  processed 130200/155154\n",
      "  processed 130400/155154\n",
      "  processed 130600/155154\n",
      "  processed 130800/155154\n",
      "  processed 131000/155154\n",
      "  processed 131200/155154\n",
      "  processed 131400/155154\n",
      "  processed 131600/155154\n",
      "  processed 131800/155154\n",
      "  processed 132000/155154\n",
      "  processed 132200/155154\n",
      "  processed 132400/155154\n",
      "  processed 132600/155154\n",
      "  processed 132800/155154\n",
      "  processed 133000/155154\n",
      "  processed 133200/155154\n",
      "  processed 133400/155154\n",
      "  processed 133600/155154\n",
      "  processed 133800/155154\n",
      "  processed 134000/155154\n",
      "  processed 134200/155154\n",
      "  processed 134400/155154\n",
      "  processed 134600/155154\n",
      "  processed 134800/155154\n",
      "  processed 135000/155154\n",
      "  processed 135200/155154\n",
      "  processed 135400/155154\n",
      "  processed 135600/155154\n",
      "  processed 135800/155154\n",
      "  processed 136000/155154\n",
      "  processed 136200/155154\n",
      "  processed 136400/155154\n",
      "  processed 136600/155154\n",
      "  processed 136800/155154\n",
      "  processed 137000/155154\n",
      "  processed 137200/155154\n",
      "  processed 137400/155154\n",
      "  processed 137600/155154\n",
      "  processed 137800/155154\n",
      "  processed 138000/155154\n",
      "  processed 138200/155154\n",
      "  processed 138400/155154\n",
      "  processed 138600/155154\n",
      "  processed 138800/155154\n",
      "  processed 139000/155154\n",
      "  processed 139200/155154\n",
      "  processed 139400/155154\n",
      "  processed 139600/155154\n",
      "  processed 139800/155154\n",
      "  processed 140000/155154\n",
      "  processed 140200/155154\n",
      "  processed 140400/155154\n",
      "  processed 140600/155154\n",
      "  processed 140800/155154\n",
      "  processed 141000/155154\n",
      "  processed 141200/155154\n",
      "  processed 141400/155154\n",
      "  processed 141600/155154\n",
      "  processed 141800/155154\n",
      "  processed 142000/155154\n",
      "  processed 142200/155154\n",
      "  processed 142400/155154\n",
      "  processed 142600/155154\n",
      "  processed 142800/155154\n",
      "  processed 143000/155154\n",
      "  processed 143200/155154\n",
      "  processed 143400/155154\n",
      "  processed 143600/155154\n",
      "  processed 143800/155154\n",
      "  processed 144000/155154\n",
      "  processed 144200/155154\n",
      "  processed 144400/155154\n",
      "  processed 144600/155154\n",
      "  processed 144800/155154\n",
      "  processed 145000/155154\n",
      "  processed 145200/155154\n",
      "  processed 145400/155154\n",
      "  processed 145600/155154\n",
      "  processed 145800/155154\n",
      "  processed 146000/155154\n",
      "  processed 146200/155154\n",
      "  processed 146400/155154\n",
      "  processed 146600/155154\n",
      "  processed 146800/155154\n",
      "  processed 147000/155154\n",
      "  processed 147200/155154\n",
      "  processed 147400/155154\n",
      "  processed 147600/155154\n",
      "  processed 147800/155154\n",
      "  processed 148000/155154\n",
      "  processed 148200/155154\n",
      "  processed 148400/155154\n",
      "  processed 148600/155154\n",
      "  processed 148800/155154\n",
      "  processed 149000/155154\n",
      "  processed 149200/155154\n",
      "  processed 149400/155154\n",
      "  processed 149600/155154\n",
      "  processed 149800/155154\n",
      "  processed 150000/155154\n",
      "  processed 150200/155154\n",
      "  processed 150400/155154\n",
      "  processed 150600/155154\n",
      "  processed 150800/155154\n",
      "  processed 151000/155154\n",
      "  processed 151200/155154\n",
      "  processed 151400/155154\n",
      "  processed 151600/155154\n",
      "  processed 151800/155154\n",
      "  processed 152000/155154\n",
      "  processed 152200/155154\n",
      "  processed 152400/155154\n",
      "  processed 152600/155154\n",
      "  processed 152800/155154\n",
      "  processed 153000/155154\n",
      "  processed 153200/155154\n",
      "  processed 153400/155154\n",
      "  processed 153600/155154\n",
      "  processed 153800/155154\n",
      "  processed 154000/155154\n",
      "  processed 154200/155154\n",
      "  processed 154400/155154\n",
      "  processed 154600/155154\n",
      "  processed 154800/155154\n",
      "  processed 155000/155154\n",
      "  processed 155154/155154\n",
      "[save] Annotated CSV -> chr22.processed.filtered.vep.csv (rows=155154)\n",
      "[warn] Some VCF columns missing in CSV view and will be left blank in output: ['CHROM']\n",
      "[save] Annotated VCF -> chr22.processed.filtered.vep.vcf (rows=155154)\n",
      "\n",
      "=== CSQ Annotation Summary ===\n",
      "          ID CSQ_gene    CSQ_most_severe CSQ_impact\n",
      "rs1451962360          intergenic_variant           \n",
      "rs1362063374          intergenic_variant           \n",
      "rs1411356411          intergenic_variant           \n",
      "rs1156673197          intergenic_variant           \n",
      "rs1401473738          intergenic_variant           \n",
      "rs1454784984          intergenic_variant           \n",
      "rs1262312921          intergenic_variant           \n",
      "rs1179390713          intergenic_variant           \n",
      "rs1385181893          intergenic_variant           \n",
      "rs1408615897          intergenic_variant           \n"
     ]
    }
   ],
   "source": [
    "# --- Annotate rsIDs with functional consequences (CSQ) via Ensembl VEP REST API ---\n",
    "\n",
    "import os\n",
    "import time\n",
    "import json\n",
    "import requests\n",
    "import pandas as pd\n",
    "from typing import Iterable, List, Dict, Any\n",
    "\n",
    "# ========= Inputs/Outputs =========\n",
    "IN_PROCESSED_CSV = \"chr22.processed.filtered.csv\"\n",
    "IN_PROCESSED_VCF = \"chr22.processed.filtered.vcf\"   # to reuse header/meta lines\n",
    "OUT_ANNOTATED_CSV = \"chr22.processed.filtered.vep.csv\"\n",
    "OUT_ANNOTATED_VCF = \"chr22.processed.filtered.vep.vcf\"\n",
    "# ==================================\n",
    "\n",
    "VEP_URL = \"https://rest.ensembl.org/vep/human/id\"\n",
    "BATCH_SIZE = 200          # Ensembl handles a few hundred IDs per POST safely\n",
    "MAX_RETRIES = 5\n",
    "BACKOFF_SECS = 2.0        # exponential backoff base\n",
    "\n",
    "def batched(iterable: Iterable[Any], n: int) -> Iterable[List[Any]]:\n",
    "    \"\"\"Yield lists of size n from iterable.\"\"\"\n",
    "    batch = []\n",
    "    for item in iterable:\n",
    "        batch.append(item)\n",
    "        if len(batch) == n:\n",
    "            yield batch\n",
    "            batch = []\n",
    "    if batch:\n",
    "        yield batch\n",
    "\n",
    "def vep_query_rsids(rsids: List[str]) -> Dict[str, dict]:\n",
    "    \"\"\"\n",
    "    Query Ensembl VEP REST API for a list of rsIDs.\n",
    "    Returns a mapping rsid -> VEP result (first match).\n",
    "    \"\"\"\n",
    "    headers = {\"Content-Type\": \"application/json\", \"Accept\": \"application/json\"}\n",
    "    payload = {\"ids\": rsids}\n",
    "    for attempt in range(1, MAX_RETRIES + 1):\n",
    "        try:\n",
    "            r = requests.post(VEP_URL, headers=headers, data=json.dumps(payload), timeout=60)\n",
    "            if r.status_code == 429:\n",
    "                # Rate limited — honor Retry-After if present\n",
    "                wait = float(r.headers.get(\"Retry-After\", BACKOFF_SECS * attempt))\n",
    "                time.sleep(wait)\n",
    "                continue\n",
    "            r.raise_for_status()\n",
    "            data = r.json()\n",
    "            out = {}\n",
    "            for entry in data:\n",
    "                # Each entry typically has 'id' (rsID) and annotations\n",
    "                rsid = entry.get(\"id\")\n",
    "                if not rsid:\n",
    "                    continue\n",
    "                out[rsid] = entry\n",
    "            return out\n",
    "        except Exception as e:\n",
    "            if attempt == MAX_RETRIES:\n",
    "                raise\n",
    "            time.sleep(BACKOFF_SECS * (2 ** (attempt - 1)))\n",
    "    return {}\n",
    "\n",
    "def summarize_vep(entry: dict) -> dict:\n",
    "    \"\"\"\n",
    "    Reduce a VEP response entry to key fields for CSQ.\n",
    "    Chooses the transcript_consequence matching 'most_severe_consequence' if available,\n",
    "    else uses the first transcript consequence.\n",
    "    \"\"\"\n",
    "    most_severe = entry.get(\"most_severe_consequence\")\n",
    "    tcs = entry.get(\"transcript_consequences\") or []\n",
    "    pick = None\n",
    "    if most_severe and tcs:\n",
    "        for tc in tcs:\n",
    "            cons = tc.get(\"consequence_terms\") or []\n",
    "            if most_severe in cons:\n",
    "                pick = tc\n",
    "                break\n",
    "    if pick is None and tcs:\n",
    "        pick = tcs[0]\n",
    "\n",
    "    gene_symbol = None\n",
    "    impact = None\n",
    "    cons_terms = entry.get(\"consequence_terms\")  # sometimes present at top level\n",
    "    if pick:\n",
    "        gene_symbol = pick.get(\"gene_symbol\") or pick.get(\"gene_id\")\n",
    "        impact = pick.get(\"impact\")\n",
    "        if not cons_terms:\n",
    "            cons_terms = pick.get(\"consequence_terms\")\n",
    "\n",
    "    # Fallback: also check 'colocated_variants' for consequence_terms\n",
    "    if not cons_terms:\n",
    "        for cv in entry.get(\"colocated_variants\", []) or []:\n",
    "            if cv.get(\"consequence_terms\"):\n",
    "                cons_terms = cv.get(\"consequence_terms\")\n",
    "                break\n",
    "\n",
    "    # Normalize\n",
    "    csq_terms = \",\".join(sorted(set(cons_terms or [])))\n",
    "    return {\n",
    "        \"CSQ_most_severe\": most_severe or \"\",\n",
    "        \"CSQ_gene\": gene_symbol or \"\",\n",
    "        \"CSQ_impact\": impact or \"\",\n",
    "        \"CSQ_terms\": csq_terms,\n",
    "    }\n",
    "\n",
    "# ---------- 1) Load processed CSV ----------\n",
    "df = pd.read_csv(IN_PROCESSED_CSV, low_memory=False)\n",
    "if \"ID\" not in df.columns:\n",
    "    raise RuntimeError(\"Expected 'ID' column in processed CSV.\")\n",
    "\n",
    "rsids_all = (\n",
    "    df[\"ID\"]\n",
    "    .astype(str)\n",
    "    .str.strip()\n",
    "    .replace({\"nan\": \".\", \"NaN\": \".\"})\n",
    ")\n",
    "rsids = sorted(set(x for x in rsids_all if x and x != \".\"))\n",
    "print(f\"[annotate] unique rsIDs to query: {len(rsids)}\")\n",
    "\n",
    "# ---------- 2) Query VEP in batches ----------\n",
    "rsid_to_summary: Dict[str, dict] = {}\n",
    "queried = 0\n",
    "for chunk in batched(rsids, BATCH_SIZE):\n",
    "    resp = vep_query_rsids(chunk)\n",
    "    for k, v in resp.items():\n",
    "        rsid_to_summary[k] = summarize_vep(v)\n",
    "    queried += len(chunk)\n",
    "    print(f\"  processed {queried}/{len(rsids)}\")\n",
    "\n",
    "# ---------- 3) Merge CSQ annotations back to dataframe ----------\n",
    "def map_field(series, key):\n",
    "    return series.apply(lambda x: rsid_to_summary.get(str(x), {}).get(key, \"\") if str(x) in rsid_to_summary else \"\")\n",
    "\n",
    "df[\"CSQ_most_severe\"] = map_field(df[\"ID\"], \"CSQ_most_severe\")\n",
    "df[\"CSQ_gene\"]         = map_field(df[\"ID\"], \"CSQ_gene\")\n",
    "df[\"CSQ_impact\"]       = map_field(df[\"ID\"], \"CSQ_impact\")\n",
    "df[\"CSQ_terms\"]        = map_field(df[\"ID\"], \"CSQ_terms\")\n",
    "\n",
    "# Optional: If you want a compact single INFO-style CSQ string:\n",
    "def make_csq_row(r):\n",
    "    parts = []\n",
    "    if r[\"CSQ_gene\"]:\n",
    "        parts.append(f\"Gene={r['CSQ_gene']}\")\n",
    "    if r[\"CSQ_most_severe\"]:\n",
    "        parts.append(f\"MostSevere={r['CSQ_most_severe']}\")\n",
    "    if r[\"CSQ_impact\"]:\n",
    "        parts.append(f\"Impact={r['CSQ_impact']}\")\n",
    "    if r[\"CSQ_terms\"]:\n",
    "        parts.append(f\"Terms={r['CSQ_terms']}\")\n",
    "    return \"|\".join(parts) if parts else \"\"\n",
    "\n",
    "df[\"CSQ\"] = df.apply(make_csq_row, axis=1)\n",
    "\n",
    "# ---------- 4) Save annotated CSV ----------\n",
    "df.to_csv(OUT_ANNOTATED_CSV, index=False)\n",
    "print(f\"[save] Annotated CSV -> {OUT_ANNOTATED_CSV} (rows={df.shape[0]})\")\n",
    "\n",
    "# ---------- 5) Write annotated VCF (adds CSQ=... to INFO) ----------\n",
    "if not os.path.exists(IN_PROCESSED_VCF):\n",
    "    raise FileNotFoundError(f\"Could not find {IN_PROCESSED_VCF} to copy VCF headers.\")\n",
    "\n",
    "# Collect meta lines and the header line from the processed VCF\n",
    "meta_lines = []\n",
    "header_line = None\n",
    "with open(IN_PROCESSED_VCF, \"r\") as fin:\n",
    "    for line in fin:\n",
    "        if line.startswith(\"##\"):\n",
    "            meta_lines.append(line.rstrip(\"\\n\"))\n",
    "        elif line.startswith(\"#CHROM\"):\n",
    "            header_line = line.rstrip(\"\\n\")\n",
    "            break\n",
    "\n",
    "if header_line is None:\n",
    "    raise RuntimeError(\"Malformed VCF: missing header line starting with #CHROM.\")\n",
    "\n",
    "# Add a CSQ INFO definition (VEP-like short schema description)\n",
    "csq_info_def = '##INFO=<ID=CSQ,Number=.,Type=String,Description=\"Functional consequence summary: Gene|MostSevere|Impact|Terms\">'\n",
    "# Avoid duplicates\n",
    "if not any(l.startswith(\"##INFO=<ID=CSQ,\") for l in meta_lines):\n",
    "    meta_lines.append(csq_info_def)\n",
    "\n",
    "# Determine the column order from the processed VCF header\n",
    "kept_header_cols = header_line.lstrip(\"#\").split(\"\\t\")  # e.g., ['CHROM','POS','ID',...]\n",
    "# Ensure we can pull the same columns from df\n",
    "missing_cols = [c for c in kept_header_cols if c not in df.columns]\n",
    "if missing_cols:\n",
    "    # If FORMAT+samples are present in VCF but not in CSV (rare), warn\n",
    "    print(f\"[warn] Some VCF columns missing in CSV view and will be left blank in output: {missing_cols}\")\n",
    "\n",
    "with open(OUT_ANNOTATED_VCF, \"w\") as vout:\n",
    "    for ml in meta_lines:\n",
    "        vout.write(ml + \"\\n\")\n",
    "    vout.write(header_line + \"\\n\")\n",
    "\n",
    "    # Iterate rows in df in the same columns as VCF expects\n",
    "    for _, row in df.iterrows():\n",
    "        # Build INFO with CSQ appended\n",
    "        info_val = str(row.get(\"INFO\", \".\"))\n",
    "        csq_val = row.get(\"CSQ\", \"\")\n",
    "        if csq_val:\n",
    "            if info_val in (\".\", \"\"):\n",
    "                info_val = f\"CSQ={csq_val}\"\n",
    "            else:\n",
    "                info_val = f\"{info_val};CSQ={csq_val}\"\n",
    "\n",
    "        # Emit line in VCF column order, replacing INFO with updated value\n",
    "        out_parts = []\n",
    "        for col in kept_header_cols:\n",
    "            if col == \"INFO\":\n",
    "                out_parts.append(info_val)\n",
    "            else:\n",
    "                out_parts.append(str(row.get(col, \".\")))\n",
    "        vout.write(\"\\t\".join(out_parts) + \"\\n\")\n",
    "\n",
    "print(f\"[save] Annotated VCF -> {OUT_ANNOTATED_VCF} (rows={df.shape[0]})\")\n",
    "\n",
    "# ---------- 6) Quick summary ----------\n",
    "print(\"\\n=== CSQ Annotation Summary ===\")\n",
    "print(\n",
    "    df[[\"ID\", \"CSQ_gene\", \"CSQ_most_severe\", \"CSQ_impact\"]]\n",
    "      .head(10)\n",
    "      .to_string(index=False)\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "993c35ab-d935-4ddf-8146-dffb8fb616f0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[save] Filtered CSV -> chr22.processed.filtered.vep.selected.csv (rows=4230)\n",
      "[save] Filtered VCF -> chr22.processed.filtered.vep.selected.vcf (rows=4230)\n",
      "\n",
      "=== Functional Consequence Filtering Summary ===\n",
      "Variants before filtering: 155154\n",
      "Variants after filtering:  4230\n",
      "\n",
      "Counts by consequence type:\n",
      "  missense_variant     900\n",
      "  stop_gained          51\n",
      "  5_prime_UTR_variant  950\n",
      "  3_prime_UTR_variant  2322\n",
      "  stop_lost            3\n",
      "  start_lost           4\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "\n",
    "# ========= Inputs/Outputs =========\n",
    "IN_ANNOTATED_CSV = \"chr22.processed.filtered.vep.csv\"   # from previous step\n",
    "IN_ANNOTATED_VCF = \"chr22.processed.filtered.vep.vcf\"\n",
    "OUT_FILTERED_CSV = \"chr22.processed.filtered.vep.selected.csv\"\n",
    "OUT_FILTERED_VCF = \"chr22.processed.filtered.vep.selected.vcf\"\n",
    "# ==================================\n",
    "\n",
    "# ---------- 1) Load ----------\n",
    "df = pd.read_csv(IN_ANNOTATED_CSV, low_memory=False)\n",
    "if \"CSQ_terms\" not in df.columns:\n",
    "    raise RuntimeError(\"Expected 'CSQ_terms' column in annotated CSV.\")\n",
    "\n",
    "n_before = df.shape[0]\n",
    "\n",
    "# ---------- 2) Define allowed consequences ----------\n",
    "keep_terms = {\n",
    "    \"missense_variant\",\n",
    "    \"3_prime_UTR_variant\",\n",
    "    \"5_prime_UTR_variant\",\n",
    "    \"start_lost\",\n",
    "    \"stop_lost\",\n",
    "    \"stop_gained\",\n",
    "}\n",
    "\n",
    "# ---------- 3) Filter ----------\n",
    "def has_keep_term(csq_str: str) -> bool:\n",
    "    if not isinstance(csq_str, str):\n",
    "        return False\n",
    "    terms = [t.strip() for t in csq_str.split(\",\")]\n",
    "    return any(t in keep_terms for t in terms)\n",
    "\n",
    "mask = df[\"CSQ_terms\"].apply(has_keep_term)\n",
    "df_filt = df[mask].copy()\n",
    "n_after = df_filt.shape[0]\n",
    "\n",
    "# ---------- 4) Count per consequence ----------\n",
    "counts = {}\n",
    "for term in keep_terms:\n",
    "    counts[term] = df_filt[\"CSQ_terms\"].str.contains(term, na=False).sum()\n",
    "\n",
    "# ---------- 5) Save CSV ----------\n",
    "df_filt.to_csv(OUT_FILTERED_CSV, index=False)\n",
    "print(f\"[save] Filtered CSV -> {OUT_FILTERED_CSV} (rows={df_filt.shape[0]})\")\n",
    "\n",
    "# ---------- 6) Save VCF ----------\n",
    "if not os.path.exists(IN_ANNOTATED_VCF):\n",
    "    raise FileNotFoundError(f\"Could not find {IN_ANNOTATED_VCF} to copy VCF headers.\")\n",
    "\n",
    "meta_lines = []\n",
    "header_line = None\n",
    "with open(IN_ANNOTATED_VCF, \"r\") as fin:\n",
    "    for line in fin:\n",
    "        if line.startswith(\"##\"):\n",
    "            meta_lines.append(line.rstrip(\"\\n\"))\n",
    "        elif line.startswith(\"#CHROM\"):\n",
    "            header_line = line.rstrip(\"\\n\")\n",
    "            break\n",
    "\n",
    "if header_line is None:\n",
    "    raise RuntimeError(\"Malformed VCF: missing #CHROM header.\")\n",
    "\n",
    "kept_header_cols = header_line.lstrip(\"#\").split(\"\\t\")\n",
    "with open(OUT_FILTERED_VCF, \"w\") as vout:\n",
    "    for ml in meta_lines:\n",
    "        vout.write(ml + \"\\n\")\n",
    "    vout.write(header_line + \"\\n\")\n",
    "    for _, row in df_filt.iterrows():\n",
    "        out_parts = [str(row.get(col, \".\")) for col in kept_header_cols]\n",
    "        vout.write(\"\\t\".join(out_parts) + \"\\n\")\n",
    "\n",
    "print(f\"[save] Filtered VCF -> {OUT_FILTERED_VCF} (rows={df_filt.shape[0]})\")\n",
    "\n",
    "# ---------- 7) Summary ----------\n",
    "print(\"\\n=== Functional Consequence Filtering Summary ===\")\n",
    "print(f\"Variants before filtering: {n_before}\")\n",
    "print(f\"Variants after filtering:  {n_after}\")\n",
    "print(\"\\nCounts by consequence type:\")\n",
    "for k, v in counts.items():\n",
    "    print(f\"  {k:20s} {v}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51e9ad14-62cc-45ea-b004-c55592ff0dda",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "=== Processing chr6 ===\n",
      "=== Processing chr12 ===\n",
      "=== Processing chr17 ===\n",
      "\n",
      "\n",
      "  chr17: annotating 1663075 rsIDs in 8316 chunks (parallelized, 60 threads)\n",
      "  chr17: completed 10/8316 chunks\n",
      "  chr17: completed 20/8316 chunks\n",
      "  chr17: completed 30/8316 chunks\n",
      "⏱️ Timeout on attempt 1/7 for chunk 92/8316. Retrying in 4.9s...\n",
      "  chr17: completed 40/8316 chunks\n",
      "  chr17: completed 50/8316 chunks\n",
      "  chr17: completed 60/8316 chunks\n",
      "  chr17: completed 70/8316 chunks\n",
      "  chr12: annotating 2710352 rsIDs in 13552 chunks (parallelized, 60 threads)\n",
      "  chr17: completed 80/8316 chunks\n",
      "  chr12: completed 10/13552 chunks\n",
      "  chr17: completed 90/8316 chunks\n",
      "  chr12: completed 20/13552 chunks\n",
      "  chr17: completed 100/8316 chunks\n",
      "  chr12: completed 30/13552 chunks\n",
      "  chr17: completed 110/8316 chunks\n",
      "  chr12: completed 40/13552 chunks\n",
      "  chr12: completed 50/13552 chunks\n",
      "  chr17: completed 120/8316 chunks\n",
      "  chr6: annotating 3398240 rsIDs in 16992 chunks (parallelized, 60 threads)\n",
      "  chr12: completed 60/13552 chunks\n",
      "  chr17: completed 130/8316 chunks\n",
      "  chr12: completed 70/13552 chunks\n",
      "  chr6: completed 10/16992 chunks\n",
      "  chr17: completed 140/8316 chunks\n",
      "  chr6: completed 20/16992 chunks\n",
      "  chr12: completed 80/13552 chunks\n",
      "⚠️ Error on attempt 1/7 for chunk 99/13552: 504 Server Error: Gateway Time-out for url: https://rest.ensembl.org/vep/human/id. Retrying in 3.1s...\n",
      "⚠️ Error on attempt 1/7 for chunk 100/13552: 504 Server Error: Gateway Time-out for url: https://rest.ensembl.org/vep/human/id. Retrying in 4.8s...\n",
      "⚠️ Error on attempt 1/7 for chunk 174/8316: 504 Server Error: Gateway Time-out for url: https://rest.ensembl.org/vep/human/id. Retrying in 4.9s...\n",
      "  chr17: completed 150/8316 chunks\n",
      "  chr6: completed 30/16992 chunks\n",
      "  chr12: completed 90/13552 chunks\n",
      "  chr17: completed 160/8316 chunks\n",
      "  chr12: completed 100/13552 chunks\n",
      "  chr17: completed 170/8316 chunks\n",
      "  chr6: completed 40/16992 chunks\n",
      "⚠️ Error on attempt 1/7 for chunk 67/16992: 504 Server Error: Gateway Time-out for url: https://rest.ensembl.org/vep/human/id. Retrying in 3.4s...\n",
      "⚠️ Error on attempt 1/7 for chunk 68/16992: 504 Server Error: Gateway Time-out for url: https://rest.ensembl.org/vep/human/id. Retrying in 3.2s...\n",
      "⚠️ Error on attempt 1/7 for chunk 130/13552: 504 Server Error: Gateway Time-out for url: https://rest.ensembl.org/vep/human/id. Retrying in 3.3s...\n",
      "  chr12: completed 110/13552 chunks\n",
      "  chr17: completed 180/8316 chunks\n",
      "  chr6: completed 50/16992 chunks\n",
      "  chr12: completed 120/13552 chunks\n",
      "  chr17: completed 190/8316 chunks\n",
      "  chr6: completed 60/16992 chunks\n",
      "  chr12: completed 130/13552 chunks\n",
      "⚠️ Error on attempt 1/7 for chunk 88/16992: 504 Server Error: Gateway Time-out for url: https://rest.ensembl.org/vep/human/id. Retrying in 4.4s...\n",
      "  chr17: completed 200/8316 chunks\n",
      "  chr12: completed 140/13552 chunks\n",
      "  chr6: completed 70/16992 chunks\n",
      "  chr12: completed 150/13552 chunks\n",
      "  chr6: completed 80/16992 chunks\n",
      "  chr17: completed 210/8316 chunks\n",
      "  chr12: completed 160/13552 chunks\n",
      "  chr6: completed 90/16992 chunks\n",
      "  chr17: completed 220/8316 chunks\n",
      "  chr12: completed 170/13552 chunks\n",
      "  chr6: completed 100/16992 chunks\n",
      "  chr12: completed 180/13552 chunks\n",
      "  chr17: completed 230/8316 chunks\n",
      "  chr6: completed 110/16992 chunks\n",
      "  chr12: completed 190/13552 chunks\n",
      "  chr17: completed 240/8316 chunks\n",
      "  chr6: completed 120/16992 chunks\n",
      "⚠️ Error on attempt 1/7 for chunk 270/8316: 504 Server Error: Gateway Time-out for url: https://rest.ensembl.org/vep/human/id. Retrying in 3.5s...\n",
      "  chr12: completed 200/13552 chunks\n",
      "  chr17: completed 250/8316 chunks\n",
      "  chr6: completed 130/16992 chunks\n",
      "⚠️ Error on attempt 1/7 for chunk 157/16992: 504 Server Error: Gateway Time-out for url: https://rest.ensembl.org/vep/human/id. Retrying in 4.6s...\n",
      "  chr12: completed 210/13552 chunks\n",
      "  chr17: completed 260/8316 chunks\n",
      "  chr6: completed 140/16992 chunks\n",
      "  chr12: completed 220/13552 chunks\n",
      "  chr17: completed 270/8316 chunks\n",
      "  chr6: completed 150/16992 chunks\n",
      "  chr12: completed 230/13552 chunks\n",
      "  chr17: completed 280/8316 chunks\n",
      "  chr6: completed 160/16992 chunks\n",
      "  chr12: completed 240/13552 chunks\n",
      "  chr17: completed 290/8316 chunks\n",
      "  chr6: completed 170/16992 chunks\n",
      "  chr12: completed 250/13552 chunks\n",
      "  chr17: completed 300/8316 chunks\n",
      "  chr6: completed 180/16992 chunks\n",
      "  chr12: completed 260/13552 chunks\n",
      "  chr17: completed 310/8316 chunks\n",
      "  chr6: completed 190/16992 chunks\n",
      "  chr12: completed 270/13552 chunks\n",
      "  chr17: completed 320/8316 chunks\n",
      "  chr6: completed 200/16992 chunks\n",
      "  chr12: completed 280/13552 chunks\n",
      "  chr6: completed 210/16992 chunks\n",
      "  chr17: completed 330/8316 chunks\n",
      "  chr12: completed 290/13552 chunks\n",
      "  chr6: completed 220/16992 chunks\n",
      "  chr17: completed 340/8316 chunks\n",
      "  chr17: completed 350/8316 chunks\n",
      "  chr12: completed 300/13552 chunks\n",
      "⚠️ Error on attempt 1/7 for chunk 377/8316: 504 Server Error: Gateway Time-out for url: https://rest.ensembl.org/vep/human/id. Retrying in 4.7s...\n",
      "  chr6: completed 230/16992 chunks\n",
      "⚠️ Error on attempt 1/7 for chunk 332/13552: 504 Server Error: Gateway Time-out for url: https://rest.ensembl.org/vep/human/id. Retrying in 4.4s...\n",
      "⚠️ Error on attempt 1/7 for chunk 255/16992: 504 Server Error: Gateway Time-out for url: https://rest.ensembl.org/vep/human/id. Retrying in 4.6s...\n",
      "  chr17: completed 360/8316 chunks\n",
      "⚠️ Error on attempt 1/7 for chunk 337/13552: 500 Server Error: Internal Server Error for url: https://rest.ensembl.org/vep/human/id. Retrying in 5.0s...\n",
      "  chr6: completed 240/16992 chunks\n",
      "  chr12: completed 310/13552 chunks\n",
      "  chr17: completed 370/8316 chunks\n",
      "⚠️ Error on attempt 1/7 for chunk 270/16992: 504 Server Error: Gateway Time-out for url: https://rest.ensembl.org/vep/human/id. Retrying in 3.0s...\n",
      "⚠️ Error on attempt 1/7 for chunk 271/16992: 504 Server Error: Gateway Time-out for url: https://rest.ensembl.org/vep/human/id. Retrying in 3.7s...\n",
      "⚠️ Error on attempt 1/7 for chunk 346/13552: 504 Server Error: Gateway Time-out for url: https://rest.ensembl.org/vep/human/id. Retrying in 4.1s...\n",
      "⚠️ Error on attempt 1/7 for chunk 351/13552: 504 Server Error: Gateway Time-out for url: https://rest.ensembl.org/vep/human/id. Retrying in 4.3s...\n",
      "⚠️ Error on attempt 1/7 for chunk 352/13552: 504 Server Error: Gateway Time-out for url: https://rest.ensembl.org/vep/human/id. Retrying in 3.8s...\n",
      "  chr6: completed 250/16992 chunks\n",
      "  chr17: completed 380/8316 chunks\n",
      "  chr12: completed 320/13552 chunks\n",
      "⚠️ Error on attempt 1/7 for chunk 285/16992: 504 Server Error: Gateway Time-out for url: https://rest.ensembl.org/vep/human/id. Retrying in 3.4s...\n",
      "⚠️ Error on attempt 1/7 for chunk 286/16992: 504 Server Error: Gateway Time-out for url: https://rest.ensembl.org/vep/human/id. Retrying in 3.7s...\n",
      "  chr6: completed 260/16992 chunks\n",
      "  chr12: completed 330/13552 chunks\n",
      "  chr17: completed 390/8316 chunks\n",
      "  chr6: completed 270/16992 chunks\n",
      "  chr12: completed 340/13552 chunks\n",
      "  chr17: completed 400/8316 chunks\n",
      "  chr12: completed 350/13552 chunks\n",
      "  chr6: completed 280/16992 chunks\n",
      "  chr17: completed 410/8316 chunks\n",
      "  chr12: completed 360/13552 chunks\n",
      "  chr6: completed 290/16992 chunks\n",
      "  chr12: completed 370/13552 chunks\n",
      "  chr17: completed 420/8316 chunks\n",
      "  chr6: completed 300/16992 chunks\n",
      "  chr12: completed 380/13552 chunks\n",
      "  chr17: completed 430/8316 chunks\n",
      "  chr6: completed 310/16992 chunks\n",
      "  chr12: completed 390/13552 chunks\n",
      "  chr6: completed 320/16992 chunks\n",
      "  chr12: completed 400/13552 chunks\n",
      "  chr17: completed 440/8316 chunks\n",
      "  chr12: completed 410/13552 chunks\n",
      "  chr17: completed 450/8316 chunks\n",
      "  chr6: completed 330/16992 chunks\n",
      "  chr12: completed 420/13552 chunks\n",
      "  chr6: completed 340/16992 chunks\n",
      "  chr17: completed 460/8316 chunks\n",
      "  chr12: completed 430/13552 chunks\n",
      "  chr6: completed 350/16992 chunks\n",
      "  chr17: completed 470/8316 chunks\n",
      "  chr17: completed 480/8316 chunks\n",
      "  chr12: completed 440/13552 chunks\n",
      "  chr6: completed 360/16992 chunks\n",
      "  chr17: completed 490/8316 chunks\n",
      "  chr12: completed 450/13552 chunks\n",
      "  chr6: completed 370/16992 chunks\n",
      "  chr17: completed 500/8316 chunks\n",
      "  chr12: completed 460/13552 chunks\n",
      "  chr6: completed 380/16992 chunks\n",
      "  chr12: completed 470/13552 chunks\n",
      "  chr17: completed 510/8316 chunks\n",
      "  chr6: completed 390/16992 chunks\n",
      "  chr12: completed 480/13552 chunks\n",
      "  chr6: completed 400/16992 chunks\n",
      "  chr12: completed 490/13552 chunks\n",
      "  chr17: completed 520/8316 chunks\n",
      "  chr6: completed 410/16992 chunks\n",
      "  chr17: completed 530/8316 chunks\n",
      "  chr12: completed 500/13552 chunks\n",
      "  chr6: completed 420/16992 chunks\n",
      "  chr12: completed 510/13552 chunks\n",
      "  chr17: completed 540/8316 chunks\n",
      "  chr6: completed 430/16992 chunks\n",
      "  chr12: completed 520/13552 chunks\n",
      "  chr17: completed 550/8316 chunks\n",
      "  chr12: completed 530/13552 chunks\n",
      "  chr6: completed 440/16992 chunks\n",
      "  chr12: completed 540/13552 chunks\n",
      "  chr6: completed 450/16992 chunks\n",
      "  chr17: completed 560/8316 chunks\n",
      "  chr12: completed 550/13552 chunks\n",
      "  chr6: completed 460/16992 chunks\n",
      "  chr17: completed 570/8316 chunks\n",
      "  chr12: completed 560/13552 chunks\n",
      "  chr6: completed 470/16992 chunks\n",
      "  chr12: completed 570/13552 chunks\n",
      "  chr17: completed 580/8316 chunks\n",
      "  chr6: completed 480/16992 chunks\n",
      "  chr12: completed 580/13552 chunks\n",
      "  chr17: completed 590/8316 chunks\n",
      "  chr6: completed 490/16992 chunks\n",
      "  chr12: completed 590/13552 chunks\n",
      "  chr17: completed 600/8316 chunks\n",
      "  chr6: completed 500/16992 chunks\n",
      "  chr12: completed 600/13552 chunks\n",
      "  chr6: completed 510/16992 chunks\n",
      "  chr12: completed 610/13552 chunks\n",
      "  chr17: completed 610/8316 chunks\n",
      "  chr6: completed 520/16992 chunks\n",
      "  chr12: completed 620/13552 chunks\n",
      "  chr6: completed 530/16992 chunks\n",
      "  chr17: completed 620/8316 chunks\n",
      "  chr12: completed 630/13552 chunks\n",
      "  chr6: completed 540/16992 chunks\n",
      "  chr12: completed 640/13552 chunks\n",
      "  chr17: completed 630/8316 chunks\n",
      "  chr6: completed 550/16992 chunks\n",
      "  chr12: completed 650/13552 chunks\n",
      "  chr17: completed 640/8316 chunks\n",
      "  chr12: completed 660/13552 chunks\n",
      "  chr6: completed 560/16992 chunks\n",
      "  chr12: completed 670/13552 chunks\n",
      "  chr17: completed 650/8316 chunks\n",
      "  chr6: completed 570/16992 chunks\n",
      "  chr12: completed 680/13552 chunks\n",
      "  chr6: completed 580/16992 chunks\n",
      "  chr17: completed 660/8316 chunks\n",
      "  chr12: completed 690/13552 chunks\n",
      "  chr6: completed 590/16992 chunks\n",
      "  chr12: completed 700/13552 chunks\n",
      "  chr17: completed 670/8316 chunks\n",
      "  chr6: completed 600/16992 chunks\n",
      "  chr12: completed 710/13552 chunks\n",
      "  chr17: completed 680/8316 chunks\n",
      "  chr6: completed 610/16992 chunks\n",
      "  chr12: completed 720/13552 chunks\n",
      "  chr17: completed 690/8316 chunks\n",
      "  chr6: completed 620/16992 chunks\n",
      "  chr12: completed 730/13552 chunks\n",
      "  chr6: completed 630/16992 chunks\n",
      "  chr12: completed 740/13552 chunks\n",
      "  chr17: completed 700/8316 chunks\n",
      "  chr12: completed 750/13552 chunks\n",
      "  chr6: completed 640/16992 chunks\n",
      "  chr17: completed 710/8316 chunks\n",
      "  chr12: completed 760/13552 chunks\n",
      "  chr6: completed 650/16992 chunks\n",
      "  chr17: completed 720/8316 chunks\n",
      "  chr12: completed 770/13552 chunks\n",
      "  chr6: completed 660/16992 chunks\n",
      "  chr12: completed 780/13552 chunks\n",
      "  chr17: completed 730/8316 chunks\n",
      "  chr6: completed 670/16992 chunks\n",
      "  chr12: completed 790/13552 chunks\n",
      "  chr17: completed 740/8316 chunks\n",
      "  chr6: completed 680/16992 chunks\n",
      "  chr12: completed 800/13552 chunks\n",
      "  chr12: completed 810/13552 chunks\n",
      "  chr6: completed 690/16992 chunks\n",
      "  chr12: completed 820/13552 chunks\n",
      "  chr12: completed 830/13552 chunks\n",
      "  chr17: completed 750/8316 chunks\n",
      "  chr12: completed 840/13552 chunks\n",
      "  chr12: completed 850/13552 chunks\n",
      "  chr12: completed 860/13552 chunks\n",
      "  chr12: completed 870/13552 chunks\n",
      "  chr17: completed 760/8316 chunks\n",
      "  chr12: completed 880/13552 chunks\n",
      "  chr6: completed 700/16992 chunks\n",
      "  chr12: completed 890/13552 chunks\n",
      "  chr12: completed 900/13552 chunks\n",
      "  chr12: completed 910/13552 chunks\n",
      "  chr12: completed 920/13552 chunks\n",
      "  chr6: completed 710/16992 chunks\n",
      "  chr12: completed 930/13552 chunks\n",
      "  chr12: completed 940/13552 chunks\n",
      "  chr12: completed 950/13552 chunks\n",
      "  chr17: completed 770/8316 chunks\n",
      "  chr12: completed 960/13552 chunks\n",
      "  chr12: completed 970/13552 chunks\n",
      "  chr12: completed 980/13552 chunks\n",
      "  chr6: completed 720/16992 chunks\n",
      "  chr12: completed 990/13552 chunks\n",
      "  chr12: completed 1000/13552 chunks\n",
      "  chr17: completed 780/8316 chunks\n",
      "  chr12: completed 1010/13552 chunks\n",
      "  chr12: completed 1020/13552 chunks\n",
      "  chr6: completed 730/16992 chunks\n",
      "  chr12: completed 1030/13552 chunks\n",
      "  chr12: completed 1040/13552 chunks\n",
      "  chr17: completed 790/8316 chunks\n",
      "  chr12: completed 1050/13552 chunks\n",
      "  chr6: completed 740/16992 chunks\n",
      "  chr12: completed 1060/13552 chunks\n",
      "  chr12: completed 1070/13552 chunks\n",
      "  chr12: completed 1080/13552 chunks\n",
      "  chr12: completed 1090/13552 chunks\n",
      "  chr12: completed 1100/13552 chunks\n",
      "  chr17: completed 800/8316 chunks\n",
      "  chr6: completed 750/16992 chunks\n",
      "  chr12: completed 1110/13552 chunks\n",
      "  chr17: completed 810/8316 chunks\n",
      "  chr6: completed 760/16992 chunks\n",
      "  chr12: completed 1120/13552 chunks\n",
      "  chr17: completed 820/8316 chunks\n",
      "  chr6: completed 770/16992 chunks\n",
      "  chr12: completed 1130/13552 chunks\n",
      "  chr6: completed 780/16992 chunks\n",
      "  chr17: completed 830/8316 chunks\n",
      "  chr12: completed 1140/13552 chunks\n",
      "  chr6: completed 790/16992 chunks\n",
      "  chr12: completed 1150/13552 chunks\n",
      "  chr17: completed 840/8316 chunks\n",
      "  chr12: completed 1160/13552 chunks\n",
      "  chr12: completed 1170/13552 chunks\n",
      "  chr6: completed 800/16992 chunks\n",
      "  chr12: completed 1180/13552 chunks\n",
      "  chr17: completed 850/8316 chunks\n",
      "  chr12: completed 1190/13552 chunks\n",
      "  chr12: completed 1200/13552 chunks\n",
      "  chr6: completed 810/16992 chunks\n",
      "  chr12: completed 1210/13552 chunks\n",
      "  chr12: completed 1220/13552 chunks\n",
      "  chr17: completed 860/8316 chunks\n",
      "  chr6: completed 820/16992 chunks\n",
      "  chr12: completed 1230/13552 chunks\n",
      "  chr6: completed 830/16992 chunks\n",
      "  chr17: completed 870/8316 chunks\n",
      "  chr12: completed 1240/13552 chunks\n",
      "  chr6: completed 840/16992 chunks\n",
      "  chr17: completed 880/8316 chunks\n",
      "  chr12: completed 1250/13552 chunks\n",
      "  chr6: completed 850/16992 chunks\n",
      "  chr12: completed 1260/13552 chunks\n",
      "  chr17: completed 890/8316 chunks\n",
      "  chr6: completed 860/16992 chunks\n",
      "  chr12: completed 1270/13552 chunks\n",
      "  chr17: completed 900/8316 chunks\n",
      "  chr6: completed 870/16992 chunks\n",
      "  chr12: completed 1280/13552 chunks\n",
      "  chr17: completed 910/8316 chunks\n",
      "  chr6: completed 880/16992 chunks\n",
      "  chr12: completed 1290/13552 chunks\n",
      "  chr6: completed 890/16992 chunks\n",
      "  chr12: completed 1300/13552 chunks\n",
      "  chr17: completed 920/8316 chunks\n",
      "  chr6: completed 900/16992 chunks\n",
      "  chr12: completed 1310/13552 chunks\n",
      "  chr17: completed 930/8316 chunks\n",
      "  chr6: completed 910/16992 chunks\n",
      "  chr12: completed 1320/13552 chunks\n",
      "  chr17: completed 940/8316 chunks\n",
      "  chr6: completed 920/16992 chunks\n",
      "  chr12: completed 1330/13552 chunks\n",
      "  chr17: completed 950/8316 chunks\n",
      "  chr6: completed 930/16992 chunks\n",
      "  chr12: completed 1340/13552 chunks\n",
      "  chr17: completed 960/8316 chunks\n",
      "  chr12: completed 1350/13552 chunks\n",
      "  chr6: completed 940/16992 chunks\n",
      "  chr12: completed 1360/13552 chunks\n",
      "  chr17: completed 970/8316 chunks\n",
      "  chr6: completed 950/16992 chunks\n",
      "  chr12: completed 1370/13552 chunks\n",
      "  chr17: completed 980/8316 chunks\n",
      "  chr6: completed 960/16992 chunks\n",
      "  chr12: completed 1380/13552 chunks\n",
      "  chr6: completed 970/16992 chunks\n",
      "  chr17: completed 990/8316 chunks\n",
      "  chr12: completed 1390/13552 chunks\n",
      "  chr6: completed 980/16992 chunks\n",
      "  chr17: completed 1000/8316 chunks\n",
      "  chr12: completed 1400/13552 chunks\n",
      "  chr6: completed 990/16992 chunks\n",
      "  chr12: completed 1410/13552 chunks\n",
      "  chr17: completed 1010/8316 chunks\n",
      "  chr6: completed 1000/16992 chunks\n",
      "  chr12: completed 1420/13552 chunks\n",
      "  chr6: completed 1010/16992 chunks\n",
      "  chr17: completed 1020/8316 chunks\n",
      "  chr12: completed 1430/13552 chunks\n",
      "  chr6: completed 1020/16992 chunks\n",
      "  chr17: completed 1030/8316 chunks\n",
      "  chr12: completed 1440/13552 chunks\n",
      "  chr6: completed 1030/16992 chunks\n",
      "  chr17: completed 1040/8316 chunks\n",
      "  chr12: completed 1450/13552 chunks\n",
      "  chr6: completed 1040/16992 chunks\n",
      "  chr12: completed 1460/13552 chunks\n",
      "  chr17: completed 1050/8316 chunks\n",
      "  chr6: completed 1050/16992 chunks\n",
      "  chr12: completed 1470/13552 chunks\n",
      "  chr17: completed 1060/8316 chunks\n",
      "  chr6: completed 1060/16992 chunks\n",
      "  chr12: completed 1480/13552 chunks\n",
      "  chr6: completed 1070/16992 chunks\n",
      "  chr6: completed 1080/16992 chunks\n",
      "  chr17: completed 1070/8316 chunks\n",
      "  chr12: completed 1490/13552 chunks\n",
      "  chr6: completed 1090/16992 chunks\n",
      "  chr12: completed 1500/13552 chunks\n",
      "  chr17: completed 1080/8316 chunks\n",
      "  chr6: completed 1100/16992 chunks\n",
      "  chr12: completed 1510/13552 chunks\n",
      "  chr17: completed 1090/8316 chunks\n",
      "  chr6: completed 1110/16992 chunks\n",
      "  chr12: completed 1520/13552 chunks\n",
      "  chr17: completed 1100/8316 chunks\n",
      "  chr6: completed 1120/16992 chunks\n",
      "  chr12: completed 1530/13552 chunks\n",
      "  chr17: completed 1110/8316 chunks\n",
      "  chr12: completed 1540/13552 chunks\n",
      "  chr6: completed 1130/16992 chunks\n",
      "⚠️ Error on attempt 1/7 for chunk 1141/16992: 504 Server Error: Gateway Time-out for url: https://rest.ensembl.org/vep/human/id. Retrying in 4.8s...\n",
      "  chr12: completed 1550/13552 chunks\n",
      "  chr17: completed 1120/8316 chunks\n",
      "  chr6: completed 1140/16992 chunks\n",
      "  chr12: completed 1560/13552 chunks\n",
      "  chr17: completed 1130/8316 chunks\n",
      "  chr6: completed 1150/16992 chunks\n",
      "  chr12: completed 1570/13552 chunks\n",
      "  chr6: completed 1160/16992 chunks\n",
      "  chr17: completed 1140/8316 chunks\n",
      "  chr12: completed 1580/13552 chunks\n",
      "  chr17: completed 1150/8316 chunks\n",
      "  chr12: completed 1590/13552 chunks\n",
      "  chr6: completed 1170/16992 chunks\n",
      "  chr17: completed 1160/8316 chunks\n",
      "  chr12: completed 1600/13552 chunks\n",
      "  chr6: completed 1180/16992 chunks\n",
      "  chr12: completed 1610/13552 chunks\n",
      "⚠️ Error on attempt 2/7 for chunk 1141/16992: 504 Server Error: Gateway Time-out for url: https://rest.ensembl.org/vep/human/id. Retrying in 6.0s...\n",
      "  chr17: completed 1170/8316 chunks\n",
      "  chr12: completed 1620/13552 chunks\n",
      "  chr6: completed 1190/16992 chunks\n",
      "  chr17: completed 1180/8316 chunks\n",
      "  chr12: completed 1630/13552 chunks\n",
      "  chr17: completed 1190/8316 chunks\n",
      "  chr12: completed 1640/13552 chunks\n",
      "  chr6: completed 1200/16992 chunks\n",
      "  chr12: completed 1650/13552 chunks\n",
      "  chr6: completed 1210/16992 chunks\n",
      "  chr17: completed 1200/8316 chunks\n",
      "  chr12: completed 1660/13552 chunks\n",
      "  chr17: completed 1210/8316 chunks\n",
      "  chr6: completed 1220/16992 chunks\n",
      "  chr12: completed 1670/13552 chunks\n",
      "  chr17: completed 1220/8316 chunks\n",
      "  chr6: completed 1230/16992 chunks\n",
      "  chr12: completed 1680/13552 chunks\n",
      "  chr12: completed 1690/13552 chunks\n",
      "  chr6: completed 1240/16992 chunks\n",
      "  chr17: completed 1230/8316 chunks\n",
      "  chr12: completed 1700/13552 chunks\n",
      "  chr6: completed 1250/16992 chunks\n",
      "  chr17: completed 1240/8316 chunks\n",
      "  chr12: completed 1710/13552 chunks\n",
      "  chr6: completed 1260/16992 chunks\n",
      "  chr17: completed 1250/8316 chunks\n",
      "  chr12: completed 1720/13552 chunks\n",
      "  chr12: completed 1730/13552 chunks\n",
      "  chr17: completed 1260/8316 chunks\n",
      "  chr6: completed 1270/16992 chunks\n",
      "  chr12: completed 1740/13552 chunks\n",
      "  chr17: completed 1270/8316 chunks\n",
      "  chr6: completed 1280/16992 chunks\n",
      "  chr12: completed 1750/13552 chunks\n",
      "  chr17: completed 1280/8316 chunks\n",
      "  chr6: completed 1290/16992 chunks\n",
      "  chr12: completed 1760/13552 chunks\n",
      "  chr6: completed 1300/16992 chunks\n",
      "  chr17: completed 1290/8316 chunks\n",
      "  chr12: completed 1770/13552 chunks\n",
      "  chr12: completed 1780/13552 chunks\n",
      "  chr6: completed 1310/16992 chunks\n",
      "  chr17: completed 1300/8316 chunks\n",
      "  chr12: completed 1790/13552 chunks\n",
      "  chr6: completed 1320/16992 chunks\n",
      "  chr17: completed 1310/8316 chunks\n",
      "  chr12: completed 1800/13552 chunks\n",
      "  chr17: completed 1320/8316 chunks\n",
      "  chr6: completed 1330/16992 chunks\n",
      "  chr12: completed 1810/13552 chunks\n",
      "  chr17: completed 1330/8316 chunks\n",
      "  chr12: completed 1820/13552 chunks\n",
      "  chr6: completed 1340/16992 chunks\n",
      "  chr12: completed 1830/13552 chunks\n",
      "  chr17: completed 1340/8316 chunks\n",
      "  chr6: completed 1350/16992 chunks\n",
      "  chr12: completed 1840/13552 chunks\n",
      "  chr17: completed 1350/8316 chunks\n",
      "  chr12: completed 1850/13552 chunks\n",
      "  chr6: completed 1360/16992 chunks\n",
      "  chr12: completed 1860/13552 chunks\n",
      "  chr17: completed 1360/8316 chunks\n",
      "  chr6: completed 1370/16992 chunks\n",
      "  chr12: completed 1870/13552 chunks\n",
      "  chr17: completed 1370/8316 chunks\n",
      "  chr6: completed 1380/16992 chunks\n",
      "  chr12: completed 1880/13552 chunks\n",
      "  chr17: completed 1380/8316 chunks\n",
      "  chr12: completed 1890/13552 chunks\n",
      "  chr6: completed 1390/16992 chunks\n",
      "  chr12: completed 1900/13552 chunks\n",
      "  chr17: completed 1390/8316 chunks\n",
      "  chr6: completed 1400/16992 chunks\n",
      "  chr12: completed 1910/13552 chunks\n",
      "  chr17: completed 1400/8316 chunks\n",
      "  chr6: completed 1410/16992 chunks\n",
      "  chr12: completed 1920/13552 chunks\n",
      "  chr17: completed 1410/8316 chunks\n",
      "  chr12: completed 1930/13552 chunks\n",
      "  chr6: completed 1420/16992 chunks\n",
      "  chr17: completed 1420/8316 chunks\n",
      "  chr12: completed 1940/13552 chunks\n",
      "  chr6: completed 1430/16992 chunks\n",
      "  chr12: completed 1950/13552 chunks\n",
      "  chr17: completed 1430/8316 chunks\n",
      "  chr12: completed 1960/13552 chunks\n",
      "  chr6: completed 1440/16992 chunks\n",
      "  chr17: completed 1440/8316 chunks\n",
      "  chr12: completed 1970/13552 chunks\n",
      "  chr6: completed 1450/16992 chunks\n",
      "  chr12: completed 1980/13552 chunks\n",
      "  chr17: completed 1450/8316 chunks\n",
      "  chr12: completed 1990/13552 chunks\n",
      "  chr17: completed 1460/8316 chunks\n",
      "  chr6: completed 1460/16992 chunks\n",
      "  chr12: completed 2000/13552 chunks\n",
      "  chr6: completed 1470/16992 chunks\n",
      "  chr17: completed 1470/8316 chunks\n",
      "  chr12: completed 2010/13552 chunks\n",
      "  chr6: completed 1480/16992 chunks\n",
      "  chr17: completed 1480/8316 chunks\n",
      "  chr12: completed 2020/13552 chunks\n",
      "  chr17: completed 1490/8316 chunks\n",
      "  chr12: completed 2030/13552 chunks\n",
      "  chr6: completed 1490/16992 chunks\n",
      "  chr12: completed 2040/13552 chunks\n",
      "  chr17: completed 1500/8316 chunks\n",
      "  chr6: completed 1500/16992 chunks\n",
      "  chr12: completed 2050/13552 chunks\n",
      "  chr17: completed 1510/8316 chunks\n",
      "  chr12: completed 2060/13552 chunks\n",
      "  chr6: completed 1510/16992 chunks\n",
      "  chr17: completed 1520/8316 chunks\n",
      "  chr12: completed 2070/13552 chunks\n",
      "  chr6: completed 1520/16992 chunks\n",
      "  chr17: completed 1530/8316 chunks\n",
      "  chr12: completed 2080/13552 chunks\n",
      "  chr12: completed 2090/13552 chunks\n",
      "  chr6: completed 1530/16992 chunks\n",
      "  chr17: completed 1540/8316 chunks\n",
      "  chr12: completed 2100/13552 chunks\n",
      "  chr6: completed 1540/16992 chunks\n",
      "  chr17: completed 1550/8316 chunks\n",
      "  chr12: completed 2110/13552 chunks\n",
      "  chr12: completed 2120/13552 chunks\n",
      "  chr17: completed 1560/8316 chunks\n",
      "  chr6: completed 1550/16992 chunks\n",
      "  chr12: completed 2130/13552 chunks\n",
      "  chr6: completed 1560/16992 chunks\n",
      "  chr17: completed 1570/8316 chunks\n",
      "  chr12: completed 2140/13552 chunks\n",
      "  chr17: completed 1580/8316 chunks\n",
      "  chr6: completed 1570/16992 chunks\n",
      "  chr12: completed 2150/13552 chunks\n",
      "  chr6: completed 1580/16992 chunks\n",
      "  chr12: completed 2160/13552 chunks\n",
      "  chr17: completed 1590/8316 chunks\n",
      "  chr12: completed 2170/13552 chunks\n",
      "  chr6: completed 1590/16992 chunks\n",
      "  chr17: completed 1600/8316 chunks\n",
      "  chr12: completed 2180/13552 chunks\n",
      "  chr17: completed 1610/8316 chunks\n",
      "  chr12: completed 2190/13552 chunks\n",
      "  chr6: completed 1600/16992 chunks\n",
      "  chr17: completed 1620/8316 chunks\n",
      "  chr12: completed 2200/13552 chunks\n",
      "  chr6: completed 1610/16992 chunks\n",
      "  chr12: completed 2210/13552 chunks\n",
      "  chr17: completed 1630/8316 chunks\n",
      "  chr6: completed 1620/16992 chunks\n",
      "  chr12: completed 2220/13552 chunks\n",
      "  chr12: completed 2230/13552 chunks\n",
      "  chr17: completed 1640/8316 chunks\n",
      "  chr6: completed 1630/16992 chunks\n",
      "  chr12: completed 2240/13552 chunks\n",
      "  chr17: completed 1650/8316 chunks\n",
      "  chr12: completed 2250/13552 chunks\n",
      "  chr6: completed 1640/16992 chunks\n",
      "  chr12: completed 2260/13552 chunks\n",
      "  chr17: completed 1660/8316 chunks\n",
      "  chr12: completed 2270/13552 chunks\n",
      "  chr6: completed 1650/16992 chunks\n",
      "  chr17: completed 1670/8316 chunks\n",
      "  chr12: completed 2280/13552 chunks\n",
      "  chr12: completed 2290/13552 chunks\n",
      "  chr17: completed 1680/8316 chunks\n",
      "  chr6: completed 1660/16992 chunks\n",
      "  chr12: completed 2300/13552 chunks\n",
      "  chr17: completed 1690/8316 chunks\n",
      "  chr12: completed 2310/13552 chunks\n",
      "  chr6: completed 1670/16992 chunks\n",
      "  chr12: completed 2320/13552 chunks\n",
      "  chr17: completed 1700/8316 chunks\n",
      "  chr6: completed 1680/16992 chunks\n",
      "  chr12: completed 2330/13552 chunks\n",
      "  chr17: completed 1710/8316 chunks\n",
      "  chr12: completed 2340/13552 chunks\n",
      "  chr6: completed 1690/16992 chunks\n",
      "  chr12: completed 2350/13552 chunks\n",
      "  chr17: completed 1720/8316 chunks\n",
      "  chr12: completed 2360/13552 chunks\n",
      "  chr6: completed 1700/16992 chunks\n",
      "  chr12: completed 2370/13552 chunks\n",
      "  chr17: completed 1730/8316 chunks\n",
      "  chr12: completed 2380/13552 chunks\n",
      "  chr6: completed 1710/16992 chunks\n",
      "  chr17: completed 1740/8316 chunks\n",
      "  chr12: completed 2390/13552 chunks\n",
      "  chr17: completed 1750/8316 chunks\n",
      "  chr6: completed 1720/16992 chunks\n",
      "  chr12: completed 2400/13552 chunks\n",
      "  chr12: completed 2410/13552 chunks\n",
      "  chr17: completed 1760/8316 chunks\n",
      "  chr6: completed 1730/16992 chunks\n",
      "  chr12: completed 2420/13552 chunks\n",
      "  chr17: completed 1770/8316 chunks\n",
      "  chr12: completed 2430/13552 chunks\n",
      "  chr6: completed 1740/16992 chunks\n"
     ]
    }
   ],
   "source": [
    "## LOOP IN ALL CHROMOSOMES OF INTEREST. **NOTE ENSEMBLE API BATCH LIMITS, AND TIMEOUT SETTINGS**\n",
    "import os\n",
    "import gzip\n",
    "import json\n",
    "import time\n",
    "import random\n",
    "import requests\n",
    "import pandas as pd\n",
    "from typing import Iterable, List, Dict, Any\n",
    "from concurrent.futures import ThreadPoolExecutor, ProcessPoolExecutor, as_completed\n",
    "from threading import Semaphore\n",
    "import multiprocessing\n",
    "\n",
    "# ========= USER INPUTS =========\n",
    "PHENO_PATH = \"ImmPort_URECA_Cross_Black.csv\"\n",
    "PHENO_ID_COL = \"id\"\n",
    "\n",
    "VCF_DIR = \"/sbgenomics/project-files\"\n",
    "VCF_TEMPLATE = os.path.join(VCF_DIR, \"URECA.WGS.annotated.recoded.chr{chrom}.vcf.gz\")\n",
    "\n",
    "# Chromosomes to process\n",
    "CHROMS = [\"6\", \"12\", \"17\"]\n",
    "OUTDIR = \"per_chrom_outputs\"\n",
    "os.makedirs(OUTDIR, exist_ok=True)\n",
    "\n",
    "QUAL_MIN = 70\n",
    "DP_MIN = 10\n",
    "MAF_MIN = 0.05\n",
    "\n",
    "KEEP_TERMS = {\n",
    "    \"missense_variant\",\n",
    "    \"3_prime_UTR_variant\",\n",
    "    \"5_prime_UTR_variant\",\n",
    "    \"start_lost\",\n",
    "    \"stop_lost\",\n",
    "    \"stop_gained\",\n",
    "}\n",
    "\n",
    "VEP_URL = \"https://rest.ensembl.org/vep/human/id\"\n",
    "BATCH_SIZE = 200\n",
    "MAX_RETRIES = 7\n",
    "BACKOFF_SECS = 3.0\n",
    "TIMEOUT_SECS = 600       # ⬅️ Increase timeout to 10 min\n",
    "FALLBACK_BATCH_SIZE = 50 # ⬅️ Split slow chunks into smaller pieces\n",
    "# ==================================\n",
    "\n",
    "# ---------- helpers ----------\n",
    "def log_step(name: str, before_df: pd.DataFrame, after_df: pd.DataFrame, log_list: list, extra: dict | None = None):\n",
    "    info = {\n",
    "        \"step\": name,\n",
    "        \"variants_before\": before_df.shape[0],\n",
    "        \"variants_after\":  after_df.shape[0],\n",
    "        \"variants_removed\": before_df.shape[0] - after_df.shape[0],\n",
    "        \"samples_before\": max(len(before_df.columns) - 9, 0),\n",
    "        \"samples_after\":  max(len(after_df.columns) - 9, 0),\n",
    "        \"samples_removed\": max(len(before_df.columns) - 9, 0) - max(len(after_df.columns) - 9, 0),\n",
    "    }\n",
    "    if extra:\n",
    "        info.update(extra)\n",
    "    log_list.append(info)\n",
    "    print(f\"[{name}] variants: {info['variants_before']} -> {info['variants_after']} \"\n",
    "          f\"(−{info['variants_removed']}); samples: {info['samples_before']} -> {info['samples_after']} \"\n",
    "          f\"(−{info['samples_removed']})\")\n",
    "\n",
    "def batched(iterable: Iterable[Any], n: int) -> Iterable[List[Any]]:\n",
    "    batch = []\n",
    "    for item in iterable:\n",
    "        batch.append(item)\n",
    "        if len(batch) == n:\n",
    "            yield batch\n",
    "            batch = []\n",
    "    if batch:\n",
    "        yield batch\n",
    "\n",
    "def vep_query_rsids(rsids: List[str], timeout=TIMEOUT_SECS, batch_label=\"\"):\n",
    "    headers = {\"Content-Type\": \"application/json\", \"Accept\": \"application/json\"}\n",
    "    payload = {\"ids\": rsids}\n",
    "    for attempt in range(1, MAX_RETRIES + 1):\n",
    "        try:\n",
    "            r = requests.post(VEP_URL, headers=headers, data=json.dumps(payload), timeout=timeout)\n",
    "            if r.status_code == 429:\n",
    "                wait = float(r.headers.get(\"Retry-After\", BACKOFF_SECS * attempt))\n",
    "                print(f\"⚠️ Rate-limited. Waiting {wait}s before retrying {batch_label}...\")\n",
    "                time.sleep(wait)\n",
    "                continue\n",
    "            r.raise_for_status()\n",
    "            data = r.json()\n",
    "            return {entry[\"id\"]: entry for entry in data if entry.get(\"id\")}\n",
    "        except (requests.exceptions.ReadTimeout, requests.exceptions.ConnectionError) as e:\n",
    "            wait = BACKOFF_SECS * (2 ** (attempt - 1)) + random.uniform(0, 2)\n",
    "            print(f\"⏱️ Timeout on attempt {attempt}/{MAX_RETRIES} for {batch_label}. Retrying in {wait:.1f}s...\")\n",
    "            time.sleep(wait)\n",
    "        except Exception as e:\n",
    "            wait = BACKOFF_SECS * (2 ** (attempt - 1)) + random.uniform(0, 2)\n",
    "            print(f\"⚠️ Error on attempt {attempt}/{MAX_RETRIES} for {batch_label}: {e}. Retrying in {wait:.1f}s...\")\n",
    "            time.sleep(wait)\n",
    "    return {}\n",
    "\n",
    "def summarize_vep(entry: dict) -> dict:\n",
    "    most_severe = entry.get(\"most_severe_consequence\")\n",
    "    tcs = entry.get(\"transcript_consequences\") or []\n",
    "    pick = None\n",
    "    if most_severe and tcs:\n",
    "        for tc in tcs:\n",
    "            if most_severe in (tc.get(\"consequence_terms\") or []):\n",
    "                pick = tc\n",
    "                break\n",
    "    if pick is None and tcs:\n",
    "        pick = tcs[0]\n",
    "    gene_symbol = pick.get(\"gene_symbol\") if pick else None\n",
    "    impact = pick.get(\"impact\") if pick else None\n",
    "    cons_terms = entry.get(\"consequence_terms\")\n",
    "    if not cons_terms and pick:\n",
    "        cons_terms = pick.get(\"consequence_terms\")\n",
    "    if not cons_terms:\n",
    "        for cv in entry.get(\"colocated_variants\", []) or []:\n",
    "            if cv.get(\"consequence_terms\"):\n",
    "                cons_terms = cv.get(\"consequence_terms\")\n",
    "                break\n",
    "    csq_terms = \",\".join(sorted(set(cons_terms or [])))\n",
    "    return {\n",
    "        \"CSQ_most_severe\": most_severe or \"\",\n",
    "        \"CSQ_gene\": gene_symbol or \"\",\n",
    "        \"CSQ_impact\": impact or \"\",\n",
    "        \"CSQ_terms\": csq_terms,\n",
    "    }\n",
    "\n",
    "# ---------- main chromosome worker ----------\n",
    "def process_chromosome(chrom: str) -> dict:\n",
    "    print(f\"\\n=== Processing chr{chrom} ===\")\n",
    "    vcf_input = VCF_TEMPLATE.format(chrom=chrom)\n",
    "    if not os.path.exists(vcf_input):\n",
    "        print(f\"[skip] Missing {vcf_input}\")\n",
    "        return {}\n",
    "\n",
    "    # --- Load phenotype and subset VCF ---\n",
    "    pheno = pd.read_csv(PHENO_PATH)\n",
    "    subset_ids_set = set(pheno[PHENO_ID_COL].astype(str).str.strip())\n",
    "\n",
    "    # Step 1: Subset VCF\n",
    "    out_csv1 = os.path.join(OUTDIR, f\"chr{chrom}.subset.csv\")\n",
    "    meta_lines, header_parts, filtered_rows = [], None, []\n",
    "    base_col_count = 9\n",
    "    with gzip.open(vcf_input, \"rt\") as f:\n",
    "        for line in f:\n",
    "            if line.startswith(\"##\"):\n",
    "                meta_lines.append(line.rstrip(\"\\n\"))\n",
    "                continue\n",
    "            if line.startswith(\"#CHROM\"):\n",
    "                header_parts = line.rstrip(\"\\n\").split(\"\\t\")\n",
    "                all_samples = header_parts[base_col_count:]\n",
    "                kept_samples = [s for s in all_samples if s in subset_ids_set]\n",
    "                sample_name_to_idx = {s: base_col_count + i for i, s in enumerate(all_samples)}\n",
    "                keep_col_idxs = list(range(base_col_count)) + [sample_name_to_idx[s] for s in kept_samples]\n",
    "                break\n",
    "    with gzip.open(vcf_input, \"rt\") as fin:\n",
    "        for line in fin:\n",
    "            if line.startswith(\"#CHROM\"):\n",
    "                break\n",
    "        for line in fin:\n",
    "            if line.startswith(\"#\"):\n",
    "                continue\n",
    "            parts = line.rstrip(\"\\n\").split(\"\\t\")\n",
    "            filtered_rows.append([parts[i] for i in keep_col_idxs])\n",
    "    vcf_df = pd.DataFrame(filtered_rows, columns=header_parts[:base_col_count] + kept_samples)\n",
    "    vcf_df.to_csv(out_csv1, index=False)\n",
    "\n",
    "    # Step 2: QC filtering\n",
    "    df = vcf_df\n",
    "    df = df[~df[\"ALT\"].astype(str).str.contains(\",\", na=True)]\n",
    "    df = df[df[\"REF\"].str.len() == 1]\n",
    "    df = df[df[\"ALT\"].str.len() == 1]\n",
    "    df[\"QUAL\"] = pd.to_numeric(df[\"QUAL\"], errors=\"coerce\")\n",
    "    df = df[df[\"QUAL\"] >= QUAL_MIN]\n",
    "\n",
    "    # Step 3: Parallel VEP annotation with adaptive fallback\n",
    "    rsids = sorted(set(x for x in df[\"ID\"].astype(str) if x not in {\".\", \"nan\", \"NaN\"}))\n",
    "    chunks = list(batched(rsids, BATCH_SIZE))\n",
    "    rsid_to_summary = {}\n",
    "\n",
    "    MAX_THREADS = max(32, multiprocessing.cpu_count() - 4)\n",
    "    RATE_LIMIT = 10\n",
    "    semaphore = Semaphore(RATE_LIMIT)\n",
    "\n",
    "    def vep_worker(chunk, idx):\n",
    "        with semaphore:\n",
    "            result = vep_query_rsids(chunk, batch_label=f\"chunk {idx}/{len(chunks)}\")\n",
    "            if not result and len(chunk) > FALLBACK_BATCH_SIZE:\n",
    "                print(f\"⚠️ Slow chunk detected, splitting into {FALLBACK_BATCH_SIZE}-sized sub-chunks...\")\n",
    "                subresults = {}\n",
    "                for subchunk in batched(chunk, FALLBACK_BATCH_SIZE):\n",
    "                    subresults.update(vep_query_rsids(subchunk, timeout=TIMEOUT_SECS * 2))\n",
    "                    time.sleep(0.5)\n",
    "                return subresults\n",
    "            return result\n",
    "\n",
    "    print(f\"  chr{chrom}: annotating {len(rsids)} rsIDs in {len(chunks)} chunks (parallelized, {MAX_THREADS} threads)\")\n",
    "\n",
    "    with ThreadPoolExecutor(max_workers=MAX_THREADS) as executor:\n",
    "        futures = {executor.submit(vep_worker, chunk, idx): chunk for idx, chunk in enumerate(chunks)}\n",
    "        for i, future in enumerate(as_completed(futures), 1):\n",
    "            try:\n",
    "                result = future.result()\n",
    "                rsid_to_summary.update({k: summarize_vep(v) for k, v in result.items()})\n",
    "                if i % 10 == 0:\n",
    "                    print(f\"  chr{chrom}: completed {i}/{len(chunks)} chunks\")\n",
    "            except Exception as e:\n",
    "                print(f\"  chr{chrom}: ERROR on chunk {i}: {e}\")\n",
    "\n",
    "    # Map annotation\n",
    "    def map_field(series, key):\n",
    "        return series.apply(lambda x: rsid_to_summary.get(str(x), {}).get(key, \"\"))\n",
    "\n",
    "    df[\"CSQ_most_severe\"] = map_field(df[\"ID\"], \"CSQ_most_severe\")\n",
    "    df[\"CSQ_gene\"] = map_field(df[\"ID\"], \"CSQ_gene\")\n",
    "    df[\"CSQ_impact\"] = map_field(df[\"ID\"], \"CSQ_impact\")\n",
    "    df[\"CSQ_terms\"] = map_field(df[\"ID\"], \"CSQ_terms\")\n",
    "\n",
    "    out_csv3 = os.path.join(OUTDIR, f\"chr{chrom}.vep.csv\")\n",
    "    df.to_csv(out_csv3, index=False)\n",
    "\n",
    "    # Step 4: functional filter\n",
    "    mask = df[\"CSQ_terms\"].apply(lambda s: any(t in KEEP_TERMS for t in str(s).split(\",\")))\n",
    "    df_filt = df[mask].copy()\n",
    "    out_csv4 = os.path.join(OUTDIR, f\"chr{chrom}.vep.selected.csv\")\n",
    "    df_filt.to_csv(out_csv4, index=False)\n",
    "\n",
    "    return {\n",
    "        \"chrom\": chrom,\n",
    "        \"variants_before\": len(df),\n",
    "        \"variants_after\": len(df_filt)\n",
    "    }\n",
    "\n",
    "# ---------- run across chromosomes in parallel ----------\n",
    "summary_rows = []\n",
    "with ProcessPoolExecutor(max_workers=len(CHROMS)) as executor:\n",
    "    futures = {executor.submit(process_chromosome, chrom): chrom for chrom in CHROMS}\n",
    "    for future in as_completed(futures):\n",
    "        result = future.result()\n",
    "        if result:\n",
    "            summary_rows.append(result)\n",
    "\n",
    "# ---------- summary ----------\n",
    "summary_df = pd.DataFrame(summary_rows)\n",
    "summary_path = os.path.join(OUTDIR, \"all_chrom_summary.csv\")\n",
    "summary_df.to_csv(summary_path, index=False)\n",
    "print(\"\\n=== Pipeline complete ===\")\n",
    "print(f\"Summary written to {summary_path}\")\n",
    "print(summary_df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "0717d61f-dd25-4ad0-b919-e91f3afe6a2c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "=== chr5 ===\n",
      "=== chr6 ===\n",
      "\n",
      "\n",
      "=== chr12 ===\n",
      "\n",
      "=== chr17 ===\n",
      "  Step1 Subset: 2,191,121 rows; samples kept: 119\n",
      "  Step1b MAF(local): 2,191,121 → 371,771 (MAF >= 0.05)\n",
      "  Step2 QC: start 371,771 → SNP 256,162 → QUAL 256,035 → DP 256,002 → MAF 231,153\n",
      "  Step3 Annotate: matches by key 231,178 | by rsID 375,714\n",
      "  Top DB_INFO tokens (sample): {'SSR': 19997, 'VC': 19997, 'SNV': 19997, 'GNO': 19997, 'RS': 19997, 'FREQ': 19994, '0': 19981, 'dbSNPBuildID': 19958, 'COMMON': 19671, 'GENEINFO': 15469, 'INT': 13867, '.': 8289, 'R5': 5985, 'R3': 3285, '116': 2775, '132': 2210, '119': 1819, '129': 1769, '120': 1561, '0.5': 1366}\n",
      "  Functional rows with ≥1 mapped term: 8,548\n",
      "  Step4 Filter(KEEP_TERMS): 231,153 → 8,548\n",
      "  Clinical summary written (CLNSIG).\n",
      "  Step1 Subset: 3,514,606 rows; samples kept: 119\n",
      "  Step1b MAF(local): 3,514,606 → 580,561 (MAF >= 0.05)\n",
      "  Step2 QC: start 580,561 → SNP 419,377 → QUAL 418,996 → DP 418,995 → MAF 379,189\n",
      "  Step3 Annotate: matches by key 378,864 | by rsID 625,432\n",
      "  Top DB_INFO tokens (sample): {'FREQ': 20000, 'SSR': 20000, 'VC': 20000, 'SNV': 20000, 'RS': 20000, 'GNO': 20000, 'dbSNPBuildID': 19997, '0': 19968, 'COMMON': 19780, 'GENEINFO': 15976, 'INT': 15101, '.': 7834, 'R5': 5429, '120': 4770, '116': 2728, '132': 2014, 'R3': 1973, 'CACNA1C:775': 1567, '129': 1472, '0.5': 1392}\n",
      "  Functional rows with ≥1 mapped term: 9,223\n",
      "  Step4 Filter(KEEP_TERMS): 379,189 → 9,223\n",
      "  Clinical summary written (CLNSIG).\n",
      "  Step1 Subset: 4,278,287 rows; samples kept: 119\n",
      "  Step1 Subset: 4,591,492 rows; samples kept: 119\n",
      "  Step1b MAF(local): 4,278,287 → 763,652 (MAF >= 0.05)\n",
      "  Step2 QC: start 763,652 → SNP 560,432 → QUAL 560,363 → DP 560,358 → MAF 514,643\n",
      "  Step3 Annotate: matches by key 514,534 | by rsID 834,507\n",
      "  Top DB_INFO tokens (sample): {'SSR': 19994, 'VC': 19994, 'RS': 19994, 'SNV': 19994, 'GNO': 19993, 'FREQ': 19992, 'dbSNPBuildID': 19990, '0': 19951, 'COMMON': 19791, 'GENEINFO': 12469, 'INT': 11640, '.': 7232, 'R3': 4930, 'R5': 3720, '119': 3321, '116': 2698, '132': 2246, 'FARS2:10667': 1543, 'GMDS:2762': 1479, '120': 1433}\n",
      "  Functional rows with ≥1 mapped term: 10,813\n",
      "  Step4 Filter(KEEP_TERMS): 514,643 → 10,813\n",
      "  Clinical summary written (CLNSIG).\n",
      "  Step1b MAF(local): 4,591,492 → 756,456 (MAF >= 0.05)\n",
      "  Step2 QC: start 756,456 → SNP 556,842 → QUAL 556,749 → DP 556,741 → MAF 504,397\n",
      "  Step3 Annotate: matches by key 504,423 | by rsID 820,103\n",
      "  Top DB_INFO tokens (sample): {'FREQ': 20000, 'SSR': 20000, 'GNO': 20000, 'VC': 20000, 'RS': 20000, 'SNV': 20000, 'dbSNPBuildID': 19999, '0': 19933, 'COMMON': 19755, 'GENEINFO': 7972, '.': 7765, 'INT': 7046, '116': 2614, '132': 2060, 'R3': 1833, '129': 1705, 'R5': 1611, '119': 1608, '120': 1421, '0.5': 1387}\n",
      "  Functional rows with ≥1 mapped term: 8,711\n",
      "  Step4 Filter(KEEP_TERMS): 504,397 → 8,711\n",
      "  Clinical summary written (CLNSIG).\n",
      "\n",
      "=== Pipeline complete ===\n",
      "  chrom  variants_before  variants_after\n",
      "1    12           379189            9223\n",
      "0    17           231153            8548\n",
      "3     5           504397            8711\n",
      "2     6           514643           10813\n"
     ]
    }
   ],
   "source": [
    "# =============================================================\n",
    "# End-to-end pipeline using dbSNP b157 archive with RefSeq CHROMs\n",
    "# Order preserved:\n",
    "#   1) Subset → 1b) MAF on kept samples → 2) QC (QUAL, DP, MAF from INFO)\n",
    "#   → 3) Functional annotation (dbSNP flags)\n",
    "#   → 4) Filter (KEEP_TERMS) → 5) Clinical annotation (CLNSIG) → 6) Summary\n",
    "# Parallel per chromosome via ThreadPool (Jupyter-safe)\n",
    "# =============================================================\n",
    "import os, re, gzip, pandas as pd\n",
    "import numpy as np\n",
    "from concurrent.futures import ThreadPoolExecutor, as_completed\n",
    "\n",
    "# ---------- USER INPUTS ----------\n",
    "PHENO_PATH   = \"ImmPort_URECA_Cross_Black.csv\"\n",
    "PHENO_ID_COL = \"id\"\n",
    "\n",
    "VCF_DIR      = \"/sbgenomics/project-files\"\n",
    "VCF_TEMPLATE = os.path.join(VCF_DIR, \"URECA.WGS.annotated.recoded.chr{chrom}.vcf.gz\")\n",
    "\n",
    "# New dbSNP archive (b157, GRCh38.p14) that you inspected in R (RefSeq accessions in CHROM)\n",
    "DBSNP_PATH   = os.path.join(VCF_DIR, \"GCF_000001405.40.gz\")\n",
    "\n",
    "CHROMS       = [\"5\", \"6\", \"12\", \"17\"]\n",
    "OUTDIR       = \"per_chrom_outputs_new\"\n",
    "os.makedirs(OUTDIR, exist_ok=True)\n",
    "\n",
    "QUAL_MIN = 70\n",
    "DP_MIN   = 10\n",
    "MAF_MIN  = 0.05\n",
    "\n",
    "# Functional classes you care about (VEP-like)\n",
    "KEEP_TERMS = {\n",
    "    \"missense_variant\",\n",
    "    \"3_prime_UTR_variant\",\n",
    "    \"5_prime_UTR_variant\",\n",
    "    \"start_lost\",\n",
    "    \"stop_lost\",\n",
    "    \"stop_gained\",\n",
    "}\n",
    "\n",
    "# Map dbSNP flag IDs → VEP-like terms\n",
    "FLAG_TO_TERM = {\n",
    "    \"NSM\": \"missense_variant\",\n",
    "    \"NSN\": \"stop_gained\",\n",
    "    \"U3\":  \"3_prime_UTR_variant\",\n",
    "    \"U5\":  \"5_prime_UTR_variant\",\n",
    "    # (dbSNP sometimes has start/stop lost flags in other builds; map if present)\n",
    "    \"SL\":  \"start_lost\",\n",
    "    \"STL\": \"stop_lost\",\n",
    "}\n",
    "\n",
    "# For debugging, also treat these as functional-ish even if not in KEEP_TERMS\n",
    "OTHER_FUNCTION_FLAGS = {\"NSF\",\"ASS\",\"DSS\",\"SYN\",\"INT\",\"R3\",\"R5\"}\n",
    "\n",
    "# ---------- helpers ----------\n",
    "def refseq_to_chr_name(acc: str) -> str | None:\n",
    "    \"\"\"NC_000006.12 → '6', NC_000023.11 → 'X', NC_000024.10 → 'Y', NC_012920.1 → 'MT'.\"\"\"\n",
    "    if not isinstance(acc, str): return None\n",
    "    m = re.match(r\"^NC_(\\d{6})\\.\\d+$\", acc)\n",
    "    if not m:\n",
    "        return acc.replace(\"chr\",\"\")\n",
    "    num = int(m.group(1))\n",
    "    if 1 <= num <= 22: return str(num)\n",
    "    if num == 23: return \"X\"\n",
    "    if num == 24: return \"Y\"\n",
    "    if acc.startswith(\"NC_012920\"): return \"MT\"\n",
    "    return None\n",
    "\n",
    "def extract_float(info_str, key):\n",
    "    if not isinstance(info_str, str): return None\n",
    "    pref = key + \"=\"\n",
    "    for kv in info_str.split(\";\"):\n",
    "        if kv.startswith(pref):\n",
    "            try: return float(kv[len(pref):])\n",
    "            except: return None\n",
    "    return None\n",
    "\n",
    "def get_info_value(info_str, key):\n",
    "    if not isinstance(info_str, str): return None\n",
    "    pref = key + \"=\"\n",
    "    for kv in info_str.split(\";\"):\n",
    "        if kv.startswith(pref):\n",
    "            return kv[len(pref):]\n",
    "    return None\n",
    "\n",
    "def tokens_from_info(info_str):\n",
    "    toks=set()\n",
    "    if not isinstance(info_str,str) or not info_str: return toks\n",
    "    for kv in info_str.split(\";\"):\n",
    "        if not kv: continue\n",
    "        if \"=\" in kv:\n",
    "            k,v = kv.split(\"=\",1)\n",
    "            toks.add(k.strip()); toks.add(v.strip())\n",
    "        else:\n",
    "            toks.add(kv.strip())\n",
    "    return toks\n",
    "\n",
    "def index_has_cols(path, cols):\n",
    "    try:\n",
    "        with open(path, \"r\") as f:\n",
    "            header = f.readline().rstrip(\"\\n\").split(\"\\t\")\n",
    "        return set(cols).issubset(header)\n",
    "    except:\n",
    "        return False\n",
    "\n",
    "# ---- compute MAF on actually kept sample columns ----\n",
    "def maf_from_sample_cols(row, sample_cols):\n",
    "    doses = []\n",
    "    for c in sample_cols:\n",
    "        g = row[c]\n",
    "        if not isinstance(g, str) or g == \"\" or g == \".\":\n",
    "            continue\n",
    "        gt = g.split(\":\", 1)[0]\n",
    "        if gt in {\"./.\", \".|.\", \".\"}:\n",
    "            continue\n",
    "        if \"/\" in gt:\n",
    "            a1, a2 = gt.split(\"/\")\n",
    "        elif \"|\" in gt:\n",
    "            a1, a2 = gt.split(\"|\")\n",
    "        else:\n",
    "            continue\n",
    "        if a1 not in {\"0\",\"1\"} or a2 not in {\"0\",\"1\"}:\n",
    "            continue\n",
    "        doses.append((a1 == \"1\") + (a2 == \"1\"))\n",
    "    if not doses:\n",
    "        return None\n",
    "    doses = np.array(doses, dtype=float)\n",
    "    af = doses.sum() / (2.0 * len(doses))\n",
    "    maf = min(af, 1 - af)\n",
    "    return float(maf)\n",
    "\n",
    "# ---- NEW: parse GENEINFO → gene symbols / IDs ----\n",
    "def geneinfo_to_fields(geneinfo: str):\n",
    "    \"\"\"\n",
    "    dbSNP b157 GENEINFO format is typically like:\n",
    "      'IL13:3596|RAD50:10111'\n",
    "    Returns (symbols_str, ids_str).\n",
    "    \"\"\"\n",
    "    if not isinstance(geneinfo, str) or geneinfo == \"\":\n",
    "        return \"\", \"\"\n",
    "    syms = []\n",
    "    ids  = []\n",
    "    for part in geneinfo.split(\"|\"):\n",
    "        part = part.strip()\n",
    "        if not part:\n",
    "            continue\n",
    "        if \":\" in part:\n",
    "            sym, gid = part.split(\":\", 1)\n",
    "            sym = sym.strip()\n",
    "            gid = gid.strip()\n",
    "            if sym:\n",
    "                syms.append(sym)\n",
    "            if gid:\n",
    "                ids.append(gid)\n",
    "        else:\n",
    "            # just a symbol with no numeric ID\n",
    "            syms.append(part)\n",
    "    syms = sorted(set(syms))\n",
    "    ids  = sorted(set(ids))\n",
    "    return \",\".join(syms), \",\".join(ids)\n",
    "\n",
    "# ---------- Step 0: Build compact per-chrom dbSNP index (normalize CHROM) ----------\n",
    "# Index columns: CHROM  POS  REF  ALT  ID  INFO  CLNSIG\n",
    "def build_dbsnp_index(dbsnp_path, chroms, outdir):\n",
    "    opener = gzip.open if dbsnp_path.endswith(\".gz\") else open\n",
    "    writers = {c: open(os.path.join(outdir, f\"dbsnp_chr{c}.tsv\"), \"w\") for c in chroms}\n",
    "    for w in writers.values():\n",
    "        w.write(\"CHROM\\tPOS\\tREF\\tALT\\tID\\tINFO\\tCLNSIG\\n\")\n",
    "    print(f\"[dbSNP] Streaming {dbsnp_path} to build normalized per-chrom indexes …\")\n",
    "\n",
    "    have_clnsig = False\n",
    "    with opener(dbsnp_path, \"rt\") as fh:\n",
    "        for line in fh:\n",
    "            if line.startswith(\"##INFO=<ID=CLNSIG\"):\n",
    "                have_clnsig = True\n",
    "            if line.startswith(\"#\"):\n",
    "                continue\n",
    "            p = line.rstrip(\"\\n\").split(\"\\t\")\n",
    "            if len(p) < 5: continue\n",
    "            chrom_raw, pos, vid, ref, alts = p[0], p[1], (p[2] if len(p)>2 else \"\"), p[3], p[4]\n",
    "            info = p[7] if len(p)>7 else \"\"\n",
    "            norm_chr = refseq_to_chr_name(chrom_raw)\n",
    "            if norm_chr not in chroms:  # only keep requested chromosomes\n",
    "                continue\n",
    "            clnsig = get_info_value(info, \"CLNSIG\") or \"\"\n",
    "            if not vid or vid == \".\": vid = \"\"\n",
    "            for a in alts.split(\",\"):\n",
    "                writers[norm_chr].write(\n",
    "                    f\"{norm_chr}\\t{pos}\\t{ref}\\t{a}\\t{vid}\\t{info.replace(',', ';')}\\t{clnsig.replace(',', ';')}\\n\"\n",
    "                )\n",
    "    for w in writers.values():\n",
    "        w.close()\n",
    "    print(\"[dbSNP] Indexes built. CLNSIG present:\", have_clnsig)\n",
    "\n",
    "# Build/rebuild if needed\n",
    "need = False\n",
    "for c in CHROMS:\n",
    "    idx = os.path.join(OUTDIR, f\"dbsnp_chr{c}.tsv\")\n",
    "    if not os.path.exists(idx) or not index_has_cols(idx, [\"CHROM\",\"POS\",\"REF\",\"ALT\",\"ID\",\"INFO\",\"CLNSIG\"]):\n",
    "        need = True; break\n",
    "if need:\n",
    "    build_dbsnp_index(DBSNP_PATH, CHROMS, OUTDIR)\n",
    "\n",
    "# ---------- Pipeline (per chromosome) ----------\n",
    "PHENO_IDS = set(pd.read_csv(PHENO_PATH)[PHENO_ID_COL].astype(str).str.strip())\n",
    "\n",
    "def process_chromosome(chrom: str) -> dict:\n",
    "    print(f\"\\n=== chr{chrom} ===\")\n",
    "    vcf_input = VCF_TEMPLATE.format(chrom=chrom)\n",
    "    if not os.path.exists(vcf_input):\n",
    "        print(\"  [skip] missing\", vcf_input)\n",
    "        return {}\n",
    "\n",
    "    # -------- Step 1: Subset to phenotype IDs --------\n",
    "    base=9\n",
    "    with gzip.open(vcf_input, \"rt\") as f:\n",
    "        for line in f:\n",
    "            if line.startswith(\"#CHROM\"):\n",
    "                header=line.rstrip(\"\\n\").split(\"\\t\")\n",
    "                all_samples=header[base:]\n",
    "                kept=[s for s in all_samples if s in PHENO_IDS]\n",
    "                idx={s:base+i for i,s in enumerate(all_samples)}\n",
    "                keep_idx=list(range(base))+[idx[s] for s in kept]\n",
    "                break\n",
    "\n",
    "    rows=[]\n",
    "    with gzip.open(vcf_input,\"rt\") as f:\n",
    "        for line in f:\n",
    "            if line.startswith(\"#CHROM\"): break\n",
    "        for line in f:\n",
    "            if line.startswith(\"#\"): continue\n",
    "            p=line.rstrip(\"\\n\").split(\"\\t\")\n",
    "            rows.append([p[i] for i in keep_idx])\n",
    "    df=pd.DataFrame(rows, columns=header[:base]+kept)\n",
    "    print(f\"  Step1 Subset: {len(df):,} rows; samples kept: {len(kept)}\")\n",
    "    df.to_csv(os.path.join(OUTDIR, f\"chr{chrom}.subset.csv\"), index=False)\n",
    "\n",
    "    # -------- Step 1b: local MAF on the actually kept samples --------\n",
    "    sample_cols = kept\n",
    "    df[\"MAF_local\"] = df.apply(lambda r: maf_from_sample_cols(r, sample_cols), axis=1)\n",
    "    n_before_local = len(df)\n",
    "    df = df[df[\"MAF_local\"].fillna(0) >= MAF_MIN]\n",
    "    n_after_local = len(df)\n",
    "    print(f\"  Step1b MAF(local): {n_before_local:,} → {n_after_local:,} (MAF >= {MAF_MIN})\")\n",
    "\n",
    "    # -------- Step 2: QC (QUAL, DP, MAF) --------\n",
    "    n0=len(df)\n",
    "    df=df[~df[\"ALT\"].astype(str).str.contains(\",\", na=True)]\n",
    "    df=df[df[\"REF\"].str.len()==1]\n",
    "    df=df[df[\"ALT\"].str.len()==1]\n",
    "    n_snp=len(df)\n",
    "    df[\"QUAL\"]=pd.to_numeric(df[\"QUAL\"], errors=\"coerce\")\n",
    "    df=df[df[\"QUAL\"]>=QUAL_MIN]\n",
    "    n_qual=len(df)\n",
    "    if \"INFO\" in df.columns:\n",
    "        df[\"DP_val\"]=df[\"INFO\"].apply(lambda s: extract_float(s,\"DP\"))\n",
    "        df[\"AF_val\"]=df[\"INFO\"].apply(lambda s: extract_float(s,\"AF\") or extract_float(s,\"MAF\"))\n",
    "        df=df[(df[\"DP_val\"].fillna(0)>=DP_MIN)]\n",
    "        n_dp=len(df)\n",
    "        df=df[(df[\"AF_val\"].fillna(0)>=MAF_MIN)]\n",
    "        n_maf=len(df)\n",
    "    else:\n",
    "        n_dp=n_maf=len(df)\n",
    "    print(f\"  Step2 QC: start {n0:,} → SNP {n_snp:,} → QUAL {n_qual:,} → DP {n_dp:,} → MAF {n_maf:,}\")\n",
    "\n",
    "    # -------- Step 3: Functional + Clinical annotation from dbSNP index --------\n",
    "    idx_path=os.path.join(OUTDIR, f\"dbsnp_chr{chrom}.tsv\")\n",
    "    usecols=[\"POS\",\"REF\",\"ALT\",\"ID\",\"INFO\",\"CLNSIG\"]\n",
    "\n",
    "    df[\"__KEY\"]=df.apply(lambda r: f\"{r['POS']}\\t{r['REF']}\\t{r['ALT']}\", axis=1)\n",
    "    need_keys=set(df[\"__KEY\"])\n",
    "    need_ids =set(x for x in df[\"ID\"] if x not in {\".\",\"nan\",\"NaN\",\"\"})\n",
    "\n",
    "    key_INFO = {}\n",
    "    key_CLN  = {}\n",
    "    id_INFO  = {}\n",
    "    id_CLN   = {}\n",
    "    matched_keys = 0\n",
    "    matched_ids  = 0\n",
    "\n",
    "    for chunk in pd.read_csv(idx_path, sep=\"\\t\", dtype=str, chunksize=1_000_000, usecols=usecols):\n",
    "        chunk[\"__KEY\"]=chunk.apply(lambda r: f\"{r['POS']}\\t{r['REF']}\\t{r['ALT']}\", axis=1)\n",
    "        mkey=chunk[chunk[\"__KEY\"].isin(need_keys)]\n",
    "        if not mkey.empty:\n",
    "            key_INFO.update(dict(zip(mkey[\"__KEY\"], mkey[\"INFO\"])))\n",
    "            key_CLN.update(dict(zip(mkey[\"__KEY\"], mkey[\"CLNSIG\"])))\n",
    "            matched_keys+=len(mkey)\n",
    "        mid=chunk[chunk[\"ID\"].isin(need_ids)]\n",
    "        if not mid.empty:\n",
    "            id_INFO.update(dict(zip(mid[\"ID\"], mid[\"INFO\"])))\n",
    "            id_CLN.update(dict(zip(mid[\"ID\"], mid[\"CLNSIG\"])))\n",
    "            matched_ids+=len(mid)\n",
    "\n",
    "    df[\"DB_INFO\"]=df[\"__KEY\"].map(key_INFO).fillna(df[\"ID\"].map(id_INFO)).fillna(\"\")\n",
    "    df[\"CLNSIG\"]=df[\"__KEY\"].map(key_CLN).fillna(df[\"ID\"].map(id_CLN)).fillna(\"\")\n",
    "    print(f\"  Step3 Annotate: matches by key {matched_keys:,} | by rsID {matched_ids:,}\")\n",
    "\n",
    "    # ---- NEW: extract gene information from DB_INFO (GENEINFO field) ----\n",
    "    df[\"GENEINFO_raw\"] = df[\"DB_INFO\"].apply(lambda s: get_info_value(s, \"GENEINFO\") or \"\")\n",
    "    df[[\"GENE_SYMBOLS\",\"GENE_IDS\"]] = df[\"GENEINFO_raw\"].apply(\n",
    "        lambda s: pd.Series(geneinfo_to_fields(s))\n",
    "    )\n",
    "\n",
    "    # Show top tokens so you can verify presence of flags like NSM/NSN/U3/U5\n",
    "    tokens=[]\n",
    "    for s in df[\"DB_INFO\"].head(20000):\n",
    "        tokens.extend(tokens_from_info(s))\n",
    "    if tokens:\n",
    "        top=pd.Series(tokens).value_counts().head(20).to_dict()\n",
    "        print(\"  Top DB_INFO tokens (sample):\", top)\n",
    "\n",
    "    # Parse functional flags → VEP-like terms\n",
    "    def flags_to_terms(info):\n",
    "        if not isinstance(info,str) or not info: return []\n",
    "        flags=set()\n",
    "        for kv in info.split(\";\"):\n",
    "            if not kv: continue\n",
    "            k = kv.split(\"=\",1)[0]  # left side or whole flag\n",
    "            if k in FLAG_TO_TERM:\n",
    "                flags.add(FLAG_TO_TERM[k])\n",
    "            elif k in OTHER_FUNCTION_FLAGS:\n",
    "                # keep for debugging but not part of KEEP_TERMS unless they map above\n",
    "                pass\n",
    "        return sorted(flags)\n",
    "\n",
    "    df[\"CSQ_terms\"]     = df[\"DB_INFO\"].apply(flags_to_terms)\n",
    "    df[\"CSQ_terms_str\"] = df[\"CSQ_terms\"].apply(lambda xs: \",\".join(xs))\n",
    "    mapped_nonempty = int((df[\"CSQ_terms\"].apply(len)>0).sum())\n",
    "    print(f\"  Functional rows with ≥1 mapped term: {mapped_nonempty:,}\")\n",
    "\n",
    "    # -------- Step 4: functional filter (KEEP_TERMS) --------\n",
    "    mask = df[\"CSQ_terms\"].apply(lambda terms: any(t in KEEP_TERMS for t in terms))\n",
    "    df_filt = df[mask].copy()\n",
    "    print(f\"  Step4 Filter(KEEP_TERMS): {len(df):,} → {len(df_filt):,}\")\n",
    "\n",
    "    # -------- Step 5: clinical annotation output (CLNSIG) ON FILTERED SET --------\n",
    "    if (df_filt[\"CLNSIG\"] != \"\").any():\n",
    "        df_filt[\"CLNSIG\"].value_counts().rename_axis(\"CLNSIG\").reset_index(name=\"count\") \\\n",
    "          .to_csv(os.path.join(OUTDIR, f\"chr{chrom}.clinical_summary.csv\"), index=False)\n",
    "        print(\"  Clinical summary written (CLNSIG) on filtered set.\")\n",
    "\n",
    "    # Write per-chrom outputs\n",
    "    # 'annotated' now contains ONLY the functionally filtered variants, with functional + clinical + gene info\n",
    "    df_filt.to_csv(os.path.join(OUTDIR, f\"chr{chrom}.annotated.csv\"), index=False)\n",
    "    df_filt.to_csv(os.path.join(OUTDIR, f\"chr{chrom}.selected.csv\"), index=False)\n",
    "\n",
    "    return {\"chrom\": chrom, \"variants_before\": len(df), \"variants_after\": len(df_filt)}\n",
    "\n",
    "# ---------- Run per chromosome in parallel (threads) ----------\n",
    "summary_rows=[]\n",
    "max_workers=min(len(CHROMS), os.cpu_count() or 1)\n",
    "with ThreadPoolExecutor(max_workers=max_workers) as ex:\n",
    "    futs={ex.submit(process_chromosome, c): c for c in CHROMS}\n",
    "    for fut in as_completed(futs):\n",
    "        res=fut.result()\n",
    "        if res: summary_rows.append(res)\n",
    "\n",
    "# ---------- Summary ----------\n",
    "summary_df=pd.DataFrame(summary_rows).sort_values(\"chrom\")\n",
    "summary_df.to_csv(os.path.join(OUTDIR, \"all_chrom_summary.csv\"), index=False)\n",
    "print(\"\\n=== Pipeline complete ===\")\n",
    "print(summary_df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6e74b090-963f-458d-bd17-07a0b364a54a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>term</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3_prime_UTR_variant</td>\n",
       "      <td>22120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5_prime_UTR_variant</td>\n",
       "      <td>9122</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>missense_variant</td>\n",
       "      <td>6654</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>stop_gained</td>\n",
       "      <td>351</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  term  count\n",
       "0  3_prime_UTR_variant  22120\n",
       "1  5_prime_UTR_variant   9122\n",
       "2     missense_variant   6654\n",
       "3          stop_gained    351"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA90AAAHqCAYAAAAZLi26AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy81sbWrAAAACXBIWXMAAA9hAAAPYQGoP6dpAABuu0lEQVR4nO3dd3xN9+PH8ffNFCQxIolUjCqpWa3ardp7FEVRSlN7k9qbUnsWqVl7FVVB7dXapa3VVjUtJWhEYkTm+f3hl/vNFRTNkYTX8/HI4+Ge8zn3fs51P8l9n884FsMwDAEAAAAAgGRnl9IVAAAAAADgeUXoBgAAAADAJIRuAAAAAABMQugGAAAAAMAkhG4AAAAAAExC6AYAAAAAwCSEbgAAAAAATELoBgAAAADAJIRuAAAAAABMQugGgFTEYrE81s/u3btNr0twcLBq166tLFmyyGKxqEePHqa/5sPcuXNHw4YNe+B5L1y4UBaLRcHBwc+8Xg9SoUIFVahQIaWrgVToxo0b8vDw0IoVK5Ls27dvn5o0aaKXXnpJTk5Ocnd3V9myZTVr1izdvn3bWu727dsaO3asXnvtNbm5ucnV1VV58+ZVkyZNtGfPniTPe+bMGbVu3Vo5c+aUs7OzsmXLpjp16mjr1q1PdQ6jR4/W+vXrn+rY1GDevHl66aWXbN5TADCbxTAMI6UrAQC45+DBgzaPR44cqV27dmnnzp022wsWLCg3NzdT69KgQQPt27dPc+fOlbe3t7Jnz65cuXKZ+poP888//yhbtmwaOnSohg0bZrPv2rVr+v333/X666/L2dk5ReqXWELgfhYXRpC29OzZUzt27NCPP/4oi8Vi3T506FCNGDFCZcuWlb+/v/Lmzas7d+7o+++/1xdffKHmzZtr8uTJiouLU/ny5fXzzz/rk08+UcmSJSVJv/32m7755hu9/fbbGjRokPV5165dq+bNm+vll19Wz5495efnpytXrmjBggXavHmzBg0apJEjRz7ROWTMmFHvvfeeFi5cmCzvybMWGxurggULqlmzZho+fHhKVwfAC8IhpSsAAPif0qVL2zzOli2b7Ozskmx/Fk6ePKmSJUvq3Xfffeav/SSyZcumbNmypXQ1gEe6fv26AgMDNXnyZJvAvXr1ao0YMUL+/v6aM2eOzb6aNWuqT58+OnDggCRp7969+v777zV//ny1adPGWq569erq0qWL4uPjrdt+//13tWzZUkWKFNHu3buVIUMG677GjRurY8eOGjVqlN544w01aNDAzFP/V3FxcYqNjX0mF80cHBzUvn17jRw5Un379lX69OlNf00AYHg5AKQx169fV6dOnazDUF9++WUNHDhQUVFRNuUsFou6dOmiwMBA5c+fX87OzipYsOADh7Ymtnv3blksFp07d06bN2+2DmkPDg5+6FDuhGMS9+5WqFBBhQsX1pEjR/T2228rffr0evnll/XZZ5/ZhAPp3rDb3r176+WXX5azs7M8PT1Vq1YtnT17VsHBwdZQPXz4cGt9WrduLenhw8vnz5+v1157TenSpVOWLFnUoEEDnTlzxqZM69atlTFjRp07d061atVSxowZ5evrq969eyd5P4cPH65SpUopS5YscnNz0xtvvKF58+bpvwwYW7ZsmcqUKaOMGTMqY8aMKlasmObNm2fqecyaNUuvvfaaMmbMKFdXV7366qsaMGCATZmQkBC1b99eOXLkkJOTk/LkyaPhw4crNjbWptylS5fUpEkTubq6yt3dXU2bNtXBgwdlsVhsekIfNuS+devWyp07t8226OhojRo1Sq+++qp1OHSbNm107do1m3K5c+dWnTp1tGXLFr3xxhtycXHRq6++qvnz5yd5nb///lvt2rWTr6+vnJyc5OPjo/fee09XrlyxlomIiFBAQIDy5MkjJycnvfTSS+rRo0eSYcirV69WqVKl5O7ubv1Mf/TRR0le834LFy5UbGysmjZtarN9xIgRypw5s6ZNm2YTuBO4urqqWrVqkqTQ0FBJUvbs2R/4GnZ2//taN3nyZN25c0fTp0+3CdwJJk6cqEyZMj1RT7fFYtHt27f15ZdfWtth4v/Xx/ncBAcHy2KxaNy4cRo1apTy5MkjZ2dn7dq1S8OGDZPFYtFPP/2kxo0by93dXVmyZFGvXr0UGxurX375RTVq1JCrq6ty586tcePG2dQvPj5eo0aNkp+fn1xcXJQpUyYVLVpUU6dOtSnXokULRURE/OvvQgBILvR0A0AacvfuXVWsWFG///67hg8frqJFi2rfvn0aM2aMTpw4oaCgIJvyGzZs0K5duzRixAhlyJBBM2fOVLNmzeTg4KD33nvvga/xxhtv6MCBA2rQoIHy5s2rCRMmSHr4F/1HCQkJUYsWLdS7d28NHTpU69atU//+/eXj46NWrVpJkm7evKm33npLwcHB6tu3r0qVKqVbt25p7969unz5ssqWLastW7aoRo0a8vf318cffyxJj+zdHjNmjAYMGKBmzZppzJgxCg0N1bBhw1SmTBkdOXJE+fLls5aNiYlRvXr15O/vr969e2vv3r0aOXKk3N3dNWTIEGu54OBgtW/fXjlz5pR0bypA165d9ffff9uUe1xDhgzRyJEj1bBhQ/Xu3Vvu7u46efKk/vzzT9POY8WKFerUqZO6du2qCRMmyM7OTufOndPp06dt/s9KliwpOzs7DRkyRHnz5tWBAwc0atQoBQcHa8GCBZKkyMhIValSRZcuXdKYMWOUP39+BQUFJQmVTyI+Pl7169fXvn371KdPH5UtW1Z//vmnhg4dqgoVKujo0aNycXGxlv/xxx/Vu3dv9evXT15eXpo7d678/f31yiuvqHz58pLuBe4SJUooJiZGAwYMUNGiRRUaGqpvv/1WYWFh8vLy0p07d/TOO+/o4sWL1jKnTp3SkCFD9PPPP2v79u2yWCw6cOCAmjZtqqZNm2rYsGFKly6d/vzzzyTTPx4kKChIr7/+ujJlymTddvnyZZ08eVJNmzZ9rB7XN998U46OjurevbuGDBmiSpUqPbRdbtu2TV5eXg8dJZM+fXpVq1ZNq1at0tWrV+Xp6fmvr3/gwAFVqlRJFStW1ODBgyXJOs3lcT83CaZNm6b8+fNrwoQJcnNzU758+azTa5o0aaIPPvhA7du317Zt2zRu3DjFxMRo+/bt6tSpkwICArRs2TL17dtXr7zyiho2bChJGjdunIYNG6ZBgwapfPnyiomJ0dmzZ3Xjxg2b1/b29tarr76qoKCgx7pgAgD/mQEASLU+/PBDI0OGDNbHs2fPNiQZq1atsik3duxYQ5KxdetW6zZJhouLixESEmLdFhsba7z66qvGK6+88q+vnStXLqN27do22xYsWGBIMv744w+b7bt27TIkGbt27bJue+eddwxJxqFDh2zKFixY0Khevbr18YgRIwxJxrZt2x5al2vXrhmSjKFDhybZd3+dwsLCDBcXF6NWrVo25f766y/D2dnZaN68uXXbhx9++MD3s1atWoafn99D6xMXF2fExMQYI0aMMLJmzWrEx8fbnPc777zz0GMNwzDOnz9v2NvbGy1atHhoGTPOo0uXLkamTJkeWbf27dsbGTNmNP7880+b7RMmTDAkGadOnTIMwzBmzZplSDK+/vprm3Jt27Y1JBkLFiywbnvYe/Lhhx8auXLlsj5evny5Icn46quvbModOXLEkGTMnDnTui1XrlxGunTpbOoZGRlpZMmSxWjfvr1120cffWQ4Ojoap0+ffug5jxkzxrCzszOOHDlis33NmjWGJGPTpk0278GNGzce+lwPkz59eqNDhw422w4ePGhIMvr16/fYzzNv3jwjY8aMhiRDkpE9e3ajVatWxt69e23KpUuXzihduvQjn6tv376GpCTn/SgZMmQwPvzwwyTbH/dz88cffxiSjLx58xrR0dE2ZYcOHWpIMiZOnGizvVixYoYkY+3atdZtMTExRrZs2YyGDRtat9WpU8coVqzYY51HixYtDC8vr8cqCwD/FcPLASAN2blzpzJkyJCklzphqPWOHTtstleuXFleXl7Wx/b29mratKnOnTunixcvml5fb29v62JPCYoWLWrTm7t582blz59fVapUSZbXPHDggCIjI63vSQJfX19VqlQpyXtksVhUt27dR9ZRuvfeV6lSRe7u7rK3t5ejo6OGDBmi0NBQXb169YnquG3bNsXFxalz587P9DxKliypGzduqFmzZvr666/1zz//JHndjRs3qmLFivLx8VFsbKz1p2bNmpJkXSF7165dcnV1Vb169WyOb968+b+/AQ+xceNGZcqUSXXr1rV57WLFisnb2zvJ4nTFihWzjjyQpHTp0il//vxJPl8VK1ZUgQIFHvm6hQsXVrFixWxet3r16jbTJkqUKCHpXk/sqlWr9Pfffz/Wed24cUN37tx5rN7kf/PRRx/p4sWLWrZsmbp16yZfX18tWbJE77zzjsaPH/9Ez2X8/9SIBw1rf1KP+7lJUK9ePTk6Oj7wuerUqWPzuECBArJYLNbnku7NzX7llVeSfL5//PFHderUSd9++60iIiIeWl9PT09dvXo1yZQJADADoRsA0pDQ0FB5e3sn+ZLs6ekpBwcH65zPBN7e3kmeI2Hb/WXNkDVr1iTbnJ2dFRkZaX187do15ciRI9le81HzXn18fJKcd/r06ZUuXbokdbx796718eHDh63zaufMmaPvvvtOR44c0cCBAyXJ5nweR8L85Eedtxnn0bJlS82fP19//vmnGjVqJE9PT5UqVUrbtm2zlrly5Yq++eYbOTo62vwUKlRIkqxBPTQ01OaCToIHfeYe15UrV3Tjxg05OTklef2QkJAkFwmS6/N15coV/fTTT0le09XVVYZhWF+3fPnyWr9+vWJjY9WqVSvlyJFDhQsX1vLlyx/5/An1uf//J+GCwR9//PHI4+/n7u6uZs2aaerUqTp06JB++ukneXl5aeDAgdah1Dlz5vzX501YB8HX1/eJXv9BHvdzk+BR01WyZMli89jJyemBn28nJyebz3f//v01YcIEHTx4UDVr1lTWrFlVuXJlHT16NMlrpEuXToZh2BwPAGZhTjcApCFZs2bVoUOHZBiGTfBO6LHx8PCwKR8SEpLkORK2PSiw/JuEL733L871oB7Tx5UtW7Zk7XVPOK/Lly8n2Xfp0qUk79HjWLFihRwdHbVx40abL/5Pe7/ihPnoFy9efGjgMeM8JKlNmzZq06aNbt++rb1792ro0KGqU6eOfv31V+XKlUseHh4qWrSoPv300wce7+PjY63f4cOHk+x/0GcuXbp0Cg8PT7L9/s+Nh4eHsmbNqi1btjzwtV1dXf/1/O73OJ8vDw8Pubi4PHARtoT9CerXr6/69esrKipKBw8e1JgxY9S8eXPlzp1bZcqUeeDxCf+X169ft9mePXt2FSlSRFu3btWdO3eeeiXtQoUK6f3339eUKVP066+/qmTJkqpWrZpmzJihgwcPPnBe9507d7Rt2zYVKlQoWXrgH/dzkyA5etfv5+DgoF69eqlXr166ceOGtm/frgEDBqh69eq6cOGCzft7/fp1OTs7K2PGjMleDwC4Hz3dAJCGVK5cWbdu3UoS9hYtWmTdn9iOHTtsVmiOi4vTypUrlTdv3qfqXU5Yafqnn36y2b5hw4Ynfq4ENWvW1K+//vrIxagSbiX0OD3KZcqUkYuLi5YsWWKz/eLFi9q5c2eS9+hxWCwWOTg4yN7e3rotMjJSixcvfuLnkqRq1arJ3t5es2bNemgZM84jsQwZMqhmzZoaOHCgoqOjderUKUn3hvaePHlSefPm1ZtvvpnkJyE8VaxYUTdv3kzyf79s2bIkr5U7d279+uuvNhdrQkND9f3339uUq1OnjkJDQxUXF/fA1/bz83vi86xZs6Z27dqlX3755aFl6tSpo99//11Zs2Z94Ovev8K6dO8z+c4772js2LGSpOPHjz/0+RPuMvD7778n2Td48GCFhYWpW7duD1wJ/9atW9q6dauke+9ZdHT0A1/j7Nmzkv4Xbnv06KH06dOra9euSVZgl6SAgACFhYWpR48eD633g9w/kiDB435unpVMmTLpvffeU+fOnXX9+vUkdzc4f/68ChYs+EzrBODFRU83AKQhrVq10ueff64PP/xQwcHBKlKkiPbv36/Ro0erVq1aSeZFe3h4qFKlSho8eLB19fKzZ88+9a1ySpQoIT8/PwUEBCg2NlaZM2fWunXrtH///qc+px49emjlypWqX7+++vXrp5IlSyoyMlJ79uxRnTp1VLFiRbm6uipXrlz6+uuvVblyZWXJkkUeHh4PDEOZMmXS4MGDNWDAALVq1UrNmjVTaGiohg8frnTp0mno0KFPXMfatWtr0qRJat68udq1a6fQ0FBNmDDhqe8rnDt3bg0YMEAjR45UZGSkmjVrJnd3d50+fVr//POPhg8fbsp5tG3bVi4uLipXrpyyZ8+ukJAQjRkzRu7u7tb5yiNGjNC2bdtUtmxZdevWTX5+frp7966Cg4O1adMmzZ49Wzly5FCrVq00efJktWrVSp9++qny5cunTZs26dtvv03yui1btlRgYKA++OADtW3bVqGhoRo3bpx15esE77//vpYuXapatWqpe/fuKlmypBwdHXXx4kXt2rVL9evXf+J7So8YMUKbN29W+fLlNWDAABUpUkQ3btzQli1b1KtXL7366qvq0aOHvvrqK5UvX149e/ZU0aJFFR8fr7/++ktbt25V7969VapUKQ0ZMkQXL15U5cqVlSNHDt24cUNTp06Vo6Oj3nnnnUfWo0KFCtq8eXOS7Y0bN9bgwYM1cuRInT17Vv7+/sqbN6/u3LmjQ4cOKTAwUE2bNlW1atW0a9cude/eXS1atFDZsmWVNWtWXb16VcuXL9eWLVusQ94lKW/evFq0aJFatGihEiVKqFevXvLz89OVK1c0f/58bd68WW3atLHeDeBxJdz3+5tvvlH27Nnl6uoqPz+/x/7cmKlu3boqXLiw3nzzTWXLlk1//vmnpkyZoly5ctms9B8fH6/Dhw/L39/f1PoAgFWKLuMGAHik+1cvNwzDCA0NNTp06GBkz57dcHBwMHLlymX079/fuHv3rk05SUbnzp2NmTNnGnnz5jUcHR2NV1991Vi6dOljvfaDVi83DMP49ddfjWrVqhlubm5GtmzZjK5duxpBQUEPXL28UKFCDzynxCtWG8a9lbq7d+9u5MyZ03B0dDQ8PT2N2rVrG2fPnrWW2b59u/H6668bzs7OhiTrCsoPW1F97ty5RtGiRQ0nJyfD3d3dqF+/vnUF5cR1uf/9NYz/raKc2Pz58w0/Pz/D2dnZePnll40xY8YY8+bNS/Laj7N6eYJFixYZJUqUMNKlS2dkzJjReP31121W/U7u8/jyyy+NihUrGl5eXoaTk5Ph4+NjNGnSxPjpp59sjrt27ZrRrVs3I0+ePIajo6ORJUsWo3jx4sbAgQONW7duWctdvHjRaNSokZExY0bD1dXVaNSokfH9998nWb084bULFChgpEuXzihYsKCxcuXKB34WYmJijAkTJhivvfaa9X159dVXjfbt2xu//fabtdzDPp8Pev8vXLhgfPTRR4a3t7fh6OhoPe8rV65Yy9y6dcsYNGiQ4efnZ32vixQpYvTs2dN6B4CNGzcaNWvWNF566SXDycnJ8PT0NGrVqmXs27cvST3ut2PHDkOScfjw4Qfu37Nnj/Hee+8Z2bNnNxwdHQ03NzejTJkyxvjx442IiAjreQwaNMgoV66c4e3tbTg4OBiurq5GqVKljOnTpxuxsbFJnvfkyZNGq1atjBw5chgODg6GJMNisRjz5s371zo/yIkTJ4xy5coZ6dOnNyTZvNeP87lJWL18/PjxSZ474fN67do1m+0P+3zf/ztm4sSJRtmyZQ0PDw/DycnJyJkzp+Hv728EBwfbHJfwf3Hs2LGneg8A4ElZDOMBY5kAAGmexWJR586dNWPGjJSuCl4gwcHBypMnjxYsWJBk5fUXXdGiRVWuXLlHTisw244dO1SrVi01bNhQS5culZ3dizfTsGXLljp//ry+++67lK4KgBfEi/ebFgAAIAWMGzdOCxcufCa363uYypUra+HChVq5cqXatWv3wHnkz7Pff/9dK1eutM7FB4BngTndAAAAz0CNGjU0fvx4/fHHH6bPb36UZs2aqVmzZjbb/u1+1XZ2ds9Fr/hff/2lGTNm6K233krpqgB4gTC8HAAA4AX3b7fw+vDDD7Vw4cJnUxkAeM7Q0w0AAPCCO3LkyCP3P+194QEA9HQDAAAAAGCatD85BwAAAACAVIrh5ckoPj5ely5dkqur67/OjQIAAAAApF2GYejmzZvy8fF55GKThO5kdOnSJfn6+qZ0NQAAAAAAz8iFCxceeVcKQncycnV1lXTvTXdzc0vh2gAAAAAAzBIRESFfX19rDnwYQncyShhS7ubmRugGAAAAgBfAv00tZiE1AAAAAABMQugGAAAAAMAkhG4AAAAAAExC6AYAAAAAwCSEbgAAAAAATELoBgAAAADAJIRuAAAAAABMQugGAAAAAMAkhG4AAAAAAExC6AYAAAAAwCSEbgAAAAAATELoBgAAAADAJIRuAAAAAABM4pDSFcCzlbtfUEpXAXhswZ/VTukqAAAAAP8JPd0AAAAAAJiE0A0AAAAAgEkI3QAAAAAAmITQDQAAAACASQjdAAAAAACYhNANAAAAAIBJCN0AAAAAAJiE0A0AAAAAgEkI3QAAAAAAmITQDQAAAACASQjdAAAAAACYhNANAAAAAIBJCN0AAAAAAJiE0A0AAAAAgEkI3QAAAAAAmITQDQAAAACASQjdAAAAAACYhNANAAAAAIBJCN0AAAAAAJiE0A0AAAAAgEkI3QAAAAAAmITQDQAAAACASQjdAAAAAACYhNANAAAAAIBJUjR0jxkzRiVKlJCrq6s8PT317rvv6pdffrEpYxiGhg0bJh8fH7m4uKhChQo6deqUTZmoqCh17dpVHh4eypAhg+rVq6eLFy/alAkLC1PLli3l7u4ud3d3tWzZUjdu3LAp89dff6lu3brKkCGDPDw81K1bN0VHR5ty7gAAAACA51+Khu49e/aoc+fOOnjwoLZt26bY2FhVq1ZNt2/ftpYZN26cJk2apBkzZujIkSPy9vZW1apVdfPmTWuZHj16aN26dVqxYoX279+vW7duqU6dOoqLi7OWad68uU6cOKEtW7Zoy5YtOnHihFq2bGndHxcXp9q1a+v27dvav3+/VqxYoa+++kq9e/d+Nm8GAAAAAOC5YzEMw0jpSiS4du2aPD09tWfPHpUvX16GYcjHx0c9evRQ3759Jd3r1fby8tLYsWPVvn17hYeHK1u2bFq8eLGaNm0qSbp06ZJ8fX21adMmVa9eXWfOnFHBggV18OBBlSpVSpJ08OBBlSlTRmfPnpWfn582b96sOnXq6MKFC/Lx8ZEkrVixQq1bt9bVq1fl5ub2r/WPiIiQu7u7wsPDH6t8SsjdLyilqwA8tuDPaqd0FQAAAIAHetz8l6rmdIeHh0uSsmTJIkn6448/FBISomrVqlnLODs765133tH3338vSTp27JhiYmJsyvj4+Khw4cLWMgcOHJC7u7s1cEtS6dKl5e7ublOmcOHC1sAtSdWrV1dUVJSOHTv2wPpGRUUpIiLC5gcAAAAAgASpJnQbhqFevXrprbfeUuHChSVJISEhkiQvLy+bsl5eXtZ9ISEhcnJyUubMmR9ZxtPTM8lrenp62pS5/3UyZ84sJycna5n7jRkzxjpH3N3dXb6+vk962gAAAACA51iqCd1dunTRTz/9pOXLlyfZZ7FYbB4bhpFk2/3uL/Og8k9TJrH+/fsrPDzc+nPhwoVH1gkAAAAA8GJJFaG7a9eu2rBhg3bt2qUcOXJYt3t7e0tSkp7mq1evWnulvb29FR0drbCwsEeWuXLlSpLXvXbtmk2Z+18nLCxMMTExSXrAEzg7O8vNzc3mBwAAAACABCkaug3DUJcuXbR27Vrt3LlTefLksdmfJ08eeXt7a9u2bdZt0dHR2rNnj8qWLStJKl68uBwdHW3KXL58WSdPnrSWKVOmjMLDw3X48GFrmUOHDik8PNymzMmTJ3X58mVrma1bt8rZ2VnFixdP/pMHAAAAADz3HFLyxTt37qxly5bp66+/lqurq7Wn2d3dXS4uLrJYLOrRo4dGjx6tfPnyKV++fBo9erTSp0+v5s2bW8v6+/urd+/eypo1q7JkyaKAgAAVKVJEVapUkSQVKFBANWrUUNu2bRUYGChJateunerUqSM/Pz9JUrVq1VSwYEG1bNlS48eP1/Xr1xUQEKC2bdvSgw0AAAAAeCopGrpnzZolSapQoYLN9gULFqh169aSpD59+igyMlKdOnVSWFiYSpUqpa1bt8rV1dVafvLkyXJwcFCTJk0UGRmpypUra+HChbK3t7eWWbp0qbp162Zd5bxevXqaMWOGdb+9vb2CgoLUqVMnlStXTi4uLmrevLkmTJhg0tkDAAAAAJ53qeo+3Wkd9+kGkhf36QYAAEBqlSbv0w0AAAAAwPOE0A0AAAAAgEkI3QAAAAAAmITQDQAAAACASQjdAAAAAACYhNANAAAAAIBJCN0AAAAAAJiE0A0AAAAAgEkI3QAAAAAAmITQDQAAAACASQjdAAAAAACYhNANAAAAAIBJCN0AAAAAAJiE0A0AAAAAgEkI3QAAAAAAmITQDQAAAACASQjdAAAAAACYhNANAAAAAIBJCN0AAAAAAJiE0A0AAAAAgEkI3QAAAAAAmITQDQAAAACASQjdAAAAAACYhNANAAAAAIBJCN0AAAAAAJiE0A0AAAAAgEkI3QAAAAAAmITQDQAAAACASQjdAAAAAACYhNANAAAAAIBJCN0AAAAAAJiE0A0AAAAAgEkI3QAAAAAAmITQDQAAAACASQjdAAAAAACYhNANAAAAAIBJCN0AAAAAAJiE0A0AAAAAgEkI3QAAAAAAmITQDQAAAACASQjdAAAAAACYhNANAAAAAIBJCN0AAAAAAJiE0A0AAAAAgEkI3QAAAAAAmITQDQAAAACASQjdAAAAAACYhNANAAAAAIBJCN0AAAAAAJiE0A0AAAAAgEkI3QAAAAAAmITQDQAAAACASQjdAAAAAACYhNANAAAAAIBJCN0AAAAAAJiE0A0AAAAAgEkI3QAAAAAAmITQDQAAAACASQjdAAAAAACYJEVD9969e1W3bl35+PjIYrFo/fr1Nvtbt24ti8Vi81O6dGmbMlFRUeratas8PDyUIUMG1atXTxcvXrQpExYWppYtW8rd3V3u7u5q2bKlbty4YVPmr7/+Ut26dZUhQwZ5eHioW7duio6ONuO0AQAAAAAviBQN3bdv39Zrr72mGTNmPLRMjRo1dPnyZevPpk2bbPb36NFD69at04oVK7R//37dunVLderUUVxcnLVM8+bNdeLECW3ZskVbtmzRiRMn1LJlS+v+uLg41a5dW7dv39b+/fu1YsUKffXVV+rdu3fynzQAAAAA4IXhkJIvXrNmTdWsWfORZZydneXt7f3AfeHh4Zo3b54WL16sKlWqSJKWLFkiX19fbd++XdWrV9eZM2e0ZcsWHTx4UKVKlZIkzZkzR2XKlNEvv/wiPz8/bd26VadPn9aFCxfk4+MjSZo4caJat26tTz/9VG5ubsl41gAAAACAF0Wqn9O9e/dueXp6Kn/+/Grbtq2uXr1q3Xfs2DHFxMSoWrVq1m0+Pj4qXLiwvv/+e0nSgQMH5O7ubg3cklS6dGm5u7vblClcuLA1cEtS9erVFRUVpWPHjj20blFRUYqIiLD5AQAAAAAgQaoO3TVr1tTSpUu1c+dOTZw4UUeOHFGlSpUUFRUlSQoJCZGTk5MyZ85sc5yXl5dCQkKsZTw9PZM8t6enp00ZLy8vm/2ZM2eWk5OTtcyDjBkzxjpP3N3dXb6+vv/pfAEAAAAAz5cUHV7+b5o2bWr9d+HChfXmm28qV65cCgoKUsOGDR96nGEYslgs1seJ//1fytyvf//+6tWrl/VxREQEwRsAAAAAYJWqe7rvlz17duXKlUu//fabJMnb21vR0dEKCwuzKXf16lVrz7W3t7euXLmS5LmuXbtmU+b+Hu2wsDDFxMQk6QFPzNnZWW5ubjY/AAAAAAAkSFOhOzQ0VBcuXFD27NklScWLF5ejo6O2bdtmLXP58mWdPHlSZcuWlSSVKVNG4eHhOnz4sLXMoUOHFB4eblPm5MmTunz5srXM1q1b5ezsrOLFiz+LUwMAAAAAPIdSdHj5rVu3dO7cOevjP/74QydOnFCWLFmUJUsWDRs2TI0aNVL27NkVHBysAQMGyMPDQw0aNJAkubu7y9/fX71791bWrFmVJUsWBQQEqEiRItbVzAsUKKAaNWqobdu2CgwMlCS1a9dOderUkZ+fnySpWrVqKliwoFq2bKnx48fr+vXrCggIUNu2bem9BgAAAAA8tRQN3UePHlXFihWtjxPmR3/44YeaNWuWfv75Zy1atEg3btxQ9uzZVbFiRa1cuVKurq7WYyZPniwHBwc1adJEkZGRqly5shYuXCh7e3trmaVLl6pbt27WVc7r1atnc29we3t7BQUFqVOnTipXrpxcXFzUvHlzTZgwwey3AAAAAADwHLMYhmGkdCWeFxEREXJ3d1d4eHiq7SHP3S8opasAPLbgz2qndBUAAACAB3rc/Jem5nQDAAAAAJCWELoBAAAAADAJoRsAAAAAAJMQugEAAAAAMAmhGwAAAAAAkxC6AQAAAAAwCaEbAAAAAACTELoBAAAAADAJoRsAAAAAAJMQugEAAAAAMAmhGwAAAAAAkxC6AQAAAAAwCaEbAAAAAACTELoBAAAAADAJoRsAAAAAAJMQugEAAAAAMAmhGwAAAAAAkxC6AQAAAAAwCaEbAAAAAACTELoBAAAAADAJoRsAAAAAAJMQugEAAAAAMMlThe6XX35ZoaGhSbbfuHFDL7/88n+uFAAAAAAAz4OnCt3BwcGKi4tLsj0qKkp///33f64UAAAAAADPA4cnKbxhwwbrv7/99lu5u7tbH8fFxWnHjh3KnTt3slUOAAAAAIC07IlC97vvvitJslgs+vDDD232OTo6Knfu3Jo4cWKyVQ4AAAAAgLTsiUJ3fHy8JClPnjw6cuSIPDw8TKkUAAAAAADPgycK3Qn++OOP5K4HAAAAAADPnacK3ZK0Y8cO7dixQ1evXrX2gCeYP3/+f64YAAAAAABp3VOF7uHDh2vEiBF68803lT17dlksluSuFwAAAAAAad5The7Zs2dr4cKFatmyZXLXBwAAAACA58ZT3ac7OjpaZcuWTe66AAAAAADwXHmq0P3xxx9r2bJlyV0XAAAAAACeK081vPzu3bv64osvtH37dhUtWlSOjo42+ydNmpQslQMAAAAAIC17qtD9008/qVixYpKkkydP2uxjUTUAAAAAAO55qtC9a9eu5K4HAAAAAADPnaea0w0AAAAAAP7dU/V0V6xY8ZHDyHfu3PnUFQIAAAAA4HnxVKE7YT53gpiYGJ04cUInT57Uhx9+mBz1AgAAAAAgzXuq0D158uQHbh82bJhu3br1nyoEAAAAAMDzIlnndH/wwQeaP39+cj4lAAAAAABpVrKG7gMHDihdunTJ+ZQAAAAAAKRZTzW8vGHDhjaPDcPQ5cuXdfToUQ0ePDhZKgYAAAAAQFr3VKHb3d3d5rGdnZ38/Pw0YsQIVatWLVkqBgAAAABAWvdUoXvBggXJXQ8AAAAAAJ47TxW6Exw7dkxnzpyRxWJRwYIF9frrrydXvQAAAAAASPOeKnRfvXpV77//vnbv3q1MmTLJMAyFh4erYsWKWrFihbJly5bc9QQAAAAAIM15qtXLu3btqoiICJ06dUrXr19XWFiYTp48qYiICHXr1i256wgAAAAAQJr0VD3dW7Zs0fbt21WgQAHrtoIFC+rzzz9nITUAAAAAAP7fU/V0x8fHy9HRMcl2R0dHxcfH/+dKAQAAAADwPHiq0F2pUiV1795dly5dsm77+++/1bNnT1WuXDnZKgcAAAAAQFr2VKF7xowZunnzpnLnzq28efPqlVdeUZ48eXTz5k1Nnz49uesIAAAAAECa9FRzun19ffXDDz9o27ZtOnv2rAzDUMGCBVWlSpXkrh8AAAAAAGnWE/V079y5UwULFlRERIQkqWrVquratau6deumEiVKqFChQtq3b58pFQUAAAAAIK15otA9ZcoUtW3bVm5ubkn2ubu7q3379po0aVKyVQ4AAAAAgLTsiUL3jz/+qBo1ajx0f7Vq1XTs2LH/XCkAAAAAAJ4HTxS6r1y58sBbhSVwcHDQtWvX/nOlAAAAAAB4HjxR6H7ppZf0888/P3T/Tz/9pOzZs//nSgEAAAAA8Dx4otBdq1YtDRkyRHfv3k2yLzIyUkOHDlWdOnWSrXIAAAAAAKRlTxS6Bw0apOvXryt//vwaN26cvv76a23YsEFjx46Vn5+frl+/roEDBz728+3du1d169aVj4+PLBaL1q9fb7PfMAwNGzZMPj4+cnFxUYUKFXTq1CmbMlFRUeratas8PDyUIUMG1atXTxcvXrQpExYWppYtW8rd3V3u7u5q2bKlbty4YVPmr7/+Ut26dZUhQwZ5eHioW7duio6OfpK3BwAAAAAAG08Uur28vPT999+rcOHC6t+/vxo0aKB3331XAwYMUOHChfXdd9/Jy8vrsZ/v9u3beu211zRjxowH7h83bpwmTZqkGTNm6MiRI/L29lbVqlV18+ZNa5kePXpo3bp1WrFihfbv369bt26pTp06iouLs5Zp3ry5Tpw4oS1btmjLli06ceKEWrZsad0fFxen2rVr6/bt29q/f79WrFihr776Sr17936StwcAAAAAABsWwzCMpzkwLCxM586dk2EYypcvnzJnzvzfKmKxaN26dXr33Xcl3evl9vHxUY8ePdS3b19J93q1vby8NHbsWLVv317h4eHKli2bFi9erKZNm0qSLl26JF9fX23atEnVq1fXmTNnVLBgQR08eFClSpWSJB08eFBlypTR2bNn5efnp82bN6tOnTq6cOGCfHx8JEkrVqxQ69atdfXq1QfeIu1BIiIi5O7urvDw8Mc+5lnL3S8opasAPLbgz2qndBUAAACAB3rc/PdEPd2JZc6cWSVKlFDJkiX/c+B+kD/++EMhISGqVq2adZuzs7Peeecdff/995KkY8eOKSYmxqaMj4+PChcubC1z4MABubu7WwO3JJUuXVru7u42ZQoXLmwN3JJUvXp1RUVFPfIWaFFRUYqIiLD5AQAAAAAgwVOHbrOFhIRIUpLh6l5eXtZ9ISEhcnJyShL67y/j6emZ5Pk9PT1tytz/OpkzZ5aTk5O1zIOMGTPGOk/c3d1dvr6+T3iWAAAAAIDnWaoN3QksFovNY8Mwkmy73/1lHlT+acrcr3///goPD7f+XLhw4ZH1AgAAAAC8WFJt6Pb29pakJD3NV69etfZKe3t7Kzo6WmFhYY8sc+XKlSTPf+3aNZsy979OWFiYYmJiHrkwnLOzs9zc3Gx+AAAAAABIkGpDd548eeTt7a1t27ZZt0VHR2vPnj0qW7asJKl48eJydHS0KXP58mWdPHnSWqZMmTIKDw/X4cOHrWUOHTqk8PBwmzInT57U5cuXrWW2bt0qZ2dnFS9e3NTzBAAAAAA8vxxS8sVv3bqlc+fOWR//8ccfOnHihLJkyaKcOXOqR48eGj16tPLly6d8+fJp9OjRSp8+vZo3by5Jcnd3l7+/v3r37q2sWbMqS5YsCggIUJEiRVSlShVJUoECBVSjRg21bdtWgYGBkqR27dqpTp068vPzkyRVq1ZNBQsWVMuWLTV+/Hhdv35dAQEBatu2Lb3XAAAAAICnlqKh++jRo6pYsaL1ca9evSRJH374oRYuXKg+ffooMjJSnTp1UlhYmEqVKqWtW7fK1dXVeszkyZPl4OCgJk2aKDIyUpUrV9bChQtlb29vLbN06VJ169bNusp5vXr1bO4Nbm9vr6CgIHXq1EnlypWTi4uLmjdvrgkTJpj9FgAAAAAAnmNPfZ9uJMV9uoHkxX26AQAAkFqZfp9uAAAAAADwaIRuAAAAAABMQugGAAAAAMAkhG4AAAAAAExC6AYAAAAAwCSEbgAAAAAATJKi9+kGgOcFt+NDWsGt+AAAeLbo6QYAAAAAwCSEbgAAAAAATELoBgAAAADAJIRuAAAAAABMQugGAAAAAMAkhG4AAAAAAExC6AYAAAAAwCSEbgAAAAAATELoBgAAAADAJIRuAAAAAABMQugGAAAAAMAkhG4AAAAAAExC6AYAAAAAwCSEbgAAAAAATELoBgAAAADAJIRuAAAAAABMQugGAAAAAMAkhG4AAAAAAExC6AYAAAAAwCSEbgAAAAAATELoBgAAAADAJIRuAAAAAABMQugGAAAAAMAkhG4AAAAAAExC6AYAAAAAwCSEbgAAAAAATELoBgAAAADAJIRuAAAAAABMQugGAAAAAMAkhG4AAAAAAExC6AYAAAAAwCSEbgAAAAAATELoBgAAAADAJA4pXQEAAIAHyd0vKKWrADy24M9qp3QVAKRS9HQDAAAAAGASQjcAAAAAACYhdAMAAAAAYBJCNwAAAAAAJiF0AwAAAABgEkI3AAAAAAAmIXQDAAAAAGASQjcAAAAAACYhdAMAAAAAYBJCNwAAAAAAJiF0AwAAAABgEkI3AAAAAAAmIXQDAAAAAGASQjcAAAAAACYhdAMAAAAAYBJCNwAAAAAAJiF0AwAAAABgEkI3AAAAAAAmSdWhe9iwYbJYLDY/3t7e1v2GYWjYsGHy8fGRi4uLKlSooFOnTtk8R1RUlLp27SoPDw9lyJBB9erV08WLF23KhIWFqWXLlnJ3d5e7u7tatmypGzduPItTBAAAAAA8x1J16JakQoUK6fLly9afn3/+2bpv3LhxmjRpkmbMmKEjR47I29tbVatW1c2bN61levTooXXr1mnFihXav3+/bt26pTp16iguLs5apnnz5jpx4oS2bNmiLVu26MSJE2rZsuUzPU8AAAAAwPPHIaUr8G8cHBxsercTGIahKVOmaODAgWrYsKEk6csvv5SXl5eWLVum9u3bKzw8XPPmzdPixYtVpUoVSdKSJUvk6+ur7du3q3r16jpz5oy2bNmigwcPqlSpUpKkOXPmqEyZMvrll1/k5+f37E4WAAAAAPBcSfU93b/99pt8fHyUJ08evf/++zp//rwk6Y8//lBISIiqVatmLevs7Kx33nlH33//vSTp2LFjiomJsSnj4+OjwoULW8scOHBA7u7u1sAtSaVLl5a7u7u1zMNERUUpIiLC5gcAAAAAgASpOnSXKlVKixYt0rfffqs5c+YoJCREZcuWVWhoqEJCQiRJXl5eNsd4eXlZ94WEhMjJyUmZM2d+ZBlPT88kr+3p6Wkt8zBjxoyxzgN3d3eXr6/vU58rAAAAAOD5k6pDd82aNdWoUSMVKVJEVapUUVBQkKR7w8gTWCwWm2MMw0iy7X73l3lQ+cd5nv79+ys8PNz6c+HChX89JwAAAADAiyNVh+77ZciQQUWKFNFvv/1mned9f2/01atXrb3f3t7eio6OVlhY2CPLXLlyJclrXbt2LUkv+v2cnZ3l5uZm8wMAAAAAQII0FbqjoqJ05swZZc+eXXny5JG3t7e2bdtm3R8dHa09e/aobNmykqTixYvL0dHRpszly5d18uRJa5kyZcooPDxchw8ftpY5dOiQwsPDrWUAAAAAAHgaqXr18oCAANWtW1c5c+bU1atXNWrUKEVEROjDDz+UxWJRjx49NHr0aOXLl0/58uXT6NGjlT59ejVv3lyS5O7uLn9/f/Xu3VtZs2ZVlixZFBAQYB2uLkkFChRQjRo11LZtWwUGBkqS2rVrpzp16rByOQAAAADgP0nVofvixYtq1qyZ/vnnH2XLlk2lS5fWwYMHlStXLklSnz59FBkZqU6dOiksLEylSpXS1q1b5erqan2OyZMny8HBQU2aNFFkZKQqV66shQsXyt7e3lpm6dKl6tatm3WV83r16mnGjBnP9mQBAAAAAM8di2EYRkpX4nkREREhd3d3hYeHp9r53bn7BaV0FYDHFvxZ7ZSuwmOjbSGtoF0B5khLbQtA8njc/Jem5nQDAAAAAJCWELoBAAAAADAJoRsAAAAAAJMQugEAAAAAMAmhGwAAAAAAkxC6AQAAAAAwCaEbAAAAAACTELoBAAAAADAJoRsAAAAAAJMQugEAAAAAMAmhGwAAAAAAkxC6AQAAAAAwCaEbAAAAAACTELoBAAAAADAJoRsAAAAAAJMQugEAAAAAMAmhGwAAAAAAkxC6AQAAAAAwCaEbAAAAAACTELoBAAAAADAJoRsAAAAAAJMQugEAAAAAMAmhGwAAAAAAkxC6AQAAAAAwCaEbAAAAAACTELoBAAAAADAJoRsAAAAAAJMQugEAAAAAMAmhGwAAAAAAkxC6AQAAAAAwCaEbAAAAAACTELoBAAAAADAJoRsAAAAAAJMQugEAAAAAMAmhGwAAAAAAkxC6AQAAAAAwCaEbAAAAAACTELoBAAAAADAJoRsAAAAAAJMQugEAAAAAMAmhGwAAAAAAkxC6AQAAAAAwCaEbAAAAAACTELoBAAAAADAJoRsAAAAAAJMQugEAAAAAMAmhGwAAAAAAkxC6AQAAAAAwCaEbAAAAAACTELoBAAAAADAJoRsAAAAAAJMQugEAAAAAMAmhGwAAAAAAkxC6AQAAAAAwCaEbAAAAAACTELoBAAAAADAJoRsAAAAAAJMQugEAAAAAMIlDSlcAAAAAwLOTu19QSlcBeCzBn9VO6SokC3q67zNz5kzlyZNH6dKlU/HixbVv376UrhIAAAAAII0idCeycuVK9ejRQwMHDtTx48f19ttvq2bNmvrrr79SumoAAAAAgDSI0J3IpEmT5O/vr48//lgFChTQlClT5Ovrq1mzZqV01QAAAAAAaRBzuv9fdHS0jh07pn79+tlsr1atmr7//vsHHhMVFaWoqCjr4/DwcElSRESEeRX9j+Kj7qR0FYDHlprb0v1oW0graFeAOWhbQPJL7e0qoX6GYTyyHKH7//3zzz+Ki4uTl5eXzXYvLy+FhIQ88JgxY8Zo+PDhSbb7+vqaUkfgReM+JaVrADx/aFeAOWhbQPJLK+3q5s2bcnd3f+h+Qvd9LBaLzWPDMJJsS9C/f3/16tXL+jg+Pl7Xr19X1qxZH3oMnj8RERHy9fXVhQsX5ObmltLVAZ4btC0g+dGugORHu3pxGYahmzdvysfH55HlCN3/z8PDQ/b29kl6ta9evZqk9zuBs7OznJ2dbbZlypTJrCoilXNzc+MXLWAC2haQ/GhXQPKjXb2YHtXDnYCF1P6fk5OTihcvrm3bttls37Ztm8qWLZtCtQIAAAAApGX0dCfSq1cvtWzZUm+++abKlCmjL774Qn/99Zc6dOiQ0lUDAAAAAKRBhO5EmjZtqtDQUI0YMUKXL19W4cKFtWnTJuXKlSulq4ZUzNnZWUOHDk0y1QDAf0PbApIf7QpIfrQr/BuL8W/rmwMAAAAAgKfCnG4AAAAAAExC6AYAAAAAwCSEbgAAAAAATELoBgAAAADAJIRuAAAAAABMQugGAAAAgEQmTpyo8PDwlK4GnhOEbuAxxMfHp3QVgOcSbQtIHnFxcTaPuSMs8PR+++03ffLJJ+rQoYNu3ryZ0tXBc4DQDdzn3Llzun37tiRp1KhRunbtmuzsaCrAf0XbAsxjb28vSTp58qQkyWKxpGR1gDQtX758OnDggLZt2yZ/f396vPGf8W0HSOTIkSOqWLGi1q1bp65du2rIkCH6559/UrpaQJpH2wLMsX79evXo0UOS1KNHD/Xr108REREpWykgDUsYgVWqVCl98803CgoK0oABA3Tjxo2UrRjSNIeUrgCQmpQoUUK1a9dWjx49dPfuXe3fv18FChRQfHw8PXLAf0DbApJfdHS0/v77b82ZM0eHDx/WyZMndfDgQbm5uaV01YA0yTAM69+kYcOGKTQ0VO7u7po1a5bCw8M1c+ZM2heeCt90gP8XGxsrSSpdurSio6OVJUsWnTt3Tjdu3JCdnR3z44CnRNsCzOHk5KT27durVKlSOnjwoBo2bKiCBQtKSjrHG8C/S5iWMXbsWE2fPl0NGjTQsmXLNH/+fH3zzTdq164dI0nwVAjdeOElDCNycLg38KNSpUr67bffVKtWLY0cOVIrV65UeHh4kvlxLAAFPBptCzBH4gtVMTExqlChgvr166egoCD17t1b0r053jExMQ89DsCDGYahI0eOqGXLlqpUqZIqVKig1q1ba/369QoKClKPHj0UFhaW0tVEGkPoxgst8dDWo0eP6ueff5ajo6O8vLw0e/Zsvf3225owYYLWrFljXb3y448/VmhoKENigUegbQHmiI+Pt7lQ5eLioiFDhmjgwIEaNmyYFi5caA3ejo6OkqSDBw9KYnE14N8YhqG4uDhdvHhRoaGh1u1xcXGqWLGi/P39tXDhQrVp08a6MCjwOCwGlz0B9enTRytXrtSNGzdUsWJFNW3aVM2aNZMk+fv7a//+/Xr77bcVHBys48eP68qVK9beOwAPR9sCko9hGNbgPGnSJP3888+6ffu2Pv30U+XLl083b97U4sWLNWTIEDVr1kwjR45Us2bNlDVrVi1evJjQDdznYeuKzJkzR8OGDdPs2bNVt25d6/ZJkybpu+++U2RkpDZu3MhFYjw2Pil4ISW+1rRnzx6tX79eS5YsUWBgoDJkyKBJkyZp/vz5kqR58+apSZMmioqKkqenp0JCQuTg4MB8OeABaFuAORL3cI8YMUKffvqpLBaLzp07p1KlSmnjxo1ydXVVq1atNG7cOC1atEjFixdXSEiIFixYQOAG7pM4cO/atUtLlixRUFCQLly4oCZNmlhHZK1bt06SdP36de3cuVM1atTQpk2bZGdnx3QoPDZ6uvFCW7t2rbZs2aKcOXNq0KBBkqSffvpJ06ZN0/Hjx9WlSxe1adNGknT37l2lS5dO0r2FoeiNAx6OtgWY48qVKxo+fLhatWql0qVLS5Jat26ttWvXaunSpdZeuQsXLujnn39W9erVZW9vT9sCHqJPnz5au3atXF1d5eHhod9//13btm1TRESEpk2bptWrV+ull15SbGysMmTIoB9++IG2hCdGTzdeWH/++admzJih1atX6+rVq9btRYsWVffu3fX6669r1qxZmjlzpiRZQ4FhGPyyBR6BtgWYY9myZcqePbv27NkjZ2dn6/aFCxeqUaNG+uCDD7Rx40bFxsbK19dXtWrVkr29veLi4mhbwAPMmzdPX375pZYsWaLjx4+rRo0aCg4O1o8//qjXX39dEyZM0Pbt2+Xv769+/fpZAzcjsvCkCN14YeXKlUvDhg1T2bJltXHjRgUFBVn3FSlSRD169JCvr6+OHj1qM2SWIXrAo9G2AHPUq1dPjRs31pkzZxQSEiLpf1M6FixYoPfee0/16tXToUOHbI6zt7d/5nUF0oJTp06pbdu2Kl26tL7++msNGzZMgYGBatiwoW7fvq3Y2FiVLl1affr0Udu2ba2BmzaFJ8XwcrwQEs/buX/RjL1792rixIm6deuWAgICVLNmTeu+8+fPK3fu3NZ7CRMKAFu0LcAcD1vg6fbt22rcuLF+/PFHffPNN3rjjTds9o8ePVp9+vShZxt4DP7+/ipQoIBeffVVNWvWTOPHj1eHDh0UHx+vxYsXKyIiQu3atbMZWQI8DUI3nnuJv7jMmTNHhw4dkp2dncqVK6cPP/xQkrRjxw5Nnz5dN2/e1CeffKIaNWo89DkA3EPbAsxx/y33oqOjlSlTJhUsWFCSFBkZqXfffVenT5/Whg0b9Prrrye5eMUcbuB/Hva3ZsCAAfryyy918+ZNjRs3Th06dJB0b9G0Zs2a6e2337auSwL8F4RuPNcSfwnp27evFi5cqObNmyskJERnzpxRlSpVNGHCBEnSzp07NWPGDJ07d05ffPGFdYEaAEnRtgBzJG5bgwYN0pIlS+Tk5KQ///xTo0ePVsuWLeXp6am7d+/q3Xff1dmzZ7Vy5UqVKlUqhWsOpE6JA/e2bdtkb28vFxcXlSlTRpJUsWJF/fTTT9qyZYteeuklRUdHq2PHjgoNDdX333/PxSskCz5FeK4lfHFZsGCB1q5dq40bN6pEiRJauXKl1q5dqxs3bujWrVuaPXu2KlWqpKioKO3Zs0clSpRI4ZoDqRttC0h+icPBp59+qnnz5mn58uWqUKGCunXrpv79+ys0NFQ9e/ZUtmzZtH79epUrV06fffaZ9bZGAGwltKmAgAAtWLBAGTJkkCR16tRJ/fr105o1a1SzZk01bdpU4eHhyp8/vyTpu+++Yw43kg2hG8+dChUqaMCAAapWrZp1W1hYmBo3bqwSJUro66+/VseOHTV69GiFh4dr2rRpypgxoyZMmKCaNWta553ySxawRdsCzBEYGKi2bdtaw8G5c+d06NAhzZo1SxUqVLDe775Zs2b67LPPJEldu3ZV9uzZdeDAAXrigPskDORNuEB8/vx5bd26VTt27FBMTIz27Nmjvn37KiYmRoMHD9bhw4e1ZcsWRUREyMfHR2XKlOFWe0heBvCcmTBhgnH37t0k24ODg43Lly8bRYsWNcaPH28YhmGcPn3a8PT0NDJkyGB8+umnz7qqQJpC2wKS36xZs4wWLVoYsbGx1m2XL182Fi9ebNy5c8f47rvvjBw5chjTpk0zDMMw2rdvb6RLl87o0qWLcf36desxcXFxz7zuQGp0f1uYMGGC0bp1a6N79+7WbREREcaUKVMMOzs7Y9iwYY/1PMB/waUbPHd69+4t6d4Krjly5FCrVq0k3buN0e7du3X79m299957kqSoqCi98847eu+996zbADwYbQtIfi1btlTbtm1lb2+vXbt2qXz58vL29lbdunXl4uKiNWvWqHz58mrXrp0kKXPmzCpVqpSOHz8ud3d36/OwICEgvffee8qUKZPmzp0rSYqIiNCFCxe0atUqVapUyVrO1dVV/v7+slgsCggIUGRkpHUUSQLaFJITnyY8N+Lj463/vn79um7cuKHWrVtrzZo11u1ubm6Kj4/X0qVLFRwcrAEDBsjFxUWNGzeWnZ2d4uLiUqLqQKpG2wLMERMTowwZMsje3l579uxR+/bt1b9/f8XFxcnd3V3R0dH67bffZGdnJwcHBxmGoTNnzmj06NHav3+/9ZZ7AO4ZPny4Zs6caX3s5uam7t27q3fv3goKCtKcOXOs+zJmzKiPPvpIw4YN03fffUdbgqno6cZzI+GK5IABAxQaGqopU6ZIkj744AMZhqHGjRvr5ZdfVsOGDTVr1izNmjVL2bNn19dffy2LxSLDMJhnCjwAbQswh6OjoyRp1apVKl26tBo0aKD9+/dr8ODBGjlypJycnFSrVi117txZYWFh+uuvvxQfH6+SJUtKEve4B+5TqFAhSdKMGTP0xRdf6MSJE8qTJ4/atWun6Oho9erVSxaLRR9//LGke8G7Z8+e6t+/v/XvFW0KZiB0I81L/Aty27ZtWr9+vb788ku5uLho3LhxiouL0wcffKD4+Hg1bdpUgwYNkr+/vy5duqQKFSqwUAbwELQtwByJVykfP368+vbtqz///FP9+/fXuHHjtHPnTsXHx2vUqFHq2LGjHBwcdOjQIfn5+Wns2LGsqAz8izfeeENhYWGqXLmyduzYoRw5cqhz586yWCzq3bu37Ozs9NFHH0mSXFxcJHERC+biPt14bqxcuVKHDh2SnZ2dJkyYYPNlv3fv3vr888+1aNEiNWnSxOY4vrgAj0bbAsxx9OhRff/998qbN69q164t6d4c1M8++0w7duxQpUqVNHLkSDk4OCgmJsbaM87FLOB/El/EShAXF6cffvhBTZo0ka+vr3bv3i07OztduHBBs2fP1pgxY7RhwwbVqVMnhWqNFw2/sfFciIuL0+TJk3X48GHr7YwcHBysv4gnTpwoOzs7vf/++8qWLZsqVqxoPZZQADwcbQswx549e1SxYkW5urpa10eIiYmRm5ub+vfvL0navXu3unXrphkzZlgDtyQCN/D/Egfuc+fOycHBQRkzZpSHh4eKFy+uVatWqUmTJqpQoYJ2794tX19fffzxx/L19VWNGjVSuPZ4kbCQGtKk+wdoJCxC07BhQ506dUpLlixRdHS07OzsrItAjR8/XtOmTdPbb7+dElUG0gTaFvBs5MmTR4MHD1ZcXJyOHTsm6d4c79jYWLm6uqp///564403FBsby5BX4AEMw7AG7hEjRqh+/fqqWrWq3njjDe3bt092dnYqUaKEVq1apYsXL6py5cqKi4tTnjx51KFDBzk4OCg2NjaFzwIvCoaXI81JfFUzLi5OhmFYr/rfvXtX9evXV2hoqAYMGKC6devK0dExydAjhuYBSdG2AHM8aPirJP3111+aNWuWJkyYoM8//9x6W7CEdnTnzh25uLiwwBPwCEOHDlVgYKC++OILFS5cWO3bt9fx48c1e/Zs6y0rjxw5onfeeUdt2rTR559/nsI1xouIb0ZIUxJ/cZk8ebKOHTumX3/9VT179lTJkiWVN29erV+/XvXq1dOYMWNksVhUp04dm2F5EkPzgPvRtgBzJG5bO3fu1K1bt2QYhurXr6+cOXOqc+fOsrOz0yeffCKLxaK2bdtaF0pLnz69JBZ4Ah7m6NGj2rlzp5YsWaIqVapow4YNOnr0qIoUKaIPPvhAFotFjRo1UokSJXTs2DHlz58/pauMFxQ93UiT+vfvr3nz5qlr1676559/FBQUpJo1a6p9+/YqXLiwIiMj1aBBA505c0aLFi3SO++8k9JVBtIE2haQfBKH5f79+2v16tWys7OTo6Oj8ubNqw0bNkiSLly4oMDAQH3++ecaOnSoevTokYK1BlKv+y9AnTlzRlu3blX37t21c+dOtWjRQgMHDlSnTp1Uvnx5/fbbb5owYYJatmxpPYZFPpESmNONNGfFihVatWqVNm3apMGDB+v999/X+fPntWXLFk2bNk1nz56Vi4uL1q5dqwYNGuitt95K6SoDaQJtC0heCeFg3LhxWrBggZYuXapff/1VrVu31saNG1WlShXFxsbK19dXHTp00AcffKCgoKAkaysAuCehTZ08eVKSVKBAATVv3lySNG/ePL333nvq1KmTDMOQr6+vMmTIoHnz5tk8B4EbKYHQjVQvYbGmBOnSpVOHDh305ptvasOGDapVq5bmz5+vnj17atGiRZo6dap++OEHpU+fXlOmTJG9vb3i4uJSqPZA6kXbAsyRODT/+eefOnr0qL744guVKlVKQUFBGjVqlPr06aNffvlFtWvXVmxsrHLkyKHBgwdr69at1jncAO5J/Pfqq6++Urt27bR8+XJJUrZs2RQREaFTp07J19fXutBndHS0Vq9erV27dqVUtQErhpcjzdiyZYvKlSsnwzAUGRkpOzs71atXT++995569+6tu3fv6tVXX9Xdu3cVEBCggIAA5sEBj4G2BZhryZIlqlq1qv788081btxY/fv3V4cOHTRkyBCNGjVKr732mo4dO2ad+037Av4n8boIGzZs0I4dO7Rw4ULlz59f/fv3V8OGDSVJbdu21apVq9SxY0ft2rVL0dHROnr0qOzt7R+6mCHwrPDpQ5pw4MABde7cWadPn5abm5u8vLwUGhqqa9eu6bXXXpMkXbp0SZUqVdLo0aPVs2dPSeJLC/AvaFuAOcaMGaNOnTpJkj744AN5eXlp9+7dKlmypD744ANJUvbs2dW8eXMVLlzYpmeb9gX8T0JY7tevn9q2baucOXOqX79+Cg0N1ZQpU6w93tOmTVOLFi107Ngx5cuXT4cPH7aOyCJwI6WxzCzShJIlS8rV1VXTpk3T0qVLJUm3b9+Wk5OTvvvuOxmGoalTp8re3l5t2rSRxWJhoQzgMdC2AHNkyZJFCxYs0E8//aSiRYtKks6ePatffvlFGTNmVGRkpLZu3aq3335bvXr1ksQCT8DDnDlzRsuXL9f8+fNVu3ZtSVLDhg3VsWNHTZo0SenSpVODBg00c+ZMRUREyM3NTRK3sUTqwWUfpDoJ83YSrvpHR0fL3t5e48eP19GjR7V9+3ZJUvHixdW4cWMtXbpUbdu21Y0bN7RmzRrrXDi+uAC2aFvAs1OmTBm5uLjo8OHD1m3t2rXTpUuXlD9/fpUqVUq//fabunXrZt1P2wIeLGPGjDIMQzExMZLuXaDy8/PTnDlzdO7cOY0fP14rVqyQJGvgNgyDwI1Ug9CNVCdhCNCRI0ckSU5OTpKkV155RR4eHtq7d6+17PDhw7Vx40Z9++232rt3rxwdHRUbG8vQPOABaFuAuaKjo63/Llq0qGrXrq3Bgwfr2rVrkqRixYpp06ZNatCggZo1a6YTJ05Y78kN4OEcHBzk4OCgY8eOSbo3BSM+Pl558+ZV8eLFFRERoaVLl+rEiRPWY/h7hdSE0I1Uaffu3XrrrbdUs2ZNrVy5Unfu3FGePHnUvn17TZgwQT/++KO1bP78+eXn52ddrZKrmsDD0bYAc0yaNEmDBg3Snj17rNu6d++ul19+WUuWLFF8fLzSpUunkiVLauzYserfv78cHBwUGxtLDzfwL7Jnz64RI0Zo9OjRCgwMlJ2dnezs7BQdHa3s2bNryJAhOnbsmNauXZvSVQUeiNCNVKlo0aL64Ycf5OjoqMmTJ+vNN99UUFCQihUrpiZNmmjVqlWKjY1NchwLZQCPRtsCzBEeHq6TJ0+qfv366t69u3bs2CEvLy+VLFlSX331lbUN3d+rzcUs4PF88MEHGjNmjDp27KhmzZqpS5cuqlq1qn788Uc1adJElStX1g8//JDS1QQeiFuGIVWLjo7Wzz//rJkzZ2rfvn3KlSuXzpw5oyxZsmj//v3WeTsAngxtC0h+YWFh2r9/v0aOHKno6Gi9+uqr+vjjj1WzZk1NnTrVupo5gKe3ZcsWzZ07V3fu3JGnp6e++OILOTk5qWbNmipQoIAmTZqU0lUEkiB0I9W6/56KO3fu1PHjxzVw4EC99NJL+u233+h9A54CbQtIXvffV/vatWs6efKkBg8erOvXr+vs2bPq2LGjPv/88xSsJZD2JbS1mJgYOTo6SpJu3ryp0aNHa/78+dqzZ49effXVFK4lkBShG6ne/QHhypUr8vDwsN57kblwwNOhbQHJ7/4Avm7dOp0+fVp9+/ZlKDlwn4T2krjd3N+GHiThb1RwcLBmzZql5cuXa8OGDSpWrNgzqDXw5AjdSFH3f+l/kvKEAuDhaFtAynpQG+SewcD/JG4jFy5cUKZMmSRJrq6uj/wbljiUR0VF6fz583J1dVWOHDmeSb2Bp0HoxjP3999/KyoqSh4eHswbBZIRbQswx+P0vP3bcU/7HMDzbvDgwVqzZo3i4+Pl4eGhsWPH6q233npg2cTtaNq0abpz54769ev3LKsLPBUm7eGZWrZsmerWrasKFSqoYMGC+vbbbyXd+yX6KIn3R0VFmVpHIC2ibQHmiI+Pt37Jj4mJsVl9/FHt6/6QTeAG7kncbpYuXapZs2ZpyJAh6tq1q/LkyaPKlStrzZo1kpSkvSW0oy+++EJ9+vRR7ty5n2ndgadFTzeemdmzZ6tnz56aNGmSsmbNqi+++EInT57U4cOHlTNnzocel/iX7OTJk3Xo0CEtXrzYuoAG8KKjbQHmSDzEdeLEiTp+/Lh++eUXNWrUSHXq1FHhwoUfeFzitrVx40ZlzJhRFSpUeFbVBtKEb7/9Vhs2bFCRIkXUoUMHSfemYAwaNEiTJk3S8ePHVahQIUm2bSowMFB9+vTRwoUL1aBBgxSrP/Ak6OnGM7Fo0SJ16tRJmzZtUseOHdWkSRPVr19f165d0+HDh63l4uPjJf3vKuj9VzWHDx+uevXqEQqA/0fbAsyTELj79++vzz77TOXLl1fDhg21bNkyderUSeHh4UmOSdy2Zs2apVatWjGPG7jPwYMH1a9fPy1btszaPuLj42Vvb69BgwapXLlymj9/vgzDUFxcXJIe7vnz5xO4kaYQumG6K1euaO7cufL09FTFihWt2zdv3izDMHTo0CFNmTJFJ0+eVFhYmKT/DcNLfFXzk08+0fz589W8efNnfxJAKkTbAsx3/Phxbdy4URs2bFC7du305ptv6rffftNHH30kd3d3m6Gy9/fGDRgwQIGBgQ+dnwq8qEqXLq1WrVopU6ZMmjt3rq5cuWK9yJUxY0ZlzpxZV69elcVisS7sOXXqVPXs2VMLFixQo0aNUrL6wBMjdMNUmzdv1sWLFzVo0CDlypVLpUuXliS1aNFCv/zyi6ZPn66MGTNqx44dqlSpkmrVqqXWrVvr119/tT7HnDlzrFc1GzZsmFKnAqQqtC3g2YiMjFRMTIzKlCmjr776So0aNdLkyZPVunVr3blzR2vXrtWNGzckKUlv3Ny5c9W4ceMUrD2Q8hJGWt3/uGfPnurdu7diYmIUEBCgGzduyGKxKDY2ViEhIXJ3d7ceYxiGLl++rLlz5/L3CmkS451gmsDAQHXs2FGHDx9WtWrV5OjoqN69eyt9+vTKnTu3fvzxR2XMmNFafseOHfrpp5+0ZcsW5c2bV9K9uaqdO3fWmjVrGEYE/D/aFmCOB92mKDIyUhkyZNDixYvVpUsXjR071jr/9MiRI9q4caMKFSpkvd3RrFmz1KdPH3355ZeEA7zwErepwMBAHT58WLGxsSpTpow6dOigLl26KD4+XjNnzlSxYsX05ptvyt7eXmFhYZo8ebKk/40g+eyzz1LyVID/hIXUYIp58+apU6dOWrlypd59911J937x7tq1S6NGjdI///yj48ePy8HBQVFRUXJ2dn7g8yxbtkzOzs4MIwL+H20LMEficLBgwQJlzZpV9erVkySVKVPGOl2jW7dukqS7d++qUaNGSpcunVavXi07OzudPHlSrVu3Vr9+/fTee++l2LkAqU3fvn21YMEC1axZU3fv3tWaNWvUtGlTBQYGytXVVZ9//rmmTp2qzJkzq0uXLmrZsqUk7m2P5wehG8lu+fLlatGihcaOHatPPvlE0r1bPtjb2ysuLk579uzRJ598Int7e+3evVvp06dXTEwMCzgB/4K2BZgj8Vzsvn37auXKlfroo4/Uvn17eXl56eTJk2revLkMw1CXLl0UExOjr7/+WpcvX9aJEydsQsHvv/9uHVEC4N6iaY0aNdLKlSut6xscPHhQNWrUUOPGjTVnzhxJ0qRJk7Rp0yblzp1bkyZNkpub2wNHnwBpEZ9iJKvAwEC1aNFC2bNn1++//65jx45Jkuzt7a2rUlaoUEETJkyQYRiqXLmybt++TSgA/gVtCzBPQuCeOHGi5s+fr6+++kpDhgyRl5eXJKlw4cLavHmz/Pz8NHv2bK1evVq5c+e2jiqJjY21zlMlcONFl/je2pJ0+/ZtOTk5yc/PT9K93uvSpUtr9erVWrJkibZt2yZJ6tWrl2rXrq1ff/1Vbdu2VWhoKIEbzw0+yUg206dPV5cuXbRjxw7NmTNHR48e1bRp0/TDDz9IunfrlYQrlu+8847Gjx+vv//+W927d0/hmgOpG20LMN+dO3d06NAhDR06VMWLF9fvv/+ur776SlWqVFHbtm0VExOjNWvWaOfOndq6davmzJkjR0dH6/BXwgEgXb9+3braeML8bU9PT128eNHmYrEkFSpUSNmzZ7e59V7Pnj1Vo0YNhYeHKzo6+tmfAGASJkkgWZw/f17Lli3TokWLrLcuunnzpsaPH6+pU6eqe/fueuONN2zCwdtvv63169frtddeS+HaA6kXbQswR+Ih5ZKUPn163bp1SwsXLpSvr69mzJih2NhY5cuXT998843CwsK0Zs0aZc6c2eY5mG8K3LNr1y6NHz9ec+fO1bhx47R27VodO3ZM+fLlU7NmzfTpp58qffr0Kl++vCQpQ4YMcnFxsd52L2G61IABA9ShQwdlyZIlJU8HSFbM6cZ/tnv3bkVGRqpIkSLKkSOHzRzSlStXasKECSpYsKA1HEj/+8Wa4P7HAGhbgFkSzxNN3EYOHjyoTz75RGfOnFGXLl1UvXp1lSlTRnPmzNG6dev01VdfycXFJSWrDqRay5cv15w5c/T333/rn3/+0ZEjR/Tyyy9LuhfIp0+frl9++UVt27aVh4eHFi9erKtXr+ro0aPWNsgcbjyvCN14aoZhKDY2VgULFpSbm5umT5+u0qVLy87Ozma1yYRwUKhQIXXv3l2vv/56CtccSN1oW4B5EvdwT58+XXv37lWePHlUq1YtVahQQZJ08eJF5ciRw3pM1apVlStXLs2dOzclqgykaokvXHXo0EFffPGFKlSooLlz51pDt3Tvotb69es1d+5cvfLKK/Ly8tKaNWvk6OjIBWI89wjd+M+uX7+uqlWrytnZWePGjVO5cuVksVhsfoGuWrVKkyZNkqenpyZMmKD8+fOncK2B1I+2BSSvxIH7008/1YQJE9SgQQMdPXpUGTNm1AcffKBOnTpJkiIiIrR//35NmzZNly5d0rFjx+To6JhkWDrwIkvcM71q1SqdPn1avr6+WrVqldKlS6eRI0eqaNGiNseEh4fL0dFRLi4uslgs3BYMLwTGb+CJJb5OExsbqyxZsmj79u26deuW+vTpo/3798swDOttjCSpSZMm6tChg7Jly6ZXXnklpaoOpGq0LcBcCWH56NGjunbtmtavX6/58+dr9erVeu211zR37lzNnj1bkvTLL7/o66+/Vvr06fXDDz9YF00jcAP3GIZhDdz9+vXTwIEDlS1bNvn7++vDDz/UrVu3NHjwYP3888/WY7Zu3SpHR0elT59eFouFdRHwwqCnG08sMjJSLi4uNvcHtre3V1hYmCpUqKAMGTJo2rRpevPNNyUlXaxGYs4O8CC0LcB869at07BhwxQfH6+goCDlzJlT0r2QPXXqVB06dEhdunRRmzZtdPHiRfn4+CSZ2gHgf0aOHKlp06YpKChI+fPnV6ZMmSRJX3/9tWbPni3DMNSpUyfNmjVLV65c0bFjx7h4hRcO38zwRPbs2aNXXnlFwcHBNqEgLi5OmTNn1u7duxUaGqo+ffooJiZGkqxXMhMjFAC2aFvAs5E1a1blzp1bwcHB2rt3r3W7n5+fevTooTJlymjQoEHasGGDcuTIYb0zAIEbSOr69evau3evpkyZopIlS+r27dvatWuX2rZtq7t376pKlSrKkCGDunXrpqioKB06dOiBf7uA5x1/QfBE0qVLpzx58qhGjRraunWrcubMaQ0HsbGxypw5s7Zv365ChQpp8eLF+uijjySJK5rAv6BtAcnvQSM/ypcvLzc3N1ksFs2cOVPp06dXw4YNJUn58+dXp06dlDt3btWuXdt6DBezgAezWCw6ffq0zpw5o71792rmzJn6448/FB8fr40bN2ro0KGaN2+erl27prx58zJqBC8s/orgiZQqVUpTpkxRjhw5VKlSJf3111/W3jgHBwcZhqGsWbMqX758unPnTkpXF0gzaFtA8kocuBcvXqzRo0erXbt2OnXqlIoVK6YRI0bIx8dHU6dO1bp166zHFSxYUAEBATZrJwB4sMyZM2vEiBGaOXOm6tatq1y5cunTTz/VkSNHVLlyZR08eFCZMmVSvnz5GDWCFxqhG48tYSjQm2++qTFjxihPnjyqXLmyNRxI96543r17Vy4uLvLw8EjJ6gJpBm0LSH4JgbtPnz4aMGCAzp49q0uXLqlEiRKaOXOmihYtqk8++USenp6aPn26li1bluQ5uIUR8O/8/f114sQJHT16VGPHjlWVKlUUHx+vkJAQm1vvSYwawYuLTz4eW+I5OCVKlNDo0aOVK1culS1bVqdPn9atW7d07do1tWzZUpLUuHHjlKwukGbQtgBzrFu3TsuXL1dQUJAWLVqkwYMH6+7du/Ly8pJ0b4RJnz59JEnfffddSlYVSNNy5sypfPny6datW9q/f7/q16+vq1evatiwYSldNSBVYPVyPLHEKyb/9NNPGjJkiL799lu98sorcnNzk729vXbs2CFHR0eb+wkDeDTaFpC85s6dqx07dmj58uVatmyZOnTooLFjx6pjx46KiIjQrVu35OPjo1OnTqlAgQL0wgH/gWEY2rNnjyZOnKiYmBh98803/L0C/h+hG0/l/lsVrV69WuHh4fLw8FDdunWtiz8xbwd4MrQt4Ok8aNG0AQMG6OjRoxo4cKDq1aunzz77TB07dpQkzZkzR6dOndKYMWPk4uLy0OcA8PiioqJ0+vRpvfbaayyaBiRC6IYpuKoJmIO2BSSVOCzv27dPWbNmVcGCBfXzzz/r/fff15kzZzRz5kx16NBBkhQZGammTZvKx8dHs2bN4i4AgAm4iAX8D5eeYOP+XranPZ5QANiibQHmMAzD+sW+f//+2rRpkwICApQzZ07lzJlT9evXV3x8vM6cOaOLFy/q3LlzGjdunC5duqS1a9da11QgeAPJi8AN/A893bBKfEXyypUrunnzpl555ZUH7k8s8ZeVu3fvKl26dM+mwkAaQdsCzDd8+HDNnDlTy5cvV5kyZaxDxsPDwxUYGKg5c+bo0qVLyp8/v7Jnz66vv/6a+aYAgGeC0A1Jtl/uhwwZoq1bt+r06dOqVKmSSpcurYCAgAfOyUl83JIlS3T58mV1795dTk5Oz7T+QGpF2wLM98cff+jdd9/VyJEjVa9ePV29elXBwcEKCgpS4cKF1bhxY8XHx+vQoUPy9fWVj48P800BAM8M4z4gSdYv96NGjdLs2bM1aNAgnT17VhEREZozZ45Onz6d5JjEoSAwMFBt2rRRoUKFCAVAIrQtIPnFx8fbPHZ3d5eTk5N+++037dy5UwEBAerQoYOCgoLUqlUrTZo0SXZ2dipdurRy5MghOzs7xcfHE7gBAM8EoRuS7n3Jv3btmrZu3arZs2erTp06+uWXX3To0CENGDBARYsWVUxMjLV8fHy8TSjo27evVq5cqVq1aqXUKQCpEm0LSF6Jp2McPHhQwcHBypgxowoVKqTVq1eratWq8vDw0NixY7Vv3z7VrFlTN27ckCSbedvMNwUAPCtc4oWke19EHB0ddfPmTZUvX14bNmxQixYtNHHiRPn7++vu3btavXq13njjDRUsWND6ZSUwMFB9+vTR/Pnz1bBhwxQ+CyD1oW0BySdx4B4wYIA2btyowYMHq3Hjxpo+fbrOnz8vwzBUrFgx6zFXr16Vs7NzCtUYAABC9wvrYSu1hoeHq1OnTtq+fbvGjRtnvb3KX3/9pcWLFytz5swqVKiQpHv3OO3du7e+/PJLNWrU6JnWH0itaFuAeRIC99ChQzV//nwtXrxYZcqUkSS5urrqtddek3SvvV2+fFk9evTQrVu31Ldv3xSrMwAAjK16ASUevvrXX3/p7t27ioqKUqZMmTRixAh9++23qlSpkjp27Ki4uDjdvn1bvXr1UlxcnGrWrClJioiI0LfffqvFixcTCoD/R9sCkt+aNWtsHp8/f15fffWVAgMDVbVqVd29e1cnTpzQ+PHjtXz5cknS+vXr1aNHD8XGxurIkSNycHBQXFxcSlQfAAB6ul9ECT0FgwcP1tq1ayVJzZs3V6tWrfT+++/r3LlzGjFihN599105Ojrqn3/+UWhoqI4dOyZ7e3vFxsbKzc1NX375pTJkyJCSpwKkKrQtIHktW7ZM48aNU8OGDa3ty97eXo6OjgoPD9f27du1fPly/fDDD4qKilJkZKRiYmLUpEkTZcuWTdWrV7e2LRZNAwCkFG4Z9oJas2aNevXqpYkTJ2rbtm06ffq0cubMqXHjxilHjhzaunWrFi5cqEyZMilXrlzq3bu3HBwc+OIC/AvaFpB8oqOjZW9vL3t7ex05ckQlSpRQdHS03n33XYWEhOjHH39Ut27dVKNGDZUqVUrvvfeeKlasqIEDB1qfI/E8cAAAUgKh+wVx/5eOJUuWKCQkRAEBAZKkuXPnatGiRXrppZc0atQo5c2bN0kIiIuLk729/TOvO5Ca0bYA8x04cEDlypXT+PHj1bt3b0VFRengwYPKmDGjihcvbi331ltvqV69eurTp89D11cAAOBZI3S/ABJ/8Zg7d64uXbqkX3/9VaVLl1aXLl2s5ebNm2cNB0OHDpWfn1+S4wH8D20LMMf9F7MMw9Do0aM1fPhwffbZZ+rVq5d1361bt3T16lV17txZly9f1tGjRxk1AgBIVfir9JxL/MWlX79+CgwM1CuvvKLff/9d3333nWrVqqWXX35ZkuTv7y87OztNmDBBy5Yt0/DhwyWJUAA8AG0LMEfitrV582bdvHlTxYoV08CBA+Xi4qKAgAA5ODioW7dukqRFixZp2bJlcnZ2tlk0jdEjAIDUgtD9HDMMw/rF5e+//9bdu3e1Y8cOvfHGG/r66681depUtWnTRgsXLlSePHkkSW3atFG2bNmsKykDSIq2BZgnoW31799f06ZNk4+Pj4KDgzV16lS1aNFCFotFPXr0kCR169ZN/v7+8vb2Vv369Vk0DQCQKrGyyHNo6dKlkv7Xi7Zs2TLlzZtX+/btk4eHhySpfv366tGjhxwdHdW6dWv98ccf1uPr1Kkje3t7bq8C3Ie2BZgnYbabYRgKDg7W/v37tX37dh06dEhjxoxRly5dtHDhQjVr1kyTJk1SQECARo4cKWdnZzVs2NDatgjcAIDUhtD9nFm9erUmTJig+Ph46xeYHDlyqGrVqjp79qyioqKsZevVq6fu3bsrXbp0ql27ti5fvmzzXAzNA/6HtgWYJ/E97sPCwhQTE6O33npLJUuWVJYsWRQQEKDJkyerf//+WrRokd5//30NHjxYW7duVeKlaWhbAIDUiIXUnjPR0dFycHCQnZ2dvvvuO5UrV06SdOjQIQUEBOjixYvat2+fcuTIYT1m1apV+v777zVx4kS+sAAPQdsCzDdw4EBt27ZNv/zyi3Lnzq1Vq1ZZFx6UpKlTpyogIED9+vVTz549lTlzZlksFhYlBACkaoTu50jixWdOnDihN954Q6NGjdKAAQMk3QsH/fv318WLF7Vz506bcJCAxWeApGhbgDkSt60VK1aoV69e6t+/v86fP68vvvhCnTp1UpcuXZQrVy7rMZ9++qk2b96sffv2EbgBAGkCofs5tHTpUuXOnVs//PCDAgICNGLECPXt21eSdPDgQQ0aNEgXL17Ut99+a/NFBsCj0bYAc+zZs0erVq1SqVKl1KpVK0nSzJkzNWbMGLVo0UIdO3a0aVMJQZvADQBIC5jT/RyIj4+3/nvcuHHq06ePMmXKpI8//ljjx4/XgAEDNHbsWElS6dKl9emnn8rZ2dkaFgA8GG0LMF9ISIj8/f21aNEihYWFWbd36tRJ/fr109KlS/XFF1/o/Pnz1n0EbgBAWsISn8+BhKF5586d08WLFzVjxgwVKlRIktS+fXtJUs+ePSVJffv2ValSpbRs2TIVKFAgZSoMpBG0LcB83t7eWrt2rRo3bqygoCBVqlRJRYoUkSR17txZdnZ26tq1q3x9fdWhQwfrcQRuAEBaQeh+TnzzzTeqX7++smXLplq1alm3Ozs7q3379rJYLOrdu7du3rypUaNGWYMD80yBR6NtAeYrWrSoVq1apTZt2mj69Onq3r27tS117NhRXl5eql+/fgrXEgCAp8Pw8udE3bp11bVrV127dk1Hjx7V7du3rfucnZ3Vrl07DRs2THv27OH2KsAToG0Bz8Zrr72mefPm6dixY5o6dapOnz5t3Zf4PtwAAKQ1LKSWBiVe7fV+bdu21ZIlS7Rw4UK9++67cnZ2tu6LiYmRg4MDc+GAh6BtASnv+PHjat++vXLlyqVx48YpT548KV0lAAD+E4aXpzGGYVhDwcKFC/Xzzz/L29tbhQoVUq1atTRnzhzFxsbK399fkmzCgaOjo/U5CAWALdoWkDq8/vrrmjFjhmbPns1dAAAAzwV6utOQxF/oBw0apClTpqhixYo6d+6cDMNQ9erVNXXqVEnSRx99pK+++kpTpkzRBx98YA0FAJKibQGpT0K7fNQIFAAA0gL+iqUhCaHgxIkTOnz4sLZs2aJvvvlGe/bsUadOnfTNN9+of//+kqT58+eratWqWrp0KaEA+Be0LSD1SZiuQeAGAKR1DC9PYz7//HMFBQUpOjpahQsXliR5enqqRYsWunHjhjZv3qzz58/r5Zdf1po1a2zuMwzg4WhbQOrDdA0AwPOAy8dpjKenpw4fPqyjR4/q+PHj1u1Zs2ZVzZo1dfToUV24cMG63c7OjnAAPAbaFgAAAMxA6E5jGjdurBUrVsjd3V1z5syxCQfZs2fXyy+/rJiYGJtjGJoH/DvaFgAAAMzAQmppSOLFnr755ht16tRJhQsXVsOGDZUzZ05Nnz5dwcHB+vHHH7lHMPAEaFsAAAAwC6E7jUkcDoKCgvTxxx/rypUratKkiTw9PTV58mTZ29srLi6OcAAk8m+386JtAQAAwAyMjUxjElZzlaTatWtr8eLF8vHxkY+Pj9q3by97e3sZhkEoAO4TFRX1yP20LQAAAJiB0J2KnDlzRrt379YPP/ygW7duPbRc4nBQpUoVzZgxQ2vWrNH06dP1448/storcJ/PP/9clSpVUnR09CPL0bYAAACQ3LhlWCqxcOFCjRw5Uo6Ojvrjjz80dOhQDRgw4KHlE8JBfHy83n33XUlS8+bN5eTkpAkTJsjJyekZ1RxI3QIDA9WzZ08tXbrUpl08bLg5bQsAAADJiTndqcCSJUvUqVMnzZo1S7Vq1dKcOXM0fPhwXbx4UZkzZ7aWuz8kGIYhwzBkZ2enmJgYnTp1ShkyZFC+fPlS4jSAVGfOnDnq0qWLlixZosaNG+v27duyt7dXVFSU3N3dH3ocbQsAAADJheHlKeznn3/W5MmTNWnSJLVo0UKZM2dWs2bNVKVKFe3Zs0c7duzQ2bNnJSlJ4LZYLLKzs9PUqVOVPn16+fj4EAqA/3fkyBG1b99enTt3VuPGjXXmzBk1b95cJUuW1CuvvKJPPvlER48eTXIcbQsAAADJieHlKczLy0v+/v6qV6+edVvnzp21b98+Xb58WeHh4fL19dWYMWNUokQJSbY93oGBgRo5cqSWLFkiT0/PFDkHIDXy8fFRkyZN9N1332natGmaNWuWypcvrzp16igiIkLLly/XhQsXNHbsWOXKlUsSbQsAAADJj+HlqUBkZKRcXFwkSSNHjtSsWbO0adMmFStWTAcPHtTHH3+s9u3bq2vXrja3KwoMDFSfPn00f/58NWrUKCVPAUiVLl++rD59+mjt2rXy9/e33vZLktavX6/WrVtrzZo1qlKliuLj42Vnd2/wD20LAAAAyYXQncpcvHhRTk5ONj1rZcuWVZkyZTRx4kTrtsDAQPXr109z584lFACPcOnSJS1evFh16tRRoUKFbMK1j4+P+vXrp27dulnL07YAAACQnBhensrkyJHD5vGVK1fk7OysokWLWrdt3rxZHTt21OrVqwkFwL/w8fFRly5dlCFDBkmyBu7z58/L29tbfn5+1rK0LQAAACQ3erpTKcMwdOfOHb3//vsKCwvTnj17rMNiz5w5o+vXr6tcuXIpXEsg7Unctm7fvq1t27bRtgAAAGAaQncqFBcXpxkzZmjLli26cuWKDh06JEdHR5v53ACeXFxcnKZMmaLNmzfrn3/+0ZEjR2hbAAAAMBW3DEuF7O3tVaVKFRUqVEiHDx+Wo6OjYmNjCQXAf2Rvb6+qVavKz89PR48epW0BAADAdPR0pwH0wgHmoG0BAADAbIRuAAAAAABMwvByAAAAAABMQugGAAAAAMAkhG4AAAAAAExC6AYAAAAAwCSEbgAAAAAATELoBgAAAADAJIRuAAAAAABMQugGAAAAAMAkhG4AAAAAAExC6AYAAAAAwCT/B/UYKqUvWjcoAAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 1000x500 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "## Visualize distribution of functional variants\n",
    "import os, glob, pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "OUTDIR = \"per_chrom_outputs_new\"\n",
    "\n",
    "# Load all annotated tables\n",
    "paths = sorted(glob.glob(os.path.join(OUTDIR, \"chr*.annotated.csv\")))\n",
    "dfs = [pd.read_csv(p, dtype=str) for p in paths]\n",
    "annot = pd.concat(dfs, ignore_index=True) if dfs else pd.DataFrame()\n",
    "\n",
    "# Guard if nothing loaded\n",
    "if annot.empty or \"CSQ_terms_str\" not in annot.columns:\n",
    "    print(\"No annotated data or CSQ_terms_str column not found.\")\n",
    "else:\n",
    "    # explode comma-separated CSQ_terms into rows\n",
    "    csq = (\n",
    "        annot.assign(CSQ_terms_str=annot[\"CSQ_terms_str\"].fillna(\"\"))\n",
    "             .assign(term=lambda d: d[\"CSQ_terms_str\"].str.split(\",\"))\n",
    "             .explode(\"term\")\n",
    "    )\n",
    "    csq[\"term\"] = csq[\"term\"].fillna(\"\").str.strip()\n",
    "    csq = csq[csq[\"term\"] != \"\"]\n",
    "\n",
    "    # counts overall\n",
    "    csq_counts = csq[\"term\"].value_counts().rename_axis(\"term\").reset_index(name=\"count\")\n",
    "    display(csq_counts.head(20))\n",
    "\n",
    "    # optional: counts by chromosome\n",
    "    if \"CHROM\" in csq.columns:\n",
    "        csq_by_chr = (\n",
    "            csq.groupby([\"CHROM\",\"term\"]).size().reset_index(name=\"count\")\n",
    "               .sort_values([\"CHROM\",\"count\"], ascending=[True, False])\n",
    "        )\n",
    "        display(csq_by_chr.head(50))\n",
    "\n",
    "    # quick bar chart of top 15 terms\n",
    "    topN = 15\n",
    "    top = csq_counts.head(topN)\n",
    "    plt.figure(figsize=(10,5))\n",
    "    plt.bar(top[\"term\"], top[\"count\"])\n",
    "    plt.xticks(rotation=45, ha=\"right\")\n",
    "    plt.title(\"Top functional consequences (CSQ_terms)\")\n",
    "    plt.ylabel(\"Count\")\n",
    "    plt.tight_layout()\n",
    "    plt.show()\n",
    "\n",
    "    # save summaries\n",
    "    csq_counts.to_csv(os.path.join(OUTDIR, \"functional_counts_overall.csv\"), index=False)\n",
    "    if \"CHROM\" in csq.columns:\n",
    "        csq_by_chr.to_csv(os.path.join(OUTDIR, \"functional_counts_by_chrom.csv\"), index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b1dc3936-d270-4ba4-a0ca-ec7dfcc3b7f7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CLNSIG_label</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Unknown code: nan</td>\n",
       "      <td>1612694</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Benign</td>\n",
       "      <td>24053</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Likely benign</td>\n",
       "      <td>801</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>253</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>other</td>\n",
       "      <td>157</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Benign/Likely benign</td>\n",
       "      <td>110</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>not provided</td>\n",
       "      <td>71</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Risk factor</td>\n",
       "      <td>56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Drug response</td>\n",
       "      <td>49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Association</td>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Conflicting interpretations of pathogenicity</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Protective</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Likely risk allele</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Likely pathogenic</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Affects</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Uncertain risk allele</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Confers sensitivity</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    CLNSIG_label    count\n",
       "0                              Unknown code: nan  1612694\n",
       "1                                         Benign    24053\n",
       "2                                  Likely benign      801\n",
       "3                         Uncertain significance      253\n",
       "4                                          other      157\n",
       "5                           Benign/Likely benign      110\n",
       "6                                   not provided       71\n",
       "7                                    Risk factor       56\n",
       "8                                  Drug response       49\n",
       "9                                    Association       31\n",
       "10                                    Pathogenic       28\n",
       "11  Conflicting interpretations of pathogenicity       22\n",
       "12                                    Protective       15\n",
       "13                            Likely risk allele       12\n",
       "14                             Likely pathogenic       10\n",
       "15                                       Affects        8\n",
       "16                         Uncertain risk allele        2\n",
       "17                           Confers sensitivity        1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CLNSIG_bucket</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Other</td>\n",
       "      <td>1613094</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Benign (any)</td>\n",
       "      <td>24964</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Uncertain</td>\n",
       "      <td>255</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Pathogenic (any)</td>\n",
       "      <td>38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Conflicting</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      CLNSIG_bucket    count\n",
       "0             Other  1613094\n",
       "1      Benign (any)    24964\n",
       "2         Uncertain      255\n",
       "3  Pathogenic (any)       38\n",
       "4       Conflicting       22"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA+UAAAHpCAYAAADknp3YAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy81sbWrAAAACXBIWXMAAA9hAAAPYQGoP6dpAAD28ElEQVR4nOzdd1QU19sH8GdtgEgRFURFUbBhRbEAgiKCgL3Ekoi9JXZjx64/scXeYkVjr7H33ntNjC2iBsEuiEr/vn9wdt5dFo3Ydtd8P+dwdGdnZ5+7U5+5d+5VAYAQERERERER0VeXSd8BEBEREREREf1XMSknIiIiIiIi0hMm5URERERERER6wqSciIiIiIiISE+YlBMRERERERHpCZNyIiIiIiIiIj1hUk5ERERERESkJ0zKiYiIiIiIiPSESTkRERERERGRnjApJ6Jvkkql+qC/Q4cO6TtUERFxdHSUtm3bKq/Dw8NFpVJJWFjYZ1ne51ajRg2pUaPGZ1mWSqWSkSNHfpZlvcuhQ4fSXd8zZ84UZ2dnyZYtm6hUKnn58qW0bdtWHB0dv2g8hiIsLExUKpWEh4dn+LMjR44UlUolT58+/WzxqJf5uSQmJkqJEiVk/Pjxn22Zxi4jv/HQoUOlYMGCkiVLFrG2tv6ygRmBz719akpMTBQnJyeZNm3aF1k+ERm2LPoOgIjoSzh58qTW6zFjxsjBgwflwIEDWtNdXFy+ZlgfzN7eXk6ePClOTk4f9flNmzaJpaXlZ47qyzh58qQUKFDgi35HhQoV5OTJk1rr+9KlS9KzZ0/p2LGjtGnTRrJkySIWFhYybNgw6dWr1xeNh76OOXPmyIsXL6RHjx76DsXobN68Wf73v/9JSEiIBAYGiomJib5D+qZlzZpVhg8fLn369JHg4GDJlSuXvkMioq+ISTkRfZOqVq2q9TpPnjySKVMmnemGysTE5JNidXV1/YzRfFlfY51YWlrqfM8ff/whIiKdOnWSypUrK9M/9kYIGZakpCSZNGmStG/fXszNzfUdjtG5du2aiIj07NlTbG1t3zvv27dvxczM7GuE9U1r2bKl9O3bV3799VcZMmSIvsMhoq+IzdeJ6D/r+fPn8tNPP0n+/PklW7ZsUqRIEQkJCZH4+Hit+VQqlXTv3l1+/fVXKVasmJiYmIiLi4usXr36g74nPj5eRo8eLSVLlhRTU1PJlSuX+Pj4yIkTJ975mfSar6ubTv7xxx/SsmVLsbKyEjs7O2nfvr1ER0drfT695usvX76Un3/+WYoUKSImJiZia2srQUFB8tdffynzjBo1SqpUqSI2NjZiaWkpFSpUkEWLFgmADyprWgcOHJAaNWpIrly5xMzMTAoWLChNmjSRN2/eKPOk13z92LFj4u7uLqamppI/f34ZNmyYLFy4UKeptaOjo9StW1d27dolFSpUEDMzMylRooQsXrxYa3lpm6/XqFFDWrVqJSIiVapUEZVKpfxe6TVfT0lJkZkzZ0r58uXFzMxMrK2tpWrVqrJlyxZlnjVr1oi/v7/Y29uLmZmZlCxZUgYNGiSvX7/WWlbbtm0lR44ccvv2bQkKCpIcOXKIg4OD/Pzzzzrb3odsOwBkzpw5Smw5c+aUpk2byt9///2v6yc9e/fulQYNGkiBAgXE1NRUnJ2dpUuXLu9spv7gwQNp3LixWFpaipWVlbRq1UqePHmiM9+aNWvE3d1dzM3NJUeOHFK7dm25ePHiv8bzIdtQerZs2SIRERESHBysNV29H125ckW+++47sbKyEhsbG+nbt68kJSXJjRs3JCAgQCwsLMTR0VEmTpyo9fm4uDj5+eefpXz58spn3d3dZfPmzToxfOixQ/0Ywd69e6Vdu3ZiY2Mj5ubmUq9evXTX4759+8TX11csLS0le/bs4unpKfv379eZb/v27VK+fHkxMTGRwoULy+TJk9/7m6k5OjrK0KFDRUTEzs5Oax9V73MbN24UV1dXMTU1lVGjRolIaiLfoEEDyZkzp5iamkr58uVl6dKlWstW74srV66UgQMHir29veTIkUPq1asnjx49klevXknnzp0ld+7ckjt3bmnXrp3ExsZ+UNy7du0SX19fsbKykuzZs0vJkiUlNDRUa54tW7aIu7u7ZM+eXSwsLMTPz0+nZVVGfrsP3f8uXrwodevWFVtbWzExMZF8+fJJnTp15J9//lHmyZYtmzRv3lzmz5//0cdcIjJOrCknov+kuLg48fHxkTt37sioUaOkbNmycvToUQkNDZVLly7J9u3btebfsmWLHDx4UEaPHi3m5uYyZ84cadmypWTJkkWaNm36zu9JSkqSwMBAOXr0qPTu3Vtq1qwpSUlJcurUKbl//754eHhkOPYmTZpI8+bNpUOHDnL16lUZPHiwiIhOIqrp1atXUq1aNQkPD5eBAwdKlSpVJDY2Vo4cOSKRkZFSokQJEUm9GdClSxcpWLCgiIicOnVKevToIRERETJ8+PAMxRkeHi516tQRLy8vWbx4sVhbW0tERITs2rVLEhISJHv27Ol+7sqVK+Ln5yfFihWTpUuXSvbs2WXevHmyfPnydOe/fPmy/PzzzzJo0CCxs7OThQsXSocOHcTZ2Vm8vb3T/cycOXNk1apVMnbsWFmyZImUKFFC8uTJ886ytG3bVpYvXy4dOnSQ0aNHS7Zs2eTChQtaNwhu3bolQUFB0rt3bzE3N5e//vpLJkyYIGfOnNF5bCIxMVHq168vHTp0kJ9//lmOHDkiY8aMESsrK+V3/tBtp0uXLhIWFiY9e/aUCRMmyPPnz2X06NHi4eEhly9fFjs7u3eWKz137twRd3d36dixo1hZWUl4eLhMmTJFqlWrJlevXpWsWbNqzd+oUSNp1qyZdO3aVf744w8ZNmyY/Pnnn3L69Gll3nHjxsnQoUOlXbt2MnToUElISJBJkyaJl5eXnDlz5p2PkXzsNiSSmlTZ2tq+c9nNmjWTVq1aSZcuXWTv3r0yceJESUxMlH379slPP/0k/fr1UxJHZ2dnady4sYik3ih5/vy59OvXT/Lnzy8JCQmyb98+ady4sSxZskRat26t9T0ZOXZ06NBB/Pz8ZOXKlfLgwQMZOnSo1KhRQ65cuaI80718+XJp3bq1NGjQQJYuXSpZs2aVX3/9VWrXri27d+8WX19fERHZv3+/NGjQQNzd3WX16tWSnJwsEydOlEePHr3zN1PbtGmTzJ49WxYtWiS7du0SKysrrUdMLly4INevX5ehQ4dK4cKFxdzcXG7cuCEeHh5ia2srM2bMkFy5csny5culbdu28ujRIxkwYIDWdwwZMkR8fHwkLCxMwsPDpV+/fsrvUq5cOVm1apVcvHhRhgwZIhYWFjJjxoz3xrxo0SLp1KmTVK9eXebNmye2trZy8+ZNpcZfRGTlypXyww8/iL+/v6xatUri4+Nl4sSJUqNGDdm/f79Uq1Ytw7/dh+x/r1+/Fj8/PylcuLDMnj1b7OzsJCoqSg4ePCivXr3SWl6NGjVk7ty5cu3aNSlTpsy/risi+kaAiOg/oE2bNjA3N1dez5s3DyKCtWvXas03YcIEiAj27NmjTBMRmJmZISoqSpmWlJSEEiVKwNnZ+b3fu2zZMogIFixY8N75ChUqhDZt2iiv7969CxHBkiVLlGkjRoyAiGDixIlan/3pp59gamqKlJSUdy5v9OjREBHs3bv3vXFoSk5ORmJiIkaPHo1cuXJpLb969eqoXr36ez+/fv16iAguXbr03vlEBCNGjFBef/fddzA3N8eTJ0+0YnFxcYGI4O7du8r0QoUKwdTUFPfu3VOmvX37FjY2NujSpYsy7eDBgxARHDx4UJm2ZMkSiAjOnj2rFU+bNm1QqFAh5fWRI0cgIggJCXlvOTSlpKQgMTERhw8fhojg8uXLWstPb9sLCgpC8eLFldcfsu2cPHkSIoJffvlFa/qDBw9gZmaGAQMGvDdO9W+g+ZumV4579+5BRLB582blPfX22KdPH63PrFixAiKC5cuXAwDu37+PLFmyoEePHlrzvXr1Cnnz5kWzZs10lqn2odtQekqWLImAgACd6ervSPublS9fHiKCjRs3KtMSExORJ08eNG7c+J3fk5SUhMTERHTo0AGurq5a733osUO9Hho1aqT1+ePHj0NEMHbsWADA69evYWNjg3r16mnNl5ycjHLlyqFy5crKtCpVqiBfvnx4+/atMi0mJgY2Njb4kMs/9e+kuR8Cqftc5syZcePGDa3pLVq0gImJCe7fv681PTAwENmzZ8fLly8B/P++mLYMvXv3hoigZ8+eWtMbNmwIGxub98b66tUrWFpaolq1alrHKU3JycnIly8fypQpg+TkZK3P2trawsPDQ5n2ob/dh+5/586dg4jg999/f285AODWrVsQEcydO/df5yWibwebr/+LI0eOSL169SRfvnyiUqnk999/z/AyAMjkyZOVpmsODg4ybty4zx8sEX2wAwcOiLm5uU5NlboJc9qmoL6+vlo1jpkzZ5bmzZvL7du3tZofprVz504xNTWV9u3bf7bY69evr/W6bNmyEhcXJ48fP35vHMWKFZNatWq9d9kHDhyQWrVqiZWVlWTOnFnpfOjZs2fvXX56ypcvL9myZZPOnTvL0qVLP7g59eHDh6VmzZqSO3duZVqmTJmkWbNm7/wedc2+iIipqakUK1ZM7t27l6F432Xnzp0iItKtW7f3zvf333/L999/L3nz5lV+u+rVq4uIyPXr17XmValUUq9ePa1pZcuW1Yr5Q7adbdu2iUqlklatWklSUpLylzdvXilXrtxHjS7w+PFj6dq1qzg4OEiWLFkka9asUqhQoXTLISLyww8/aL1u1qyZZMmSRQ4ePCgiIrt375akpCRp3bq1VoympqZSvXr198b4sduQiMjDhw/f+yx03bp1tV6XLFlSVCqVBAYGKtOyZMkizs7OOtvSunXrxNPTU3LkyKH8RosWLUr398nIsSPtb+nh4SGFChVSfssTJ07I8+fPpU2bNlq/ZUpKigQEBMjZs2fl9evX8vr1azl79qw0btxYTE1NleVZWFjobHcfo2zZslKsWDGtaQcOHBBfX19xcHDQmt62bVt58+aNThPx9H5/EZE6deroTH/+/Pl7m7CfOHFCYmJi5Keffnpn7+g3btyQhw8fSnBwsGTK9P+Xvzly5JAmTZrIqVOn5M2bNxn67T50/3N2dpacOXPKwIEDZd68efLnn3++syzqbTYiIuKd8xDRt4dJ+b94/fq1lCtXTmbNmvXRy+jVq5csXLhQJk+eLH/99Zds3bpVq1MhIvr6nj17Jnnz5tW5gLO1tZUsWbLIs2fPtKbnzZtXZxnqaWnn1fTkyRPJly+f1kXgp0rbK6+6V+S3b9++N45/6+H8zJkz4u/vLyIiCxYskOPHj8vZs2clJCTkX5efHicnJ9m3b5/Y2tpKt27dxMnJSZycnGT69Onv/dyzZ8/SbXL9rmbY6fVSbGJikuF43+XJkyeSOXPmdLcBtdjYWPHy8pLTp0/L2LFj5dChQ3L27FnZuHGjiOj+dtmzZ9e64FfHHBcXp/W9/7btPHr0SACInZ2dZM2aVevv1KlTGR6uLCUlRfz9/WXjxo0yYMAA2b9/v5w5c0ZOnTqVbjlEdPeNLFmySK5cuZT9Qt3kt1KlSjoxrlmz5r0xfuw2pI417W+sycbGRut1tmzZ0l0v2bJl01ovGzdulGbNmkn+/Pll+fLlcvLkSTl79qy0b99eaz61jBw73jVv2t+yadOmOr/lhAkTBIA8f/5cXrx4ISkpKe/97k9hb2+vM+3Zs2fpTs+XL5/yvqb0fv/3TU/vt1VT92HwvmOc+vvfFWNKSoq8ePEiQ7/dh+5/VlZWcvjwYSlfvrwMGTJESpUqJfny5ZMRI0ZIYmKi1jLV29/nOn4RkXHgM+X/IjAwUOuueVoJCQkydOhQWbFihbx8+VJKly4tEyZMUMbvvX79uvJsUPHixb9S1ET0b3LlyiWnT58WAFqJ+ePHjyUpKUmrllZEJCoqSmcZ6mnvG7omT548cuzYMUlJSfmsiXlG5cmT5701+iIiq1evlqxZs8q2bdu0EpOPaSGk5uXlJV5eXpKcnCznzp2TmTNnSu/evcXOzk5atGiR7mdy5cqV7rOb6a2DryFPnjySnJwsUVFR6V7Qi6TWEj58+FAOHTqk1I6LpHau9ynf+2/bTu7cuUWlUsnRo0fTHbIqo8NYXbt2TS5fvixhYWHSpk0bZfrt27ff+ZmoqCjJnz+/8jopKUmePXum7BfqfWn9+vVKjXtGfMw2pP7e58+fZ/j7/s3y5culcOHCsmbNGq1jR9pO+tQycux417zOzs4i8v+/5cyZM985aoGdnZ0kJiaKSqV673d/ivRqo3PlyiWRkZE60x8+fCgionNM/ZzU/UG87xin/q3fFWOmTJkkZ86cyjnhQ367jOx/ZcqUkdWrVwsAuXLlioSFhcno0aPFzMxMBg0apMyn3ma/5O9FRIaHNeWfqF27dnL8+HFZvXq10pNrQECA3Lp1S0REtm7dKkWKFJFt27ZJ4cKFxdHRUTp27PhFLhSI6MP5+vpKbGysTsK5bNky5X1N+/fv10oUk5OTZc2aNeLk5PTe2pnAwECJi4vT6kVdHwIDA+XmzZs6HY5pUqlUkiVLFsmcObMy7e3bt/Lbb7998vdnzpxZqlSpIrNnzxaR1I6i3qV69epy4MABrRrUlJQUWbdu3SfH8THUN2bnzp37znnUSUraC/Nff/31k77337adunXrCgCJiIgQNzc3nb+MdhT1MeVYsWKF1uu1a9dKUlKScnO6du3akiVLFrlz5066Mbq5uX1QbBnZhkRESpQoIXfu3PmgZWeESqWSbNmyaSWmUVFR6fa+LpKxY0fa3/LEiRNy79495bf09PQUa2tr+fPPP9/5W2bLlk3Mzc2lcuXKsnHjRq0a5levXsnWrVs/9SdIl6+vr3JzStOyZcske/bsX3ToQw8PD7GyspJ58+a9s9fy4sWLS/78+WXlypVa87x+/Vo2bNig9Miekd/uY/Y/lUol5cqVk6lTp4q1tbXOdqx+RONdHRQS0beJNeWf4M6dO7Jq1Sr5559/lOZZ/fr1k127dsmSJUtk3Lhx8vfff8u9e/dk3bp1smzZMklOTpY+ffpI06ZN33txTERfVuvWrWX27NnSpk0bCQ8PlzJlysixY8dk3LhxEhQUpPPsde7cuaVmzZoybNgwpQflv/7661+HRWvZsqUsWbJEunbtKjdu3BAfHx9JSUmR06dPS8mSJd9b0/c59e7dW9asWSMNGjSQQYMGSeXKleXt27dy+PBhqVu3rvj4+EidOnVkypQp8v3330vnzp3l2bNnMnny5AzXtqrNmzdPDhw4IHXq1JGCBQtKXFyc0kP8+55tDwkJka1bt4qvr6+EhISImZmZzJs3Txla7Gu3OPDy8pLg4GAZO3asPHr0SOrWrSsmJiZy8eJFyZ49u/To0UM8PDwkZ86c0rVrVxkxYoRkzZpVVqxYIZcvX/7o7/2QbcfT01M6d+4s7dq1k3Pnzom3t7eYm5tLZGSkHDt2TMqUKSM//vjjB39niRIlxMnJSQYNGiQAxMbGRrZu3Sp79+5952c2btwoWbJkET8/P6X39XLlyil9ADg6Osro0aMlJCRE/v77bwkICJCcOXPKo0eP5MyZM2Jubq4MqZXWx25DIqm9WI8ePVrevHnz3l7aM0o9HNhPP/0kTZs2lQcPHsiYMWPE3t5euSGvKSPHjnPnzknHjh3lu+++kwcPHkhISIjkz59ffvrpJxFJff555syZ0qZNG3n+/Lk0bdpUbG1t5cmTJ3L58mV58uSJcvNozJgxEhAQIH5+fvLzzz9LcnKyTJgwQczNzb9IxcCIESNk27Zt4uPjI8OHDxcbGxtZsWKFbN++XSZOnChWVlaf/TvVcuTIIb/88ot07NhRatWqJZ06dRI7Ozu5ffu2XL58WWbNmiWZMmWSiRMnyg8//CB169aVLl26SHx8vEyaNElevnwp48ePV5b3ob/dh+5/27Ztkzlz5kjDhg2lSJEiAkA2btwoL1++FD8/P62ynDp1SjJnzvzOkSOI6Bull+7ljJSIYNOmTcrrtWvXQkRgbm6u9ZclSxalN9lOnTpBRLR6KT1//jxEBH/99dfXLgLRf1ba3tcB4NmzZ+jatSvs7e2RJUsWFCpUCIMHD0ZcXJzWfCKCbt26Yc6cOXByckLWrFlRokQJrFix4oO+++3btxg+fDiKFi2KbNmyIVeuXKhZsyZOnDihzJOR3tfT9oacXg/aaZcHAC9evECvXr1QsGBBZM2aFba2tqhTp47WsWjx4sUoXrw4TExMUKRIEYSGhmLRokU6y/+Q3tdPnjyJRo0aoVChQjAxMUGuXLlQvXp1bNmyRWs+SdP7OgAcPXoUVapUgYmJCfLmzYv+/fsrPeOre3FWl7NOnTo63502vk/pfR1I7bl56tSpKF26NLJlywYrKyu4u7tj69atyjwnTpyAu7s7smfPjjx58qBjx464cOGCznpMb1sEdHseBz5s2wFS11uVKlVgbm4OMzMzODk5oXXr1jh37pzO92hKb9v5888/4efnBwsLC+TMmRPfffcd7t+/r7Oe1PGeP38e9erVQ44cOWBhYYGWLVvi0aNHOt/1+++/w8fHB5aWljAxMUGhQoXQtGlT7Nu3752/wYduQ+m5ffs2VCqVTi/379qP3rVeqlevjlKlSmlNGz9+PBwdHWFiYoKSJUtiwYIF6a6/Dz12qNfDnj17EBwcDGtra5iZmSEoKAi3bt3Sienw4cOoU6cObGxskDVrVuTPnx916tTBunXrtObbsmULypYti2zZsqFgwYIYP358unGm5329r6e3zwHA1atXUa9ePVhZWSFbtmwoV66c1rYP/P++mDbWd+2P74ojPTt27ED16tVhbm6O7Nmzw8XFBRMmTNCa5/fff0eVKlVgamoKc3Nz+Pr64vjx4zrLyshv92/7319//YWWLVvCyckJZmZmsLKyQuXKlREWFqazLC8vL52e6Yno26cC3tHOh3SoVCrZtGmTNGzYUERE1qxZIz/88IP88ccfWs09RVLv2ubNm1dGjBgh48aN0+rI4+3bt5I9e3bZs2ePzh1SIjI8KpVKunXr9kkdPtKn8/f3l/DwcLl586a+QyEjUa9ePUlKSlJ60P/aPvTYERYWJu3atZOzZ89+cHN++vbcuXNHihYtKrt37+b1IdF/DJuvfwJXV1dJTk6Wx48fi5eXV7rzeHp6SlJSkty5c0ecnJxERJQLyo/p8IaI6L+gb9++4urqKg4ODvL8+XNZsWKF7N27VxYtWqTv0MiIhIaGiqurq5w9e1YqVaqk73CI3mvs2LHi6+vLhJzoP4hJ+b+IjY3V6nX27t27cunSJbGxsZFixYrJDz/8IK1bt5ZffvlFXF1d5enTp3LgwAEpU6aM8lxqhQoVpH379jJt2jRJSUmRbt26iZ+fn84Yn0RElCo5OVmGDx8uUVFRolKpxMXFRX777Tdp1aqVvkMjI1K6dGlZsmSJ3nruJ/pQSUlJ4uTkJIMHD9Z3KESkB2y+/i8OHTokPj4+OtPbtGkjYWFhkpiYKGPHjpVly5ZJRESE5MqVS9zd3WXUqFFKr5sPHz6UHj16yJ49e8Tc3FwCAwPll19+0RmLk4iIiIiIiP5bmJQTERERERER6QnHKSciIiIiIiLSEz5Tno6UlBR5+PChWFhYiEql0nc4REREREREZEQAyKtXryRfvnySKdP768KZlKfj4cOH4uDgoO8wiIiIiIiIyIg9ePBAChQo8N55mJSnw8LCQkRSf0BLS0s9R0NERERERETGJCYmRhwcHJTc8n2YlKdD3WTd0tKSSTkRERERERF9lA95HJodvRERERERERHpCZNyIiIiIiIiIj1hUk5ERERERESkJ0zKiYiIiIiIiPSESTkRERERERGRnjApJyIiIiIiItITJuVEREREREREesKknIiIiIiIiEhPsug7APo0joO26zuEDAsfX0ffIRARERERERkE1pQTERERERER6QmTciIiIiIiIiI9YVJOREREREREpCdMyomIiIiIiIj0hEk5ERERERERkZ4wKSciIiIiIiLSEyblRERERERERHrCpJyIiIiIiIhIT5iUExEREREREekJk3IiIiIiIiIiPdFrUn7kyBGpV6+e5MuXT1Qqlfz+++/vnf/QoUOiUql0/v766y+t+TZs2CAuLi5iYmIiLi4usmnTpi9YCiIiIiIiIqKPo9ek/PXr11KuXDmZNWtWhj5348YNiYyMVP6KFi2qvHfy5Elp3ry5BAcHy+XLlyU4OFiaNWsmp0+f/tzhExEREREREX2SLPr88sDAQAkMDMzw52xtbcXa2jrd96ZNmyZ+fn4yePBgEREZPHiwHD58WKZNmyarVq1K9zPx8fESHx+vvI6JiclwTEREREREREQZZZTPlLu6uoq9vb34+vrKwYMHtd47efKk+Pv7a02rXbu2nDhx4p3LCw0NFSsrK+XPwcHhi8RNREREREREpMmoknJ7e3uZP3++bNiwQTZu3CjFixcXX19fOXLkiDJPVFSU2NnZaX3Ozs5OoqKi3rncwYMHS3R0tPL34MGDL1YGIiIiIiIiIjW9Nl/PqOLFi0vx4sWV1+7u7vLgwQOZPHmyeHt7K9NVKpXW5wDoTNNkYmIiJiYmnz9gIiIiIiIiovcwqpry9FStWlVu3bqlvM6bN69Orfjjx491as+JiIiIiIiI9M3ok/KLFy+Kvb298trd3V327t2rNc+ePXvEw8Pja4dGRERERERE9F56bb4eGxsrt2/fVl7fvXtXLl26JDY2NlKwYEEZPHiwREREyLJly0QktWd1R0dHKVWqlCQkJMjy5ctlw4YNsmHDBmUZvXr1Em9vb5kwYYI0aNBANm/eLPv27ZNjx4599fIRERERERERvY9ek/Jz586Jj4+P8rpv374iItKmTRsJCwuTyMhIuX//vvJ+QkKC9OvXTyIiIsTMzExKlSol27dvl6CgIGUeDw8PWb16tQwdOlSGDRsmTk5OsmbNGqlSpcrXKxgRERERERHRB1ABgL6DMDQxMTFiZWUl0dHRYmlpqe9w3stx0HZ9h5Bh4ePr6DsEIiIiIiKiLyYjOaXRP1NOREREREREZKyYlBMRERERERHpCZNyIiIiIiIiIj1hUk5ERERERESkJ0zKiYiIiIiIiPSESTkRERERERGRnjApJyIiIiIiItITJuVEREREREREesKknIiIiIiIiEhPmJQTERERERER6QmTciIiIiIiIiI9YVJOREREREREpCdMyomIiIiIiIj0hEk5ERERERERkZ4wKSciIiIiIiLSEyblRERERERERHrCpJyIiIiIiIhIT5iUExEREREREekJk3IiIiIiIiIiPWFSTkRERERERKQnTMqJiIiIiIiI9ESvSfmRI0ekXr16ki9fPlGpVPL777+/d/6NGzeKn5+f5MmTRywtLcXd3V12796tNU9YWJioVCqdv7i4uC9YEiIiIiIiIqKM02tS/vr1aylXrpzMmjXrg+Y/cuSI+Pn5yY4dO+T8+fPi4+Mj9erVk4sXL2rNZ2lpKZGRkVp/pqamX6IIRERERERERB8tiz6/PDAwUAIDAz94/mnTpmm9HjdunGzevFm2bt0qrq6uynSVSiV58+b94OXGx8dLfHy88jomJuaDP0tERERERET0sYz6mfKUlBR59eqV2NjYaE2PjY2VQoUKSYECBaRu3bo6NelphYaGipWVlfLn4ODwJcMmIiIiIiIiEhEjT8p/+eUXef36tTRr1kyZVqJECQkLC5MtW7bIqlWrxNTUVDw9PeXWrVvvXM7gwYMlOjpa+Xvw4MHXCJ+IiIiIiIj+4/TafP1TrFq1SkaOHCmbN28WW1tbZXrVqlWlatWqymtPT0+pUKGCzJw5U2bMmJHuskxMTMTExOSLx0xERERERESkySiT8jVr1kiHDh1k3bp1UqtWrffOmylTJqlUqdJ7a8qJiIiIiIiI9MHomq+vWrVK2rZtKytXrpQ6der86/wA5NKlS2Jvb/8VoiMiIiIiIiL6cHqtKY+NjZXbt28rr+/evSuXLl0SGxsbKViwoAwePFgiIiJk2bJlIpKakLdu3VqmT58uVatWlaioKBERMTMzEysrKxERGTVqlFStWlWKFi0qMTExMmPGDLl06ZLMnj376xeQiIiIiIiI6D30WlN+7tw5cXV1VYYz69u3r7i6usrw4cNFRCQyMlLu37+vzP/rr79KUlKSdOvWTezt7ZW/Xr16KfO8fPlSOnfuLCVLlhR/f3+JiIiQI0eOSOXKlb9u4YiIiIiIiIj+hQoA9B2EoYmJiRErKyuJjo4WS0tLfYfzXo6Dtus7hAwLH//vjx0QEREREREZq4zklEb3TDkRERERERHRt4JJOREREREREZGeMCknIiIiIiIi0hMm5URERERERER6wqSciIiIiIiISE+YlBMRERERERHpCZNyIiIiIiIiIj1hUk5ERERERESkJ0zKiYiIiIiIiPSESTkRERERERGRnjApJyIiIiIiItITJuVEREREREREesKknIiIiIiIiEhPmJQTERERERER6QmTciIiIiIiIiI9YVJOREREREREpCdMyomIiIiIiIj0hEk5ERERERERkZ4wKSciIiIiIiLSEyblRERERERERHrCpJyIiIiIiIhIT/SalB85ckTq1asn+fLlE5VKJb///vu/fubw4cNSsWJFMTU1lSJFisi8efN05tmwYYO4uLiIiYmJuLi4yKZNm75A9ERERERERESfRq9J+evXr6VcuXIya9asD5r/7t27EhQUJF5eXnLx4kUZMmSI9OzZUzZs2KDMc/LkSWnevLkEBwfL5cuXJTg4WJo1ayanT5/+UsUgIiIiIiIi+igqANB3ECIiKpVKNm3aJA0bNnznPAMHDpQtW7bI9evXlWldu3aVy5cvy8mTJ0VEpHnz5hITEyM7d+5U5gkICJCcOXPKqlWrPiiWmJgYsbKykujoaLG0tPy4An0ljoO26zuEDAsfX0ffIRAREREREX0xGckpjeqZ8pMnT4q/v7/WtNq1a8u5c+ckMTHxvfOcOHHincuNj4+XmJgYrT8iIiIiIiKiL82okvKoqCixs7PTmmZnZydJSUny9OnT984TFRX1zuWGhoaKlZWV8ufg4PD5gyciIiIiIiJKw6iScpHUZu6a1K3vNaenN0/aaZoGDx4s0dHRyt+DBw8+Y8RERERERERE6fuopLxIkSLy7NkznekvX76UIkWKfHJQ75I3b16dGu/Hjx9LlixZJFeuXO+dJ23tuSYTExOxtLTU+iMiIiIiIiL60j4qKQ8PD5fk5GSd6fHx8RIREfHJQb2Lu7u77N27V2vanj17xM3NTbJmzfreeTw8PL5YXEREREREREQfI0tGZt6yZYvy/927d4uVlZXyOjk5Wfbv3y+Ojo4fvLzY2Fi5ffu28vru3bty6dIlsbGxkYIFC8rgwYMlIiJCli1bJiKpPa3PmjVL+vbtK506dZKTJ0/KokWLtHpV79Wrl3h7e8uECROkQYMGsnnzZtm3b58cO3YsI0UlIiIiIiIi+uIylJSrhytTqVTSpk0brfeyZs0qjo6O8ssvv3zw8s6dOyc+Pj7K6759+4qISJs2bSQsLEwiIyPl/v37yvuFCxeWHTt2SJ8+fWT27NmSL18+mTFjhjRp0kSZx8PDQ1avXi1Dhw6VYcOGiZOTk6xZs0aqVKmSkaISERERERERfXEfNU554cKF5ezZs5I7d+4vEZPecZzyL4vjlBMRERER0bcsIzllhmrK1e7evftRgRERERERERHR//uopFxEZP/+/bJ//355/PixpKSkaL23ePHiTw6MiIiIiIiI6Fv3UUn5qFGjZPTo0eLm5ib29vbvHQOciIiIiIiIiNL3UUn5vHnzJCwsTIKDgz93PERERERERET/GR81TnlCQgLH/SYiIiIiIiL6RB+VlHfs2FFWrlz5uWMhIiIiIiIi+k/5qObrcXFxMn/+fNm3b5+ULVtWsmbNqvX+lClTPktwRERERERERN+yj0rKr1y5IuXLlxcRkWvXrmm9x07fiIiIiIiIiD7MRyXlBw8e/NxxEBEREREREf3nfNQz5URERERERET06T6qptzHx+e9zdQPHDjw0QERERERERER/Vd8VFKufp5cLTExUS5duiTXrl2TNm3afI64iIiIiIiIiL55H5WUT506Nd3pI0eOlNjY2E8KiIiIiIiIiOi/4rM+U96qVStZvHjx51wkERERERER0TfrsyblJ0+eFFNT08+5SCIiIiIiIqJv1kc1X2/cuLHWawASGRkp586dk2HDhn2WwIiIiIiIiIi+dR+VlFtZWWm9zpQpkxQvXlxGjx4t/v7+nyUwIiIiIiIiom/dRyXlS5Ys+dxxEBEREREREf3nfFRSrnb+/Hm5fv26qFQqcXFxEVdX188VFxEREREREdE376OS8sePH0uLFi3k0KFDYm1tLQAkOjpafHx8ZPXq1ZInT57PHScRERERERHRN+ejel/v0aOHxMTEyB9//CHPnz+XFy9eyLVr1yQmJkZ69uz5uWMkIiIiIiIi+iZ9VFK+a9cumTt3rpQsWVKZ5uLiIrNnz5adO3dmaFlz5syRwoULi6mpqVSsWFGOHj36znnbtm0rKpVK569UqVLKPGFhYenOExcXl/GCEhEREREREX1BH5WUp6SkSNasWXWmZ82aVVJSUj54OWvWrJHevXtLSEiIXLx4Uby8vCQwMFDu37+f7vzTp0+XyMhI5e/BgwdiY2Mj3333ndZ8lpaWWvNFRkZy/HQiIiIiIiIyOB+VlNesWVN69eolDx8+VKZFRERInz59xNfX94OXM2XKFOnQoYN07NhRSpYsKdOmTRMHBweZO3duuvNbWVlJ3rx5lb9z587JixcvpF27dlrzqVQqrfny5s37McUkIiIiIiIi+qI+KimfNWuWvHr1ShwdHcXJyUmcnZ2lcOHC8urVK5k5c+YHLSMhIUHOnz+vM665v7+/nDhx4oOWsWjRIqlVq5YUKlRIa3psbKwUKlRIChQoIHXr1pWLFy++dznx8fESExOj9UdERERERET0pX1U7+sODg5y4cIF2bt3r/z1118CQFxcXKRWrVofvIynT59KcnKy2NnZaU23s7OTqKiof/18ZGSk7Ny5U1auXKk1vUSJEhIWFiZlypSRmJgYmT59unh6esrly5elaNGi6S4rNDRURo0a9cGxExEREREREX0OGaopP3DggLi4uCg1yX5+ftKjRw/p2bOnVKpUSUqVKvXejtrSo1KptF4D0JmWnrCwMLG2tpaGDRtqTa9ataq0atVKypUrJ15eXrJ27VopVqzYe2vwBw8eLNHR0crfgwcPMlQGIiIiIiIioo+RoaR82rRp0qlTJ7G0tNR5z8rKSrp06SJTpkz5oGXlzp1bMmfOrFMr/vjxY53a87QAyOLFiyU4OFiyZcv23nkzZcoklSpVklu3br1zHhMTE7G0tNT6IyIiIiIiIvrSMpSUX758WQICAt75vr+/v5w/f/6DlpUtWzapWLGi7N27V2v63r17xcPD472fPXz4sNy+fVs6dOjwr98DQC5duiT29vYfFBcRERERERHR15KhZ8ofPXqU7lBoysKyZJEnT5588PL69u0rwcHB4ubmJu7u7jJ//ny5f/++dO3aVURSm5VHRETIsmXLtD63aNEiqVKlipQuXVpnmaNGjZKqVatK0aJFJSYmRmbMmCGXLl2S2bNnf3BcRERERERERF9DhpLy/Pnzy9WrV8XZ2Tnd969cuZKhGunmzZvLs2fPZPTo0RIZGSmlS5eWHTt2KL2pR0ZG6oxZHh0dLRs2bJDp06enu8yXL19K586dJSoqSqysrMTV1VWOHDkilStX/uC4iIiIiIiIiL4GFQB86Mw9evSQQ4cOydmzZ8XU1FTrvbdv30rlypXFx8dHZsyY8dkD/ZpiYmLEyspKoqOjDf75csdB2/UdQoaFj6+j7xCIiIiIiIi+mIzklBmqKR86dKhs3LhRihUrJt27d5fixYuLSqWS69evy+zZsyU5OVlCQkI+KXgiIiIiIiKi/4oMJeV2dnZy4sQJ+fHHH2Xw4MGirmRXqVRSu3ZtmTNnzr/2nE5EREREREREqTKUlIuIFCpUSHbs2CEvXryQ27dvCwApWrSo5MyZ80vER0RERERERPTNynBSrpYzZ06pVKnS54yFiIiIiIiI6D8lQ+OUExEREREREdHnw6SciIiIiIiISE+YlBMRERERERHpCZNyIiIiIiIiIj1hUk5ERERERESkJ0zKiYiIiIiIiPSESTkRERERERGRnjApJyIiIiIiItITJuVEREREREREesKknIiIiIiIiEhPmJQTERERERER6QmTciIiIiIiIiI9YVJOREREREREpCdMyomIiIiIiIj0hEk5ERERERERkZ4wKSciIiIiIiLSEyblRERERERERHrCpJyIiIiIiIhIT/SelM+ZM0cKFy4spqamUrFiRTl69Og75z106JCoVCqdv7/++ktrvg0bNoiLi4uYmJiIi4uLbNq06UsXg4iIiIiIiCjD9JqUr1mzRnr37i0hISFy8eJF8fLyksDAQLl///57P3fjxg2JjIxU/ooWLaq8d/LkSWnevLkEBwfL5cuXJTg4WJo1ayanT5/+0sUhIiIiIiIiyhAVAOjry6tUqSIVKlSQuXPnKtNKliwpDRs2lNDQUJ35Dx06JD4+PvLixQuxtrZOd5nNmzeXmJgY2blzpzItICBAcubMKatWrUr3M/Hx8RIfH6+8jomJEQcHB4mOjhZLS8uPLN3X4Thou75DyLDw8XX0HQIREREREdEXExMTI1ZWVh+UU+qtpjwhIUHOnz8v/v7+WtP9/f3lxIkT7/2sq6ur2Nvbi6+vrxw8eFDrvZMnT+oss3bt2u9dZmhoqFhZWSl/Dg4OGSwNERERERERUcbpLSl/+vSpJCcni52dndZ0Ozs7iYqKSvcz9vb2Mn/+fNmwYYNs3LhRihcvLr6+vnLkyBFlnqioqAwtU0Rk8ODBEh0drfw9ePDgE0pGRERERERE9GGy6DsAlUql9RqAzjS14sWLS/HixZXX7u7u8uDBA5k8ebJ4e3t/1DJFRExMTMTExORjwiciIiIiIiL6aHqrKc+dO7dkzpxZpwb78ePHOjXd71O1alW5deuW8jpv3ryfvEwiIiIiIiKir0FvSXm2bNmkYsWKsnfvXq3pe/fuFQ8Pjw9ezsWLF8Xe3l557e7urrPMPXv2ZGiZRERERERERF+DXpuv9+3bV4KDg8XNzU3c3d1l/vz5cv/+fenatauIpD7rHRERIcuWLRMRkWnTpomjo6OUKlVKEhISZPny5bJhwwbZsGGDssxevXqJt7e3TJgwQRo0aCCbN2+Wffv2ybFjx/RSRiIiIiIiIqJ30WtS3rx5c3n27JmMHj1aIiMjpXTp0rJjxw4pVKiQiIhERkZqjVmekJAg/fr1k4iICDEzM5NSpUrJ9u3bJSgoSJnHw8NDVq9eLUOHDpVhw4aJk5OTrFmzRqpUqfLVy0dERERERET0Pnodp9xQZWRMOX3jOOVERERERESGxSjGKSciIiIiIiL6r2NSTkRERERERKQnTMqJiIiIiIiI9IRJOREREREREZGeMCknIiIiIiIi0hMm5URERERERER6wqSciIiIiIiISE+YlBMRERERERHpCZNyIiIiIiIiIj1hUk5ERERERESkJ0zKiYiIiIiIiPSESTkRERERERGRnjApJyIiIiIiItITJuVEREREREREesKknIiIiIiIiEhPmJQTERERERER6QmTciIiIiIiIiI9YVJOREREREREpCdMyomIiIiIiIj0hEk5ERERERERkZ4wKSciIiIiIiLSE70n5XPmzJHChQuLqampVKxYUY4ePfrOeTdu3Ch+fn6SJ08esbS0FHd3d9m9e7fWPGFhYaJSqXT+4uLivnRRiIiIiIiIiDJEr0n5mjVrpHfv3hISEiIXL14ULy8vCQwMlPv376c7/5EjR8TPz0927Ngh58+fFx8fH6lXr55cvHhRaz5LS0uJjIzU+jM1Nf0aRSIiIiIiIiL6YFn0+eVTpkyRDh06SMeOHUVEZNq0abJ7926ZO3euhIaG6sw/bdo0rdfjxo2TzZs3y9atW8XV1VWZrlKpJG/evB8cR3x8vMTHxyuvY2JiMlgSIiIiIiIioozTW015QkKCnD9/Xvz9/bWm+/v7y4kTJz5oGSkpKfLq1SuxsbHRmh4bGyuFChWSAgUKSN26dXVq0tMKDQ0VKysr5c/BwSFjhSEiIiIiIiL6CHpLyp8+fSrJycliZ2enNd3Ozk6ioqI+aBm//PKLvH79Wpo1a6ZMK1GihISFhcmWLVtk1apVYmpqKp6ennLr1q13Lmfw4MESHR2t/D148ODjCkVERERERESUAXptvi6S2tRcEwCdaelZtWqVjBw5UjZv3iy2trbK9KpVq0rVqlWV156enlKhQgWZOXOmzJgxI91lmZiYiImJyUeWgIiIiIiIiOjj6C0pz507t2TOnFmnVvzx48c6tedprVmzRjp06CDr1q2TWrVqvXfeTJkySaVKld5bU05ERERERESkD3prvp4tWzapWLGi7N27V2v63r17xcPD452fW7VqlbRt21ZWrlwpderU+dfvASCXLl0Se3v7T46ZiIiIiIiI6HPSa/P1vn37SnBwsLi5uYm7u7vMnz9f7t+/L127dhWR1Ge9IyIiZNmyZSKSmpC3bt1apk+fLlWrVlVq2c3MzMTKykpEREaNGiVVq1aVokWLSkxMjMyYMUMuXboks2fP1k8hiYiIiIiIiN5Br0l58+bN5dmzZzJ69GiJjIyU0qVLy44dO6RQoUIiIhIZGak1Zvmvv/4qSUlJ0q1bN+nWrZsyvU2bNhIWFiYiIi9fvpTOnTtLVFSUWFlZiaurqxw5ckQqV678VctGRERERERE9G9UAKDvIAxNTEyMWFlZSXR0tFhaWuo7nPdyHLRd3yFkWPj4f3/sgIiIiIiIyFhlJKfU2zPlRERERERERP91TMqJiIiIiIiI9IRJOREREREREZGeMCknIiIiIiIi0hMm5URERERERER6wqSciIiIiIiISE+YlBMRERERERHpCZNyIiIiIiIiIj1hUk5ERERERESkJ0zKiYiIiIiIiPSESTkRERERERGRnjApJyIiIiIiItITJuVEREREREREesKknIiIiIiIiEhPmJQTERERERER6QmTciIiIiIiIiI9YVJOREREREREpCdMyomIiIiIiIj0hEk5ERERERERkZ4wKSciIiIiIiLSEyblRERERERERHqSRd8BzJkzRyZNmiSRkZFSqlQpmTZtmnh5eb1z/sOHD0vfvn3ljz/+kHz58smAAQOka9euWvNs2LBBhg0bJnfu3BEnJyf53//+J40aNfrSRaEvwHHQdn2HkGHh4+voOwQiIiIiIjISeq0pX7NmjfTu3VtCQkLk4sWL4uXlJYGBgXL//v105797964EBQWJl5eXXLx4UYYMGSI9e/aUDRs2KPOcPHlSmjdvLsHBwXL58mUJDg6WZs2ayenTp79WsYiIiIiIiIg+iAoA9PXlVapUkQoVKsjcuXOVaSVLlpSGDRtKaGiozvwDBw6ULVu2yPXr15VpXbt2lcuXL8vJkydFRKR58+YSExMjO3fuVOYJCAiQnDlzyqpVqz4orpiYGLGyspLo6GixtLT82OJ9Fd96TfK3Xj4iIiIiIvr2ZCSn1Fvz9YSEBDl//rwMGjRIa7q/v7+cOHEi3c+cPHlS/P39tabVrl1bFi1aJImJiZI1a1Y5efKk9OnTR2eeadOmvTOW+Ph4iY+PV15HR0eLSOoPaehS4t/oO4QMy8jv+q2XT0Sk9IjdXyiSL+faqNofPC/LZ3gyUj6R/0YZiYiIiD4ndU7wIXXgekvKnz59KsnJyWJnZ6c13c7OTqKiotL9TFRUVLrzJyUlydOnT8Xe3v6d87xrmSIioaGhMmrUKJ3pDg4OH1ocygCrafqO4Mv61ssn8u2XkeUzfv+FMhIREZHhe/XqlVhZWb13Hr139KZSqbReA9CZ9m/zp52e0WUOHjxY+vbtq7xOSUmR58+fS65cud77uW9ZTEyMODg4yIMHDwy+Cf/HYPmM37deRpbP+H3rZWT5jN+3XkaWz/h962Vk+b5tAOTVq1eSL1++f51Xb0l57ty5JXPmzDo12I8fP9ap6VbLmzdvuvNnyZJFcuXK9d553rVMERETExMxMTHRmmZtbf2hRfmmWVpaftM7Ectn/L71MrJ8xu9bLyPLZ/y+9TKyfMbvWy8jy/ft+rcacjW99b6eLVs2qVixouzdu1dr+t69e8XDwyPdz7i7u+vMv2fPHnFzc5OsWbO+d553LZOIiIiIiIhIX/TafL1v374SHBwsbm5u4u7uLvPnz5f79+8r444PHjxYIiIiZNmyZSKS2tP6rFmzpG/fvtKpUyc5efKkLFq0SKtX9V69eom3t7dMmDBBGjRoIJs3b5Z9+/bJsWPH9FJGIiIiIiIionfRa1LevHlzefbsmYwePVoiIyOldOnSsmPHDilUqJCIiERGRmqNWV64cGHZsWOH9OnTR2bPni358uWTGTNmSJMmTZR5PDw8ZPXq1TJ06FAZNmyYODk5yZo1a6RKlSpfvXzGzMTEREaMGKHTrP9bwfIZv2+9jCyf8fvWy8jyGb9vvYwsn/H71svI8pGaXscpJyIiIiIiIvov09sz5URERERERET/dUzKiYiIiIiIiPSESTkRERERERGRnjApJyIiIiIiItITJuVEREREREREesKknIgMwn9lIIj/Sjm/Ra1atZIGDRroO4wvTr2NJiYm6jkSIiLjoz6Gap7vv6Vzf0pKivL/uLg4PUbybWFS/h+hPhho7khkPL719ZaSkiIqlUpEUg/w3/JBXqVSybJly2T27Nn6DuWz+5YuOtICIA0bNpQjR45I+/bt9R3OF6VSqWTDhg0yZ84cef36tb7D+ay+9WPpf0HadfitHnfGjRsnU6ZM0XcYlEGa1zMvXryQN2/eyJs3b0SlUn0Tx5+UlBTJlCk1fZw0aZKMHj1a7t27p+eovg1Myv8DAIhKpZI9e/ZI3759pV27dnLt2jVJSEjQd2ifTXoHum/h4CeSuv7UB8AVK1bIiBEjZM2aNXL//n09R/Z5aB7gJ0yYIK1atZLy5cvL0KFD5ejRo3qO7vNRXzjeuXNHunXrJrGxsXqO6PPSvBC5d++exMbGflM3V1QqlTRt2lSWL18uW7ZskQ4dOug7pM9Ocxtt1aqVZM+eXbJnz67nqD4fzWPNmTNnZMeOHXLt2jV59uyZniP7fL7VBFVNcx1u2rRJbt++rRx3viUAJDo6Wnbt2iXPnz/Xdzj0gTSv18aPHy/NmzcXLy8vad68uVy5ckV5z5ipyzBw4ECZMmWKODg4iImJiZ6j+jao8K0fwUlERHbv3i3169eXunXryvXr1yUqKkqmTJkiTZo0EQsLC32H90k0T9L79u2TN2/eSJEiRaR06dJ6juzTqW+oiIgMGjRIFi1aJAULFpRXr16Ji4uLjBo1SsqVK6fnKD+PIUOGyPz582Xq1Kny+vVrWbBggQCQ3bt3S548efQd3mdx5swZOXLkiDx69EgmTZqk73C+iOHDh8uGDRskOTlZ6tatK127dhVnZ2d9h/VJkpOTJXPmzCIicujQITl06JCMHj1aevToIdOnT9dzdJ/X0aNH5dmzZ3L69GkJDQ3VdzifjeaxdPDgwbJ27VoxMTERlUolxYsXlyFDhoibm5ueo/w06jIePHhQNm/eLHnz5hV3d3epXr26vkP7LDTX4cCBA2X16tUSHBwsISEhYmZmpufoPr+dO3dKx44dZf369eLu7q51rWPMNNfjtyokJETmz58vM2fOFEtLSwkJCZHIyEi5fv265MyZU9/hfTL1trlp0yapXLmyvsP5doC+eS9evEC/fv3w66+/KtN+/PFH2NraYuHChYiJidFjdJ/PgAEDYGFhAScnJ2TJkgWzZs3C27dv9R3WZ3H58mW0bNkSZ8+eBQCsW7cOAQEBqFmzJi5duqTn6D7dtWvXUL58eRw9ehQAsG/fPpiZmWHx4sUAgOTkZH2G91k8ffoUDRs2RPbs2dG8eXMAQEpKClJSUvQc2afRjH/t2rWws7PD2rVr0bt3b9SsWRMBAQG4fv26HiP8fPr374/ixYujR48eqFq1KjJnzox27drpO6zP5u3bt6hZsyZUKhUCAgL0Hc4XMXv2bOTNmxdHjhwBAPz888+wtLTEvn379BzZ57Ft2zaYmJjAz88PpUqVQqlSpbBkyRJ9h/VJ0h7/J0+ejFy5cuH8+fPfzPWLWtrzQfPmzVGzZk3Ex8frKaLPS12+o0ePYtSoUejRowfWr1+v56g+D3XZ7t+/jypVqmD//v0AgC1btsDa2hpz5szRms9Yzv2zZ89GUlKS1rRFixahatWqiIuLQ2JiIoD/L09ycrLRlM3QMCn/xp0/fx65c+dG+fLlsXXrVq33unXrhjx58mDRokV4+fKlniL8eJo7/aVLl1C+fHmcPn0a4eHhmDZtGlQqFUJDQ/HmzRs9RvnpVq9eDU9PT/j7+yM2NlaZvnnzZgQEBMDX19foE/OLFy/C2dkZSUlJ2LBhA3LkyIG5c+cCAN68eYPVq1fj0aNHeo7y023fvh1BQUGwsLDAuXPnABjPifnf7Ny5E/369VNupACpSbqfnx/8/f2NPjE/dOgQrKyscPjwYQBAdHQ0VqxYgRw5cqB9+/Z6ju7zuXHjBpo3b46cOXPi6tWrAL6Nm2Lq/ax169YYPXo0AOD333+HpaUl5s2bByD1psTTp0/1FuOnioiIwIQJEzB//nwAwNWrV9G3b1/kz58fixYtUuYzpmNO2m0vLi4OTZo0wYQJE7TeN9ZtVDPuCRMmYOrUqTh//rwybfv27ShfvjyuXLmiM7+x2rBhA2xsbNC4cWP89NNPUKlU6Nu3r1Feh37//fdKsq125coV5M6dG7Gxsdi2bZvW9czr168xY8YMPHv2TB/hZtj+/ftRoUIFnaR84sSJyJ07tzJd898DBw7g/v37Xz3WbwGT8v+Axo0bQ6VSYerUqUhISNB6r2fPnsiUKROWLVtmVCdqTRMmTEDfvn3Rs2dPrelz5swxysRcfdJVr49ffvkFZcuWhb29PR48eKA175YtW1CnTh2ULVsWt27d+uqxfoz0LipOnTqFUqVKYd68ebCyssLs2bOV944fP47vv/8ely9f/pphfjL1+nv79q1Wbc6pU6dQq1YtVKhQARcuXNCa11idPn0a5cuXh42NDcLCwrTeW7duHfz9/REYGKgkecZow4YNKFCgAF69eqVMe/v2rXKc6devnx6j+ziaNRtxcXHK9AcPHqBGjRooUKAAwsPDlXmMTdr9KikpCUFBQdi5cycOHTqEHDlyKAl5YmIiFixYgM2bNxtlWa9evQoXFxeUKlUKhw4dUqbfvn1bScyNrca8VatWaNu2LYD/X5evXr2Co6MjhgwZosynfu/NmzeIiorSmmbINK9LYmJiEBISgmrVqiF37twICQnB8ePHAQBubm7o0qWLvsL8rG7duoXChQsrSWpsbCzMzMwwYMAAPUeWcc+ePUPnzp1hZWWFpUuXKtMfPHiAgIAADB06FBYWFlqtVK9evYpGjRopLXWMgXpf2rdvn9Ji4+zZs3BxccHQoUO1bqbExMTAx8cHCxYs0Eusxo5J+TdM86TUpEkT5MyZE9u3b9dJzPv3748bN2587fA+WtqTba9evaBSqeDj46PTXH3OnDnIkiULhgwZYnTNv9RN1QEgLCwMZcuWRdOmTfH3339rzbdmzRr07dvXKC4kNWOcMWMGpk2bpmyPderUgUqlUmpAgNSLljp16qB+/fpGUT419Ta6bds2BAUFoXTp0vjuu++wdetWpKSk4PDhw6hXrx7c3Nxw8eJFrc8Yg/RinTlzJooXL47q1avjn3/+0Xpv/fr1cHV1RZ8+fb5WiJ8kvfL98ccfsLCwwMaNG7WmX7t2DTY2NlCpVEoNrDFQl3Hnzp0IDg6Gu7s7QkJClITu4cOHqF69utEm5pqxat6w7Ny5M6ytrWFubo7ffvtNmf7s2TPUrFkTkydP/qpxfi6nTp1C8+bNYW5ujpUrV2q9d+fOHfTv3x+mpqZYvny5niLMmKSkJJw4cUI5P6i319jYWLRs2RLNmzdHRESE1mdOnz6Npk2b4uHDh1893ozatm0bpk6dCiB1myxdujSA1NYOq1atQuXKlVG6dGk0a9YMgwcPRuHChY26RZx6/V2+fBnVqlUDkLpd5s+fH507d1bmM6ZrUQD4559/lEcn1Tekk5KS0KBBA6hUKvTv31+Z9/Xr1wgMDERQUJBRHUsB4M8//1RaNACpZezTpw/c3d3RuXNnXLt2DQcOHEBQUBAqVqyoNGmnjGFS/g1RH/T+/vtvXLt2TafmtH79+sidOze2b9/+Tewwms1/xo4di0yZMqVbEzBp0iR4enoafNKjeZA+duwYbG1ttZoCz5s3D97e3mjRogXu3r2b7jLSNjEyVP3790f+/Pkxfvx4JYG7c+cOvL294eDggMmTJ2PcuHHw9fVFqVKllAszYzqRbdu2Debm5hg2bBgOHToET09PFC5cWLnZsm/fPjRs2BBOTk5G1QpAcz9asGCBUuMBpD575u7ujjZt2uhcMB88eNAo1p9mjPHx8UhISEBKSgri4+PRpEkT1K1bFwcOHFDmefDgAYKDg3Hw4EGj2f/UNm/ejGzZsqFz587o2bMnihcvjho1aigXlw8ePICfnx+yZ89uVM0RNdfh6NGj4e/vj127dgFIvYiuVasW8uXLh5iYGMTExCAqKgoBAQGoUqWK0a1DTefPn8d3332HwoUL6zyudvPmTYSEhODmzZt6iu7j/frrryhVqpRy3bJixQqYmJhg2LBhuHPnDoDUPjsaNGgAPz8/ozjOdOvWDfny5UPNmjWRO3dupXm6+vj6+PFjnDlzBoGBgShatChUKhVmzZoFwLjOg+ryqNfdvn37UKhQIRw5cgSFCxdG586dlX3uxIkTCAwM1Kl4MESa58EHDx6gf//+sLCwUB4TSUhIQMWKFeHi4oIuXbpg+PDhqF69OsqUKWMU1zPpXS+vWrUKpqamSmKekJCAcePGoWrVqlCpVChTpgxq1qyplM+Yj6X6wqT8G6HegTZt2oTChQujePHiMDExwYgRI/Dnn38q89WvXx/29vbYtGmT0SXmmgewqVOnwtvbG3/88YcybdCgQciaNatW7YeaoXesoVm2ZcuWoWvXrsiRIwfy5cuHhQsXKu/NnTsX3t7e+OGHH4ymuXpaq1evhq2trVZLALUXL16gQ4cO8PDwQK1atdCtWzdlOzWW7TU5ORkxMTHw9/fHuHHjAKTeIS9QoAB69OihtQ3u3r0bLVq0MIqLEEB7Oz179izq1asHe3t7rFu3Tpk+ffp0eHp6ppuYp12GodGMbdKkSWjRogXKli2L0aNH49atW7h8+TJq1qyJatWqITQ0FFu2bEGtWrXg6+urc/Fp6J4+fQoPDw+tlik3b95E69atUaNGDZw6dQpAai1z/fr1jTKZGzBgAHLnzo1t27bh3r17AFIvFA8fPgxXV1fkzJkTLi4uqFSpEipVqmQ0F5Pqbe3q1as4cOAANm7cqGy7V65cQatWreDi4oJt27ZpfS5tKzljkJKSgq1bt6JkyZLw9vZW9q+5c+fC3t4elSpVQuXKlVGpUiWULVvWoBOegQMH4uTJk8prdTIzcOBArRYBaWO/ePEievfuDXt7e51WSMbg1KlT8PT0xKtXr/DixQvUqVMH5ubmSoenaoMHD0b16tXx+PFjPUX6YdLbtu7fv4+ff/4ZOXLkUJpux8fH4+eff0ZgYCDq1auHn3/+2SiuZzTL9/r1a633Vq1ahSxZsiiJuXp7PXPmDO7evat81pDLZ8iYlH9Ddu3aBWtra8ycORPx8fGYMWMGTE1N0b17d61nOatXrw5nZ2etTsMMneZB4vjx45g4cSJUKhVatGih1dxp4MCByJYtG1asWKGzDENMyNMe3AcOHIi8efNi3rx5mDRpEmrUqAFnZ2fluUcgtdagZMmSGDFixFeO9vMYN24cGjVqBOD/D9xpL4JjYmK0fhtjO8AnJiaiSpUquH79Oh48eIB8+fJpNdHbvn27krCmPekZg8GDByMgIAC1atWCtbU1ChcurPVM3YwZM+Dt7Y169erhyZMneoz04wwaNAi5c+fGokWL8Msvv6BMmTKoWrUqgNReg3/++WfY2NigTJky8Pb21mlia4jS9vT/5s0blCxZEjNnzlTeB/7/mc/Q0FBlXmPb/4DUjvmcnJyUm39xcXGIiorC7t278ebNGyQkJGDBggWYP38+Nm7cqByDDL2s6vW0bt065M2bF87OzrC2toaLi4vSGuDy5csIDg5GuXLlsGHDBn2Gm2HpJTxxcXHYs2cPSpcuDXd3d2UdHThwAAsXLkSfPn0wd+5cg054Ll68iDZt2mjF2KFDB3z//fcoUqQIpk2bprT+U/8GmufFGzduoHz58kY5SsDvv/+OkiVLKvvinDlzUKJECbRo0QJnzpzBiRMn0K9fP1hZWSktBgyV5vZ58+ZNnDp1Ci9fvkRycjJevXqFvn37wsLCQqsyJTExUevYa8g3/TTjnDRpEho3bowGDRrg8OHDSs6gTsz79euX7o0+Q7whZiyYlH8jXrx4gRYtWmDkyJEAgPDwcDg7O6N69eqwtLRE+/bttRJzY2qKqGnAgAHIly8fxowZgzZt2sDCwgIBAQFaifngwYOhUqmwe/duPUaacXfu3EHJkiW1nlm9cuUKOnfujMKFC2t1oLVp0yaDPrC/T7du3VCqVCnltfoAnpCQgIMHD+pcUBlyopOe5ORkvH37FiVKlMDAgQPh7OyMzp07KyevyMhINGzYEKtWrdJzpB9O8yQbFhaGHDly4NixY4iNjcXx48fRqlUrFCtWDMuWLVPmCw0NRdeuXY3mBK2O8/LlyyhTpozSydLevXthamqq1Xs1kHrMjYyMNOgacs3ncNWOHTuGs2fP4tWrV6hQoQIGDhwIIDV+9W8QHByM+vXrG+0xBki98VWgQAG8efMGf/75J4YMGQInJyeYm5ujSpUq6X7GWMp79uxZWFtbY+nSpQgPD8eTJ09Qu3ZtFC1aVBmG6cyZM2jYsCGqVq2K2NhYoziOah4rdu3ahQ0bNiiJXFJSkpKYe3h4vHN/M+R1qF4Hq1evxokTJ5Tp3bt3h6Ojo1ZiDkDptE6tSJEiWp2gGgt1U+6mTZsq02bNmgV/f39kzpwZZcuWRaVKlQz+mXnNfWjIkCEoVaoUbG1tUaVKFfTo0QPPnj3Do0eP0L9/f1haWup0emroNPe/yZMnw9LSEoMGDUKZMmWUG7jR0dEAUrdhU1NTdOnSxaD3OWPDpPwb8fr1a6xfvx53797F06dPUaZMGXTo0AFAau/dOXLkQLt27Yyq9+O0nbadOXMGuXLl0nqe89KlS8iZMycCAwPx119/KdPnzJljkBfJak2bNtXpLT4iIgK5cuXSeo4cSE0SihQpAjs7O627r4BhX4C8KxlbvXo1ihcvjt9++01rHT99+hQ1atTApk2bvlKEny4pKUkpp7pXbvWJe968ebC0tIS7u7vWZ0JCQlCyZEml8yxDNnToUJ1pffv21RnD+sKFCwgMDESBAgWwdu1aZbp6+zTUxHzo0KE6j7ucPn0axYoVA5Da47qFhYVWT8Hr1q3TaV5pqOUDUm8C2dvb49y5c9i2bRuyZ8+uJG5LliyBSqXS6RisYcOGOscnQ5be7//nn3+iQoUKKFGiBGxtbdGxY0csXrwYd+/eRbZs2bQeuTBkO3fu1HkMZPny5XB1dUV0dLRW2X19fVGuXDnl9cWLF9N9hMTQDRw4EDly5EDRokWRKVMmTJw4UTnW7tmzR2mhYsjneE2aw7aFh4fDxcUFderUUVo2AECPHj3g5OSECRMm4Pr16/D19YWHh4fyvnr4PkMeWvJ941MfPnwYjo6O2LJlizItNjYWFy9exMOHD/H8+fOvFeYnmzRpEmxtbZVWCy1atECePHmURxMiIiIwcOBAqFQq7NixQ5+hfpQ///wTHTp00LrW7tixI8qVK4fp06crifnixYvh7e1tFDf8jAWTciOV3k6gHl9V/dyxelznefPmoXjx4ihTpgwiIyO/apwfq0WLFvj999+1pp04cQL58uXD7du3Afz/83EnT55EtmzZ0KpVK63EHDDM2quEhATs379fpzf4x48fo3bt2ujdu7dOk9/mzZvD29sblSpV0jqRGyrNC8Vt27Zh1apVyvjOsbGxCAwMhKenJ6ZNm4bIyEhcuXIFderUMZqOlg4ePKh1sbt161YEBQXBy8sLCxYswMOHD/Hy5Ut0794dNjY26NmzJ8aOHYsOHTrA0tJS6XHdkO3btw+1a9fW2YcmT56M8uXLp5soZMqUCSVKlNB6fMRQT9i3bt1CzZo1Ub16da0mvgcPHkT58uWxePFineH5jhw5guDgYK2+LAzd8+fP8dNPP8Hc3BxZs2ZVbpqo99ERI0ZApVKhR48eGDVqFHr06AELCwujKWPafg7Onj2rDEF47tw5jB07Fps3b8aLFy8AAE+ePEHlypVx8OBBPUT74VJSUnDgwAEUK1ZM57w9ZcoUODg4KK/Vj8DcvXsX1tbWyk0XY3T16lW4urrizJkzePDgAebNmweVSoVhw4YprTn27t0LW1tb/Pjjj/oO919pbp/qFiuHDh2Ct7c3GjZsqHU+//nnn1G0aFEUKVIElStX1rpGOHPmjMH2PTJo0CCtbW7Xrl1o1aoVDh06pBz/IyIiULNmTQwaNAiA4Z4X0tK8eZ6UlKSMCKMe5mznzp3IkSMH5s+fDyD1OfLk5GQ8fPgQs2fPNshr0PdZuXIlHBwcULRoUZw7d07rvU6dOqFcuXKYMWOGcjxVM5b1aeiYlBsh9cZ/5swZrFmzRqdmccyYMahWrZpyIh84cCCWLl2q3N0yBoMHD1ZOSOrk++7duzAxMVGakaqfkXzy5AmKFy+OLFmyoEmTJkZ1cJg5cya8vb2V1zNmzICVlRWmTJmi3FSJiYlB06ZNMWfOHFSrVk1pbmoM1DUeJUuWhEqlQkhICAAgOjoarVu3RqlSpZApUyaULVsW7u7uRtHR0pEjR+Do6IiBAwciJiYGly5dgqmpKQYMGIC6deuibNmy6NSpEyIiIvDq1SssWLAA5cqVQ40aNdCmTRujSXbi4uKUfUkzad22bRscHBwwdepUrdqN/fv3o0mTJujSpQsCAgKM4gbg2bNn0bRpU/j4+GjV8FerVg0qlQpTpkxRpr19+xZ16tRBo0aNDLpmPD07duyASqVC1qxZlWb5mmVYuXIlfHx8ULlyZQQFBRnVaABq/fr1Q6FChZAtWzYEBQXp1ITHx8cjMjIS9erVM5qbfwCU88DNmzfx4MEDAKmPn+XKlQv9+vXTmvePP/6Ak5MTzp8//9Xj/BzGjx+P7t2746efftKarm7RoZmYnzlzxuDXoeY+Nn78eLRr107pqO3w4cPw9PTUScyPHTuGAwcOKGUz9M75njx5gp9++knrWfC9e/eiWLFiqFy5Mry9vXH69GkAqY+UZM2a1WiOL/369UPDhg214n379i08PDzwxx9/YNeuXciRI4fSkio+Ph4LFy5UjrFqxpSYJyQkoFGjRsiWLRumTZuGuLg4rfe7dOmCvHnzap0v6fNhUm6k1q1bBysrKxQpUgQODg7w9/dX3luxYgXMzc3RqFEjBAUFwdzcHNeuXdNjtB8u7cXunDlzMHv2bLx8+RJAarLu4OCgdcH16tUrdOvWDXv37oWJiYnWEE2GLCkpCatWrYKDgwPq16+vTB89ejRsbW1Rp04dtG/fHu7u7nB1dQUAtGnTBj4+PkaRFFy/fh2VKlXCuXPnEBUVhZUrVyJz5szo1asXgNQTWFRUFLZs2YJz584ZVa+dQ4cORcWKFTF8+HCMHDkSkyZNUt6bM2cOqlativbt2ytD16kvrAz9Ais9f/zxB7Jnz650zgekNsG3trbG2LFjcezYMTx48ABBQUEYPHgwtm7diixZsii9dxsizW1s48aNqF+/PipXrqy0zrlx4wYqVKiA4sWLY9asWZgyZYoyPJ/6s4a+D6pvqERHR+Phw4fYsWMHevTogWzZsin9bSQlJSnzJSQkIDk52Wg6Hkw7ioGLiwsOHTqEPXv2oF69evD29lZqr4DUppbqGw/GcPNPcxu9f/8+bG1tMXDgQKU/mOnTp8PZ2Rk///wzkpKS8OjRI4wcORJFihQxinG60zNkyBCoVCpUq1ZN2Q7V6zksLAxZs2ZFr169tNabIa9DtQEDBsDe3h5z5sxRRgEA/j8xb9SoUbp94BhD2QAoFSh79uxRyvH69Wvs3LkTDRo0QMGCBVGvXj0sX74c9evXR58+fXSSPUM0Z84cuLm5oX379lqJuaenJ8qXLw8rKyutRwr/+ecf1KxZM92heQ1R2nOYZieEdevWRZkyZbB27VqdVp3jx483mm3T2DApNyLqk9ObN2/QpEkTLFu2DJGRkdi2bRuKFSsGNzc3Zd65c+fi+++/x/fff29Uz5GnVa9ePTg5OSEsLAzx8fG4d+8eunTpAhsbGwwcOBAzZ85EzZo14ebmhqSkJHh5eaF79+76Djtd6V3Ex8bGKsPYBQUFKdNXr16Nfv36ISAgAN26dVNOYE2aNEHPnj0NPiEYN24cOnTogE6dOmldPK9fvx6ZM2dGnz59lGewNRl6uTST6pEjR6JSpUooWrQoZsyYoTWfOjHv1KmTVs24MbTiSBtjbGwsVq1ahcKFC6NJkybK9NGjR6NChQqwsLCAk5OTkrA+fPgQJUqUMIraupCQEDRq1AgVK1ZE1qxZUaFCBaXl0ZMnT9C0aVO4ubmhRo0aWp31GfqNI/U63LZtG4KDg5Uxq589e4bOnTsjW7ZsWr04r1mzJt0hCo3B9u3b8eOPP+J///ufMu3vv/9Gy5Yt4eXlpVw079+/H9OnTzfoHro1nz1WU3diOmHCBDg6OmLYsGF4+vQpYmJiMGvWLOTJkwe2trZwcXGBvb29Uex3wLuP9ZMmTYJKpUr35vrs2bPh6elp8MdRzfgOHDiA/Pnz48iRI1rvq+c5cuQIvL294eXlZXT7oObwbcnJyfjhhx+QKVMmnUfsNmzYgJ49e8LMzAwqlQru7u4GffNPs/Xp0qVLUbFiRbRr107Zt86cOQNnZ2dUqlQJQGqrspcvXyqPsBlDwqq5/y1ZsgQ9evRAx44dlc5aExMTERQUhPLly6ebmAPGc9PImDApNzKHDx+Gj48PvvvuO+WOa3JyMo4fP46iRYuiQoUKyrxxcXFGtdO86ySt7tl5yZIlSo3AzJkz4eTkhCpVqqBu3brKxXL16tUxduzYrxn2B9Es28GDB7Fjxw5l3N/4+Hhs3LhRJzHX/ExUVBSGDBkCGxsbo2j+PGHCBKhUKri5uSnPd6ovQjZu3AgTExN06NDBKIblS2+7VO97EyZMQK5cudCwYUOlmamaeui67t27G00Nedph6NQn4tjYWKxZswYODg5o3LixMs/169dx7NgxHDx4UPls7969UaJECZ3fwxBolm/+/PmwsLDA0aNH8fjxY+zfvx/+/v7w9PTUuih79uyZ1vozxGQuPZs2bYKJiQkmTJigdcx4/vw5unTpgmzZsmHmzJno3bs3rK2tlb46jElkZCRcXV1hamqqdGyqpk7Mvb29dRI8Qz4v3r17F35+fgCALVu2oFChQkpfKZMnT0a+fPmUxBxIvXm0dOlSbN++XasW1pBp7ocXL17EwYMHcfHiRWXfGj58ODJnzqyM96xJfR4xxMQ8vY4x1Z3yaT4OpP5XvR3u2bPHaEaqUMf+4sULpQXj9u3b8ezZM9y7dw+dO3dGzpw50635v3z5Mnr27Ik///zzq8acEb/++iuKFCmi1fotLCwMFStWRNu2bZUa86VLl8LKygplypSBl5eXUntuDK1wNPXv3x8FChRA586dlZGLRo8eDSD1XFenTh1UrFgRS5cuNZpznzFjUm5EUlJSsHbtWhQpUgS5c+fGmzdvlPfUiXnJkiXh5OSkxyg/TtqT9K1bt5RnrwCgZcuWSmKuLverV6+0PjdgwADkz58ft27d+nqBZ9DAgQNhaWkJR0dHmJqaKj2tJyQkYOPGjShSpAgaNGig9Znnz5+jQ4cOKFKkiEF2EKZ5caT5//nz50OlUmHixIk6n1mxYgW8vLwM8sIqPbdv30azZs0ApN71L1SokJLEjB07FmXKlMGQIUN0hrBR9/ZsbEJDQ9G4cWP4+Pjg0KFDAFJv8qkTc82hbdSOHj2K77//Hrlz5za47XTChAk603788UfUrVtXa9qxY8fg6uoKV1dXbN68WeczxrK93rt3D2XLltVpwaEWExODQYMGoXDhwqhSpYrR1K6ml5BdvnwZfn5+KFOmjNZwkkBqglu7dm38+OOPRrPuDh06hBIlSsDFxQWZMmXC6tWrtd7XTMyNJQnXpLkeBg0ahNKlSyNfvnzw8fGBn5+f0ips9OjRyJIli85QhGmXYSjOnDkDLy8vncTlt99+g42NjbKu1LXkycnJ+P3333XWoTEk5pGRkahUqRJ+++03LF26FCqVSrmReevWLbRv3x45c+bEnj17AKSWSf27GHqyGh4ejp9++gnu7u5a1y5hYWGoUKEC2rZtq/SAHx4ejoEDB2LkyJGYP3++UjZjSV737duHQoUKKcPz7dq1CyqVSmsEoISEBFSqVAlt27bVV5j/KUzKjczr16+xYcMG5M2bV+eCMjk5GYcOHULFihUNtpfOfzNgwAAUKlQI1tbWaNq0qVYHUy1btkSJEiWwZMkS5Q4tAJw/fx69e/eGvb09Lly4oI+w30nz4uHs2bMoU6YMTpw4gVu3bmHMmDHIlCkTpk2bBiD14Pf7778je/bsOp25PXjwQOnkx5BoXkC8evVKa4xVAJg6dapOh1lpL6iM4SLk+PHjyvjGKpVKZxit4cOHw9XVFYMHD9ZJzI2B5joYN24c8uTJg169eqF27drImjWrVs+ya9asgaOjI2rVqqW1jBs3biA4ONjg+q9Ys2YNmjZtqnMxOHz4cHh5eel0gDl//nyYmZmhZMmSWk28jcmNGzfg4OCg02Q27f/v37+v04uuodLcRp8/f464uDhlSMXz58/Dx8cHgYGBOjdTHj58qHzWEJO59IwbNw4qlQpFixZVpmkOHzl58mQ4Ojri559/Nsohz4DUHuRz586N48ePIyUlRRlCSnOfGz16NFQqldYwWoZKs0m65nXLqVOnULJkSQwbNkyroiEuLg7e3t7p3jA0Bu3bt0eBAgWQKVMmzJs3T+s9zcR87969AIzjPK9efw8ePMCPP/6IKlWqvDMxf9eNZ0O/6aBpxYoVCAwMBJD6aGGOHDmUXuVfvnyp9AujOfQrfVlMyo1QXFwc1q9fj0KFCml1vgSkHvg0a9ANneZF0v79+1GkSBEcOnQIYWFhaNasGSpVqoTly5cr87Rq1Qo5c+bEtm3blGkxMTHYvn27QY/7PHnyZAwZMkSnt9wJEyYgU6ZMmD59OoDUpOfw4cNaB3ZDvZBM27Ost7c3ihcvjuDgYPzzzz9K3FOmTEHmzJmVmw/GSn2BqO50D4DWc1bDhw9H5cqV0bNnT51xrI3FvXv3MGDAAKV2HPj/pqTqk3V8fDzCwsLS7YXcEJvpx8bGKnFu375dmb5+/XqYmZlh6dKlWuVYs2YN/Pz8EBoaarQXIsePH4eNjY1y4ajZqdK5c+ewceNGgz2upEcz1nHjxqF69epwc3NDYGCgUmt14cIF+Pj4ICgoSHmGXpMxrEt1jOvWrcPw4cNRqVIlVK5cWTmna57bx48fjxIlShjlsSYhIQE//PCDckzZtm0bcuTIoTRXf/36tXIOXLx4sdHUPAKpj02oVCqtx3xCQkJQokQJdOnSBTt37sSePXvg5+cHV1dXoyob8P9J55kzZ5A1a1bY2tpi2bJlWhUlQGpi3rlzZ6hUKoMfflDThyTmbm5u6NChA86cOaPzOUOV3vFv9erV8PHxwaJFi2BhYaH1mM+WLVvQokULrcogYziGGjsm5Ubq7du3WL9+PRwdHdNtSmpsNm7ciK5duyI0NFSZdv78ebRu3Rpubm5aifnIkSOVE4OhHwjVOnToAJVKhcDAQJ3EZcKECciaNatWR0WA8dxxDQkJgb29PaZNm4YDBw7A2toajRo1woULF5T1M23aNKhUKp2mmIZOc/tas2YNhgwZojPagWbCM3jwYHh6ehrlhfLWrVuhUqlQoEABZUx5tREjRiBLlixKjbnmNmzIJ2rNfejs2bOwt7dH+/btlWn9+/dHtmzZMHv2bFy4cAGPHj1CnTp1MGzYMJ3nPo1N1apVtfoYUevduzd+/PFHo7p5qzZ06FDkypULc+bMwahRo+Dv7w9LS0vlJtKZM2dQq1YtVKpUCceOHdNztB/uXeexPXv2wNXVVWfMavUQU8bYygFI3aeqV6+OtWvXKgm5OiFITEzE3LlzdYa0M9TkNe26S0pKwu7du2Fra6uVmP/yyy+oXbs2VCoVKlSoAD8/P6N7/lhd1ufPn+P58+c4duwYunbtCmdnZ8ybN08nMX/w4AG6dOmi9IlgqN51DgsPD0fXrl11EvNly5bBwcFB55rNUKW9jlE7deoUXF1dYWJignHjxinT1WOxt2vXzmiusb8VTMoN1IfsCG/fvsXGjRthaWmJVq1afYWovow7d+7A29sb1tbWOjXJ58+fR5s2bVClShWt4W0Awz2RvesAP3DgQGTNmjXd8R2HDh2KatWqGd0BUD0UkbqZ7KFDh2BmZgYrKytUrlwZFy9eVMq0evVqg72wSo867suXL2Pnzp3YsmULoqOjcfjwYTg5OSmdMampH51I24TfWKSkpKBXr15QqVTKkC6a2+OoUaOgUqmUYcOMya1bt5CUlISpU6fC1dUVHTt2VN4bNmwY8ufPjzx58qBIkSJwcXFRLpYNfX9Ux3f9+nUcPHgQJ06cUDpWPHnyJJydnVG2bFkcPnwYW7duRb9+/WBpaWmUI3JERESgXLlyWjf2YmNj0aZNG1hZWSnNuE+fPo0ePXoY9M0iTep1eObMGUyfPh0zZ85UauASEhKwd+9euLq6ws3NDXfu3EFISAicnZ0NsiPFf6OO+e3bt2jevDnc3d1hbW2NOXPmKPNEREQgICBAp0m0IdLcxl6/fq11s3LPnj2wsbHRSszfvn2LGzduICIiQlnvxnJOVMe7detWVK5cWavVUYcOHeDs7IwFCxYoifmsWbOM4lyouQ6PHDmCtWvX4sSJE0pHinfv3k03Md+xY4fBXoNq0izfvXv3oFKp0Lp1a2XapEmTYGdnh27dumH79u3YsWMH/P39UbZsWWXbNPTz4LeESbkB0twB0t55TOvNmzfYsmWLQXdullZ6O/ju3btRq1YtODk5Kc8gqV24cAH16tVTargM+QCheQD8448/cPr0aaV5JQB07doVZmZmOh0SAYbdq6xa2udSDx06hNmzZwNIXYc2NjbKUH05c+ZE/fr1dcarNpaLECC1GamNjQ3Kly+vjJ87bdo0HD58GM7OzvD398fdu3cREhKCokWLGs2F8vsSlnbt2sHc3Bw7d+7UeW/hwoVGtf6A1Oc7PT098fr1a7x48QLTpk1D2bJltRLz8+fP4+DBg9i8ebPRdNajOZqBvb09ypYtC3Nzc3To0AEnT54EkHoMqlmzJgoUKIAiRYqgSpUqBtcJ34e6efMmcuTIodSKq7fhx48fw9XVFf/73/90tmtDT8w1n0HOmzcvPDw8UKtWLVhZWSlN8BMSEnD48GG4ubnBzs4Ojo6OWs1mDZnm7799+3YULVpUGQng6tWryJMnD1xdXREZGYm4uDg8evQIgYGB8PDwMIqER+1///sf6tWrh6pVq2L37t1KKxR1Yv6u1oyGvn2m9fvvv8PMzAwTJ05UjjFqHTt2RIkSJdC3b1/06NEDKpXK4PoXSUvzembgwIFwcnKCg4MDPDw88N133+H+/fsAUhPzH3/8ER4eHhg2bJjWMgx5O9Us39ixY9G+fXsUKlRI5/GKiRMnwt/fH1mzZoWnpycaNGhgdK04vhVMyg2Qekfy9vZGv379dDoiMmaaJ6HHjx8rBz0g9TnIwMBA+Pn5Yf/+/Vqfu3HjhsGfwDQPgIMHD0b58uVha2sLX19fNG/eXHmvR48eyJ49e7o1joackGv+/uoOa+Lj4/HPP//g9evX8PHxwahRowCk3kyqUKECVCqVVvJjTC5cuIDcuXNj4cKFeP78OSIjI9G6dWv4+Phg1qxZOHnyJAoVKoTChQsjX758RjPGrOZ6XLRoEbp3744uXbpoPU/Wrl075MiRI93EHDD8hFXTnj17YGJiopTlXYm5JkO9EPntt9+0ekrfu3cvbGxslJrGVatWwczMDPXq1dPqF+CPP/7AP//8YxQ1V8C7j4Pu7u7o2LGj8shISkoKEhMT4enpif79+3/NED9a2rIdO3YMtra2yvPVFy9ehEqlgkqlUh7bSk5OxqtXr3D48GGj6dhN8zizefNm/PTTT8icOTOqVauGK1euAEgdw9vKygoVKlRA8eLF4enpiQoVKhh8QqBZtsmTJ8PGxgZDhw5FYGAgsmfPjmnTpikVKnv27EGePHl0OsY0Nk+ePIGbm5tWM2dA+1Gmfv36ITAwEFWqVMGlS5e+dogfbeLEiciXLx+OHj0KIPWxJlNTU/j4+Cijp4SHh6Nly5bo1KmTQV+npWfcuHFKp3tHjhzB7NmzYWNjg3r16inzxMTE4ObNm3jx4oXRteL4ljApN2A///wz/Pz88Pr1a32H8lloHsjGjBmDSpUqoXDhwqhUqZLSY+7+/ftRp04d+Pn54cCBAzrLMPTEHPj/sauPHDmCmJgYpUmwZk/I6jvJaZ/dNVSav/uoUaNQo0YNrdqaR48eoVSpUspzgLGxsejWrZvSbNgYrVixAi4uLoiOjla23cjISLRs2RLVq1dHcnIyYmNjsXfvXqO5UNbUv39/2NnZoV+/fujWrRty586NTp06Ke937NgR1tbWWmN2GzrNXrY1t1n1EDfqnvGjo6Mxffp0uLq64rvvvtNLrBk1aNAgWFhYKNtabGwsunTpoozUcPfuXTg5OSEgIAAuLi7w9fVN9xhq6DTXW0xMjNaNhMmTJ+s0I01ISICnp6dOsmCIpk2bptwcSklJQVxcHMaNG4eQkBAAqc/gFixYEB06dED37t2RKVMmo3xcRFPfvn1RtGhRjBgxAq1atYKzszMqV66sjPccHh6O+fPnY/z48Vi3bp3RtFQBUofK7N27t1aP8cOGDYO1tTWmTJmiJOZbtmxBQECAUVy/vMudO3eQL18+pdM2zd7m046M8OrVK32E+FHu378PX19f5bHCnTt3IkeOHOjcuTNcXV1Rq1YtpfIoMjLS6EZyiIuLQ4MGDbRq+BMSErB9+3ZYWFigRYsWynTN9WjM26oxY1JuANQ7d1xcnPI8oNq3VEuuNmrUKOTNmxdr167F06dPUapUKbi4uCjDuO3duxf16tWDq6ur0Yyfq/bmzRs0bdoUy5YtAwDlwKfuVVazc6UpU6YYxYWHpiFDhsDOzg5r1qzRGl81NjYWjo6OqFevHpYsWQI/Pz9UrFjRqDvLWrVqFZycnBAZGQng/y8S7969C5VKhV27dukzvE9y6NAhFClSRGmCuHbtWpibmyvbqVqjRo2MsoYnNjZW6/XmzZtRvnx5pSYESD22/u9//0Pr1q0N/gIkMjISHh4eWLp0KYDUZABIHRv++vXrePHiBVxdXZVHfNTrs2bNmkaZmAOpvf5XqVIFdnZ2+PHHH3Ho0CEkJyejb9++KFeuHDw9PTFgwAB4eHjAxcXFoI+lKSkpePPmDfz9/XHz5k2t965du4bjx48jNjYWnp6eyo2xc+fOIUuWLFCpVDqdnhmLM2fOoECBAlrb4KZNm1CrVi1UqVJF6dsgvY7gDN2WLVugUqmQP39+nZZ9w4YNQ86cOTF16lQ8f/5c6z1DP9a8y4MHD+Dk5ISFCxcq09Rl2bNnD9avX6+v0D7Zzp07ce/ePZw5cwb58+dXWo2pK1RKly6tdePdmNZhUlISKlasiJYtW2pNT0hIwE8//QSVSoXvv/9emW5MZfsWMSnXM83OM+rXrw8nJyd07txZ6WhJ0549e4yyV2e1pKQkPH78GO7u7so4nvv27YOFhYXSdE9N3SmRsR0g4uPj4ebmhh07duj0KpuQkIDZs2drDecGGHaNgObd4KtXr6Jo0aLvjP/KlSsoUqQIXF1d4evrazSdZb3L7du3YWJiotRiqYWHh6N06dI6z8obkxUrVqBq1aoAUp9ntbCwUDpWiomJwe7du5V5jW0fXL58OXLmzIklS5ZoPdNYq1YteHt7a837+vXrdGt7DE10dDSKFCmCrl27Ys2aNVCpVPj777+Vm3xr1qxB1apV8fDhQwCpiU+FChVQr149rbGRDZnm7//LL78gT548mDp1KqZMmQI3Nzd4eXlhw4YNSE5OxoYNG/DDDz+gUaNG6Natm3IMMtRkLu02duLECWzZskUr3itXrsDNzU1JVG/duoWWLVti9OjR+PPPP79+0J/B8ePHkSNHDpw7d05r+ooVK2BpaQkPDw/lGXNjPE+oW7xNmTJFZzSDESNGGOWII4D2ulBvo2/evIGfnx88PT2VVg5qffr0Qe3atY2qhhzQ3eZGjRqFli1bKiMdzJgxA4GBgRg6dKjBHls0vescNmvWLJQtW1ZnqMjp06fj+++/R968edGnT5+vESL9CyblBmDr1q3Inj07Ro0ahR07diAgIAAFCxbE8ePHlXl2794NlUqFSZMmGfTFY1rx8fFaze/Dw8Ph5OSEuLg47Nq1SytpjY2NTXdYDUMtb3pxxcbGon79+vDz80POnDm1epW9d+8eAgMD073hYmgaNGigNPVVO3DgAOzt7ZWaY83ma+qT2OvXr/H48eNv5pmk5cuXI1u2bBg8eDBu3bqFR48eISQkBA4ODkbZZF29za5fvx5NmjTBqlWrkCNHDq2ejnft2oVu3boZzfikaTtIPHv2LPr374/ixYujSpUqGDp0KGJjY3H69Gl4eHgotY5pOy00VOrYrly5AhMTE5iZmSktGtTvLV68GCVLllSe4xwyZAjGjBljlC2tLl26hBEjRmjVvN24cQOtWrWCt7e31vBKmtulMRxrEhMTkZiYiKpVq6Js2bLYvn27UoZDhw5pPeYUEhKCWrVqGU2io7kPqct048YNlC9fHr/++qvWkG7Jyclwc3NDuXLl4Ovrq9XqyhC9rwPBdu3aIXv27FizZo3W8JgAsGDBAqPYLjWp1+OePXvQvXt3+Pj4YPr06fjnn38QFRWFggULwsfHB1OmTMGmTZvw448/wsrKSuknwJj17NkTpUuXVo6bjRs3xi+//KK8b8iJueY2ee3aNZw8eRJPnjxBSkoK7t69i8DAQNSrV0/pZPjFixeoX78+pk6ditDQUJQtW9ZobuB+y5iUf2XqE5P6mcfo6GgEBAQoz8jFxsYib9686N27t85nJ02apNxVNgbr169H48aN4erqijFjxgBIPXBUqFABTZs21WrWDaTWTFarVk3nbp4h0jz53rlzB1FRUcqYsUePHoWZmRmqVauG58+fIzk5GU+fPkVQUBCqVatm0Ad2ILVGvGfPnloXUUDqxXLu3Lm1asrVZVm/fr1OM1lDTuQ+VEpKClauXAkLCwsULFgQxYoVQ4ECBYzisYq0z1Vrunz5MiwsLKBSqTBz5kxl+ps3bxAQEIA2bdoYdKKqlnZIIk2nTp3CzJkzkSdPHlSvXh3NmzeHm5ub8gy2sTlx4gRUKhWyZMmiM3Tk4cOHUbx4cbi7u8PLyws5cuQwyovkU6dOKZ2cLVq0SOu927dvw87ODrNmzdJTdB9PvS+pH0978uQJatSoAQ8PD2zdulU5jv7www9QqVRwdXWFhYWF0XSWpbkfpqSkaJ07goODUbRoUezZs0cp5+PHj9G0aVNMnToV5cqVw4oVK756zB9Ks2zz589Hhw4d0LZtW4SGhirT1SNWpJeYA8Zxw0jTpk2bYGFhgZ9++gljx45FgQIF4OPjg9evX+P+/fto2rQpSpUqBWdnZ1SvXt3gt1PNdfi+669NmzahatWqKFasGCpUqIASJUoYxbBgmrENGjQIjo6OsLOzQ86cOdG9e3c8ePAAf/75Jxo2bIgCBQqgaNGiKFGiBEqVKgUAWLlyJYoWLaoMA0f6w6T8K1q8eDFCQkK0ai/Ud80vXbqE8PBw5MuXT6uzpV27dhnlxdW8efNgaWmJPn36oFevXsicObNyMTVx4kTY2tqiYcOGyvxv3rxROngz5KR1zJgxWsNeDR48GI6OjihcuDBKlCihdIq1ZcsWmJiYwMPDAxUqVEC1atVQvnx5g+9VNiAgACNHjlQO8tOnT0d4eDiA1A5RPD090aJFC62expOSkuDr64sePXroJeavITw8HLt27cL27du1apAN1du3b7VeL126FGPGjMH//vc/3LhxA0DqNpopUyb06NEDGzZswK5du1CrVi2UKVPGKC5ENC+0pkyZgjp16qB27dro0KGDVvmfPn2KcePGoUGDBlCpVHBzc9NHuJ/sxIkTOHToEHbv3g1TU1Od/W3btm0YNGgQevToYTQ3b9MmcwDw66+/InPmzPjpp5/w5s0brW2wTp066Ny581eP81Oo49+xYwfq16+vNOV+/vw5qlWrBnd3d+VG5+vXr7FixQrMnTtX6TfA0KXdDxs2bIhq1aph4MCByrVOUFAQihUrhm7dumH27NmoXr260ldF6dKl0aVLF73EnhEDBgxA3rx5MWDAAIwcORIqlUrpwwFIHavbysoKS5Ys0eqR3Nj8888/cHV1VVr5JScnw9LSEv3799fqhC82NhYPHz40+JYcmtvnxIkT39sXTHJyMjZu3Ijhw4cjJCTE4B+LSWvGjBnInTs3du/ejfv372PmzJmoVq0amjdvjqioKDx+/BjHjx/HiBEjsHDhQmU77datG2rXrq3TpxV9fUzKv6L27dujfPnyCA0NVU5WT58+RcmSJTF27Fg4OzujY8eOygEgIiICLVq0wPr16w364jitBQsWIGvWrFq9Nrds2RLTp0/Hq1evcPfuXXTu3BmlS5dGYGAgOnfuDC8vL5QpU0Y5SBhiLevRo0dRrFgxBAUFISYmBrt27ULu3LmxYcMGrFmzBj/++KPWzYfLly9j6tSpGDZsGH777TeD71W2V69eKFy4sPL68ePHcHNzg729vdL76LZt21CyZEkEBARgzJgxWLx4MWrUqKGVyJF+DRw4EF5eXkpnZ71794a1tbXSZNbExARhYWEAUse4LlWqFOzt7VG1alU0bNjQ4G8cpTV48GDkypULo0ePRvfu3VGiRAkULVpUGcoGSD2eJCcnY+3atUZxw+HfbNiwId3E3JhoHuPXrl2LzZs3K9verFmzoFKpEBoaqjzO9Pr1a5QuXVqnjwdjsHHjRpibm2Po0KE4evSosu2pE/OqVati586dRrPPpWfQoEHInTs3+vfvjwEDBiBnzpzw8fFRHjcICQlBQEAAypYti0aNGinPYPv7+2s1ETZEx48fh7OzM44dOwYgtUbVzMxMaxhJAGjYsCF8fX31EeJH0XwETe3hw4dwdXVFTEwMbt26hfz582tVFJ04cULnEUNDtGHDBuU6OykpCbGxsShSpMg7a/Xfdc1pDPtkSkoKkpKS0KhRI/z8889a761evRqlSpXClClTdD73559/om/fvrC0tNTpJ4D0g0n5V5SSkoK+ffuiSpUqGDNmjNLcedasWcicOTN8fHy05g8JCUGJEiWUmkpjcPDgQahUKmW8arVy5cqhTJkyMDc3R1BQEEaNGoW1a9eifv36aNu2LYYPH65cLBtqcpecnIxVq1ahWrVqCAoKwpAhQzB16lStecaNGweVSqWcvNMy1AN8UlISunXrpozbPHXqVGzfvh23bt1CzZo14eDgoDz3d+DAAXTt2hUFChSAp6cnmjVrZnSJ3LcqKSkJc+fOhbu7Oxo1aoSrV6+iYcOGuHDhAhISEpCYmIgBAwYgW7Zsyk2zR48e4f79+4iKijK6vgD++usvFC1aVOuRl4cPH8LLywsuLi7KtLQXXMZSvvfZuHEjTE1N033UydBpJgIDBgxAgQIFsHjxYqW/CiB1+DCVSoWaNWuia9euqF+/vtaNW2Nx7949FCtWTOeiWF2O58+fo0aNGihZsqTRjuhw7do1ODo6Yu/evcq08PBwFC9eHP7+/sq0hIQEpTYuJSUFQ4cORZ48eXR6pDc069atQ+XKlQGk7nea/XDExMRoHX8MsUIhLXWMmvvSnTt38OzZM1y/fh12dnbYvXs3nJ2d0alTJ+W8fvXqVTRv3lyrpZwhWrBgAQoWLIjx48crNfmRkZGwsbHBxYsX9RvcF6A+njZs2FBpSaR5LdatWzeUKFFCpxn/L7/8Aj8/PybkBoRJ+VeiPvhFRESgefPmKFeuHEJDQ/Hq1Su8evVKGXph4MCBGDZsGDp16gRLS0ujO4DcvHkTXl5eqF+/vnLgbty4MZydnbF69Wrs2LEDLi4uKF++fLqdShhqUqfZe+6KFSvg6+sLc3NzTJgwAYD2yS0oKAgtWrRAUlKSwZYnPaGhobC3t0dQUBBUKpXSTPvOnTuoXr06ChQooCTmycnJeP36tdbQU99CovMtSExMxLJly+Dp6YmqVauicuXKePLkidYJuVu3brC3t8eTJ090Pm/INchpLypu3rwJS0tLpTZOHfuNGzfg6Oio81yyMbl3796/XuBv2rRJOW8Yo8mTJyNv3rzK0Hxq6mPJzJkzoVKp4OXlhW3btinHU2NKzM+fP4/ChQtrbaNpe2N/8uQJAgICtFp3GJPLly8jf/78Ss/x6mfKb926BVNTU2WIULU7d+6gQYMGKFCgAC5cuPDV432f9Pa5PXv2oE6dOpg/f75Ox5gHDhxA69atcevWrfcuw9Dcv38fDRs2REpKCrZs2YJChQopjza1bdsWKpUKTZo00frMkCFD4ObmZvCdnCYnJ6NHjx6oXLkyxo0bh9jYWCQmJqJIkSJKq7+EhARlPRnD+lJ7X18x/fv3R548eXQq8ubMmQMfHx+dfoKSkpJ0huwj/WJS/hWtWrUKPj4+8Pf3h62tLXLmzInx48fjzZs3iI+Px6+//orKlSujRo0aaNeundZwPsbk5s2bCAgIQJ06deDp6YkKFSpoXWycP38eKpVKq3m7MdC8kFq1ahVKly6N4sWLK4mN+kKyffv2OiczQ7Vu3TqtJMzJyQlmZmYYOXKk1nzqxLxgwYLpPlNtyIncf4l6PSQkJGDJkiVwd3eHtbW1cuJVP2t9/vx55M+fH6dPn9ZbrJ9iwIAB6Nu3L169egVnZ2edljnR0dFwcXEx+Gaxamkvls6ePYuCBQvqjICQnq1bt+L69etfKrQv5u3bt6hTpw7Gjh0LAPj777+xefNmNG7cGN9//71yXJ09ezYyZcqk1DQby41O9b544cIF5M6dW6sWWX1RvXPnTuXmtbEkBukd6x88eABzc3NlaNOUlBQkJiYiISEBFSpU0GlRBqQOh2poz81rlm3hwoVK55EXLlyAo6MjVCqV0ikvkNoXTmBgIIKDg43uHLh582a4u7vD1dUVWbNmxapVq5T3jhw5gsDAQBQpUgTbtm3D6tWr0bt3b6PofFDz8aRu3bqhYsWKCA0NxdWrV+Ht7Y1nz57pfCZtHyzGYu/evThw4ADOnDmjTKtatSpKliyJa9eu4enTp3j79i1q1qyJZs2aaX3W2LbX/wom5V/JxYsXYW1tjUWLFiEqKgoJCQlo3bo1ypQpo9XERt20y5hqAtJz8+ZN1KpVC1ZWVli7di2A1IuOlJQUnD9/Hi4uLu9s4m3INBPzdevWwc3NDTVq1FAO9ElJSfD09ES7du30GeYHWblyJcqXL4/k5GQkJiYqzSwbNmwIOzs7LFy4UKtTwjt37sDHxwcmJiZand2RYdFsgr58+XIUKVIEvr6+Wuvy+vXryJcvnzL8kqHTvIDYvXs3SpYsiWPHjiEuLg59+vRB9erVtYYajIuLQ6VKlTB79mw9RJsx//vf/zBmzBitTna2bdumjCP/rUpISMB3332H7777DnPmzEFgYCD8/PzQsGFDVKxYEbVq1VLOg9OnT4eJiYmSwBuq9C507927BxcXF7Rr106ppVPr3LkzfvjhB8TFxRnFRbLmdYl6uCX1TZKQkBAUKFBAayi7+Ph4lClTRqtm2VBp3hS5efMm8ufPDy8vL+XZ97Vr10KlUqFz585YunQptmzZAl9fX5QtW9Zo+6lQj6VeqlQpnU6+jh8/jnbt2iFnzpwoW7YsateubfDNnNMbuq5bt27w9PRE69atoVKpUKtWLeU4o+6UUHMEEkPVvXt3pWUmkDo2vK2tLfLkyaM1wlFkZCS8vLyQK1cuFCtWDOXLl0fp0qWVfdfYttH/GiblX8mWLVtQoEABrZNyUlISWrVqBWtra0yYMEGrGcm3sOPcvn0btWvXRmBgoNbFf926dVGjRg2Drxl4V3yaifnq1atRunRp2NraombNmmjdujVcXFyM5gCoLqO6xlR9gdW5c2fY2tpi0aJFWsncjRs30K1bN6Oprfqv0kzMly1bhgoVKqBKlSo4fPgwdu3ahaCgIFSoUMHo1uPWrVvRqVMnDBo0SJkWHh6O5s2bo0KFCmjWrBmmTJmC6tWro3Tp0kbxSIX6wnjKlClKPyPLly+Hl5eXfgP7jDSPpZr/X7VqFWrXro2cOXNi5MiROHXqFABg9OjRaN68udYyxo8fDxsbm3RrugyBep87e/Ysli5dimnTpik1wTt27ICZmRnatWuHTZs24fTp0+jVqxdy5sxpFC3i1q9fr3V9MmbMGOXxmIULF+LFixeIiopCly5dYG1tjb59+2LcuHGoVauWUeyHmufpsWPHomnTpnBxcYFKpULVqlWVx7RWrlwJX19f5MqVC9WqVUOTJk2Msj+VxMREpKSkYMmSJRg4cCD8/Pzg4+OTbiu4iIgIvHnzxqh6Wf/tt9+wf/9+ZXr37t3h5OQEV1dXDBgwAOPGjUNoaChGjhyJ/v37G/z2GRkZifbt26NEiRKYN28ewsPD4erqiosXL+LkyZMYNWoUChYsiGHDhimf+e233/Drr79i4cKFBt/JMP0/JuVfmPpgv3PnTjg4OCjNDNVjWUZHRyNPnjwoXLgwJk2aZPBJXEapm7IHBQXh6NGjaNy4MYoVK2bQvaynpT5RpTd8j7rG3NPTEzY2Nti9e7cyn7EcAE+ePAmVSoWpU6dqxdylSxfY2tpi8eLFWom5mjFdhHyr3rf/aCbmK1asUB5N+O677zBw4ECju5h89OgRKlasCFNTU7Rs2VLrvX/++QezZ89GtWrV4Ofnh+DgYKMq36RJk6BSqTB58mQAqR0Vubu76zmqz0NzG505cyY6deoEf39/zJ8/H9HR0UhJSdHpX8Tf3x8dO3bUOR8a+vOP69atg7W1NcqXL48iRYoge/bsmDFjBoDUpqaenp7IkycPihYtinLlyhlFnzGLFi1CwYIFMW7cOCQmJmLJkiWwsbHBrFmzULduXVSoUAG9evXCixcvEB0djdmzZ6N06dKoVasWfvjhB6PbDy0sLLBv3z5cvnwZc+fORalSpVCxYkUlMX/69CkePnyIFy9eGF3HmO+6vty4cSO8vb3h4+OjtS+eO3fOKMau1izXwIED4ejoiNGjRyuPwKSkpKB79+6oWrUqJk+erLR+0GTo2+etW7fQt29fuLi4oHXr1ujWrZvyXlRUFCZMmAAHBwcMHjw43c8bevkoFZPyLyC9A19sbCwKFiyIRo0aaU3/+++/ERQUhC5duhhVL+sZcfPmTdSpUwdZs2ZF8eLFlZO0MZzIli1bBmdn53Tf00zMFy5ciJ49expFxyHpbZ8TJkxAtmzZMGPGDK310rVrV+TLlw8zZ85Unq8jwzN//nylB+P0bh4lJSVh+fLlcHFxQb9+/ZT3DXkfVMeuub1evXoVtWvXRpEiRbB69ep0P6f5fLYhlw/QvlAKDQ2FSqXCggULMGHCBNSrVw///PMPrly5gvDwcERGRuLYsWPpds5nDAYMGIA8efJg8uTJ6NOnD4oWLYqGDRsqz3PGxMTgyJEjqF27ttYQi8Zyo/ratWuws7NDWFiY0hQ4JCQEuXLlUsZ8fvz4McLDw/HXX38Z/A0GTb169YKbmxsmTJiAnj174vfff1feCw0NReXKldGzZ0+lD4S05wpD3w+B1IqSpk2barXCSUxMxPbt21G4cGFUq1ZNSeY0t0lj2T7VcR44cAA9evRAcHAwfvnlF+Vmw+bNm1GjRg1Ur14d165dw4gRI+Ds7GxUx5vx48cjd+7cOHfunM61WEpKCnr37o1KlSph0KBBWp3UGjLNc0RUVBR69+6NfPnyoU6dOlrzPXr0CBMmTEDhwoWNeqjM/zom5Z+ZZhO2efPmISwsDEePHgUAnDp1Crlz50b9+vVx7tw53Lx5E0OHDkVAQEC6NZHfkuvXr6NHjx4GP+xZWhEREShUqJAyrnNa6SUOhnxHUjNhU7fWUJs4cSIyZcqkk5g3b94cdevWNZqLj/+a5ORk5MuXD126dEn3fc3O33bs2KFsn4a8PjW306ioKLx580a50L98+TJ8fX1Ru3ZtreQg7THFkMsH/H98msOAhYaGIlOmTChUqBDMzMxQvHhxZM+eHQUKFEChQoWQO3duPHz4UF8hf7QTJ06gePHiSvP0HTt2wNTUVOu4euLECTRu3NhomwQfOnQIxYoVQ3h4uNb2O2jQIFhYWKTbNNjQaZ4j+vTpg4oVK6JAgQLYvXu31nyhoaGoUqUKevbsqfPcvKHvh5rq1q2LgIAAnem9e/eGSqWCt7e3kpgb8o33d1EPo9ikSRM0bdoU2bNnR1BQEK5cuQIg9fGgGjVqIHfu3ChcuLBWB2KGZvjw4VrPuL98+RL169dXRtz4+++/sXXrVtStWxd9+/bFo0ePkJKSguDgYHTo0MEotsu//vpL2d5GjBiBS5cuITw8HL169UKOHDmUVjhqjx49wtChQ9GoUSOjKB/pYlL+Baxfvx7W1tZwc3NDiRIlkCVLFqXHzjNnzsDJyQkODg7Inz8/7O3tce7cOT1H/HUZakKe3ljGr169QosWLZTxu9M70BnLwU8zzhkzZqB169bo2rUrLl26pKyTdyXmmnebyXCo18uCBQvg5eWFO3fupDtf2vVmLMnOiBEjULZsWZQrVw5+fn7K87eXLl2Cr68vAgICsHnzZj1HmXHq9bFt2zY0bdpUq3Os2bNnQ6VSoXv37vjzzz9x9+5dPHz4EP/884/RdLCYdvvas2cPypYtCyD1/GhhYYG5c+cCSG1FtnPnTgCpF6HG9viP2u+//47s2bPj8ePHAKBcTMfFxaFAgQI6w4IZE3XyM3jwYFhbW6NLly54+fKl1jwTJ05EkSJFdMZjN0TpJdQpKSmYM2cO3NzcsGnTJq1teNGiRWjZsiW8vb3Rrl07ozl+aoqIiICLiwumT5+uTPvrr7/g7OyMunXrKi2MoqKicPz4cYO+iXTs2DF8//33OscIDw8P+Pv7K/2mVKtWDS1btvy/9u48rsa8/QP453TaSYhkX0LWpKSy77IMQ9kZjD0kQmVfiqyhBokQSQgRssRME8aSZYiRpbG2WFJpP+dcvz/6nXvOkZlnFpxz6nr/8zzus7y+Z+67731f3+W6qGzZsjRhwgQiUi4ppq7PMzKZjO7evUsikYgCAwNp6tSpZGBgQPfu3SOiogGHmTNnkoWFRbFkpu/evfvkZBHTDByUf2b3798nU1NT2rp1KxUUFFBKSgr98MMPpK2tLZTnycnJoYsXL9KPP/6o1h1fafFxOaKPz0l0dDSJxWIhcYgmUuycV6xYQWXLlqXJkydT1apVydbWlkJCQoTZqdWrV5OOjg6tWLFC6eFDE2cGSpo/OwcJCQlkZmZG27dv/8ot+rwUr9MdO3ZQhQoVaPv27bRy5Urq3bs3lS1bVigtFR8fT927d6dWrVppTBZ5RREREWRgYEBr1qwpVtLM19eXtLS0aOPGjWqfYOmvrF+/nq5cuUKHDh2ijh07CgG5fDk3UVHAPmHCBHr69KlwTN37Gvl1evv2beHak0ql1KpVK+rRo4dSos/Xr19T48aNNWrw6PDhw0LeBjc3N7K2thZmzd3d3ally5bk4+NTbIXf3r171T5gVby2rl27Rj///LMw2Pfu3Tvq3r07de3alUJDQykvL4/S09Opf//+tHz5clqxYgU1adKkWA4EdaXYn6amplK9evXozJkzRPRHJv379++Tvr6+UM5OU8jPY0REBP34449EVFSZw8rKioyNjWn+/PnC3+aqVauof//+SnvJ1b2PISrqP/X09MjQ0JAuXbpERH+c04cPH9LMmTOpUaNGwgCnIg7INRMH5Z/ZhQsXqEmTJsVmNDZu3EgGBgZqvRyoNBo0aBBt2rRJWBobFBRE1tbW5ObmRsnJyUInPnz4cJo+fTrl5uZqRGf+Z+7du0cjR44UblaFhYXUv39/srOzo927dws36oULF1K7du24Y1cTP/30k1It1aNHj9LBgweV3rNkyRJq3rx5seWjmujs2bM0depUpTJn2dnZNG7cODIyMhJ+47Vr15RyOairj+uNP3jwgOrXr09BQUFEVPQAJZPJ6PLly8Jv8fX1JZFIRP7+/mr/++QU2xkUFERisZju379PWVlZQp1nxfJYubm51Lt3bxo6dKjG9DXydkZERAgJpZKSkkgmk9HRo0fJxsaGunbtSklJScLe3KpVq2pMzhiZTEZRUVGkq6tLVlZWVK5cuWIZ4mfMmEE2NjafDMyJ1HcljuI1NnfuXKpRowbVqlWLxGIxjR07lh49ekRv3ryhfv36UbNmzahy5crUtGlTsrCwIKKifqlevXoaNZmya9cuWrlyJX348IEqVKggBHDyUqhERN26daNZs2apspl/m+LsuHygwdnZWVhxKpFIiv2tdenSRSkxmrpTHHDQ1tYmLS0t2rJlS7HVKQ8fPiR3d3cyNjamw4cPq6Kp7DPjoPwz++mnn0gkEtGNGzeI6I8/ridPnlDt2rWV9kAy1Rs/fjzp6+tTcHAwERXNvm3cuJHq169PNjY2NHLkSHr48CGtXr2aGjVqJCTn0ZQHSEU7duygJk2akJWVlZAUjKgowdK3334rBObylQO8BEo9BAYGkkgkotDQUMrOzqZnz56RnZ0dVa5cmfr06UP79++njIwMevDgAbVq1Yqio6OJSH0fjP+X2NhYsrS0pAoVKlB4eDgR/dGPvnnzhlq1akULFy4s9vvUNXBdtGgR9ezZU2l/7o0bN6h+/fr08OFDkkqltHHjRmrXrh2ZmJhQ8+bNlepzJyQkqKrp/9q5c+dox44dtHfvXuFYbGws1ahRgxwdHWn//v0UGhparGSWpvQ10dHRZGhoSJs3by6Wyfn06dPUunVrMjAwoAYNGlC9evUoPj5eRS399/r27UsikYj69esnHFOsU+7m5katW7cmDw8PjUmaJbd582aqXLkyxcbG0vPnzykqKoosLS3J2dmZXr58SZmZmXT16lXy8/Oj0NBQ4fp0cXGhjh07qnUOIPkAHxFRUlISVaxYkXx8fIiIhAGi48ePK32mS5cutHjx4q/d1P/Ezc2NEhIS6NSpU+Tg4EBDhgyhuLg44fXMzEw6c+YMOTo6akziyI/bJpVKKTc3l9auXStUyPn42ktJSaENGzZo7P2eKeOg/D+Q/wFdv36dIiMjKTs7mzIzM6l79+40YsQI+u2334T3ZmVlkaWlpfCQyVRLsfObPXs26ejoUFBQkFJG9cDAQBo4cCBVqVKFpk6dSiKRiObOnauqJv9jHwcpz58/pzZt2pCBgQGFhIQo/TfIysoiJycnqlu3Lp08eVI4rs43sNLExcWFypQpQ3v27CEiovT0dLp//z717duX2rVrRw0aNKCTJ09So0aNimVl1TS5ubm0bNkyMjMzo549ewoP/PK9gN26dSMXFxcVt/LvS01NpTt37hARCUvRHz16RPb29tSlSxeqV68e9evXj+bNm0dXr14lExMTWrVqlSqb/J/cv3+fRCIRiUQipWXqEomEbt26Rfb29tSwYUOyt7en4cOHa1RSN5lMRjk5OeTk5EQeHh5EVPTwf+fOHVq6dKmwRY2oKMv1rVu3NDIxH1HRYODGjRupfPnyNGrUKOG44oqdiRMn0ujRozXuPvHdd9/RuHHjlI6dP3+eatasSQsWLCj2/l9++YXc3NyofPnydOvWra/VzL/tUwOSv/zyCy1fvpxmzpwpHHvw4AGNHz+eqlSpQuvWraP9+/cLM60PHjz4mk3+xxSvsaioKNLR0aFz584REdGxY8eodevWNHToUGGZd2xsLI0ZM4a++eYbjaj4o3gOP3z4UGx11fLly4WVU/LAfPTo0cK9hUgz+lD21zgo/5fkHcShQ4eocuXKtHTpUmFfYHBwMLVt25aGDBlCFy9epEePHpGnpyeZmZkp7ZtjqqXYgc2aNYt0dHRox44dxUYiDxw4QHPnzqUqVapQy5Ythdc15UHk3Llzwg03NTWV7O3tqW3btsKMqlxmZiZ5eXlxx65GFM/FlClTSE9Pj/bs2SNstygsLKTHjx/TjBkzqFOnTlSzZk0SiUTCvmtNuUbl5A8m+fn5tGLFCrKysqIpU6YIxyUSCbVu3Zrc3d1V2cy/TfEhMCYmhtq2bSvMfB87doy8vLzI29tbabllly5daN++fV+9rZ9LQUEBRUREULVq1Wjw4MHCccVzmJKSQu/fv9e4Os9yw4YNo379+tH9+/dp4sSJ1LVrV2rWrBmZmJiQk5OTqpv3j/3VKpNjx45RuXLllAJzIhKqymjSiiqJREIymYwGDBhA3333HREVXa/yfnbNmjVUtWpVev/+vdJ/k+DgYOrcubOQpVydyNv57NkzCgoKog0bNtD+/fvJ1dWVypQpQ998843S+x8+fEg+Pj5kZmZGzZs3J3t7e7p586YKWv7vhIeH08KFC4slOJMH5sOGDRNWqspXIhGpdx+jeK2tWLGCOnfuTDVq1KDp06crDQJ5e3uTjo4OjR49mhwcHKhu3bpq/bvYP8dB+X9w8eJFMjY2psDAwGLlpXbu3Em9e/cmkUhETZs2pTp16ggdBVOtP3sAmTFjBunq6tKOHTs+uRzvyZMnZGxsXKwMhbpR/H3yMj0zZswQBoRevXpFtra21KFDh2KlbeQ4MFcfiktGJ02aRAYGBrR3795i12hCQgJFRkZS1apVaeLEiV+7mf/Yn11j8us3Ly+Pli1bRubm5tSsWTP67rvvaNCgQdSwYUO1fhD5s/4lJSWFjIyMqEOHDvTw4cNir+fl5dGiRYuoWrVqf5pFX1MUFBTQwYMHydDQUGlVw6fOm7oHc/L2JSQk0JUrV4ioaPlz+/btSUtLi5ydnSk8PJzy8/Np48aN1KlTJ6XZZHWneL2Gh4fTmjVraMmSJcL+ValUSlFRUVS+fHkaMmQIJSQkkKOjI3Xv3l2tA3LFLNsf27hxI4nFYmFbgWIVizZt2hR7niMitVyyLm/37du3qU6dOtSyZUsqV64cVa5cmZycnGjKlClkYGBAP/30U7HPpqenU2Zmplr+rj9z9+5dsra2JgMDA9qwYQMRKd8fjx8/Tvb29tS9e3chUzmR+m5t+tiCBQvIzMyM/P39KTo6mipWrEhOTk5KSYa3bNkirPTQpFVG7O/hoPxfkN+AFi1aVGwUUrGDkEgkdO3aNbp582axpShMNRQ75+vXr1NcXJww4k+kHJgr7hWUn1dXV1caO3bs12vwP6T4cLRhwwaaM2cOmZqaUtmyZcnd3Z2ePHlCREXlUezs7Khz584alRW4tFC8Tj9+uBw/frwQmCs+/MvP/dmzZ8nU1FSt9yIr/p6EhASlflPx9by8PPL19aVatWqRvb290h5ldQzM5e1OTU2lq1ev0oULF5ReT01NperVq1O7du2Uzs/hw4eFZaUlZfBWIpHQgQMHyMDAQKOSLCn6OKnb6tWr6c2bNySRSOjx48dK9w6iokGzAQMGFKvooa4U/w49PDyoZs2a1LlzZ7K3t6caNWrQtWvXiKjov0NMTAxVqVKFLCwsyNbWttjfrDo7deoUHTp0iK5duyb8ZicnJ6pYsSLFxsbS27dvKTMzk3r06FGsxrM6DjgQKQfkhoaG5OHhQe/evaNLly7R2LFjqUaNGhQQEED9+/cnS0tLYb+1TCbT2CBOKpVSaGgotWjRgpo0aUKvX78mIuXn7gMHDtD48eM1JhCXO336NDVq1EjoU65cuUK6urpUpUoV6tixo5BhnoiUnk3V8T7I/j0Oyv+DCRMmUK9evYio+EhcfHy8sMSUqQfFm6unpyc1a9aMateuTba2ttSzZ0/hNXd3d9LT06Pg4OBi59DR0ZH69eunljc1xd/n7e1N5cqVo8jISIqNjSVXV1dq2LAhzZo1S1gq++rVK6pTp47GPjCXVIp9SWBgII0dO5ZGjRpFy5YtE47LZ8xDQ0OLBeaPHj2iJk2aqO2SxOjoaBo9ejQREU2fPp26d+9OmZmZxd6nuJR96dKl1KVLF3JzcxOOq9tDl7w9v/76K1lZWVG9evXIxMREqW8hKpoxr169OrVv314IzCMjI8nLy0spD4k6+1RCok+RSCR08OBBMjIyohEjRnyNpn12p06dojJlylBAQMAnr1OZTEYPHz6kWbNmUfny5dVyifP/EhAQQFWqVBFmjiMiIkgkElGNGjWUBh7S09OVKgSoY0Dg4eGhlPvFzc2NzMzMqGLFimRtbU0eHh4klUopNTWVRo4cSTo6OtSgQQNq3LgxWVpaKpWzU3fPnj2jSpUq0aBBg5SOHz58mMqXL083b96kK1eukLOzM7Vo0YIuXrxIRJrx2z6muCrjwIEDZGtrS46OjvTmzRsiok8OEqnbPeKv/PLLL0IOjujoaKpQoQLt3buXkpKSyMDAgAYMGKCU74dIM88j+2sclP8HK1asIGNjY2EWXP4HkpWVRXPmzBHqQTL1snbtWjIxMaHLly9TXl4eLVmyhEQiEZ0/f154z6xZs0gkElFUVJRw7MmTJ2RpaSmU3lAXH9dozszMJHt7eyHjqtzSpUvJ1NSU3N3dhcD89evXajnAwIpK9piZmZGXl5eQ5EVxX+eUKVOobNmytG3bNqWZuR9++IFEIpFalkYrKCigDRs2UNOmTcna2prKly//lwmGPl7K3rZtWxo3bpzaDXjK23nr1i0qU6YMzZ07l65duyZkzfX09CQiEs6TPDDv1KmTkItEU2ZXFR8Eg4ODhTKff/aAKJFIaPfu3dSlSxeNekiWyWSUnZ1Nffv2FZK6ZWVlUWJiIq1du1ZYPnvt2jUaO3YsWVpaqmUSsE85f/48LVq0iNzd3Wnnzp20YcMGYRVKZGQkGRkZkb+/P/Xt25dq1qwpJM9SpI7nMiMjg8aMGUP29va0YsUKunnzJrVv357i4+Pp8ePHtHjxYmrVqhVNnz5daP+JEydoz549tG/fPuFeqI6DDZ+SlJREtra21K9fP6XBk4sXL5KRkZEwyBIbG0tDhgyh2rVr0y+//KKq5v5tf3ZtKSbiDQsLo7Zt21Lv3r2FwFxTztunfl9mZiYlJydTVlYWdevWjZYvX04ymYwKCwvJysqKxGIxzZkzRwWtZV8TB+X/grxjeP/+PdnZ2VHjxo3p5cuXJJPJqKCggObNm0c1atTgpG5qSCKR0KhRo4QSaJGRkVSuXDnatm0bEZHSTMimTZuUOvnCwsJPzpSokpeXF33//fdKD8R5eXnUoUMHmjdvHhEp36gGDBhApqamNHfuXKVaqxyYq5dLly5RgwYNhCWHR44cEUowKRo8eDB17txZ+LdUKqVTp04pZWRVNzKZjHr16kUikYgGDhwoHP87e8w9PT2pe/fuarkd6OHDh6Svr6+Uvfn169dkYmJSLEkWEVFycjLp6emRo6OjxiwFVnyYvHHjBjk4OJC1tbUw4/9ngbni59QxmPu4TYr/Hjx4MI0bN47u3LlDLi4u1LVrV6pbty7VqFGDvv/+eyIiunDhAr18+fKrtvnfCgoKIlNTU+rQoQNVr16djI2NadiwYZSXl0cJCQnUoEED8vf3J6KiGVeRSEQ6Ojp0+/ZtFbf870lNTSU3Nzdq3749DR8+nCZMmCC8lpWVRb6+vmRjY0NTp079ZJ+jaffCxMREcnR0pB49etC9e/coMzOTTE1Nafbs2Urvi4mJodGjR2tUvgr5M8rH27nkx8LCwqhDhw7UunVrjdkbr/hbHj9+TCkpKZSeni4ce/v2LbVo0YJ27dpFRETZ2dk0adIk+vnnnzXu2mT/HAfl/9Evv/xCHTt2pHLlypGDgwN16NCBKleuXGL2BWq6jx8S8/PzqWXLlhQSEkLR0dFUtmxZIdApLCyktWvXUlhYmNJn1Hn09fbt20L7FGccR48eTQ0aNKC3b98S0R83gjlz5pC9vT1ZWVnRjh07iIiXQKmDj4OCiIgIsrGxIaKigLxs2bK0detWIioaODp69Gixz2rCeZRIJJSRkUErV64kLy8vsra2FgIbok8vQSRSXsou30eoTqRSKXl5eVHlypXJz89POO7r60sikYhsbW1p8eLFtGzZMkpOThYewlJTUykxMVE1jf6HFK+vZcuWkZOTE1lbW5OOjg7Z2toK2yU04Tr8lPv375Onpyc9fvxY6eF30aJFZG9vT2KxmJydnWnfvn2UmZlJixcvpr59+6qwxf9cUFAQ6erq0sGDB6mwsJB+/fVXGj58ONWpU4devHhB4eHh1L59e0pOTiaiomW0M2bMoMWLF6v1fZBIOZjOyMigWbNmUc2aNalNmzZK78vKyqJVq1aRnZ0dDR8+XC0Hif6pxMRE6tWrF3Xs2JEqVKhAbm5uwmuK501xL7K6CwkJofr163/yNcXAfMeOHTR58mSNO49eXl5Up04dqlu3LjVq1IiOHz9OBQUF9Pr1a7KwsKDBgwfT5s2bqUePHmRrayv8Zg7MSzYOyv+BP3vYKCwspC1bttCSJUto7dq19OjRo6/cMva/LFu2jIKCgkgikdDUqVOpW7duZGxsTFu2bBHe8+rVK+rTp4/SMU1x8OBBatGiBR08eJCIikZXGzVqRA4ODvTixQv68OEDyWQycnZ2plOnTtF3331HFhYWGncjK4kU+xX5zHhMTAz17t2bAgMDlQJyoqJZuVGjRikFc+p8Hv+sbXl5ebRp0yZq3ry5UmBORHT16tU/Tf6mrl6+fEkzZswgOzs72rx5M61atYoqVqxIK1asoBMnTpCrqyvZ2dlRzZo1qWHDhsLqHE2zadMmKlOmDJ0/f56ePXtGO3fupC5dulCrVq2E5duaFpjn5+eTra0tiUQiql+/Prm5uSmVpUtMTBTKDMqvwwkTJtCgQYMoPz9fI37vhQsXSCQS0dKlS4noj3O0c+dOqlSpEv3+++/k7+9PZcuWpWfPnlFqaip988035OrqKnyHugYEirP4S5YsoVOnTtG7d++EwHzJkiVK78/KyqL58+fTuHHj1L5f+bsSExOpS5cuVLt2baVs6zKZTCOuz4+9fPmSateuLcwYf+xTmf/V+VwqtvPw4cNUqVIlioiIoPDwcJoyZQqJxWIKCAggoqLngEaNGlGrVq2oW7duGpXngP03HJT/CfnFn5qaqlRDlqm/kJAQoROTSqWUmZlJNjY2QsATGxtLhoaG5ODgIJQmSk5Opt69e1ObNm3U9sFD0cc3nwsXLlD//v2pS5cuFBERQUREv/32GzVr1oxq1KhBdnZ21Lx5czI3Nyciot27d1Pz5s01auS8pPn4Brt//35q0qQJERVlJK9Xrx6JRCJatWqV8J6cnBzq1asXjRgxQiNu0IrX6bZt22jatGk0fPhwioiIIKlUSnl5eRQQEEBWVlY0cuRISk1Npe7duyvVt9YkycnJNG3aNLKwsCBtbW2lUjZyERERNHfuXLXeYvBnCgsLadiwYTR58mSl41FRUWRtbU2tW7f+n0vZ1dXq1atp/fr1dPbsWVq8eDEZGxvT4MGDacuWLUrXcVJSErm7u1P58uU16hwmJiZS+/btqX///kpB26ZNm6hatWqUkpJCOTk55ODgQDo6OkIpQnXfWvHkyRMSiUS0ZMkSmj59OpUvX164Bl+/fk1ubm5kb29P3t7eSp/LyclRmnEtCR4+fEiOjo7Us2dP4XlHE3z837+wsJCysrJo6NChNH78eCL6dH+iaX0MUVHZQU9PTyEnhdyKFStIJBIJ5+3Nmzf07t074Teq+0oV9nlwUP4XIiIiqGHDhlSrVi3q0qULXbt27X8GbJrYSZQkP//8s5BYSd6JFRYWkoWFhVJJiejoaKpcuTLZ2NhQ48aNycHBgWxsbDSi7qPiDez06dOUlZVFRESXL1+mgQMHUocOHZTKnK1fv56WLl1KPj4+wn+T77//nnr27KlR9XRLEicnJ3J3d1c6l1u3bqWuXbsK/z5x4gRpaWnR999/Tzt27KCjR49S165dqXnz5sJ51JT+Zvbs2VS5cmUaPHgw9e3bl7S0tMjV1ZXS0tIoJyeHtm3bRhYWFlStWjVq3bq12gcCfyUlJYVcXV3J0tKS1q5dKxxXrH2sKeftU1sjJkyYQF27di1Wy3nBggUkEonI3t5eY/YfK7pw4QIZGxsLZcBevXpFS5YsIT09PWrdujVt3ryZ1q5dS7Nnz1br6gZ/RXH/cWJiIsXExJC+vr6wwoqoaNVASEgIhYeHa0zisxMnTpCurq5ScjN521NTU2nGjBlC8rePacrf4t+VmJhIffv2JXt7e7p8+bKqm/OXPk5uqZjnhqjoOU0sFn9ycFMT3b17l2xsbEhfX18YbFe81/Xu3ZuGDh1KhYWFGrMCgH1eHJR/RP6HcOvWLapRowb5+PjQ4cOHqWXLltSsWTOKjo5W+xtUaSXvuMLDw0lXV5c8PDyEG7OFhYWQKEN+7Pbt2xQaGkrLly+ngwcPasQDiGJHPX/+fKpVqxYFBQUJbb506ZIQmB86dKjY55OSkmjatGlUsWJFjZrlKWn8/f1JW1ubli5dKjyYrFy5kpycnIjoj/N85MgR6t69O1WqVInatWtHgwYN0oiBI0U//fQTVa1aVcjSTVT0N1qxYkXy8vIioqJZq5cvX1JMTIxG/B3+L/IZczs7O/L19RWOa8o5+5hiZueAgABq2LAhRUZGKg3qhYaG0oABA8jZ2ZnGjBlDHz58UEVT/5PZs2fTiBEjhN81ZMgQatSoEY0dO5Z69uxJYrGYJk2aRKmpqSpu6b8n338szwcgz7peWFj4yYd/db9mpVIpnTx5kkQiEYnFYlqyZInQp8p/T2pqKs2cOfMvl0OXJPfv3ydnZ2e1TjY8aNAg2rRpk1BJIygoiKytrcnNzY2Sk5OFVXzDhw+n6dOnU25ursYHp4p11i0sLIT8KIqTJfJnAFY6cVD+CTdu3KCQkBAhe7Vc+/btqVmzZnT69GmNfmAsiQYNGqQ0c7F//37S1tamWbNmUXJyMrVs2VIpCJWfv49nitX9AURu4cKFVLlyZYqLixPKgchdvXqVBg4cSJ07d1baF/n69Wvavn07tW7dWmNK95Q0CQkJwoPFrl27SEtLixYtWkRERPPmzaMhQ4YQkfLIeE5ODqWkpFBGRoZGLmWLjo4mc3NzevXqFUkkEuE37N69m3R0dD55LWrK3+FfkQfmbdu2Fc6xJrp+/TqJRCKlrPK9evWiBg0aUGhoKD19+pTev39P/fr1I19fX1q3bh2ZmpqqZUm+/+XgwYPk4OBAEomExo0bR1WqVKG7d+8SUVEizS1btghLozWZfP9xs2bNlEpkacqs8afa+fr1azpy5AiJxWKaN28eFRQUKL0vMzOT/P39S0Tf8neoe4nF8ePHk76+vlAJJz4+njZu3Ej169cnGxsbGjlyJD18+JBWr15NjRo1onfv3hGR5lyjH1PcA3/w4EFq1aoVderUSUjGK5FIqG3btjR27FhVNpOpGAflCuSzqLVq1SKRSFRsX6NUKqUOHTqQlZUVHTt2TKMejEu6ESNGCDcheecXFhZGYrGYRo4cSbVr1yYTExPq2rUrtW/fnlq0aEE2Nja0bNkyVTb7X3n58iXZ29vTkSNHiKhoFiA+Pp7c3d3p5MmTVFhYSPHx8dSpUyeaOnWq0mczMjKUym+wr2fUqFH07bffKh3bsWMHaWlpUUBAAC1ZsoR69uxJsbGxdOzYMbp48aIwQCjfokCk3g8ln2rb2bNnSVtbWxgUky97Tk9Pp9q1a39yRUdJkZycTGPGjKFu3boVGzzTFJmZmbRx40bS1dVVCsydnJyoefPmVLFiRWrcuDE1bNiQiIoers3NzenJkyeqavJ/0qFDB9LS0qJq1aqV6MFL+f5jR0dHjd1//Pbt22Kl6Pbu3UtisZgWLlwo9DXff/+90hLo0hKYqyPFe8Ts2bNJR0eHgoKClPb3BwYG0sCBA6lKlSo0depUEolENHfuXFU1+R/7O3XW9+/fT82aNaPKlStT586dacSIEdS4cWNO6lbKcVCuQJ7QLSMjg2xsbKhWrVp09erVYjVWLS0tqU2bNhq5PK+k+fjm6u/vT+fPnxeOh4eHk6GhITVs2JA2btxIwcHBtG3bNtqwYQNt27ZNIwdWnj59SsbGxrRnzx66dOkSjRkzhlq2bEn169cnU1NTCg8PJ6Ki5fmaVC6rpJNKpcLA0cuXL4VzExQURGKxmPT09KhRo0ZkZWVFVapUoYYNG5KFhQXZ2tpqxLI9xTZ+vAKlb9++1KJFC6UauampqdSgQQOKior6am1UhZSUFLWsq/5PfPjwgfz9/YVZSLmff/6Z9u7dSwcOHBD63KlTp5Ktra3GDf7J+8gTJ05Qw4YNhUHPktx3JiYmUp8+fahVq1YakQdAsY/x8fGh1q1bk7m5OfXq1YsePnwoXIN79+4lkUhEffr0odatW5OFhYVG3utLKsXntlmzZpGOjg7t2LGjWK3xAwcO0Ny5c6lKlSrUsmVL4XVN+Zv8X3XWDx48SPb29sXug3ytll4clP+/Z8+ekZaWlrDcNyMjg8zNzZXqr8pJpVLOyK5m5B2dPDFfXFyc0PFHRESQjo7On86Ka2IH6OHhQYaGhlSmTBmaNWsWnTx5koiIOnbsSDNmzFB6ryYEdCWd4kNIYGAgmZmZ0S+//CKcm7CwMNLW1iZXV1d6//49ZWRk0IcPH+j9+/efLP2ibhTb5ufnR/3796cJEyYIZaSuXr1K3bt3p7p161JYWBiFhYVR7969ydrammet1ITiOVy3bh2tX79e6XV5YC4SiYplsiYqKuMzYcIEqlixokbPMKekpFD9+vWVVgWUZPfu3aNZs2Zp1H1iwYIFVLVqVdq2bRvdvHmTqlevTl26dFG678fExNCIESPIzc1N4/JwlFR/do3NmDGDdHV1aceOHZ+c7Hry5AkZGxvTpk2bvnQTP5u/W2c9NDSUOnfuTL169RIGHTTpb5F9XqU2KP/4QTc7O5smT55Mzs7OQv1feWCuKaPIpZniA2Xbtm3J3NxcSOxGRLRv3z7S09OjCRMmqP1eq7/r2rVrxa7Lzp070+rVq1XUIvZ3SCQSatKkCTVv3pyuXLki3ICDg4NJLBbT8uXLiwXg6nyTVmzr6tWrqVy5cjRr1iyysLCgNm3aCLVX7927R2PHjqXKlStTy5YtqU+fPvywrCYUr6+7d++Sp6cnGRgYFKulnpGRQQMGDBAqXCg6c+YMde/enX799dev0uYvac+ePVSmTBm6cuWKqpvyValzPyN34cIFsrS0pAsXLhBRUfBdtmxZMjMzo4YNG1JcXJxwj1esEKCJg+8lieK1df36dYqLi1NKIKkYmCuWapXfI1xdXTVqv/XfrbMulUpp37591LFjR2rTpo2wx5yVTqU+KJcH4ERF5aVatmxJe/bsEY5lZmZSo0aNqH79+pytWg0pdvSKN107O7tigXlwcDB16NBBrWcc/40PHz7QjRs3qG/fvkrlsph6+NQ1KpFIyNLSkpo0aVIsMBeJRELyG00SHx9PEydOFPZupqen07hx48je3l5phuPZs2cam7SupPP09KRJkybRzZs3adGiRWRkZERbt25Veo+Xlxd17NiROnfuTDKZTKk/lWdS1nQvXrygTp06FSvRxFTvypUrtHnzZiIqGggyMTGhnTt3Um5uLlWvXp06depEMTExGjHAUFoo9hGenp7UrFkzql27Ntna2lLPnj2F19zd3UlPT4+Cg4OL9SWOjo7Ur18/tRzA/a911qVSKe3cuZMcHR01MkEm+3xKbVBOVNS5i0QiGjVqFN24cYOIiJYtW0YmJiZKe1syMjLI2tpaYxPXlDTnzp2j5cuXC//+q8C8fv36Skva5NQ5MP8nDxMymYxOnDhBXbp0oa5du/LMoxpRHPD7J4F5VFSUxgWq+/fvJ2tra2rSpAk9ePBAOJ6Wlkbjxo0jBwcHWrduXbHP8YOzain2g5cuXaLGjRsLpeuSk5Np/vz5VK5cOWHG/MOHDzRkyBAKCwv75HeUJB/nRWBf36f6B6lUSs+fP6f8/Hzq2bMnzZ8/n4iKrs327duTSCSioUOHfu2msr9h7dq1ZGJiQpcvX6a8vDxasmQJiUQiOn/+vPCeWbNmkUgkUtpj/eTJE7K0tKTr16+rotl/6nPUWVcMzD/eU89Kn1IdlMtLvZQrV46mTJlCs2fPpidPnpCTkxMNGzZMKbApqQ8emiYvL48mTpxIzZs3V1qmrXjzlgemREVL2Q0MDDRm+4Hi7zhz5gyFhIRQYmKi0jK8j6WlpdGPP/5YIuo7lxTx8fHUrFkzWrVqlXDsU4G5VColKysratasGcXGxir1M5p0Hh8+fEh9+vQhIyMj2rhxo9Jrr1+/pokTJ1KDBg1o//79Kmoh+yvr168nNzc3cnFxUTr+6tUr4cHZxsaGGjduTJaWlsK1yfdF9qUo9pcJCQn07NkzpVnEt2/fUvPmzYXlwfn5+TRhwgT6/fffeVBaDUkkEho1apSwCiwyMlJpwC8zM1N476ZNm5Tuf4WFhUqvq4PPWWed+1Emp4VShIgAANnZ2ZBIJLCxscGOHTvQsmVL1K1bFxKJBPb29hCLxXj48CHOnj2r4hazj+np6WHx4sXo0qULIiIisGrVKgCAlpYWZDIZAEBHR0f4/3FxcRg/fjyaNm2qsjb/E1paRX+Sc+fOxZAhQzBv3jzY29tjw4YNePny5Sc/U7lyZXTs2BFisRhSqRTa2tpfs8nsI4mJiWjWrBlat26NyMhI+Pn5AVC+RrW1tSGVSqGlpYX4+Hh8+PABP/zwA0QikfA9mnQe69evjy1btqBTp06IiIhAWFiY8FqlSpWwbNkyfP/993B2dlZhKxnwx31Q8f/fvHkTGzduxPXr1/H+/Xvh9apVq8LT0xMXLlxAp06dMGrUKMTHxwvXr+L1ytjnJL8Xenh4oHfv3mjdujW6deuGzZs3AwDKly8PkUiEoKAg+Pv7o1evXrh27Rpq1qwp3AuZ6ij2MwAglUpx9+5daGtr4/Tp0xgxYgR8fX0xYcIESCQSbNu2Dfv37wcATJ8+Hdra2pBIJACK7oVGRkZf/Tf8FWNjY8ydOxfh4eEAAGtra4wePRpRUVHo27cvJk6ciEePHsHKygpnz55Fbm4utLS0iv13AcD9KPuDSocEVODixYvk4OBAW7ZsoYyMDHrz5g1NmDCBAgMDiYjohx9+oGbNmgnL2nnEVT0lJyfTtGnTyM7Ojnx9fYXj8pHIlJQUGjRoEB09elR4TZ3PpWLb4uLiqF27dhQXF0c5OTm0ePFisrCwoAULFtCLFy9U2Er2v7x69YoqVapEp0+fptevX9PkyZPJzs5OKZO14rlOTk6mixcvFjuuqZ48eUJ9+vShzp07C5UsPlYSfqcmKygooJycHHrx4oXSqqI5c+aQSCSiwMBApURLn5rF0aRVHEyzKF5vkZGRVLVqVTp16hRFRETQ8uXLSSwWC5nxnz17RlZWVmRvb0+Ojo7C9czbYtTHsmXLKCgoiCQSCU2dOpW6detGxsbGtGXLFuE9r169oj59+igdU1eloc46U51SF5Tn5OTQ4MGDqXv37tShQwdKSkqiDRs2UNu2bYX9HNevXydvb2+6f/++ilvL/opiYL5y5Urh+KtXr6ht27bUoEEDtX94lOcykNuxYwe5uLjQlClTlI57e3uThYUFLVy4kANzNZabm0vu7u5CZuqkpCSaNGlSscCcqGjgqE2bNuTo6Cjc0EtCwPrkyRPq27cvdevWjbZv367q5jAFp0+fpilTplDNmjWpfPny1KdPH6VEbpMmTSJ9fX3as2cP76lmKnXs2DEaP358sfJ7u3fvJpFIJOQ1kEqllJ6ezokj1UBISIjSwEhmZibZ2NhQXFwcERHFxsaSoaEhOTg40MOHD4mo6Dmud+/e1KZNG425/5WWOuvs6yvxQbnixS9PyiCRSCguLo6+/fZbqlSpEgUEBJCZmRmNGDFCeK+mdA6lnWJgvnr1anrz5g117tyZmjRpovZJzyZOnEizZs0ioj+u05EjR5JIJCIHBwdKT09Xer+Pjw81bdqUXF1dKS0t7Ws3l/1NJ06coKpVqwrllD4VmL9//57at29PDRs2VJqtLCmePHlC9vb2NH36dFU3hf2/HTt2UPXq1WnWrFm0atUqCgwMpGbNmlHVqlVpzpw5wvsmTZpEhoaGtHfvXqUZc8a+lI+DlISEBLK3tydjY2MhkZtMJiOpVEpSqZRGjhxJQ4cOpdzcXKXPcrCjOj///LNQKlE+MFJYWEgWFhb0448/Cu+Ljo6mypUrCzkqHBwcyMbGRu2f14hKV511pholOiiXd9BnzpyhiRMnUq9evWj58uVKWZHXrVtHffr0oUaNGpFIJFJa7sw0Q3JyMk2fPp3s7OzI2NiYGjduLHTw6jxqfuXKFaGdipn95SOrGzZsoHfv3il9xtPTk4YNG8YPH2pu5syZ1KFDB2Hw5OnTpzR58mRycHCgZcuWUadOnTTmOv23Xr16xctI1cTWrVtJV1eXwsLClAaBEhMTafTo0VSlShWlpITTpk0jkUhE0dHRqmguY3To0CGytbWl2rVrCxUB5FxdXalr164qahn7mLyfDw8PJ11dXfLw8BCCawsLC6E0rfzY7du3KTQ0lJYvX04HDx7UiCS1pa3OOlONEh2UExEdPXqU9PT0aOTIkfT999+Tqakp9erViyIjI4X3xMfH06pVq6h27dpc9kxDJScn03fffUe9evXSuEBn586d1LVrVzp9+rRwbMqUKVSvXj3y9/cvNmMuD8g5MFe9j8+B/OHi5s2bNGDAAKWyLs+ePSMXFxfS09Ojpk2batx1+m9xYK5aR44cIZFIRMeOHSMi5ZJ8RESPHj2iTp06ka2tLf3+++/C59auXVvir02mWu3ataMzZ84oHVPsL44ePUodO3aknj17CuWwPnz4QB06dKCRI0d+1bayTxs0aBDdvHlT+Pf+/ftJW1ubZs2aRcnJydSyZUu6c+eO8Lq8T/l4e4w6z5CX9DrrTH2U6KA8NTWVrK2tlfZy/vbbb9SlSxfq3bu3Uj1dIir2R8Q0y7t374QbuiY9TEZHR1Pr1q3J2dlZ6QFlypQpZG5uTj/88AO9fftW6TMckKue4sOjVCotdrMdO3YsdenSRenY77//rhTsaNJ1yjRPXl4eTZ48mczNzcnf3184Lr9W5f3I6dOnSUtLS2nmR46vUfYljBo1ilq0aPHJ60vx/nbw4EFycHCgMmXKUIcOHWjIkCHUsmVLYTsi3wtVa8SIEcXORVhYGInFYho5ciTVrl2bTExMqGvXrtS+fXtq0aIF2djY0LJly1TZ7H+lpNVZZ+pHc2ru/Au6urrIzs5GlSpVAAASiQQWFhbYvHkzHBwcEB0djYYNGwrvNzQ0VFVT2WdQoUIFAIBMJlPbclIymUwo9SLXs2dPaGtrY9GiRdi6dSsAoHv37ti8eTOmTZuGuXPnwszMDAMHDhQ+wyU0VIuIhPO4Zs0a3Lx5ExkZGfD29kbLli0BAP7+/rC1tYWfnx9mzpwJAKhduzbc3d0BFPVH6nqdspJBT08PixYtgp6eHvbu3Yvs7Gx4eHhALBZDJpMJ/UidOnWE++XH+Bpln1tOTg6SkpIwbNgwaGtrY8eOHWjVqhVatGgBoOj+RkQQiURwdnaGnp4eVqxYgezsbHTt2lUonVVYWAgdHR1V/pRSSyqVQiwWY+/evQCAgIAANG3aFB06dMDQoUOhpaWFsWPHokaNGpg1axaMjIwgkUiQk5MDQ0NDjB07VsW/4J+RSqW4ffs21qxZA3t7exw7dgzr169HYGAgOnfujKysLBgZGWHdunWoU6cOevbsKXy2Zs2aiIuLU7uybkz9lLi7rbwjl8lkyM/PR35+Pp4/fy68Lg/MO3XqhPj4eBW2lH0pHwe96kIxIH/48CEyMzNRr149GBsbo2vXrpDJZFiyZAm2bt0KkUiEbt26ISAgAHXr1kX//v1V3Homp3gefXx8sGHDBgwaNAgvXrxA+/btsX37dgwYMABlypTBxIkTcfPmTbx9+xYmJiZK38PBDvsa5LXGfXx8cOTIEQBFtZ+1tLSEgaE7d+7AxsYGTZo0UXFrWWlARGjSpAkuXLiAy5cv48qVK7h69arSexQD82+++QZ5eXnYuXMnoqKi0LZtWzRp0oQDchUSi8UA/njm9vf3R15eHvbt2wd7e3sMHjwY2traGDp0KDIyMuDq6lrsO9R5YFr+u+Tkdda7d+8u1FlfvXq1Up316tWrY+jQoZg+fTqAP36fOtZZZ+pJPaOXf4GIAEAY6SciVKlSBdOnT8f8+fNx/Phx4Y8DAHJzc2FmZqay9rLSRTGQW7BgAZycnNC+fXsMHToUa9euRWFhIbp3744lS5YgNTUVgYGBiIqKAgC4u7tDLBZDKpWq8iew/yc/j69evcKbN29w5MgRbN68GbGxsZgwYQLGjh2LiIgIAEWrIGJjY3Hs2DFVNpmVcmZmZpg/fz5sbW1x5MgRrFq1CkDRwFBWVhaCg4PRqFEj1KhRQ8UtZaVBmTJl4OHhgcTERERHR8PLyws1a9YE8MezHPBHYA4AgwYNwtixY5GXlwcXFxfcuXNHJW1nn/bgwQPUrFkTo0ePxuXLlyGVSjFw4EDs3r0bPj4+mDhxIgoKCpQ+o64BOfDHasTly5dj+/btEIvFaNOmDUJCQjBkyBCsWbMGU6ZMAQC8fv0aFy5cwPv375W+Q51/H1NTKlo2/0WcOHGCHB0dqU+fPkqZq11cXEgkEtGiRYsoICCAZs6cSUZGRlyHnH11y5YtoypVqtCpU6coLS2N+vXrR7Vr1yZPT08h8dfZs2epQYMG5OXlpeLWsj9z6NAhEolEZG5uTpcvX1Z6zc3NjfT19Sk0NJSIiIKCgsjCwoKePn2qiqYyJlAsIblmzRoiIurfv7/S3l7eo8u+hmPHjpGBgQG1bduW+vTpIyQiJCp+DSr+e8+ePdS/f3969uzZV2srK04xp4piXgA7OzsyNzcXMq4TEQUHB1OHDh3Uvm8pLXXWmfoqMUH55cuXSVdXl+bMmUOOjo5kb29PI0aMEDJX+/v7U9OmTcna2po6depEt27dUm2DWamgeBOKj4+nVq1aCcnczp8/T4aGhtSzZ09q1KgRLViwQLghXL16lTt4NfJxBnGJRELjxo0jkUhEBw4cICLlc+3u7k4ikYhOnz5Nz58/p40bN9Ljx4+/apsZ+xR5Cck2bdqQqakpNWzYUCNqBDPN9qkqDNnZ2RQfH089e/YkR0dHpeRYfxWYZ2ZmfrmGsj917tw5Wr58ufDvvwrM69evT3FxccX6FHUNzEtDnXWm/kRECmuFNFRiYiKOHz8OIsLs2bMBANu2bcPu3btRs2ZNBAQEoFKlSnjz5g3KlSuHgoIClC1bVsWtZiXdx0ndCgsLsXfvXjg7O+PGjRsYMmQIfHx8MG7cOLRr1w5JSUn45ptvEBAQICx7kidTYerh5MmTaNasGWrVqgWJRIKRI0fi7NmzOHr0KNq3b6/03k2bNmHatGnQ0tJCSkoKb5dhaiMlJQUeHh54/fo1IiMjoaOjo9b7O5lmU7wXPnjwAOnp6ahfvz4qVKgAsViMn376Cb6+vgCA6dOno3fv3gCK7+v9+N/s68nPz4erqysuX76MUaNGYc6cOQCUz61i4r127drhxo0b+OWXX2Bpaamydv8d8t9w4MABjBo1CjNnzoSPjw/EYjEaNWqE7du3w8HBAUDRXvpff/0Vd+/exZMnT9CoUSMMGDAAYrGY+1D236l0SOAzePToEbVr146qV69OW7ZsEY5LpVIKDAykNm3a0LBhwyglJUWFrWSlTWxsLF28eJGIiMaPH08LFy4kIqKcnBwiKiqXNXPmTGFEduLEiWRtbU1ubm5c11mNKJ6Lhw8fkkgkopkzZ9Lz58+F152dncnExIRiY2P/53cwpi40tYQk0ywf13hu3LgxGRgYkI2NDY0aNYry8vKIiOjHH3+kXr16Ue/evenQoUOqai77Cy9fvqQZM2aQnZ0d+fr6Csc/Lg8qN336dLWfOS4NddaZ5tDIIR36/9HSnJwc1KpVC506dcKzZ88QFRWF77//Hrq6utDS0sKECRMgFovh5+cHLy8vbN++XW0zc7OSgYjw/v17fPfdd7C0tETZsmVx/Phx/PzzzwAAAwMDAEVJwhRHVTMyMjB79mwMHTpUqB7A16pqkULZs2XLlkFPTw+VK1dGQEAAcnNzsXDhQlSrVg3h4eEYOnQonJ2dERoaim7duil9D59Hpo40oYQk03zyme3169dj27Zt2LdvH0xMTPDzzz8jNDQU/fr1w/Hjx9GxY0eIxWLMnDkTly5dgpOTk4pbzj5WrVq1P63kIH9mef36NaZPn44RI0Zg06ZNANR7xZ+urq5QdYKIMGTIEBARRo4cibS0NLx79w6dOnWClZUVCgoKkJmZCW1tbfTv3x8LFy4Uvkddfx/TLBq3fF0ekEdFReHgwYPYvXs38vPzsX79ehw8eBAdO3aEj4+PUHOciBASEoKOHTuiTp06qm08K/HkN59nz57B3t4eKSkp2LVrF7777jvhdSLCggULEBsbi+rVqyMtLQ1v377F7du3hfrBHMipD19fX6xevVrIqP7o0SNMmTIF48aNw+LFi1GtWjXIZDL06NEDurq6OHnypIpbzBhj6iM3NxcjRoxAq1atMG/ePABAQUEBoqOjsXjxYgwcOFAIcO7cuYOmTZvyPVCNpaSkwMfHB9euXcO3334LT09PAEBycjIGDRqEtLQ03Lt3T60H+z4eKFCssy4Wi3HgwAGhzvrUqVM/WWddnX8f00wacUUdOHAAurq6+Pbbb4VR15s3bwoj/Xp6epg1axYKCwtx8uRJzJs3DytWrIChoSFEIhFGjx6tyuazUoKIhE7+9OnTsLCwgIGBASIjI1GvXj20a9dOeN3NzQ1AUb3yunXr4ty5cxyQq4G4uDi0adNGOAdSqRQ//fQTxo8fj86dOwMAOnfuDDMzMwwYMAC6urqYO3cuatasibNnz0Imk6my+YwxpnYMDAzw9u1bPHjwQDimq6uLfv36ISIiAleuXBGON2/eHEDxnCxMfchLLPr4+ODo0aMQi8X4/vvvMWLECKSnpyMhIQHa2tpqPUNe0uusM82k9j1eSkoKVqxYga1btyI6Olo4/vvvvyMvLw9A0YOznp4ePDw80Lt3b8THx2PGjBnIzc1VVbNZKUMKCWgWLFiAlStXIiQkBGfOnMG9e/ewevVqxMXFCe83MzODr68vIiIisGvXLiHREj+EqM60adOwdetWpURC+fn5eP36tVAjXiqVQiKR4JtvvsHUqVPxww8/wM/PD5mZmRCJRMLACmOMsSJEBAcHB/z++++4deuWUh/ZokULZGVlIScnR+kzfC9Ub/LAvHXr1oiIiIC5uTlSUlJw69Yt4XlGXQPyTylpddaZZlL7Xs/MzAzbtm1DYWEhAgICcOLECQBFfwzlypUDULRniYigp6eHOXPmoF27dnjy5AkyMzNV2XRWisgDuVu3biEpKQl79uxBzZo1YW5ujgMHDuDhw4dYv349fvzxRwBAhw4dsGrVKuHzRMQdvIqtWrUKO3fuhEgkQmJiIiQSCQwNDeHk5ITg4GBcvXoVYrFYeFg0MzPDN998gw0bNmDHjh3C9/DDJGOM/UEkEmHSpEl4+vQpli5disuXL6OwsBBZWVmIiopCnTp1hC2HTHOYmZlh3rx5sLCwQJs2bXD79m2NqeQgHxgSiUSQSCQAilbKVapUCWPGjBEC82HDhmHLli148OCBkFmesS9F7feUy2cgr1+/jjlz5kBPTw/u7u44ffo0GjVqhLFjxwoBjUwmQ35+PvT19fHu3TuYmJiouvmsFDlw4ADWrVsHIkJ0dDTKly8vJFG6e/cuRo4cCS0tLeTn54OIcOvWLejq6qq62QzKSyX37NkDb29vrFy5Ev3790dKSgpmzJiBBw8eYPv27bCzs0Nubi6GDBkCFxcX3L59G35+frh58ybMzMy4ZA9jrNShvyhfJl/GnJiYiAEDBkBHRwdZWVmoVKkScnJycOPGDejo6HDJMw2Vnp4OY2NjaGlpqXVAHhMTg8uXL2PBggUAlO/7iu22t7fH27dvsWvXLtjb2yvN+PM1yr4ktQ/KFf9orl27hrlz58LIyAiXLl3Cu3fv0KpVK2RmZkImk0FXVxcmJiY4fvy4MIvO2Neya9cubN26FQkJCfjxxx9hY2MDmUwmBOaPHj3CuXPnkJOTA1dXV2hra6v1Day0ysjIQJ8+fSCTyTBv3jz07dsXN27cEJK9WVlZ4f3799DR0cGdO3cQGhqKVatW4dq1azzbwxgrdRSf0168eIEaNWoUe488ME9JScHVq1dx9+5dVKlSBaNHj+Z7YQmhznkASnKddVZyqG1QLh+NysjIQLly5VBYWAhdXV1cvnwZ8+bNw/v379GiRQtMmTIF6enpePfuHcqXLw9zc3NYWFiouvmshPuzm8/x48exbNkylC9fHmvWrIGVlRWICDKZrNj+KnVOglJa/Nl5zMzMxDfffIPc3FwsWbIEvXv3RkFBAU6cOIH79++jXLlymDRpEnR0dODq6or79+/j8OHDMDIyUsGvYIwx1VDsQ1euXIkrV67Ay8sLdnZ2f/leRXwvZF/Dq1evsHr1avzyyy8YMGAAPDw8AChfl4r/39XVFX5+fnxtsq9GbYNyoCjA8fX1hZ6eHqysrDB//nyYmJjg6tWr8PDwgKGhIebMmYNOnTqpuqmsFFHstCMjI5Geno43b95g0qRJMDIywrFjx7Bx40aULVsWy5YtQ4sWLXjJkxpSPCe7d+9GQkICmjRpgtatW6NJkybIyMhAv379hJrkjo6OSnvKnj17htWrVyM0NBSxsbFC1mDGGCtt5s6di127diEwMBAtW7bkErRMLSmWc/tUYJ6amirUWe/fvz8AHjRiX4/aBuU3btxAmzZtMHfuXDx//hyPHj0CAERERMDU1BSXL1/GokWLkJubi0WLFqFHjx4qbjErbTw8PBAWFoZmzZrh8ePHICL4+fmhT58+OHDgAIKCgmBkZIR58+ahVatWqm4uU6AYkM+bNw/btm1Do0aNkJaWhvr168PT0xMdOnRARkYG+vfvj4KCAri6umLw4MHQ0tLC+/fvsX//fkRGRsLX1xctWrRQ8S9ijDHVOH36NCZPnoxDhw7BxsYGUqkUHz58wN27d9G2bVtVN48xJSWhzjormdRi84d8XED+v7/++ivu3LmDJUuWYNmyZQgODhb+aL799lukpaXBwcEBixYtQoUKFdC4cWOVtZ2VTiEhIdizZw+ioqJw8uRJ+Pn5CQNHADB48GBMmjQJjx8/xqFDh1TYUvYp8oD8119/xcuXL3Hq1CnExcXBz88POjo6mD9/Pn766ScYGxsjMjISGRkZOH/+vLBConz58hg6dCj279/PATljrFR7+/YtdHV1YWNjg3v37sHb2xutWrVCjx49MHDgQFU3jzEl8nJutra2OHr0KNasWYO3b99+ss46Y1+TSmfK5bNV2dnZKFOmDIgIL168wJAhQ3D37l14enpi3rx5AIqWlkRHR2PFihXQ0tLCgQMHYGZmhry8POjr66vqJ7BSytvbGykpKQgICMC+ffvg4uKClStXYsqUKcjKyoKOjg709fURExODzp07q23yk9IsPDwcGzZsgL6+PiIjI4XkkGfPnkVAQADS09OxfPlydOzYETk5OdDT04NYLOatCIyxUutT+8KvXbuGESNGoHz58khOTkb37t1hZ2cHKysrODg44OzZs+jatauKWszYp6WkpGDFihW4evUqfvvtN1SrVk2jyrqxkkelkYJIJEJaWhrq1KmDgwcPQiQSwcTEBCNHjkTt2rURGRmJwsLCooZqaaFXr15YsGAB0tPTMXr0aCHjOmNfkryepaJff/0VIpEI165dw6RJk4SAnIgQGBiIgIAAAEDXrl2hpaXFI65qKDMzExKJBHfv3sXLly+F4927d8e0adNQqVIlTJ48GTdv3oShoSHEYjGkUikH5IyxUkkxIL979y4ePnyIgoIC2NjYYPPmzWjTpg3WrVsHX19fTJo0CTVq1ICtrS2MjY1V3HLGitPkOuusZFL5nvI3b97Aw8MDe/fuxf79+zFgwADk5eVh//792LBhAxo2bIjdu3fDwMAAQNFNISYmBg0aNOBEIuyLU3wI+fnnn1GtWjWYm5vjxIkTcHFxwfPnzxEcHIwxY8YAALKzszF48GA0btwYa9euVWHLmaI/y/p7+PBhrFixAtWqVcPKlSvRtGlT4bWoqCjExcXBx8eHk7wwxtj/8/DwwOHDh/Hy5Us4OTlh7Nix6NKli/B6YWEhMjMzMWbMGLx79w6xsbHchzK1pSl11lnJp/KgHABev34Nb29v+Pv7IyIiAgMGDEBubi5CQ0Oxbds21K5dG3v27OFl6uyrUlym7OnpiTNnzmDMmDGYOHEiUlNTsWjRIly5cgVz5szByJEjkZiYCA8PD6SmpuLKlSvcsasJxYD8+PHjyMjIwLt37zBhwgQYGBjg0KFD+OGHH1ChQgV4e3ujSZMmxb6Ds68yxkorxT40KioKM2bMwJYtW5CSkoJdu3ZBW1sbkyZNgpOTEyQSCUJDQ7Fr1y58+PABly5dgo6OjlrXsGYMUO8666x0+KpB+ccXvOKDblpaGnx8fODv749Dhw5h4MCByM3Nxb59+7Bjxw4YGRkhMjKSA3P21a1cuRLr16/HkSNHYGlpKew9vn37NrZt24b9+/dDS0sLVatWhYmJCc6cOQMdHR0O5NSMPFt+o0aN8OjRIxgaGmL9+vXo0aMHQkNDERwcjIoVK2LBggWcvI0xxj5y9uxZHDt2DPXr18eMGTMAANevX8fSpUuRl5cHFxcXDBgwAIcPH8aDBw8wZ84caGtr8+wjY4z9DV+1l9TS0sLz589x9epVODk5CXs0xWIxTE1NMX/+fADA8OHDceTIEfTq1QsjRoxAfn4+IiIi8PbtW1SvXv1rNpmVYkSEN2/e4PTp01izZg3atWsnvCaTydCiRQv4+vrC3d0dd+7cQa1atdCiRQteAqWGdu3ahT179iA6OhqWlpY4duwYvv32W2Gv/4gRI6ClpYUVK1YgPDycg3LGGFPw5MkTzJgxA7///jumT58uHG/VqhWWLl2KJUuWYMuWLZDJZHBychJel0qlfC9kjLG/4avOlEskEnz33Xf47bffMHfuXAwdOhSA8oz5ixcvsHDhQty5cwfHjh1DtWrVUFBQgNzcXE4Wwr66d+/ewcrKCosXL8a4ceOUXsvLy0NaWhpq1aqldJyXQKmfJUuWID09HRs3bkRYWBimTJmilC1fX18fOjo6OH36NLp168YrHBhjpZp8+5biNq7z58/Dy8sLRIRVq1ahc+fOwvtv3LiBadOmwdbWFhs3blRVsxljTGN91chBW1sby5YtQ61atbBt2zbs27cPAIQZcwCoUaMGBg8ejBcvXiA9PR0AoKurywE5++I+lWW9oKAAhoaGePz4MQAoZVFPSEjA5s2b8fr1a6XPcECuPuTVG+RZVa9fv46JEycqZcvfunUrNm/eDADo2bOnUn/EGGOljUwmEwLxnJwcSKVSEBG6dOmC5cuXQywWY/PmzYiNjRU+Y21tjV27dsHPz09VzWaMMY321aOH+vXrw8/PD4aGhti+fTvCwsIAFAXm8gfoOnXqoEqVKlx6iH01irPbz58/x4cPH5Cfnw8zMzO4urpi1apV2LVrlzCDmp2djYULFyIpKQmVKlVSZdOZgpiYGPj5+WHu3LlCvXgAGD16NA4cOAA7OzsEBARgypQpAIoeOC9cuIBXr14pfQ/PlDPGSiPFe+HatWvh5OSEHj164LvvvsP79+/Ro0cPLF26FC9evIC/vz9+/vln4bMNGzaElpbWJwe4GWOM/TWVbPSpW7cu/P39MX36dAQFBaGwsBDfffed8AC9e/du6OrqwszMTBXNY6WQ/CFk8eLFCA8Ph6GhIWxtbbFixQq4uLggJSUF33//PSIjIyESifDmzRukp6fjxo0bxZb4MdXYvn075s+fj+bNm+PevXs4duwY7ty5Ax0dHTRv3hzt27fH9evXhQHAhw8fYvbs2UhNTcWxY8dU3XzGGFM5+b1w/vz5CAwMxOzZs/H69WucP38elpaWOHPmDHr06AGpVAofHx8sWbIE69evV8rDwavFGGPsn1NpSbSkpCS4u7vj5cuXaN26Ndq0aYOLFy/i4MGDOHPmDCdbYl+c4qzAgQMHMG3aNGzcuBE3btzAlStXQEQ4evQoTExMcPz4cZw6dQo5OTmoU6cOFixYwJll1URgYCCmTZuGAwcOoHv37khOTkbnzp1x+PBh2NraQiQS4e7du9iwYQMOHToEPT09VKlSBRUqVMC5c+c4Wz5jjP2/J0+eoHfv3li1ahX69+8PoKh0rbz0Z2JiInR0dHD8+HFERUVhy5YtHIgzxth/pPI65S9evEBwcDAOHz4MsViMWrVqwdvbG02bNlVls1gJ93EytiNHjuDBgweoWrUqRo8eDQA4ceIEVq5cCSLCkSNHYGpqioKCAujq6gqf40BO9Y4ePYqBAwciMjIS33zzDQAgNzcXVlZW6Nq1KxISEjBo0CAMGzYMxsbGePz4MRISElCzZk3Y2NhwtnzGWKnVvn17zJw5EwMHDhSO3bp1Cx06dMDFixfRvHlz4X6ZlJSEHj16YObMmXBxcVH6Hk5wyhhj/43Ke9AaNWpg0aJFuHXrFn7++Wfs37+fA3L2xaWkpAAoyjB78+ZNLFy4ED4+PkoBd69evTBv3jxoaWnByckJaWlpSq8DvPdY1fLz83H69GnUq1cPT548EY6PGDECWVlZMDIyQtmyZTFz5kysXbsWEokEFhYWGDhwIGxtbYX9jxyQM8ZKo++++w59+vRROtasWTPUqFFDSMYrD7ZNTU1haGiIjIyMYt/DATljjP03atGLyifrDQ0Noaenp+LWsJLu1q1bqFGjBg4fPgyRSAQLCwvMmDED1atXx5YtW5Cbmwug6CFDHpi/fv0aixYtUnHL2cf09PSwaNEi9O3bF/v378fGjRsxaNAgJCUlIS4uDqtWrcKJEycwbNgw7Ny5Ex8+fCj2HfwwyRgrrSZMmAA9PT34+Phg06ZNwvF+/fohLi4O27ZtE47p6OjAwMAAZcqUUUVTGWOsRFOLp1FOkMW+pqpVq2LixIkYNmwYDh8+DENDQ4waNQoeHh7Izs7G6NGjhcBcJBKhZ8+e2LVrF3744QcVt5x9StWqVeHp6YlWrVph48aNOH/+PKKiolCvXj3k5OQAANq1a4datWpxVmDGGPuE3NxcuLm5ISgoCNra2nBzc0PdunWxdetW9O/fH6tWrUL37t3x4cOHYkvXGWOM/XdqEZQz9jVVqVIFy5cvx+TJk+Hs7IwjR45AX18fQ4cOxZQpU/D7779j9OjRyMvLA1A0k2pvb8/1q9WYmZkZFixYgG+++QZ169YVSi0aGhpCIpHg0KFDqFevHipXrqziljLGmGpdvXoVb968AQAsX74c586dg7e3N7y9vTF58mRs3boVZmZmWLduHVxcXJCdnY0ff/wR5ubmuHnzJrS1tfleyBhjn5nKE70x9jUpJmZLS0uDt7c3AgICEBERgQEDBiA3NxdhYWEICgqCoaEhTp06VWwfOVNfKSkp8PHxwdWrVzFo0CDMnj0b/fr1w+PHj3H79m1oa2tz+TrGWKmVmJiIESNGwMrKCtra2ggMDMSdO3eEXD7e3t5YvHgxfvjhB0yePFn4XGFhoVC2lhNjMsbY58dBOSvxfvzxRzx//hyjRo0CUDww9/Hxwfbt23H48GH07NkT+fn5CAoKwq1bt7Bt2zbec6xhUlJSsGLFCsTHx+PRo0coX7487t69Cx0dHX6YZIyVelu3bsXSpUuRkZGBEydOoHPnzkp9o7e3N5YuXYotW7Zg3LhxSoOYPKjJGGNfBj+dshKLiJCbmwsfHx98+PABOjo6GDp0qLAMXSwWw9TUFO7u7nj37h1Wr14NW1tbVKxYEZMmTYK2tjZEIhGXetEwZmZmmDdvHjw8PGBsbIzIyEgOyBljpZ78Xla3bl0YGhrC1NQU+/fvR6NGjVC1alXhvrhgwQJoaWlh4sSJMDU1Rb9+/YTv4ICcMca+DJ4pZyXe48eP4e7ujszMTIwfPx7Dhw8HoDxjvnv3bnh4eCA+Ph7Vq1cXPsuzAporPT0dxsbGXIecMVaqfXwfS01NhUgkwqFDhxAaGgoLCwusWLECZmZmSp/bs2cPhg0bxn0nY4x9BTz9x0o8c3Nz+Pn5wdDQENu3bxeSgInFYhQUFAAAateujfr16xerO84BueaqUKEC1yFnjJVqMplMuI+9evUKaWlpwiy5i4sLnJ2d8eDBAyxcuBBpaWkAisqkxcbGYtSoUdDW1oZEIlHlT2CMsVKBZ8pZqZGUlITp06cjJycHo0aNwtixYwEAeXl5cHJygr6+Pg4dOsSBOGOMMY2nuPVq+fLlOH36NB4/fgxHR0cMHDgQ33zzDQDAz88PERERyMvLg5GREe7fv48XL17wYCZjjH1FHJSzUiUpKQlz587FkydP0LBhQ7Rs2RIxMTF4+/YtLl++DB0dHV6yzhhjrMRYtGgRNm/ejMDAQIjFYmzevBm///47vL29MXjwYADAwYMHcf36dWRlZWHTpk1C2bOPV48xxhj7MjgoZ6XOq1evcOjQIRw4cACVKlVC7dq1sW7dOmGZHs8OMMYYKwnOnj2LmTNnYvv27bC3t8f58+fRp08f2NjYIC0tDb6+vhg4cGCxz/G9kDHGvi4Oyhn7fzwrwBhjrCRJSkpCYGAgVq5cidOnT2PUqFHw9fWFra0tBg4cCJlMhsWLF2P06NGqbipjjJVqHJSzUomXqDPGGCtJbty4gYoVK6JOnTrw8vJC165d0a1bN2RnZ8PAwAADBw6EpaUlli5dCpFIhL59++L333+HtbU1QkJCVN18xhgr1XhtEiuVOCBnjDFWUiQmJmLUqFFwdHREZmYmduzYIZT/LFOmDDIyMnD37l20a9cOIpEI79+/R9myZbF48WI4OzuruPWMMcY4KGeMMcYY02ANGzaEq6srFi9ejPfv3yMqKgrNmzcXSqLp6Oigffv2OHr0KAoLCxETE4Ps7Gw4OTlBJBIpZWpnjDH29XEPzBhjjDGmoaRSKQCgbt26MDQ0RJ06dXD+/Hk8evQIWlpaEIlEMDQ0xMiRI1G3bl2Eh4ejbNmyiI2NhZaWFgfkjDGmBnhPOWOMMcaYhvk4mM7Ozoa2tjZ27NiBbdu2oUOHDpgxYwbMzc2VPpOdnY2yZctCJBJxlnXGGFMT3BMzxhhjjGkQxYD87NmzyMzMRG5uLkaOHAkXFxcQEbZv3w5tbW1MnToV5ubmGDJkCKZNm4b27dsL38EBOWOMqQeeKWeMMcYY00AeHh44ePAgqlSpgtTUVJQvXx67d+9G8+bN4e/vj5CQEOjr60MqleL333/H06dPoaOjo+pmM8YY+wgPkTLGGGOMaZjAwEDs2rULp0+fhpWVFfbv34/hw4cjOTkZzZs3x/Tp01G5cmUkJCTg/fv3iI2Nhba2Ni9ZZ4wxNcS9MmOMMcaYmvt4D/mDBw8wdepUWFlZITw8HJMnT8bmzZvRo0cPZGVlwcjICEOHDlX6Dg7IGWNMPXG6TcYYY4wxNUZESnvIAeDXX3+FVCrFxYsXMX78ePj6+mLy5MmQyWRYvXo1tm7dWux7OCBnjDH1xEE5Y4wxxpiaIiKIRCIAwJIlSzBx4kS8fPkSw4cPx9GjR9GlSxds2LABkydPBlCUhf3mzZt49eqVKpvNGGPsH+CgnDHGGGNMTckD8vj4eNy7dw8hISGoXr06bG1tUbFiRTRt2hRVqlQBADx8+BBDhw5FWloaFi1apMpmM8YY+wc4+zpjjDHGmBoLCQlBSEgIcnJycPz4cZiYmAAAYmNjsWbNGty4cQMAYGpqijJlyuDChQvQ0dGBVCqFWCxWZdMZY4z9Dby5iDHGGGNMjRkYGCAlJQVPnz5FfHw8evToAQDo0KED6tWrh9TUVNy5cwfm5uZo06YNxGIxJ3VjjDENwjPljDHGGGNq4uMs63LR0dHw8vJC7dq14eHhAQcHBwDKe87leIacMcY0Cw+hMsYYY4ypAcWAPDo6Grm5udDX10evXr3g6OiIgoICeHt7Y9OmTRCJRLC3ty8WkAPggJwxxjQMB+WMMcYYYyqmWPZs5syZ2Lt3L3R0dKCtrY2KFSvi+PHj6NevH6RSKVatWoWAgAAUFBSgQ4cOKm45Y4yx/4qzrzPGGGOMqZDiEvQrV64gLi4Op06dwi+//IIjR47A0NAQXbp0QUZGBgYMGABPT09cunQJ58+fV3HLGWOMfQ68p5wxxhhjTA2Eh4cjNDQUZcqUQWhoqDBznpaWhh49eqBatWo4efIkAODixYuwt7fnpeqMMVYC8Ew5Y4wxxpgKyGQy4f+np6fj5MmTuH79Op48eSIE5BKJBKamppg6dSqePXuGlJQUAEDbtm0hFoshlUpV0nbGGGOfDwfljDHGGGMqIA+8J02ahKCgICxZsgSDBw/Go0ePsHTpUgAQypqZmZkhJycHBQUFSt/BM+WMMab5ONEbY4wxxthXpLiH/P79+zhz5gyCg4NRt25deHl5obCwEMeOHUN2dja8vLyQlpaGgIAA1K5dGzVq1FBx6xljjH1uvKecMcYYY0wF1q9fj6SkJOjq6mLdunVCSbRXr15hxYoV2L59OypUqID27duDiBASEgIDA4M/rWXOGGNMM/FMOWOMMcbYV5aeno5bt24hLCwM/fr1E45LJBJUq1YNCxcuhEgkwuXLl9GgQQP4+PgAAPLy8qCvr6+qZjPGGPsCeJiVMcYYY+wLU0zqBgAVKlTA/PnzMXHiRBw5cgTHjx+HlpYWtLS0IJVKUaVKFXh6eqJ169Y4f/48NmzYAAAckDPGWAnEy9cZY4wxxr4gxeXmz549Q3p6Oho3bgxdXV28ffsWs2fPxuHDhxEWFobevXtDJpOBiCAWi/Hy5UusWrUKZ86cwbRp0zBt2jQV/xrGGGOfGy9fZ4wxxhj7QohICMgXLFiAEydO4OnTp2jRogXatGmD+fPnY9myZdDV1cXIkSMRGhqKXr16gYhARKhevTrmzJkDPT099O3bV8W/hjHG2JfAM+WMMcYYY1+Yr68v1q9fjz179qBjx44YPHgw4uPjcfz4cVhbWyMpKQm+vr4ICgrCpUuXYG9vD+CPTO1SqZTLnzHGWAnFM+WMMcYYY1+ITCZDZmYmYmJisH79evTs2RNnz57F+fPn4efnB2tra0gkEtStWxceHh6oV68eWrVqJXxeXjqNA3LGGCu5eKacMcYYY+wzkS87VyxZVlBQgLZt2+LQoUNISEjAkCFDsGbNGkyePBn5+fnYt28frKys0LJlS+EzEokE2to8d8IYY6UBZ19njDHGGPtMHjx4IATkQUFB+Omnn6ClpYWCggKMHTsWI0aMwLp16zB58mQAQHJyMkJDQ5GYmKj0PRyQM8ZY6cEz5Ywxxhhjn0FCQgJatmwJPz8/PH/+HFu2bMG1a9fQsGFDnD17FmPGjEGjRo0QExODwsJC5OXlYejQocjOzkZMTAwvUWeMsVKKg3LGGGOMsc/gzZs32Lt3L7y8vKCnp4d79+6hWrVqkMlkyMrKwt69e+Hu7g5bW1sYGhoiNzcXmZmZuHbtGnR0dDiZG2OMlVK8fJ0xxhhj7DOoVKkSTExMkJ+fD4lEgoiICACAlpYWjI2NMWnSJFy9ehUtWrSApaUlBg0ahOvXr0NHRwcSiYQDcsYYK6V4ppwxxhhj7F+SyWRKSd3S0tLw8uVLXLhwAUuWLMHChQsxZ86cv/wOniFnjLHSjbOIMMYYY4z9C4oBeUJCAqRSKSwtLWFqaooaNWogNzcX3t7eEIvFmDVrFgBg6dKl6Nq1K9q1ayfUIOeAnDHGSjeeKWeMMcYY+w88PT2xa9cuAICpqSn27t0LS0tLvH37FoGBgfD29saQIUPw4sULJCUl4cGDBxyIM8YYE/CecsYYY4yxf0AqlQr//8SJEzh06BCCg4MRFhYGMzMzODo64ueff4aJiQmmTZuGrVu34unTp6hWrRru378PsVgMmUymwl/AGGNMnfBMOWOMMcbY35CdnY0yZcoI/w4JCUFmZiYKCgqE5ekA0KdPH9y8eRMHDhxAu3btACjvG5dIJFyHnDHGmIBnyhljjDHG/oe2bdsiPDxc+Ld8v7irqyuePn2q9N4TJ06gZcuWGDZsGGJiYpQCciLigJwxxpgSnilnjDHGGPsfQkND4ezsDD09PeTn50NPTw9paWkYPnw4kpKScOLECTRq1EhI3gYAdnZ2MDMzQ2RkpIpbzxhjTJ1xUM4YY4wx9icUg2wA8Pb2Fparly9fHm/evIGjoyPy8/Nx5MgR1K9fX+kzH5dMY4wxxj7GdwnGGGOMsT+hGJADgFgshre3N7Zt24b379+jUqVKiI6Ohq6uLgYOHIjHjx8rfUZLS4uTujHGGPtLvKmJMcYYY+xv8vLyQpkyZeDm5gaZTIbJkyejUqVKOH36NHr37o02bdrgxo0bqF69uvAZnilnjDH2VzgoZ4wxxhj7G+RL0V1dXUFEmDlzJgAIgXlUVBQ8PT1hZmam4pYyxhjTJByUM8YYY4wp+LN94PKl6FpaWpgxYwYAYObMmdDS0sKECRNgamqK4OBgAMol0BhjjLG/wkE5Y4wxxpgCeUAuD6wVg/SPA3ORSAQ3NzdUq1YNI0eOFL6DA3LGGGN/F29yYowxxhgDlBKy7dq1C5aWlsjLyyuWrE3x366urti/fz+GDh361dvLGGOsZOCgnDHGGGOlnuJs+PHjx5GSkoL79+/DyckJ+fn5fxmYDx48GNra2pBIJCppO2OMMc3GQTljjDHGSj15QD537ly4ubmhoKAAgwcPxtWrV9G5c+c/DcwVaWvzrkDGGGP/nIiISNWNYIwxxhhTtfj4ePTq1QuhoaHo3r07ZDIZ4uLiMHbsWFStWhXnzp2Dvr7+nyaCY4wxxv4NvqMwxhhjjAF4//49CgsL0bRpUwBFM+Ft2rTBpk2b8Msvv8DZ2RlSqRRaWlqQSqUqbi1jjLGSgoNyxhhjjJU6n1ooaGVlBSMjI4SHhwvHtLW1YWNjAwsLC5w7dw6dO3cGwNnVGWOMfT4clDPGGGOsVJHJZBCJRACADx8+IDc3FwBgaGiIPn364NixYwgLCxPer6enBysrK4SHh+PFixcICAhQSbsZY4yVTByUM8YYY6xUke8HX758OQYOHAh7e3scPHgQBgYG8PDwQIUKFeDn54dJkyYhJCQEAwYMwKtXr9CpUycYGxvj+fPnKv4FjDHGShIOyhljjDFWKihmTt+4cSP8/f3RoUMHNG3aFCNHjsSyZctQs2ZNbNq0CYMGDcLFixfh7+8PQ0NDREdHw9jYGJUrV4aJiQmATy+BZ4wxxv4pzr7OGGOMsVLlwYMH2LFjB7p27YqePXsCAPz8/LB48WK4u7vD3d0dZcuWBREhOzsbZcuWBQB4eHggJCQEcXFxMDc3V+VPYIwxVoJwQU3GGGOMlVjz5s3DsGHD0Lx5cxARzp07h549e6JixYro2LGj8L6ZM2dCJBJh8eLFEIvFGDNmDGrUqIGyZcsiPj4eAQEBiImJwcmTJzkgZ4wx9lnx8nXGGGOMlUiJiYm4e/cuGjduDAAQiUTo3r07Fi9ejHfv3uH69ev48OGD8H43Nzd4e3tj0aJFOHv2rHDcxsYGffr0wU8//YSWLVt+9d/BGGOsZOPl64wxxhgrcYhIyLAOAAcOHEDlypWFkmaenp5Yt24dtmzZguHDh8PQ0FB478GDBzFgwABoa2tDJpMJieEYY4yxL4GXrzPGGGOsRCEiSKVSaGsXPea8e/cOU6dOhYODA3R1ddG2bVv4+vpCKpXCxcUFAJQC80GDBgEAJBKJ8B2MMcbYl8JDv4wxxhgrUVJSUoRgOiwsDGKxGGfOnMGjR4+wbt06xMXFAQDWrFkDNzc3TJ8+HUFBQcjPz1f6Hg7IGWOMfQ28fJ0xxhhjJcbVq1fRsWNHxMbGIiIiAjt37sSVK1dQp04d3LhxA8OGDUPTpk0xa9YstGvXDgAwadIk/Pbbb/jxxx+VlrwzxhhjXwMH5YwxxhgrMR4/fow1a9Zg3759EIvFuHfvHqpWrYrCwkLo6OgIgXmzZs0wa9YstG3bFsAfe9A/3ovOGGOMfWm8fJ0xxhhjGk8+x2Bubg5zc3N8+PABMpkMT548AQBoaWlBIpHA2toaYWFhuH//Pry8vPDrr78qfQcH5Iwxxr42DsoZY4wxptFkMpkQTKekpKBHjx64ePEihg8fjr59++Ls2bMQi8VC4G5tbY19+/ahevXqaNasGYCicmkckDPGGFMFzmDCGGOMMY2lWLJs+fLlePbsGUaOHImOHTvCxMQEBQUFGDJkCA4dOoQuXboAADZs2IBRo0YhLCys2HcwxhhjXxvvKWeMMcaYxvPy8sL27duxefNmtGvXDlWrVgUAPHr0CL6+vggLC8Py5ctx6tQppKSk4NatWxCLxSpuNWOMMcZBOWOMMcY03KVLlzBq1Cjs3r1byKiuuD/8+fPn+OGHH3DixAk0aNAA4eHh0NHR4RlyxhhjaoHvRIwxxhjTGIsWLcK9e/eUjsmTutWqVUs4Jg/IpVIpqlWrBl9fX/z000+IiIiAjo4OJBIJB+SMMcbUAt+NGGOMMaYRfv31V8THx6Nhw4ZKx9+9e4e0tDQYGBgAAAoLC4XXfvzxR1y4cAFSqRQVK1aESCSCTCaDtjan1WGMMaYeePk6Y4wxxtReTk4ODA0NIZVKIRaLcfjwYVStWhUODg7Iy8tD27ZtUaFCBRw7dgyGhobCZwYMGIDOnTvD09NTxb+AMcYY+zQOyhljjDGm1jw8PCAWizFv3jyULVsWz549g6WlJbp16wYvLy/Y2Njg8OHDWLFiBYgI3t7eeP36NcLCwvDq1SvEx8fzzDhjjDG1xcvXGWOMMaa2Hj9+jJcvXyImJgabN29GVlYWatWqhWPHjuHOnTvw9fXF3bt3MXDgQGzatAlmZmb4/vvv4efnBwMDA1y/fh3a2tqQSqWq/imMMcbYJ/FMOWOMMcbUUpMmTdCjRw+sXbsWM2bMwC+//AJnZ2e4uLjA2NgYcXFxGD16NKytrbFw4UJYWloCAJ4+fYoKFSrAyMgIIpEIEomEZ8oZY4ypLQ7KGWOMMaZ21q9fj127duHGjRvQ1tYGEcHT0xMxMTFwdnbGlClTlALzVq1awd3dHa1bt1b6Hi57xhhjTN3xXYoxxhhjaoeI8PbtW2hra2PWrFlYtGgRVq5cCTs7O0RERGDLli3IyMhAu3btsHv3bty8eRMLFizAgwcPlL6HA3LGGGPqjmfKGWOMMaY2iAgikQj379/H5MmT8fz5c7x79w7x8fEwNzeHVCqFq6srrl69CicnJ2HGPCYmBoGBgdi/fz8H4owxxjQKB+WMMcYYU0vOzs44fPgwrK2tce3aNYhEIgAQAvNr167ByckJEydORIUKFYTP8ZJ1xhhjmoTvWIwxxhhTKxKJBMnJyTA0NMTGjRuhr6+PTp06ISMjAwAgFouxadMm2NvbY8uWLYiKigJQNMsO8JJ1xhhjmoVnyhljjDGmcn81u3306FGsWrUKurq6OHbsGIyNjQEUzZhv3LgRM2bMgFgs/prNZYwxxj4bDsoZY4wxpja2bNmC69evo3bt2ujZsyfs7OxQWFiIkydPwtfXF/r6+oiMjES5cuWUPieVSjkwZ4wxppE4KGeMMcaYyijOkC9YsABbt25FmzZt8OrVK+Tl5WHVqlXo06ePEJivWbMG6enpuHr1KsqUKaPi1jPGGGP/nbaqG8AYY4yx0ksekN+/fx85OTk4efIkWrdujRs3biAgIAAuLi7YvHkz+vTpg969eyMvLw/nz5+Hvr6+ilvOGGOMfR48U84YY4wxlTpy5AimT5+OihUr4tSpU6hevToA4O7du/Dz80NMTAx++OEH9OnTR2mZOi9ZZ4wxVhJwelLGGGOMqZSBgQFatWqFx48f4/nz58LxZs2aYebMmejevTsGDhyIixcvKgXhHJAzxhgrCXj5OmOMMca+mk9lWXd0dETZsmWRk5ODiRMnIigoCHZ2dgCKAnMXFxfUq1cP9vb2qmgyY4wx9kXx8nXGGGOMfRWKAfnZs2eRk5ODvLw8DBkyBABw6dIlrF69Gs+ePcPWrVvRunXrYt/BS9YZY4yVNByUM8YYY+yrmj17Nvbt2wcjIyMkJyejadOmWLNmDdq1a4e4uDisX78ez58/h5+fH9q1a6fq5jLGGGNfFO8pZ4wxxthXs3PnToSEhODkyZOIjY1FYmIiZDIZXF1dcefOHbRr1w7Tpk2DoaEhgoODVd1cxhhj7IvjmXLGGGOMfXFEBJFIhHnz5iEhIQGRkZGQSCTQ1tZGXl4ebGxsUK9ePRw/fhwAcPPmTbRo0aLY/nPGGGOspOE7HWOMMca+iNu3byMyMhIXL16ESCQCAKSmpuLt27cAAG1tbeTm5kJfXx9r167FtWvX8PjxYwBAy5YtoaWlBZlMprL2M8YYY18DB+WMMcYY++xCQ0MxZswYBAcH48SJE8LxsWPH4vbt29iwYQOAonJoAFBYWIhKlSrByMhI6Xt4ppwxxlhJxyXRGGOMMfZZhYSEYPLkyQgODoajoyPKly8vvNa8eXPMmjUL/v7+yMvLw5QpU5Ceno6goCDUqFEDlStXVl3DGWOMMRXgPeWMMcYY+2wSEhIwZMgQzJgxAxMmTBCOy/eUA8CzZ88QFhaGFStWwMDAAEZGRqhQoQIuXrwIHR2dT9YyZ4wxxkoqDsoZY4wx9tmcOXMGkydPxqlTp9CwYUMhEJdTDM5fvnyJ69evw9jYGO3bt4dYLBaSvzHGGGOlBd/1GGOMMfbZxMfHIysrCxYWFgCUg3AAEIlEuH//PlJTU9GpUydUr15deE0qlXJAzhhjrNThtWGMMcYY+2zq16+P7OxsnDlzBgCKzZQDRXvO9+3bh48X64nF4q/SRsYYY0ydcFDOGGOMsc/GxsYGurq62LZtG549eyYclwfgmZmZePjwIZo3b/7JgJ0xxhgrbTgoZ4wxxthnU69ePWzduhVRUVGYN28ebt26BaBoxvzVq1cYOnQoUlJSMGXKFNU2lDHGGFMTnOiNMcYYY5+VVCrFzp074eLigipVqqBZs2aQyWTIyMiATCYTsqxLpVJess4YY6zU46CcMcYYY1/ErVu3EBwcjMTERNSoUQMtW7bE5MmTOcs6Y4wxpoCDcsYYY4x9VTxDzhhjjP2Bg3LGGGOMfTEfl0RjjDHGmDJO9MYYY4yxL4YDcsYYY+yvcVDOGGOMMcYYY4ypCAfljDHGGGOMMcaYinBQzhhjjDHGGGOMqQgH5YwxxhhjjDHGmIpwUM4YY4wxxhhjjKkIB+WMMcYYY4wxxpiKcFDOGGOMMcYYY4ypCAfljDHGGGOMMcaYinBQzhhjjDHGGGOMqQgH5YwxxhhjjDHGmIr8HxEzLkWpM+ggAAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 1000x500 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAArIAAAGGCAYAAACHemKmAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy81sbWrAAAACXBIWXMAAA9hAAAPYQGoP6dpAABMhElEQVR4nO3deVgV5f//8dcRZBEEBRVFEdyXXDLczdTcwrI9zX0tTTOXsjRL0yzLcmlxaVHJUrNyKc1MPlppLiUmmYn7gguoaIKaocL9+8Mf5+uRRUDgOPp8XNe5Luc+98z9npkzh5fDzGAzxhgBAAAAFlPA2QUAAAAAOUGQBQAAgCURZAEAAGBJBFkAAABYEkEWAAAAlkSQBQAAgCURZAEAAGBJBFkAAABYEkEWAAAAlkSQBfLJtm3b1KtXL5UrV04eHh7y9vbWXXfdpYkTJ+r06dP2fs2bN1fz5s0d5rXZbHrttddyNG56y8tNPXv2VEhISK4sKyQkRD179syVZWXk4MGDstlsCg8Pd2hfuHCh7rjjDnl6espmsykqKkqvvfaabDZbntZzs5k7d66KFy+us2fPOrsUpwkPD5fNZlNkZGS+jpv62Xz33XdzZXnz58/X1KlTczTv6tWr5e3traNHj+ZKLUBecXV2AcDt4JNPPtGAAQNUpUoVDR8+XNWrV9elS5cUGRmpmTNnauPGjVqyZEmG82/cuFFlypTJ0djTp0/Padn5bsmSJfLx8cnTMUqVKqWNGzeqQoUK9raTJ0+qW7duuu+++zR9+nS5u7urcuXK6tu3r+677748redm8u+//+rll1/WSy+9pMKFCzu7HNyg+fPna/v27RoyZEi2523ZsqXq16+vl19+WZ999lnuFwfkEoIskMc2btyoZ555Rq1bt9bSpUvl7u5uf69169Z6/vnntXLlykyX0bBhwxyPX7169RzPm9/q1KmT52O4u7un2Z67d+/WpUuX1LVrVzVr1szeXqhQoRz/B8KKPvvsM506dUp9+/Z1dinp+vfff1WoUCFnl3HbGDhwoDp27Kjx48crKCjI2eUA6eLSAiCPvfnmm7LZbPr4448dQmwqNzc3Pfjgg5ku49pLC1J/9fnTTz/pmWeeUbFixeTv769HH31Ux44dc5g3vUsLkpKSNG7cOFWrVk0eHh7y9/dXixYttGHDBnufadOm6Z577lGJEiXk5eWlmjVrauLEibp06VL2N4KkrVu36oEHHlCJEiXk7u6uwMBA3X///Tpy5Ii9T3qXFvz9999q06aNChUqpOLFi2vgwIH6/vvvZbPZ9PPPPzusZ40aNbR582Y1bdpUhQoVUvny5fXWW28pJSXF3u/aSwt69uypu+++W5LUsWNH2Ww2+/bK6NKC+fPnq1GjRvL29pa3t7fuvPNOzZo1y/5+RESEHnroIZUpU0YeHh6qWLGi+vXrp/j4eIflpC7/77//VqdOneTr66uAgAD17t1bCQkJDn1TUlL0wQcf6M4775Snp6eKFCmihg0b6rvvvnPot3DhQjVq1EheXl7y9vZW27ZttXXr1sx3zv83Y8YMtW/fXkWKFMn22CkpKZo4caKqVq0qd3d3lShRQt27d3fYvznZNn/88Ycef/xxFS1a1H4Wff/+/XryyScVGBgod3d3BQQEqGXLloqKisq1bSFJ//zzj3r16iU/Pz95eXmpffv22r9/v0OfjC6HSe+4O3PmjJ5//nmVL1/evo3atWunnTt3ZljDpUuX1KNHD3l7e2v58uWSJGOMpk+fbt8fRYsW1eOPP+5QW/PmzfX999/r0KFDstls9leqGTNmqHbt2vL29lbhwoVVtWpVvfzyyw5jt2/fXt7e3vrkk0+yusmAfEeQBfJQcnKy1qxZo9DQ0Dw5o9G3b18VLFhQ8+fP18SJE/Xzzz+ra9eumc5z+fJlhYWF6fXXX9cDDzygJUuWKDw8XI0bN1ZMTIy93759+9S5c2d9/vnnWr58ufr06aN33nlH/fr1y3ad58+fV+vWrXX8+HFNmzZNERERmjp1qsqWLZvptZixsbFq1qyZdu3apRkzZmju3Lk6e/asnn322XT7x8XFqUuXLuratau+++47hYWFaeTIkfriiy8yHOPVV1/VtGnTJF35T8fGjRszvRxj9OjR6tKliwIDAxUeHq4lS5aoR48eOnTokL3Pvn371KhRI82YMUOrVq3S6NGj9dtvv+nuu+9O9z8Cjz32mCpXrqxFixZpxIgRmj9/voYOHerQp2fPnho8eLDq1aunhQsX6ssvv9SDDz6ogwcP2vu8+eab6tSpk6pXr66vvvpKn3/+uc6ePaumTZtqx44dGa6TJB05ckR//fWXWrRokea9rIz9zDPP6KWXXlLr1q313Xff6fXXX9fKlSvVuHFjh5Ca3W3z6KOPqmLFivr66681c+ZMSVK7du20ZcsWTZw4UREREZoxY4bq1KmjM2fO5Mq2SNWnTx8VKFDAfq3p77//rubNmzuMk1Vnz57V3XffrY8++ki9evXSsmXLNHPmTFWuXFmxsbHpznPmzBm1bdtWq1at0i+//KIHHnhAktSvXz8NGTJErVq10tKlSzV9+nT9/fffaty4sY4fPy7pyiVFTZo0UcmSJbVx40b7S5K+/PJLDRgwQM2aNdOSJUu0dOlSDR06VOfPn3cY383NTY0bN9b333+f7fUF8o0BkGfi4uKMJPPkk09meZ5mzZqZZs2aObRJMmPGjLFPz5kzx0gyAwYMcOg3ceJEI8nExsZmuLy5c+caSeaTTz7Jck3Jycnm0qVLZu7cucbFxcWcPn3a/l6PHj1McHBwpvNHRkYaSWbp0qWZ9gsODjY9evSwTw8fPtzYbDbz999/O/Rr27atkWR++ukne1uzZs2MJPPbb7859K1evbpp27atffrAgQNGkpkzZ4697aeffjKSzNdff+0w75gxY8zVX5P79+83Li4upkuXLpmux9VSUlLMpUuXzKFDh4wk8+2336ZZ/sSJEx3mGTBggPHw8DApKSnGGGPWrl1rJJlRo0ZlOE5MTIxxdXU1gwYNcmg/e/asKVmypOnQoUOmdS5cuNBIMps2bXJoz8rY0dHR6X4ef/vtNyPJvPzyy+nOl5VtM3r0aId54uPjjSQzderUDOu50W2Renw98sgjDu3r1683ksz48ePtbdd+ZlNde9yNGzfOSDIREREZjpv62XznnXfMgQMHTPXq1U316tXNwYMH7X02btxoJJlJkyY5zHv48GHj6elpXnzxRXvb/fffn+6x+eyzz5oiRYpkWMfVRo0aZQoUKGDOnTuXpf5Afrutz8iuXbtW7du3V2BgoGw2m5YuXZrtZRhj9O6776py5cpyd3dXUFCQ3nzzzdwvFkjHtZck1KpVS5Iczg5e64cffpCHh4d69+6d6bK3bt2qBx98UP7+/nJxcVHBggXVvXt3JScna/fu3dmqs2LFiipatKheeuklzZw5M8tnxH755RfVqFEjzXW+nTp1Srd/yZIlVb9+fYe2WrVqZbo9siMiIkLJyckaOHBgpv1OnDih/v37KygoSK6uripYsKCCg4MlSdHR0Wn6p7cf//vvP504cULSlX0mKdNxf/zxR12+fFndu3fX5cuX7S8PDw81a9bM4TKM9KReklKiRAmH9qyM/dNPP0lSml+x169fX9WqVdPq1avtbdndNo899pjDtJ+fnypUqKB33nlHkydP1tatWx0uHZFufFuk6tKli8N048aNFRwcbF/f7Pjhhx9UuXJltWrV6rp9//jjDzVs2FABAQFav369fftI0vLly2Wz2dS1a1eHdStZsqRq166dpXWrX7++zpw5o06dOunbb79Nc1nH1UqUKKGUlBTFxcVlaT2B/HZb3+x1/vx51a5dW7169UrzZZlVgwcP1qpVq/Tuu++qZs2aSkhIyPRLAbeXYsWKqVChQjpw4ECeLN/f399hOvUa3AsXLmQ4z8mTJxUYGKgCBTL+f2xMTIyaNm2qKlWq6L333lNISIg8PDz0+++/a+DAgZkuPz2+vr765Zdf9MYbb+jll1/WP//8o1KlSumpp57SK6+8ooIFC6Y736lTp1SuXLk07QEBAen2v3Z7SFe2SXbrzcjJkyclKdMbwFJSUtSmTRsdO3ZMr776qmrWrCkvLy+lpKSoYcOG6dZyvf148uRJubi4qGTJkhmOm/or5Xr16qX7fmb7++qxPDw8HNqzMvapU6ckXXkixLUCAwPt/5HIyba5dpk2m02rV6/WuHHjNHHiRD3//PPy8/NTly5d9MYbb6hw4cI3vC1SpbfOJUuWtK9vdpw8eVJly5bNUt+IiAjFx8dr8uTJaa5XPn78uIwxGR4D5cuXv+7yu3XrpsuXL+uTTz7RY489ppSUFNWrV0/jx49X69atHfqmfh5y6xgCctttHWTDwsIUFhaW4fsXL17UK6+8onnz5unMmTOqUaOG3n77bfsF/NHR0ZoxY4a2b9+uKlWq5FPVsBIXFxe1bNlSP/zwg44cOXJT3AFfvHhx/frrr0pJScnwB/rSpUt1/vx5LV682OFs0LU302RHzZo19eWXX8oYo23btik8PFzjxo2Tp6enRowYke48/v7+9lByNWedHSpevLikK9eTZnTN8/bt2/Xnn38qPDxcPXr0sLfv3bv3hsZNTk5WXFxcumFRuvKfJkn65ptvHPZZVqXOf/r0aYcxsjJ2ahCPjY1N8xk/duyYfdk52Tbp3WwXHBxsv7lu9+7d+uqrr/Taa6/p4sWLmjlz5g1vi1Tpfc7i4uJUsWJF+7SHh4eSkpLS9IuPj7fXIV3Zjtfe+JaR4cOHa9++ffYzyt27d7e/V6xYMdlsNq1bty7dm0fTa0tPr1691KtXL50/f15r167VmDFj9MADD2j37t0O2yz1GddXrwtwM7mtLy24nl69emn9+vX68ssvtW3bNj3xxBO67777tGfPHknSsmXLVL58eS1fvlzlypVTSEiI+vbt6/Bwe2DkyJEyxuipp57SxYsX07x/6dIlLVu2LN/qCQsL03///ZfmDwJcLTU8XP1D0RiTK3cv22w21a5dW1OmTFGRIkX0xx9/ZNi3WbNm2r59e5pLEb788ssbriMn2rRpIxcXF82YMSPDPultO0n66KOPcjxu6n+4Mxu3bdu2cnV11b59+1S3bt10X5mpWrWqpCs3Y2V37HvvvVeS0txUt3nzZkVHR6tly5aS8mbbVK5cWa+88opq1qxp/yzd6LZINW/ePIfpDRs26NChQw5PIwgJCdG2bdsc+u3evVu7du1yaAsLC9Pu3bu1Zs2a645boEABffTRRxo8eLB69uzpsO0feOABGWN09OjRdNerZs2a9r5Z+W2El5eXwsLCNGrUKF28eFF///23w/v79++Xv79/hmeAAWe7rc/IZmbfvn1asGCBjhw5osDAQEnSCy+8oJUrV2rOnDl68803tX//fh06dEhff/215s6dq+TkZA0dOlSPP/54lr6scHtIvUN7wIABCg0N1TPPPKM77rhDly5d0tatW/Xxxx+rRo0aat++fb7U06lTJ82ZM0f9+/fXrl271KJFC6WkpOi3335TtWrV9OSTT6p169Zyc3NTp06d9OKLL+q///7TjBkz9M8//+RozOXLl2v69Ol6+OGHVb58eRljtHjxYp05cybNrzKvNmTIEM2ePVthYWEaN26cAgICNH/+fPvjirL6K+LcEhISopdfflmvv/66Lly4YH9k1o4dOxQfH6+xY8eqatWqqlChgkaMGCFjjPz8/LRs2TJFRETkeNymTZuqW7duGj9+vI4fP64HHnhA7u7u2rp1qwoVKqRBgwYpJCRE48aN06hRo7R//37dd999Klq0qI4fP67ff/9dXl5eGjt2bIZjNGjQQJ6entq0aZPDNbtZGbtKlSp6+umn9cEHH6hAgQIKCwvTwYMH9eqrryooKMj+BIbc2Dbbtm3Ts88+qyeeeEKVKlWSm5ub1qxZo23bttnP7N/otkgVGRmpvn376oknntDhw4c1atQolS5dWgMGDLD36datm7p27aoBAwboscce06FDhzRx4kT72ftUQ4YM0cKFC/XQQw9pxIgRql+/vi5cuGB/GkF6T4uYNGmSChcurAEDBujcuXMaPny4mjRpoqefflq9evVSZGSk7rnnHnl5eSk2Nla//vqratasqWeeeUbSld+CLF68WDNmzFBoaKgKFCigunXr6qmnnpKnp6eaNGmiUqVKKS4uThMmTJCvr2+ayzE2bdqkZs2a3XZ/4Q4W4rTbzG4yksySJUvs01999ZWRZLy8vBxerq6u9jten3rqKSPJ7Nq1yz7fli1bjCSzc+fO/F4F3OSioqJMjx49TNmyZY2bm5vx8vIyderUMaNHjzYnTpyw98vOUws2b97s0C/17vtr7+a/dnkXLlwwo0ePNpUqVTJubm7G39/f3HvvvWbDhg32PsuWLTO1a9c2Hh4epnTp0mb48OHmhx9+SLP8rDy1YOfOnaZTp06mQoUKxtPT0/j6+pr69eub8PBwh37p3QG+fft206pVK+Ph4WH8/PxMnz59zGeffWYkmT///NNhPe+44440Y19b3408tSDV3LlzTb169YyHh4fx9vY2derUcVjejh07TOvWrU3hwoVN0aJFzRNPPGFiYmLS7MfU5Z88edJh+an798CBA/a25ORkM2XKFFOjRg3j5uZmfH19TaNGjcyyZcsc5l26dKlp0aKF8fHxMe7u7iY4ONg8/vjj5n//+1+a9bhWt27dTPXq1dO0Z2Xs5ORk8/bbb5vKlSubggULmmLFipmuXbuaw4cPOyzrRrfN8ePHTc+ePU3VqlWNl5eX8fb2NrVq1TJTpkwxly9fzpVtkbr9V61aZbp162aKFCliPD09Tbt27cyePXsc+qakpJiJEyea8uXLGw8PD1O3bl2zZs2adI+7f/75xwwePNiULVvWFCxY0JQoUcLcf//99p8XVz+14GrvvPNOmic4zJ492zRo0MB4eXkZT09PU6FCBdO9e3cTGRlp73P69Gnz+OOPmyJFihibzWb/LH/22WemRYsWJiAgwLi5uZnAwEDToUMHs23bNodx9+7daySZRYsWZbq9AGeyGWNMfgbnm5XNZtOSJUv08MMPS7ryIO0uXbro77//louLi0Nfb29vlSxZUmPGjNGbb77p8OzDCxcuqFChQlq1alWmZ5oA5NzTTz+tBQsW6NSpU3Jzc3N2ObeMyMhI1atXT5s2bVKDBg2cXQ6c7NVXX9XcuXO1b98+ubryC1zcnPhkZqBOnTpKTk7WiRMn1LRp03T7NGnSRJcvX9a+ffvsf3Em9bFEN3KDAYD/M27cOAUGBqp8+fI6d+6cli9frk8//VSvvPIKITaX1a1bVx06dNDrr79u/ytSuD2dOXNG06ZN0wcffECIxU3ttv50njt3zuFu2QMHDigqKkp+fn6qXLmyunTpou7du2vSpEmqU6eO4uPjtWbNGtWsWVPt2rVTq1atdNddd6l3796aOnWqUlJSNHDgQLVu3VqVK1d24poBt46CBQvqnXfe0ZEjR3T58mVVqlRJkydP1uDBg51d2i1p0qRJmjVrls6ePavChQs7uxw4yYEDBzRy5Eh17tzZ2aUAmbqtLy34+eef073AvkePHgoPD9elS5c0fvx4zZ07V0ePHpW/v78aNWqksWPH2u8MPXbsmAYNGqRVq1bZ7/6cNGmS/Pz88nt1AAAAbiu3dZAFAACAdfEcWQAAAFgSQRYAAACWdNvd7JWSkqJjx46pcOHCPOAZAADgJmOM0dmzZxUYGHjdP3xz2wXZY8eOZfg30gEAAHBzOHz4sMqUKZNpn9suyKY+Tubw4cPy8fFxcjUAAAC4WmJiooKCgrL0CMDbLsimXk7g4+NDkAUAALhJZeUSUG72AgAAgCURZAEAAGBJBFkAAABYEkEWAAAAlkSQBQAAgCURZAEAAGBJBFkAAABYEkEWAAAAlkSQBQAAgCURZAEAAGBJBFkAAABYklOD7Nq1a9W+fXsFBgbKZrNp6dKl150nKSlJo0aNUnBwsNzd3VWhQgXNnj0774sFAADATcXVmYOfP39etWvXVq9evfTYY49laZ4OHTro+PHjmjVrlipWrKgTJ07o8uXLeVwpAAAAbjZODbJhYWEKCwvLcv+VK1fql19+0f79++Xn5ydJCgkJyaPqclfIiO+dXcJt5+Bb9zu7BAAAkIcsdY3sd999p7p162rixIkqXbq0KleurBdeeEEXLlzIcJ6kpCQlJiY6vAAAAGB9Tj0jm1379+/Xr7/+Kg8PDy1ZskTx8fEaMGCATp8+neF1shMmTNDYsWPzuVIAAADkNUudkU1JSZHNZtO8efNUv359tWvXTpMnT1Z4eHiGZ2VHjhyphIQE++vw4cP5XDUAAADygqXOyJYqVUqlS5eWr6+vva1atWoyxujIkSOqVKlSmnnc3d3l7u6en2UCAAAgH1jqjGyTJk107NgxnTt3zt62e/duFShQQGXKlHFiZQAAAMhvTg2y586dU1RUlKKioiRJBw4cUFRUlGJiYiRduSyge/fu9v6dO3eWv7+/evXqpR07dmjt2rUaPny4evfuLU9PT2esAgAAAJzEqUE2MjJSderUUZ06dSRJw4YNU506dTR69GhJUmxsrD3USpK3t7ciIiJ05swZ1a1bV126dFH79u31/vvvO6V+AAAAOI/NGGOcXUR+SkxMlK+vrxISEuTj45Nv4/Ic2fzHc2QBALCe7GQ1S10jCwAAAKQiyAIAAMCSCLIAAACwJIIsAAAALIkgCwAAAEsiyAIAAMCSCLIAAACwJIIsAAAALIkgCwAAAEsiyAIAAMCSCLIAAACwJIIsAAAALIkgCwAAAEsiyAIAAMCSCLIAAACwJIIsAAAALIkgCwAAAEsiyAIAAMCSCLIAAACwJIIsAAAALIkgCwAAAEsiyAIAAMCSCLIAAACwJIIsAAAALIkgCwAAAEsiyAIAAMCSCLIAAACwJKcG2bVr16p9+/YKDAyUzWbT0qVLszzv+vXr5erqqjvvvDPP6gMAAMDNy6lB9vz586pdu7Y+/PDDbM2XkJCg7t27q2XLlnlUGQAAAG52rs4cPCwsTGFhYdmer1+/furcubNcXFyydRYXAAAAtw7LXSM7Z84c7du3T2PGjHF2KQAAAHAip56Rza49e/ZoxIgRWrdunVxds1Z6UlKSkpKS7NOJiYl5VR4AAADykWXOyCYnJ6tz584aO3asKleunOX5JkyYIF9fX/srKCgoD6sEAABAfrFMkD179qwiIyP17LPPytXVVa6urho3bpz+/PNPubq6as2aNenON3LkSCUkJNhfhw8fzufKAQAAkBcsc2mBj4+P/vrrL4e26dOna82aNfrmm29Urly5dOdzd3eXu7t7fpQIAACAfOTUIHvu3Dnt3bvXPn3gwAFFRUXJz89PZcuW1ciRI3X06FHNnTtXBQoUUI0aNRzmL1GihDw8PNK0AwAA4Nbn1CAbGRmpFi1a2KeHDRsmSerRo4fCw8MVGxurmJgYZ5UHAACAm5jNGGOcXUR+SkxMlK+vrxISEuTj45Nv44aM+D7fxsIVB9+639klAACAbMpOVrPMzV4AAADA1QiyAAAAsCSCLAAAACyJIAsAAABLIsgCAADAkgiyAAAAsCSCLAAAACyJIAsAAABLIsgCAADAkgiyAAAAsCSCLAAAACyJIAsAAABLIsgCAADAkgiyAAAAsCSCLAAAACyJIAsAAABLIsgCAADAkgiyAAAAsCSCLAAAACyJIAsAAABLIsgCAADAkgiyAAAAsCSCLAAAACyJIAsAAABLIsgCAADAkgiyAAAAsCSCLAAAACyJIAsAAABLcmqQXbt2rdq3b6/AwEDZbDYtXbo00/6LFy9W69atVbx4cfn4+KhRo0b68ccf86dYAAAA3FScGmTPnz+v2rVr68MPP8xS/7Vr16p169ZasWKFtmzZohYtWqh9+/baunVrHlcKAACAm42rMwcPCwtTWFhYlvtPnTrVYfrNN9/Ut99+q2XLlqlOnTq5XB0AAABuZk4NsjcqJSVFZ8+elZ+fX4Z9kpKSlJSUZJ9OTEzMj9IAAACQxyx9s9ekSZN0/vx5dejQIcM+EyZMkK+vr/0VFBSUjxUCAAAgr1g2yC5YsECvvfaaFi5cqBIlSmTYb+TIkUpISLC/Dh8+nI9VAgAAIK9Y8tKChQsXqk+fPvr666/VqlWrTPu6u7vL3d09nyoDAABAfrHcGdkFCxaoZ8+emj9/vu6//35nlwMAAAAnceoZ2XPnzmnv3r326QMHDigqKkp+fn4qW7asRo4cqaNHj2ru3LmSroTY7t2767333lPDhg0VFxcnSfL09JSvr69T1gEAAADO4dQzspGRkapTp4790VnDhg1TnTp1NHr0aElSbGysYmJi7P0/+ugjXb58WQMHDlSpUqXsr8GDBzulfgAAADiPU8/INm/eXMaYDN8PDw93mP7555/ztiAAAABYhuWukQUAAAAkgiwAAAAsiiALAAAASyLIAgAAwJIIsgAAALAkgiwAAAAsiSALAAAASyLIAgAAwJIIsgAAALAkgiwAAAAsiSALAAAASyLIAgAAwJIIsgAAALAkgiwAAAAsiSALAAAASyLIAgAAwJIIsgAAALAkgiwAAAAsiSALAAAASyLIAgAAwJIIsgAAALAkgiwAAAAsiSALAAAASyLIAgAAwJIIsgAAALAkgiwAAAAsiSALAAAAS3JqkF27dq3at2+vwMBA2Ww2LV269Lrz/PLLLwoNDZWHh4fKly+vmTNn5n2hAAAAuOk4NcieP39etWvX1ocffpil/gcOHFC7du3UtGlTbd26VS+//LKee+45LVq0KI8rBQAAwM3G1ZmDh4WFKSwsLMv9Z86cqbJly2rq1KmSpGrVqikyMlLvvvuuHnvssTyqEgAAADcjS10ju3HjRrVp08ahrW3btoqMjNSlS5ecVBUAAACcwalnZLMrLi5OAQEBDm0BAQG6fPmy4uPjVapUqTTzJCUlKSkpyT6dmJiY53UCAAAg71nqjKwk2Ww2h2ljTLrtqSZMmCBfX1/7KygoKM9rBAAAQN6zVJAtWbKk4uLiHNpOnDghV1dX+fv7pzvPyJEjlZCQYH8dPnw4P0oFAABAHrPUpQWNGjXSsmXLHNpWrVqlunXrqmDBgunO4+7uLnd39/woDwAAAPnIqWdkz507p6ioKEVFRUm68nitqKgoxcTESLpyNrV79+72/v3799ehQ4c0bNgwRUdHa/bs2Zo1a5ZeeOEFZ5QPAAAAJ3LqGdnIyEi1aNHCPj1s2DBJUo8ePRQeHq7Y2Fh7qJWkcuXKacWKFRo6dKimTZumwMBAvf/++zx6CwAA4DZkM6l3S90mEhMT5evrq4SEBPn4+OTbuCEjvs+3sXDFwbfud3YJAAAgm7KT1Sx1sxcAAACQiiALAAAASyLIAgAAwJIIsgAAALAkgiwAAAAsKUdBtnz58jp16lSa9jNnzqh8+fI3XBQAAABwPTkKsgcPHlRycnKa9qSkJB09evSGiwIAAACuJ1t/EOG7776z//vHH3+Ur6+vfTo5OVmrV69WSEhIrhUHAAAAZCRbQfbhhx+WJNlsNvXo0cPhvYIFCyokJESTJk3KteIAAACAjGQryKakpEi68qdiN2/erGLFiuVJUQAAAMD1ZCvIpjpw4EBu1wEAAABkS46CrCStXr1aq1ev1okTJ+xnalPNnj37hgsDAAAAMpOjIDt27FiNGzdOdevWValSpWSz2XK7LgAAACBTOQqyM2fOVHh4uLp165bb9QAAAABZkqPnyF68eFGNGzfO7VoAAACALMtRkO3bt6/mz5+f27UAAAAAWZajSwv+++8/ffzxx/rf//6nWrVqqWDBgg7vT548OVeKAwAAADKSoyC7bds23XnnnZKk7du3O7zHjV8AAADIDzkKsj/99FNu1wEAAABkS46ukQUAAACcLUdnZFu0aJHpJQRr1qzJcUEAAABAVuQoyKZeH5vq0qVLioqK0vbt29WjR4/cqAsAAADIVI6C7JQpU9Jtf+2113Tu3LkbKggAAADIily9RrZr166aPXt2bi4SAAAASFeuBtmNGzfKw8MjNxcJAAAApCtHlxY8+uijDtPGGMXGxioyMlKvvvpqrhQGAAAAZCZHQdbX19dhukCBAqpSpYrGjRunNm3a5EphAAAAQGZyFGTnzJmT23UAAAAA2XJD18hu2bJFX3zxhebNm6etW7fmaBnTp09XuXLl5OHhodDQUK1bty7T/vPmzVPt2rVVqFAhlSpVSr169dKpU6dyNDYAAACsK0dB9sSJE7r33ntVr149Pffcc3r22WcVGhqqli1b6uTJk1lezsKFCzVkyBCNGjVKW7duVdOmTRUWFqaYmJh0+//666/q3r27+vTpo7///ltff/21Nm/erL59++ZkNQAAAGBhOQqygwYNUmJiov7++2+dPn1a//zzj7Zv367ExEQ999xzWV7O5MmT1adPH/Xt21fVqlXT1KlTFRQUpBkzZqTbf9OmTQoJCdFzzz2ncuXK6e6771a/fv0UGRmZk9UAAACAheUoyK5cuVIzZsxQtWrV7G3Vq1fXtGnT9MMPP2RpGRcvXtSWLVvS3BzWpk0bbdiwId15GjdurCNHjmjFihUyxuj48eP65ptvdP/992c4TlJSkhITEx1eAAAAsL4cBdmUlBQVLFgwTXvBggWVkpKSpWXEx8crOTlZAQEBDu0BAQGKi4tLd57GjRtr3rx56tixo9zc3FSyZEkVKVJEH3zwQYbjTJgwQb6+vvZXUFBQluoDAADAzS1HQfbee+/V4MGDdezYMXvb0aNHNXToULVs2TJby7LZbA7Txpg0bal27Nih5557TqNHj9aWLVu0cuVKHThwQP37989w+SNHjlRCQoL9dfjw4WzVBwAAgJtTjh6/9eGHH+qhhx5SSEiIgoKCZLPZFBMTo5o1a+qLL77I0jKKFSsmFxeXNGdfT5w4keYsbaoJEyaoSZMmGj58uCSpVq1a8vLyUtOmTTV+/HiVKlUqzTzu7u5yd3fP5hoCAADgZpejIBsUFKQ//vhDERER2rlzp4wxql69ulq1apXlZbi5uSk0NFQRERF65JFH7O0RERF66KGH0p3n33//laurY8kuLi6SrpzJBQAAwO0jW5cWrFmzRtWrV7ffMNW6dWsNGjRIzz33nOrVq6c77rjjus+BvdqwYcP06aefavbs2YqOjtbQoUMVExNjv1Rg5MiR6t69u71/+/bttXjxYs2YMUP79+/X+vXr9dxzz6l+/foKDAzMzqoAAADA4rJ1Rnbq1Kl66qmn5OPjk+Y9X19f9evXT5MnT1bTpk2ztLyOHTvq1KlTGjdunGJjY1WjRg2tWLFCwcHBkqTY2FiHZ8r27NlTZ8+e1Ycffqjnn39eRYoU0b333qu33347O6sBAACAW4DNZON38sHBwVq5cqXDY7eutnPnTrVp0ybDP2hwM0hMTJSvr68SEhLSDeR5JWTE9/k2Fq44+FbGj2UDAAA3p+xktWxdWnD8+PF0H7uVytXVNVt/2QsAAADIqWwF2dKlS+uvv/7K8P1t27al++QAAAAAILdlK8i2a9dOo0eP1n///ZfmvQsXLmjMmDF64IEHcq04AAAAICPZutnrlVde0eLFi1W5cmU9++yzqlKlimw2m6KjozVt2jQlJydr1KhReVUrAAAAYJetIBsQEKANGzbomWee0ciRI+3PbrXZbGrbtq2mT5+e4R8zAAAAAHJTtv8gQnBwsFasWKF//vlHe/fulTFGlSpVUtGiRfOiPgAAACBdOfrLXpJUtGhR1atXLzdrAQAAALIsWzd7AQAAADcLgiwAAAAsiSALAAAASyLIAgAAwJIIsgAAALAkgiwAAAAsiSALAAAASyLIAgAAwJIIsgAAALAkgiwAAAAsiSALAAAASyLIAgAAwJIIsgAAALAkgiwAAAAsiSALAAAASyLIAgAAwJIIsgAAALAkgiwAAAAsiSALAAAASyLIAgAAwJKcHmSnT5+ucuXKycPDQ6GhoVq3bl2m/ZOSkjRq1CgFBwfL3d1dFSpU0OzZs/OpWgAAANwsXJ05+MKFCzVkyBBNnz5dTZo00UcffaSwsDDt2LFDZcuWTXeeDh066Pjx45o1a5YqVqyoEydO6PLly/lcOQAAAJzNZowxzhq8QYMGuuuuuzRjxgx7W7Vq1fTwww9rwoQJafqvXLlSTz75pPbv3y8/P78cjZmYmChfX18lJCTIx8cnx7VnV8iI7/NtLFxx8K37nV0CAADIpuxkNaddWnDx4kVt2bJFbdq0cWhv06aNNmzYkO483333nerWrauJEyeqdOnSqly5sl544QVduHAhP0oGAADATcRplxbEx8crOTlZAQEBDu0BAQGKi4tLd579+/fr119/lYeHh5YsWaL4+HgNGDBAp0+fzvA62aSkJCUlJdmnExMTc28lAAAA4DROv9nLZrM5TBtj0rSlSklJkc1m07x581S/fn21a9dOkydPVnh4eIZnZSdMmCBfX1/7KygoKNfXAQAAAPnPaUG2WLFicnFxSXP29cSJE2nO0qYqVaqUSpcuLV9fX3tbtWrVZIzRkSNH0p1n5MiRSkhIsL8OHz6ceysBAAAAp3FakHVzc1NoaKgiIiIc2iMiItS4ceN052nSpImOHTumc+fO2dt2796tAgUKqEyZMunO4+7uLh8fH4cXAAAArM+plxYMGzZMn376qWbPnq3o6GgNHTpUMTEx6t+/v6QrZ1O7d+9u79+5c2f5+/urV69e2rFjh9auXavhw4erd+/e8vT0dNZqAAAAwAmc+hzZjh076tSpUxo3bpxiY2NVo0YNrVixQsHBwZKk2NhYxcTE2Pt7e3srIiJCgwYNUt26deXv768OHTpo/PjxzloFAAAAOIlTnyPrDDxH9vbBc2QBALAeSzxHFgAAALgRBFkAAABYEkEWAAAAlkSQBQAAgCURZAEAAGBJBFkAAABYEkEWAAAAlkSQBQAAgCURZAEAAGBJBFkAAABYEkEWAAAAlkSQBQAAgCURZAEAAGBJBFkAAABYEkEWAAAAlkSQBQAAgCURZAEAAGBJBFkAAABYEkEWAAAAlkSQBQAAgCURZAEAAGBJBFkAAABYEkEWAAAAlkSQBQAAgCURZAEAAGBJBFkAAABYEkEWAAAAlkSQBQAAgCU5PchOnz5d5cqVk4eHh0JDQ7Vu3boszbd+/Xq5urrqzjvvzNsCAQAAcFNyapBduHChhgwZolGjRmnr1q1q2rSpwsLCFBMTk+l8CQkJ6t69u1q2bJlPlQIAAOBm49QgO3nyZPXp00d9+/ZVtWrVNHXqVAUFBWnGjBmZztevXz917txZjRo1yqdKAQAAcLNxWpC9ePGitmzZojZt2ji0t2nTRhs2bMhwvjlz5mjfvn0aM2ZMlsZJSkpSYmKiwwsAAADW57QgGx8fr+TkZAUEBDi0BwQEKC4uLt159uzZoxEjRmjevHlydXXN0jgTJkyQr6+v/RUUFHTDtQMAAMD5nH6zl81mc5g2xqRpk6Tk5GR17txZY8eOVeXKlbO8/JEjRyohIcH+Onz48A3XDAAAAOfL2mnNPFCsWDG5uLikOft64sSJNGdpJens2bOKjIzU1q1b9eyzz0qSUlJSZIyRq6urVq1apXvvvTfNfO7u7nJ3d8+blQAAAIDTOO2MrJubm0JDQxUREeHQHhERocaNG6fp7+Pjo7/++ktRUVH2V//+/VWlShVFRUWpQYMG+VU6AAAAbgJOOyMrScOGDVO3bt1Ut25dNWrUSB9//LFiYmLUv39/SVcuCzh69Kjmzp2rAgUKqEaNGg7zlyhRQh4eHmnaAQAAcOtzapDt2LGjTp06pXHjxik2NlY1atTQihUrFBwcLEmKjY297jNlAQAAcHuyGWOMs4vIT4mJifL19VVCQoJ8fHzybdyQEd/n21i44uBb9zu7BAAAkE3ZyWpOf2oBAAAAkBMEWQAAAFgSQRYAAACWRJAFAACAJRFkAQAAYEkEWQAAAFgSQRYAAACWRJAFAACAJRFkAQAAYEkEWQAAAFgSQRYAAACWRJAFAACAJRFkAQAAYEkEWQAAAFgSQRYAAACWRJAFAACAJRFkAQAAYEkEWQAAAFgSQRYAAACWRJAFAACAJRFkAQAAYEkEWQAAAFgSQRYAAACWRJAFAACAJRFkAQAAYEkEWQAAAFgSQRYAAACW5PQgO336dJUrV04eHh4KDQ3VunXrMuy7ePFitW7dWsWLF5ePj48aNWqkH3/8MR+rBQAAwM3CqUF24cKFGjJkiEaNGqWtW7eqadOmCgsLU0xMTLr9165dq9atW2vFihXasmWLWrRoofbt22vr1q35XDkAAACczWaMMc4avEGDBrrrrrs0Y8YMe1u1atX08MMPa8KECVlaxh133KGOHTtq9OjRWeqfmJgoX19fJSQkyMfHJ0d150TIiO/zbSxccfCt+51dAgAAyKbsZDWnnZG9ePGitmzZojZt2ji0t2nTRhs2bMjSMlJSUnT27Fn5+fnlRYkAAAC4ibk6a+D4+HglJycrICDAoT0gIEBxcXFZWsakSZN0/vx5dejQIcM+SUlJSkpKsk8nJibmrGAAAADcVJx+s5fNZnOYNsakaUvPggUL9Nprr2nhwoUqUaJEhv0mTJggX19f+ysoKOiGawYAAIDzOS3IFitWTC4uLmnOvp44cSLNWdprLVy4UH369NFXX32lVq1aZdp35MiRSkhIsL8OHz58w7UDAADA+ZwWZN3c3BQaGqqIiAiH9oiICDVu3DjD+RYsWKCePXtq/vz5uv/+69/M4+7uLh8fH4cXAAAArM9p18hK0rBhw9StWzfVrVtXjRo10scff6yYmBj1799f0pWzqUePHtXcuXMlXQmx3bt313vvvaeGDRvaz+Z6enrK19fXaesBAACA/OfUINuxY0edOnVK48aNU2xsrGrUqKEVK1YoODhYkhQbG+vwTNmPPvpIly9f1sCBAzVw4EB7e48ePRQeHp7f5QMAAMCJnPocWWfgObK3D54jCwCA9VjiObIAAADAjSDIAgAAwJIIsgAAALAkgiwAAAAsiSALAAAASyLIAgAAwJIIsgAAALAkgiwAAAAsiSALAAAASyLIAgAAwJIIsgAAALAkgiwAAAAsiSALAAAASyLIAgAAwJIIsgAAALAkgiwAAAAsiSALAAAASyLIAgAAwJIIsgAAALAkgiwAAAAsiSALAAAASyLIAgAAwJIIsgAAALAkgiwAAAAsiSALAAAASyLIAgAAwJIIsgAAALAkV2cXAFhVyIjvnV3CbefgW/c7uwQAwE3E6Wdkp0+frnLlysnDw0OhoaFat25dpv1/+eUXhYaGysPDQ+XLl9fMmTPzqVIAAADcTJwaZBcuXKghQ4Zo1KhR2rp1q5o2baqwsDDFxMSk2//AgQNq166dmjZtqq1bt+rll1/Wc889p0WLFuVz5QAAAHA2pwbZyZMnq0+fPurbt6+qVaumqVOnKigoSDNmzEi3/8yZM1W2bFlNnTpV1apVU9++fdW7d2+9++67+Vw5AAAAnM1p18hevHhRW7Zs0YgRIxza27Rpow0bNqQ7z8aNG9WmTRuHtrZt22rWrFm6dOmSChYsmGaepKQkJSUl2acTEhIkSYmJiTe6CtmSkvRvvo6HvN/H7NP8l9/HLQAg/6V+1xtjrtvXaUE2Pj5eycnJCggIcGgPCAhQXFxcuvPExcWl2//y5cuKj49XqVKl0swzYcIEjR07Nk17UFDQDVQPK/Cd6uwKkNvYpwBw+zh79qx8fX0z7eP0pxbYbDaHaWNMmrbr9U+vPdXIkSM1bNgw+3RKSopOnz4tf3//TMfBlf8RBQUF6fDhw/Lx8XF2Ocgl7NdbD/v01sM+vfWwT7POGKOzZ88qMDDwun2dFmSLFSsmFxeXNGdfT5w4keasa6qSJUum29/V1VX+/v7pzuPu7i53d3eHtiJFiuS88NuQj48PB90tiP1662Gf3nrYp7ce9mnWXO9MbCqn3ezl5uam0NBQRUREOLRHRESocePG6c7TqFGjNP1XrVqlunXrpnt9LAAAAG5dTn1qwbBhw/Tpp59q9uzZio6O1tChQxUTE6P+/ftLunJZQPfu3e39+/fvr0OHDmnYsGGKjo7W7NmzNWvWLL3wwgvOWgUAAAA4iVOvke3YsaNOnTqlcePGKTY2VjVq1NCKFSsUHBwsSYqNjXV4pmy5cuW0YsUKDR06VNOmTVNgYKDef/99PfbYY85ahVuau7u7xowZk+bSDFgb+/XWwz699bBPbz3s07xhM1l5tgEAAABwk3H6n6gFAAAAcoIgCwAAAEsiyAIAAMCSCLJIIzw8nGft3sRCQkI0derUfBlrzZo1qlq1qlJSUvJlvFT16tXT4sWL83VMpNW8eXMNGTLE2WXkGyutb259T99zzz2aP3/+jReUDcuXL1edOnXy/Xslrxlj9PTTT8vPz082m01RUVFpPlO58f2dnz8DrIAgews7fPiw+vTpo8DAQLm5uSk4OFiDBw/WqVOn7H04IHJHz549ZbPZ7C9/f3/dd9992rZtW66PtXnzZj399NO5vtz0vPjiixo1apQKFMjfr4pXX31VI0aMsMwPuowC0NKlS2+KvyCY0+N88eLFev3113O/oBt09fFWsGBBlS9fXi+88ILOnz+fpfl//vln2Ww2nTlzJm8LzUMdO3bU7t27b2gZy5cvV1xcnJ588slcqiprHnjgAdlstnwN0HFxcRo0aJDKly8vd3d3BQUFqX379lq9enWujbFy5UqFh4dr+fLl9icxXSs7398Z/WclP38GWAFB9ha1f/9+1a1bV7t379aCBQu0d+9ezZw5U6tXr1ajRo10+vTpfK/p0qVL+T5mfrrvvvsUGxur2NhYrV69Wq6urnrggQdyfZzixYurUKFCub7ca23YsEF79uzRE088kedjXev+++9XQkKCfvzxx3wf+1Zy8eLFG5rfz89PhQsXzqVqclfq8bZ//36NHz9e06dPv62eKe7p6akSJUrc0DLef/999erVK9//oypJvXr10gcffJAvYx08eFChoaFas2aNJk6cqL/++ksrV65UixYtNHDgwFwbZ9++fSpVqpQaN26skiVLytU17RNOc+P7O79+BliGwS3pvvvuM2XKlDH//vuvQ3tsbKwpVKiQ6d+/v2nWrJmR5PAyxpg5c+YYX19fs3LlSlO1alXj5eVl2rZta44dO+awrNmzZ5uqVasad3d3U6VKFTNt2jT7ewcOHDCSzMKFC02zZs2Mu7u7mT17dt6vuJP06NHDPPTQQw5ta9euNZLMiRMn7G1HjhwxHTp0MEWKFDF+fn7mwQcfNAcOHEiznHfeeceULFnS+Pn5mQEDBpiLFy/a+wQHB5spU6bYp6Ojo02TJk2Mu7u7qVatmomIiDCSzJIlS4wx/7cvFi1aZJo3b248PT1NrVq1zIYNGzJdp0GDBpnHH3/coW3v3r3mwQcfNCVKlDBeXl6mbt26JiIiwqFPcHCweeONN0yvXr2Mt7e3CQoKMh999JH9/RYtWpiBAwc6zBMfH2/c3NzM6tWr7W09e/Y03bp1y7TGm0WzZs3M4MGD07QvWbLEflyNGTPG1K5d28ydO9cEBwcbHx8f07FjR5OYmGjvn5ycbN566y1ToUIF4+bmZoKCgsz48ePt72f18/Pmm2+aUqVKmeDg4AyP8/j4ePPkk0+a0qVLG09PT1OjRg0zf/78TNfrevs2v6R3vPXt29eULFnSGGPM559/bkJDQ423t7cJCAgwnTp1MsePHzfG/N/xcPWrR48expgr6zto0CAzfPhwU7RoURMQEGDGjBnjMM6hQ4fMgw8+aLy8vEzhwoXNE088YeLi4hz6vP7666Z48eLG29vb9OnTx7z00kumdu3aDn2y8v2Z2TGb+j19tW+//daEhoYad3d34+/vbx555JEMt+HJkyeNzWYz27dvd2ifNGmSqVGjhilUqJApU6aMeeaZZ8zZs2fTjJvRz4dffvnFuLq6mtjYWIflDhs2zDRt2tQ+ffDgQSPJ7Nu3L8Mac0tYWJgpXbq0OXfuXJr3/vnnH2PM9ffr9Y7fHj16OHymgoODjTHpH0NXf3//888/5qmnnjIlSpQw7u7u5o477jDLli0zP/30U5rPaepn8dplSDKffPKJefjhh42np6epWLGi+fbbbx3W89tvvzUVK1Y0Hh4epnnz5iY8PNxIsq+/lRFkb0GnTp0yNpvNvPnmm+m+/9RTT5miRYua+Ph4U6ZMGTNu3DgTGxtr/+KZM2eOKViwoGnVqpXZvHmz2bJli6lWrZrp3LmzfRkff/yxKVWqlFm0aJHZv3+/WbRokfHz8zPh4eHGmP/7Ig4JCbH3OXr0aN6vvJNc+4P17Nmzpl+/fqZixYomOTnZGGPM+fPnTaVKlUzv3r3Ntm3bzI4dO0znzp1NlSpVTFJSkn05Pj4+pn///iY6OtosW7bMFCpUyHz88cf2ZV/9JZacnGyqVKliWrdubaKiosy6detM/fr10w2yVatWNcuXLze7du0yjz/+uAkODjaXLl3KcJ1q165t3nrrLYe2qKgoM3PmTLNt2zaze/duM2rUKOPh4WEOHTrkUJ+fn5+ZNm2a2bNnj5kwYYIpUKCAiY6ONsYYM2/ePFO0aFHz33//2ed57733TEhIiElJSbG3TZ8+3YSEhGRjLzhPVoOst7e3efTRR81ff/1l1q5da0qWLGlefvlle/8XX3zRFC1a1ISHh5u9e/eadevWmU8++cQYk/XPj7e3t+nWrZvZvn27+euvv8ypU6fSPc6PHDli3nnnHbN161azb98+8/777xsXFxezadOmDNfrevs2v6QXZAcNGmT8/f2NMcbMmjXLrFixwuzbt89s3LjRNGzY0ISFhRljjLl8+bJZtGiRkWR27dplYmNjzZkzZ4wxV9bXx8fHvPbaa2b37t3ms88+MzabzaxatcoYY0xKSoqpU6eOufvuu01kZKTZtGmTueuuu0yzZs3sdXzxxRfGw8PDzJ492+zatcuMHTvW+Pj4OATZrH5/ZnbMXhtkly9fblxcXMzo0aPNjh07TFRUlHnjjTcy3IZLliwxXl5e9u+nVFOmTDFr1qwx+/fvN6tXrzZVqlQxzzzzjP39rPx8qFy5spk4caJ9+tKlS6ZEiRJpTmaUKFHCvs555Xo/D43J2n693vF75swZM27cOFOmTBkTGxtrP4GRWZBNTk42DRs2NHfccYdZtWqV2bdvn1m2bJlZsWKFSUpKMlOnTjU+Pj724zb1PxTpBdkyZcqY+fPnmz179pjnnnvOeHt7m1OnThljrnyeChYsaF544QWzc+dOs2DBAlO6dGmCLG5emzZtcggy15o8ebKRZI4fP57mgDDmyheVJLN3715727Rp00xAQIB9OigoKM3Zm9dff900atTIGPN/X8RTp07NnZW6yfXo0cO4uLgYLy8v4+XlZSSZUqVKmS1bttj7zJo1y1SpUsUhrCUlJRlPT0/z448/2pcTHBxsLl++bO/zxBNPmI4dO9qnr95nP/zwQ5qzHxmdkf3000/tff7++28jKdMA4uvra+bOnXvdda9evbr54IMPHOrr2rWrfTolJcWUKFHCzJgxwxhjzH///Wf8/PzMwoUL7X3uvPNO89prrzks99tvvzUFChRI84P2ZpTVIFuoUCGHM7DDhw83DRo0MMYYk5iYaNzd3e3B9VpZ/fwEBATYg22q9I7z9LRr1848//zzGa7X9fZtfrk2yP7222/G39/fdOjQId3+v//+u5FkDwKpZ7uu/SHerFkzc/fddzu01atXz7z00kvGGGNWrVplXFxcTExMjP391GPp999/N8YY06BBgzS/cWjSpIlDkM3q92dmx+y1QbZRo0amS5cu6a5/eqZMmWLKly9/3X5fffWV/T8IqeNe7+fD22+/bapVq2afXrp0qfH29k5zRrROnTppjvvc9ttvvxlJZvHixRn2ycp+vd7xa8yVbZp6JjZVZkH2xx9/NAUKFDC7du1Kt670zrpfuwxjrgTZV155xT597tw5Y7PZzA8//GCMMeall14yNWrUcFjGqFGjbpkgyzWytyHz//+YW2Y3oRQqVEgVKlSwT5cqVUonTpyQJJ08edJ+I5m3t7f9NX78eO3bt89hOXXr1s2DNbg5tWjRQlFRUYqKitJvv/2mNm3aKCwsTIcOHZIkbdmyRXv37lXhwoXt28zPz0///fefw3a744475OLiYp++ettfa9euXQoKClLJkiXtbfXr10+3b61atRyWKSnD5UrShQsX5OHh4dB2/vx5vfjii6pevbqKFCkib29v7dy50+FPSV87ls1mU8mSJe1jubu7q2vXrpo9e7YkKSoqSn/++ad69uzpsAxPT0+lpKQoKSkpwxqtJiQkxOGa06v3bXR0tJKSktSyZct0583q56dmzZpyc3O7bi3Jycl64403VKtWLfn7+8vb21urVq1Ksy+vldm+zU/Lly+Xt7e3PDw81KhRI91zzz32ay63bt2qhx56SMHBwSpcuLCaN28uSdddN8lx/aS0+ygoKEhBQUH291OPhejoaElXjslrj8Grp7Pz/ZmdYzYqKirDz0560ju+Jemnn35S69atVbp0aRUuXFjdu3fXqVOnHG6ky+zng3TlZry9e/dq06ZNkqTZs2erQ4cO8vLychjL09NT//77b5Zrzoms/LzLyn6VMj9+cyIqKkplypRR5cqVc7yMVFd/Vry8vFS4cGF7bbt27VK9evUc+mf0c8KK0l6JDMurWLGibDabduzYoYcffjjN+zt37lTRokVVrFixDJdRsGBBh2mbzWb/Qki9k/yTTz5RgwYNHPpdHcAkpfniupV5eXmpYsWK9unQ0FD5+vrqk08+0fjx45WSkqLQ0FDNmzcvzbzFixe3/zu9bZ/R3fvGmCzfFX/1clPnyeypAMWKFdM///zj0DZ8+HD9+OOPevfdd1WxYkV5enrq8ccfT3NT0fXWoW/fvrrzzjt15MgRzZ49Wy1btlRwcLDDPKdPn1ahQoXk6emZpfVzJh8fHyUkJKRpP3PmjHx8fOzTmW2X661nVj8/WT3mJk2apClTpmjq1KmqWbOmvLy8NGTIkOveIJadz2deatGihWbMmKGCBQsqMDDQXtf58+fVpk0btWnTRl988YWKFy+umJgYtW3bNks3v2W2fhkdb9e2X9sn9btTyt73Z3aO2eweJ+kd34cOHVK7du3Uv39/vf766/Lz89Ovv/6qPn36ONysm9nPB0kqUaKE2rdvrzlz5qh8+fJasWKFfv755zQ1nD592uGzmxcqVaokm82m6OjodH8eSlnfr7n92c/N77bsfm6v3l9WxxnZW5C/v79at26t6dOn68KFCw7vxcXFad68eerYsaNsNpvc3NyUnJycreUHBASodOnS2r9/vypWrOjwKleuXG6uiqXZbDYVKFDAvg/uuusu7dmzRyVKlEiz3Xx9fXM0RtWqVRUTE6Pjx4/b2zZv3pwr9depU0c7duxwaFu3bp169uypRx55RDVr1lTJkiV18ODBbC+7Zs2aqlu3rj755BPNnz9fvXv3TtNn+/btuuuuu3Jafr6qWrWqIiMj07Rv3rxZVapUydIyKlWqJE9PzwwfB3Qjn5/0jvN169bpoYceUteuXVW7dm2VL19ee/bsyVKtN4PU/zgGBwc7/BDfuXOn4uPj9dZbb6lp06aqWrVqmrNmqWess/vdV716dcXExOjw4cP2th07dighIUHVqlWTJFWpUkW///67w3xXfzby6vuzVq1a2XqUVJ06dRQXF+cQZiMjI3X58mVNmjRJDRs2VOXKlXXs2LEc1dO3b199+eWX+uijj1ShQgU1adLE4f3U3yTUqVMnR8vPKj8/P7Vt21bTpk1L9/FsZ86cydJ+zQu1atXSkSNHMnyMWk5+PqenatWqaX4upPd9ZVUE2VvUhx9+qKSkJLVt21Zr167V4cOHtXLlSvuvjN544w1JV35VsnbtWh09elTx8fFZXv5rr72mCRMm6L333tPu3bv1119/ac6cOZo8eXJerdJNLykpSXFxcYqLi1N0dLQGDRqkc+fOqX379pKkLl26qFixYnrooYe0bt06HThwQL/88osGDx6sI0eO5GjM1q1bq0KFCurRo4e2bdum9evXa9SoUZIy/1VaVrRt21a//vqrQ1vFihW1ePFi++UAnTt3zvEZib59++qtt95ScnKyHnnkkTTvr1u3Tm3atMnRsvPbgAEDtG/fPg0cOFB//vmndu/erWnTpmnWrFkaPnx4lpbh4eGhl156SS+++KLmzp2rffv2adOmTZo1a5akG/v8pHecV6xYUREREdqwYYOio6PVr18/xcXF3diGuAmULVtWbm5u+uCDD7R//3599913aZ6FGxwcLJvNpuXLl+vkyZM6d+5clpbdqlUr1apVS126dNEff/yh33//Xd27d1ezZs3sl1ENGjRIs2bN0meffaY9e/Zo/Pjx2rZtm8PxmBffn2PGjNGCBQs0ZswYRUdH66+//tLEiRMz7F+nTh0VL15c69evt7dVqFBBly9ftm+7zz//XDNnzsxRPW3btpWvr6/Gjx+vXr16pXl/06ZNcnd3V6NGjXK0/OyYPn26kpOTVb9+fS1atEh79uxRdHS03n//fTVq1ChL+zUvNGvWTPfcc48ee+wxRURE6MCBA/rhhx+0cuVKSVeO23Pnzmn16tWKj4/P8WUY/fr1086dO/XSSy9p9+7d+uqrrxQeHi7pxn9O3AwIsreoSpUqKTIyUhUqVFDHjh1VoUIFPf3002rRooU2btwoPz8/SdK4ceN08OBBVahQIVu/4unbt68+/fRThYeHq2bNmmrWrJnCw8Nv6zOyK1euVKlSpVSqVCk1aNBAmzdv1tdff22/Pq9QoUJau3atypYtq0cffVTVqlVT7969deHCBYdfP2eHi4uLli5dqnPnzqlevXrq27evXnnlFUlK9/q37Ojatat27NihXbt22dumTJmiokWLqnHjxmrfvr3atm2b47OmnTp1kqurqzp37pym1qNHj2rDhg3p/gC8GYWEhGjdunXat2+f2rRpo3r16ik8PFzh4eHZeg7vq6++queff16jR49WtWrV1LFjR/vZxBv5/KR3nL/66qu666671LZtWzVv3lwlS5bM8FevVlK8eHGFh4fr66+/VvXq1fXWW2/p3XffdehTunRpjR07ViNGjFBAQICeffbZLC3bZrNp6dKlKlq0qO655x61atVK5cuX18KFC+19unTpopEjR+qFF17QXXfdpQMHDqhnz54On/G8+P5s3ry5vv76a3333Xe68847de+99+q3337LsL+Li4t69+7tcKnKnXfeqcmTJ+vtt99WjRo1NG/ePE2YMCFH9RQoUEA9e/ZUcnKyunfvnub9BQsWqEuXLvnyPNRy5crpjz/+UIsWLfT888+rRo0aat26tVavXq0ZM2Zkab/mlUWLFqlevXrq1KmTqlevrhdffNF+FrZx48bq37+/OnbsqOLFi2f6H5PMlCtXTt98840WL16sWrVqacaMGfYTHu7u7rm2Ls5iM7fShRIAtH79et19993au3evww0ZOfHiiy8qISFBH330US5V938OHz6skJAQbd68OU0YHj58uBISEvTxxx/n+rhAfmvdurVKliypzz//3NmlODh+/LjuuOMObdmyJc016rnhqaee0vHjx/Xdd985tJ88edJ+Oc7tfPLDmd544w3NnDnT4XIKq+JmL8DilixZIm9vb1WqVEl79+7V4MGD1aRJkxsOsZI0atQoTZs2TcnJyWluRMmpS5cuKTY2ViNGjFDDhg3TPaNbokSJ2+qvNOHW8e+//2rmzJlq27atXFxctGDBAv3vf/9TRESEs0tLIyAgQLNmzVJMTEyuBtmEhARt3rxZ8+bN07fffpvm/QMHDmj69OmE2Hw0ffp01atXT/7+/lq/fr3eeeedLP8m4mbHGVnA4ubOnavXX39dhw8fVrFixdSqVStNmjRJ/v7+zi4tXT///LNatGihypUr65tvvlHNmjWdXRKQay5cuKD27dvrjz/+UFJSkqpUqaJXXnlFjz76qLNLyzfNmzfX77//rn79+mnKlCnOLgeShg4dqoULF+r06dMqW7asunXrppEjR6b7Z3SthiALAAAAS+JmLwAAAFgSQRYAAACWRJAFAACAJRFkAQAAYEkEWQAAAFgSQRYAAACWRJAFAACAJRFkAQAAYEkEWQAAAFjS/wN0hTvUXq9FpQAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 700x400 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Visualize Distribution of Clinically Annotated Variants\n",
    "import os, glob, re, pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "OUTDIR = \"per_chrom_outputs_new\"\n",
    "\n",
    "# Load all annotated per-chrom files\n",
    "paths = sorted(glob.glob(os.path.join(OUTDIR, \"chr*.annotated.csv\")))\n",
    "dfs = [pd.read_csv(p, dtype=str) for p in paths]\n",
    "annot = pd.concat(dfs, ignore_index=True) if dfs else pd.DataFrame()\n",
    "\n",
    "if annot.empty or \"CLNSIG\" not in annot.columns:\n",
    "    print(\"No annotated data or CLNSIG column not found.\")\n",
    "else:\n",
    "    # Mapping from numeric code -> label (from your VCF header)\n",
    "    CLNSIG_MAP = {\n",
    "        \"0\":  \"Uncertain significance\",\n",
    "        \"1\":  \"not provided\",\n",
    "        \"2\":  \"Benign\",\n",
    "        \"3\":  \"Likely benign\",\n",
    "        \"4\":  \"Likely pathogenic\",\n",
    "        \"5\":  \"Pathogenic\",\n",
    "        \"6\":  \"Drug response\",\n",
    "        \"8\":  \"Confers sensitivity\",\n",
    "        \"9\":  \"Risk factor\",\n",
    "        \"10\": \"Association\",\n",
    "        \"11\": \"Protective\",\n",
    "        \"12\": \"Conflicting interpretations of pathogenicity\",\n",
    "        \"13\": \"Affects\",\n",
    "        \"14\": \"Association not found\",\n",
    "        \"15\": \"Benign/Likely benign\",\n",
    "        \"16\": \"Pathogenic/Likely pathogenic\",\n",
    "        \"17\": \"Conflicting data from submitters\",\n",
    "        \"18\": \"Pathogenic, low penetrance\",\n",
    "        \"19\": \"Likely pathogenic, low penetrance\",\n",
    "        \"20\": \"Established risk allele\",\n",
    "        \"21\": \"Likely risk allele\",\n",
    "        \"22\": \"Uncertain risk allele\",\n",
    "        \"255\":\"other\",\n",
    "        \".\":  None,   # treat '.' as missing\n",
    "        \"\":   None,\n",
    "    }\n",
    "\n",
    "    # Split CLNSIG on common delimiters (/,; or |), map each code to a label, explode to 1-per-row\n",
    "    splitter = re.compile(r\"[\\/;\\|,]+\")\n",
    "    cln = (\n",
    "        annot.assign(CLNSIG=annot[\"CLNSIG\"].fillna(\"\"))\n",
    "             .assign(code_list=lambda d: d[\"CLNSIG\"].apply(lambda s: [t.strip() for t in splitter.split(s) if t.strip()]))\n",
    "             .explode(\"code_list\")\n",
    "    )\n",
    "\n",
    "    # Map numeric code to label; keep unknowns as 'Unknown code: X'\n",
    "    def code_to_label(code):\n",
    "        if code in CLNSIG_MAP:\n",
    "            return CLNSIG_MAP[code]\n",
    "        return f\"Unknown code: {code}\" if code not in (None, \"\") else None\n",
    "\n",
    "    cln[\"CLNSIG_label\"] = cln[\"code_list\"].map(code_to_label)\n",
    "    cln = cln[cln[\"CLNSIG_label\"].notna()]\n",
    "\n",
    "    # Optional: add coarse buckets for quick rollups\n",
    "    def bucket(label: str) -> str:\n",
    "        if label in (\"Pathogenic\",\"Likely pathogenic\",\"Pathogenic/Likely pathogenic\",\n",
    "                     \"Pathogenic, low penetrance\",\"Likely pathogenic, low penetrance\"):\n",
    "            return \"Pathogenic (any)\"\n",
    "        if label in (\"Benign\",\"Likely benign\",\"Benign/Likely benign\"):\n",
    "            return \"Benign (any)\"\n",
    "        if label in (\"Uncertain significance\",\"Uncertain risk allele\"):\n",
    "            return \"Uncertain\"\n",
    "        if label in (\"Conflicting interpretations of pathogenicity\",\"Conflicting data from submitters\"):\n",
    "            return \"Conflicting\"\n",
    "        return \"Other\"\n",
    "\n",
    "    cln[\"CLNSIG_bucket\"] = cln[\"CLNSIG_label\"].map(bucket)\n",
    "\n",
    "    # Counts: detailed and buckets\n",
    "    sig_counts = cln[\"CLNSIG_label\"].value_counts().rename_axis(\"CLNSIG_label\").reset_index(name=\"count\")\n",
    "    bucket_counts = cln[\"CLNSIG_bucket\"].value_counts().rename_axis(\"CLNSIG_bucket\").reset_index(name=\"count\")\n",
    "\n",
    "    display(sig_counts.head(30))\n",
    "    display(bucket_counts)\n",
    "\n",
    "    # Save tables\n",
    "    sig_counts.to_csv(os.path.join(OUTDIR, \"clinical_counts_labels_overall.csv\"), index=False)\n",
    "    bucket_counts.to_csv(os.path.join(OUTDIR, \"clinical_counts_buckets_overall.csv\"), index=False)\n",
    "\n",
    "    # Bar chart of top 15 detailed labels\n",
    "    topN = 15\n",
    "    top = sig_counts.head(topN)\n",
    "    plt.figure(figsize=(10,5))\n",
    "    plt.bar(top[\"CLNSIG_label\"], top[\"count\"])\n",
    "    plt.xticks(rotation=45, ha=\"right\")\n",
    "    plt.title(\"Top clinical significance labels (mapped from codes)\")\n",
    "    plt.ylabel(\"Count\")\n",
    "    plt.tight_layout()\n",
    "    plt.show()\n",
    "\n",
    "    # Bar chart of coarse buckets\n",
    "    plt.figure(figsize=(7,4))\n",
    "    plt.bar(bucket_counts[\"CLNSIG_bucket\"], bucket_counts[\"count\"])\n",
    "    plt.xticks(rotation=0)\n",
    "    plt.title(\"Clinical significance (coarse buckets)\")\n",
    "    plt.ylabel(\"Count\")\n",
    "    plt.tight_layout()\n",
    "    plt.show()\n",
    "\n",
    "    # (Optional) Per-chromosome breakdowns\n",
    "    if \"CHROM\" in cln.columns:\n",
    "        per_chr_labels = (\n",
    "            cln.groupby([\"CHROM\",\"CLNSIG_label\"]).size().reset_index(name=\"count\")\n",
    "               .sort_values([\"CHROM\",\"count\"], ascending=[True, False])\n",
    "        )\n",
    "        per_chr_buckets = (\n",
    "            cln.groupby([\"CHROM\",\"CLNSIG_bucket\"]).size().reset_index(name=\"count\")\n",
    "               .sort_values([\"CHROM\",\"count\"], ascending=[True, False])\n",
    "        )\n",
    "        per_chr_labels.to_csv(os.path.join(OUTDIR, \"clinical_counts_labels_by_chrom.csv\"), index=False)\n",
    "        per_chr_buckets.to_csv(os.path.join(OUTDIR, \"clinical_counts_buckets_by_chrom.csv\"), index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6dee0218-22ef-4980-8a87-1456efb61caa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[info] Resolved trait columns: {'asthma': 'asthma ', 'max_pets': 'max_pets', 'num_smokers': 'num_smokers', 'IgE_Value': 'IgE_Value', 'BMI': 'BMI'}\n",
      "[info] Variants total: 1,629,382  |  with valid clinical labels: 16,688\n",
      "[info] Overlapping samples (all subjects): 119\n",
      "[info] Asthma-only samples: 110\n",
      "[info] Non-asthma-only samples: 9\n",
      "[trait asthma | all] Interpreted as BINARY; non-missing=119\n",
      "[save] asthma (all) -> assoc_clinical_asthma.csv (variants tested: 16482)\n",
      "[trait max_pets | all] Interpreted as NUMERIC; non-missing=119\n",
      "[save] max_pets (all) -> assoc_clinical_max_pets.csv (variants tested: 16687)\n",
      "[trait num_smokers | all] Interpreted as NUMERIC; non-missing=119\n",
      "[save] num_smokers (all) -> assoc_clinical_num_smokers.csv (variants tested: 16687)\n",
      "[trait IgE_Value | all] Using log10(IgE_Value) as the quantitative trait.\n",
      "[trait IgE_Value | all] Interpreted as NUMERIC; non-missing=119\n",
      "[save] IgE_Value (all) -> assoc_clinical_IgE_Value.csv (variants tested: 16687)\n",
      "[trait BMI | all] Interpreted as NUMERIC; non-missing=119\n",
      "[save] BMI (all) -> assoc_clinical_BMI.csv (variants tested: 16687)\n",
      "[trait max_pets | asthmaOnly] Interpreted as NUMERIC; non-missing=110\n",
      "[save] max_pets (asthmaOnly) -> assoc_clinical_asthmaOnly_max_pets.csv (variants tested: 16687)\n",
      "[trait max_pets | nonAsthmaOnly] Too few samples (9) – skipping.\n",
      "[save] max_pets (nonAsthmaOnly) -> assoc_clinical_nonAsthmaOnly_max_pets.csv (variants tested: 0)\n",
      "[trait num_smokers | asthmaOnly] Interpreted as NUMERIC; non-missing=110\n",
      "[save] num_smokers (asthmaOnly) -> assoc_clinical_asthmaOnly_num_smokers.csv (variants tested: 16687)\n",
      "[trait num_smokers | nonAsthmaOnly] Too few samples (9) – skipping.\n",
      "[save] num_smokers (nonAsthmaOnly) -> assoc_clinical_nonAsthmaOnly_num_smokers.csv (variants tested: 0)\n",
      "[trait IgE_Value | asthmaOnly] Using log10(IgE_Value) as the quantitative trait.\n",
      "[trait IgE_Value | asthmaOnly] Interpreted as NUMERIC; non-missing=110\n",
      "[save] IgE_Value (asthmaOnly) -> assoc_clinical_asthmaOnly_IgE_Value.csv (variants tested: 16687)\n",
      "[trait IgE_Value | nonAsthmaOnly] Too few samples (9) – skipping.\n",
      "[save] IgE_Value (nonAsthmaOnly) -> assoc_clinical_nonAsthmaOnly_IgE_Value.csv (variants tested: 0)\n",
      "[trait BMI | asthmaOnly] Interpreted as NUMERIC; non-missing=110\n",
      "[save] BMI (asthmaOnly) -> assoc_clinical_asthmaOnly_BMI.csv (variants tested: 16687)\n",
      "[trait BMI | nonAsthmaOnly] Too few samples (9) – skipping.\n",
      "[save] BMI (nonAsthmaOnly) -> assoc_clinical_nonAsthmaOnly_BMI.csv (variants tested: 0)\n",
      "\n",
      "Done. Top rows from each output:\n",
      "\n",
      "### asthma_all ###\n",
      "         variant_id CHROM       POS REF ALT      MAF   N      beta       se     pval       r2                                           CLNSIG_labels  qval_fdr\n",
      "  chr6:33171551:C:T  chr6  33171551   C   T 0.340336 119  2.586335 0.710118 0.000270 0.317679 Benign|Benign|Benign|Benign|Benign|Benign|Benign|Benign       NaN\n",
      "  chr5:53560967:G:T  chr5  53560967   G   T 0.344538 119 -2.499854 0.692469 0.000306 0.278350                                                  Benign       NaN\n",
      "chr12:132742948:C:G chr12 132742948   C   G 0.193277 119 -1.738204 0.499646 0.000504 0.211568                                                  Benign       NaN\n",
      "chr12:132751120:T:A chr12 132751120   T   A 0.168067 119 -1.632275 0.481191 0.000693 0.188853                                                  Benign       NaN\n",
      "  chr17:5034280:T:C chr17   5034280   T   C 0.067227 119  2.420368 0.742301 0.001112 0.157910                                           Benign|Benign       NaN\n",
      "\n",
      "### max_pets_all ###\n",
      "        variant_id CHROM      POS REF ALT  MAF   N     beta       se         pval           r2 CLNSIG_labels  qval_fdr\n",
      "chr17:21415657:C:T chr17 21415657   C   T  0.5 119 0.365546 0.062498 4.513289e-08 2.220446e-16        Benign  0.000034\n",
      "chr17:21298925:G:C chr17 21298925   G   C  0.5 119 0.365546 0.062498 4.513289e-08 2.220446e-16        Benign  0.000034\n",
      "chr17:21300581:T:C chr17 21300581   T   C  0.5 119 0.365546 0.062498 4.513289e-08 2.220446e-16        Benign  0.000034\n",
      "chr17:21300622:G:A chr17 21300622   G   A  0.5 119 0.365546 0.062498 4.513289e-08 2.220446e-16 Likely benign  0.000034\n",
      "chr17:21300637:C:T chr17 21300637   C   T  0.5 119 0.365546 0.062498 4.513289e-08 2.220446e-16 Likely benign  0.000034\n",
      "\n",
      "### num_smokers_all ###\n",
      "        variant_id CHROM      POS REF ALT  MAF   N     beta       se         pval  r2 CLNSIG_labels     qval_fdr\n",
      "chr17:21415657:C:T chr17 21415657   C   T  0.5 119 0.504202 0.044852 2.243618e-20 0.0        Benign 1.701784e-17\n",
      "chr17:21415640:A:G chr17 21415640   A   G  0.5 119 0.504202 0.044852 2.243618e-20 0.0        Benign 1.701784e-17\n",
      "chr17:21416080:G:A chr17 21416080   G   A  0.5 119 0.504202 0.044852 2.243618e-20 0.0        Benign 1.701784e-17\n",
      "chr17:21415896:C:T chr17 21415896   C   T  0.5 119 0.504202 0.044852 2.243618e-20 0.0        Benign 1.701784e-17\n",
      "chr17:21415859:G:A chr17 21415859   G   A  0.5 119 0.504202 0.044852 2.243618e-20 0.0        Benign 1.701784e-17\n",
      "\n",
      "### IgE_Value_all ###\n",
      "        variant_id CHROM      POS REF ALT  MAF   N      beta     se         pval            r2 CLNSIG_labels     qval_fdr\n",
      "chr17:21415639:C:T chr17 21415639   C   T  0.5 119 -0.391872 0.0287 4.832890e-26 -2.220446e-16        Benign 3.665747e-23\n",
      "chr17:21416080:G:A chr17 21416080   G   A  0.5 119 -0.391872 0.0287 4.832890e-26 -2.220446e-16        Benign 3.665747e-23\n",
      "chr17:21415636:C:T chr17 21415636   C   T  0.5 119 -0.391872 0.0287 4.832890e-26 -2.220446e-16        Benign 3.665747e-23\n",
      "chr17:21415669:C:T chr17 21415669   C   T  0.5 119 -0.391872 0.0287 4.832890e-26 -2.220446e-16        Benign 3.665747e-23\n",
      "chr17:21415896:C:T chr17 21415896   C   T  0.5 119 -0.391872 0.0287 4.832890e-26 -2.220446e-16        Benign 3.665747e-23\n",
      "\n",
      "### BMI_all ###\n",
      "        variant_id CHROM      POS REF ALT  MAF   N     beta       se          pval  r2 CLNSIG_labels      qval_fdr\n",
      "chr17:21298925:G:C chr17 21298925   G   C  0.5 119 6.423962 0.070371 2.633672e-111 0.0        Benign 1.997640e-108\n",
      "chr17:21415657:C:T chr17 21415657   C   T  0.5 119 6.423962 0.070371 2.633672e-111 0.0        Benign 1.997640e-108\n",
      "chr17:21415639:C:T chr17 21415639   C   T  0.5 119 6.423962 0.070371 2.633672e-111 0.0        Benign 1.997640e-108\n",
      "chr17:21302148:C:T chr17 21302148   C   T  0.5 119 6.423962 0.070371 2.633672e-111 0.0 Likely benign 1.997640e-108\n",
      "chr17:21415636:C:T chr17 21415636   C   T  0.5 119 6.423962 0.070371 2.633672e-111 0.0        Benign 1.997640e-108\n",
      "\n",
      "### max_pets_asthmaOnly ###\n",
      "        variant_id CHROM      POS REF ALT  MAF   N     beta       se         pval  r2 CLNSIG_labels  qval_fdr\n",
      "chr17:21415351:G:A chr17 21415351   G   A  0.5 110 0.386364 0.066692 6.776366e-08 0.0        Benign  0.000049\n",
      "chr17:21415859:G:A chr17 21415859   G   A  0.5 110 0.386364 0.066692 6.776366e-08 0.0        Benign  0.000049\n",
      "chr17:21415798:G:A chr17 21415798   G   A  0.5 110 0.386364 0.066692 6.776366e-08 0.0        Benign  0.000049\n",
      "chr17:21415775:G:A chr17 21415775   G   A  0.5 110 0.386364 0.066692 6.776366e-08 0.0        Benign  0.000049\n",
      "chr17:21415757:G:A chr17 21415757   G   A  0.5 110 0.386364 0.066692 6.776366e-08 0.0        Benign  0.000049\n",
      "\n",
      "### max_pets_nonAsthmaOnly ###\n",
      "Empty DataFrame\n",
      "Columns: [variant_id, CHROM, POS, REF, ALT, MAF, N, beta, se, pval, r2, qval_fdr, CLNSIG_labels]\n",
      "Index: []\n",
      "\n",
      "### num_smokers_asthmaOnly ###\n",
      "        variant_id CHROM      POS REF ALT  MAF   N     beta       se         pval            r2 CLNSIG_labels     qval_fdr\n",
      "chr17:21302148:C:T chr17 21302148   C   T  0.5 110 0.490909 0.046559 2.431894e-18 -2.220446e-16 Likely benign 1.764392e-15\n",
      "chr17:21415669:C:T chr17 21415669   C   T  0.5 110 0.490909 0.046559 2.431894e-18 -2.220446e-16        Benign 1.764392e-15\n",
      "chr17:21415351:G:A chr17 21415351   G   A  0.5 110 0.490909 0.046559 2.431894e-18 -2.220446e-16        Benign 1.764392e-15\n",
      "chr17:21415657:C:T chr17 21415657   C   T  0.5 110 0.490909 0.046559 2.431894e-18 -2.220446e-16        Benign 1.764392e-15\n",
      "chr17:21415640:A:G chr17 21415640   A   G  0.5 110 0.490909 0.046559 2.431894e-18 -2.220446e-16        Benign 1.764392e-15\n",
      "\n",
      "### num_smokers_nonAsthmaOnly ###\n",
      "Empty DataFrame\n",
      "Columns: [variant_id, CHROM, POS, REF, ALT, MAF, N, beta, se, pval, r2, qval_fdr, CLNSIG_labels]\n",
      "Index: []\n",
      "\n",
      "### IgE_Value_asthmaOnly ###\n",
      "        variant_id CHROM      POS REF ALT  MAF   N      beta       se         pval           r2 CLNSIG_labels     qval_fdr\n",
      "chr17:21415555:G:A chr17 21415555   G   A  0.5 110 -0.384585 0.030784 9.130262e-23 1.110223e-16        Benign 6.624203e-20\n",
      "chr17:21415757:G:A chr17 21415757   G   A  0.5 110 -0.384585 0.030784 9.130262e-23 1.110223e-16        Benign 6.624203e-20\n",
      "chr17:21415696:G:C chr17 21415696   G   C  0.5 110 -0.384585 0.030784 9.130262e-23 1.110223e-16        Benign 6.624203e-20\n",
      "chr17:21415669:C:T chr17 21415669   C   T  0.5 110 -0.384585 0.030784 9.130262e-23 1.110223e-16        Benign 6.624203e-20\n",
      "chr17:21415657:C:T chr17 21415657   C   T  0.5 110 -0.384585 0.030784 9.130262e-23 1.110223e-16        Benign 6.624203e-20\n",
      "\n",
      "### IgE_Value_nonAsthmaOnly ###\n",
      "Empty DataFrame\n",
      "Columns: [variant_id, CHROM, POS, REF, ALT, MAF, N, beta, se, pval, r2, qval_fdr, CLNSIG_labels]\n",
      "Index: []\n",
      "\n",
      "### BMI_asthmaOnly ###\n",
      "        variant_id CHROM      POS REF ALT  MAF   N     beta       se          pval            r2 CLNSIG_labels     qval_fdr\n",
      "chr17:21415896:C:T chr17 21415896   C   T  0.5 110 6.417711 0.075335 1.446709e-101 -2.220446e-16        Benign 1.049619e-98\n",
      "chr17:21415798:G:A chr17 21415798   G   A  0.5 110 6.417711 0.075335 1.446709e-101 -2.220446e-16        Benign 1.049619e-98\n",
      "chr17:21415775:G:A chr17 21415775   G   A  0.5 110 6.417711 0.075335 1.446709e-101 -2.220446e-16        Benign 1.049619e-98\n",
      "chr17:21415757:G:A chr17 21415757   G   A  0.5 110 6.417711 0.075335 1.446709e-101 -2.220446e-16        Benign 1.049619e-98\n",
      "chr17:21415696:G:C chr17 21415696   G   C  0.5 110 6.417711 0.075335 1.446709e-101 -2.220446e-16        Benign 1.049619e-98\n",
      "\n",
      "### BMI_nonAsthmaOnly ###\n",
      "Empty DataFrame\n",
      "Columns: [variant_id, CHROM, POS, REF, ALT, MAF, N, beta, se, pval, r2, qval_fdr, CLNSIG_labels]\n",
      "Index: []\n"
     ]
    }
   ],
   "source": [
    "# =============================================================\n",
    "# Association analysis on clinically annotated SNPs\n",
    "# - Loads annotated files from per_chrom_outputs_new\n",
    "# - Maps CLNSIG numeric codes → labels\n",
    "# - Filters out missing / unknown codes\n",
    "# - Runs per-variant association per trait:\n",
    "#       * Logistic regression for binary traits (e.g. asthma)\n",
    "#       * Linear regression for quantitative traits\n",
    "# - FIXES:\n",
    "#     * Sample detection = only phenotype IDs (no extra columns)\n",
    "#     * Trait typing = numeric vs binary decided robustly\n",
    "# - UPDATED:\n",
    "#     * Treat IgE_Value and BMI as TRAITS (i.e. run association against them)\n",
    "#     * IgE_Value is analyzed on log10 scale (standard IgE practice)\n",
    "#     * Do NOT add them as covariates\n",
    "#     * ALSO run trait associations separately:\n",
    "#           - ONLY in asthma cases\n",
    "#           - ONLY in non-asthma subjects\n",
    "#     * NEW: save R^2 (linear) or pseudo-R^2 (logistic) per SNP\n",
    "# =============================================================\n",
    "import os, glob, re, warnings\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import statsmodels.api as sm\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "from statsmodels.tools.sm_exceptions import ConvergenceWarning, PerfectSeparationWarning\n",
    "\n",
    "# ---------- Silence noisy warnings ----------\n",
    "warnings.filterwarnings(\"ignore\", category=ConvergenceWarning)\n",
    "warnings.filterwarnings(\"ignore\", category=PerfectSeparationWarning)\n",
    "warnings.filterwarnings(\"ignore\", category=RuntimeWarning)\n",
    "np.seterr(divide=\"ignore\", invalid=\"ignore\", over=\"ignore\")\n",
    "\n",
    "# ========= User inputs =========\n",
    "PHENO_PATH   = \"ImmPort_URECA_Cross_Black.csv\"\n",
    "PHENO_ID_COL = \"id\"\n",
    "\n",
    "# Traits to analyze (asthma is binary; others are quantitative)\n",
    "TRAITS       = [\"asthma\", \"max_pets\", \"num_smokers\", \"IgE_Value\", \"BMI\"]\n",
    "\n",
    "OUTDIR_ANN   = \"per_chrom_outputs_new\"\n",
    "OUT_PREFIX   = \"assoc_clinical\"\n",
    "# ==============================\n",
    "\n",
    "# ---------- Helpers ----------\n",
    "def norm(s: str) -> str:\n",
    "    return re.sub(r\"\\W+\", \"_\", str(s).strip().lower())\n",
    "\n",
    "def resolve_trait_columns(pheno_df: pd.DataFrame, wanted_traits: list[str]) -> dict:\n",
    "    \"\"\"\n",
    "    Map user-friendly trait names to actual column names in the phenotype file,\n",
    "    allowing for differences in case/spacing/underscore.\n",
    "    \"\"\"\n",
    "    colmap = {norm(c): c for c in pheno_df.columns}\n",
    "    resolved = {}\n",
    "    for t in wanted_traits:\n",
    "        key = norm(t)\n",
    "        if key in colmap:\n",
    "            resolved[t] = colmap[key]\n",
    "        else:\n",
    "            # fuzzy: asthma -> asthma_status, IgE_Value -> ige_value, etc.\n",
    "            cands = [c for k, c in colmap.items() if key == k or key in k]\n",
    "            if len(cands) == 1:\n",
    "                resolved[t] = cands[0]\n",
    "            else:\n",
    "                raise KeyError(\n",
    "                    f\"Could not find trait column for '{t}'. \"\n",
    "                    f\"Available (normalized) include: {list(sorted(colmap.keys())[:20])} ...\"\n",
    "                )\n",
    "    return resolved\n",
    "\n",
    "def decide_trait(y_raw: pd.Series):\n",
    "    \"\"\"\n",
    "    Decide if a trait is binary or numeric.\n",
    "    - If coercing to numeric gives >2 unique values → numeric\n",
    "    - Else try typical binary strings\n",
    "    - Else if values are {0,1} → binary\n",
    "    - Else → numeric\n",
    "    \"\"\"\n",
    "    y_num = pd.to_numeric(y_raw, errors=\"coerce\")\n",
    "    nuniq = y_num.dropna().nunique()\n",
    "\n",
    "    sl = y_raw.astype(str).str.strip().str.lower()\n",
    "    bin_map = {\n",
    "        \"1\": 1, \"0\": 0,\n",
    "        \"yes\": 1, \"no\": 0,\n",
    "        \"true\": 1, \"false\": 0,\n",
    "        \"y\": 1, \"n\": 0,\n",
    "        \"asthma\": 1, \"no asthma\": 0,\n",
    "        \"present\": 1, \"absent\": 0,\n",
    "        \"case\": 1, \"control\": 0\n",
    "    }\n",
    "    y_bin = sl.map(bin_map).astype(float)\n",
    "    coverage = y_bin.notna().mean()\n",
    "\n",
    "    if nuniq > 2:\n",
    "        # clearly quantitative\n",
    "        return y_num.astype(float), False\n",
    "    if coverage >= 0.9:\n",
    "        # mapped well to 0/1\n",
    "        return y_bin, True\n",
    "    if set(pd.unique(y_num.dropna())).issubset({0,1}):\n",
    "        # already numeric 0/1\n",
    "        return y_num.astype(float), True\n",
    "    return y_num.astype(float), False\n",
    "\n",
    "def gt_to_dosage(field: str) -> float:\n",
    "    \"\"\"\n",
    "    Convert VCF-like GT field to dosage 0/1/2.\n",
    "    \"\"\"\n",
    "    if not isinstance(field, str) or field == \"\" or field == \".\":\n",
    "        return np.nan\n",
    "    gt = field.split(\":\", 1)[0]\n",
    "    if gt in {\"./.\", \".|.\", \".\"}:\n",
    "        return np.nan\n",
    "    if \"/\" in gt:\n",
    "        a1, a2 = gt.split(\"/\")\n",
    "    elif \"|\" in gt:\n",
    "        a1, a2 = gt.split(\"|\")\n",
    "    else:\n",
    "        return np.nan\n",
    "    if a1 not in {\"0\",\"1\"} or a2 not in {\"0\",\"1\"}:\n",
    "        return np.nan\n",
    "    return float((a1 == \"1\") + (a2 == \"1\"))\n",
    "\n",
    "def maf_from_dosage(dosages: np.ndarray) -> float | None:\n",
    "    x = dosages[~np.isnan(dosages)]\n",
    "    n = x.size\n",
    "    if n == 0:\n",
    "        return None\n",
    "    af = x.sum() / (2.0 * n)\n",
    "    return float(min(af, 1 - af))\n",
    "\n",
    "# Map CLNSIG numeric codes → labels\n",
    "CLNSIG_MAP = {\n",
    "    \"0\":  \"Uncertain significance\",\n",
    "    \"1\":  \"not provided\",\n",
    "    \"2\":  \"Benign\",\n",
    "    \"3\":  \"Likely benign\",\n",
    "    \"4\":  \"Likely pathogenic\",\n",
    "    \"5\":  \"Pathogenic\",\n",
    "    \"6\":  \"Drug response\",\n",
    "    \"8\":  \"Confers sensitivity\",\n",
    "    \"9\":  \"Risk factor\",\n",
    "    \"10\": \"Association\",\n",
    "    \"11\": \"Protective\",\n",
    "    \"12\": \"Conflicting interpretations of pathogenicity\",\n",
    "    \"13\": \"Affects\",\n",
    "    \"14\": \"Association not found\",\n",
    "    \"15\": \"Benign/Likely benign\",\n",
    "    \"16\": \"Pathogenic/Likely pathogenic\",\n",
    "    \"17\": \"Conflicting data from submitters\",\n",
    "    \"18\": \"Pathogenic, low penetrance\",\n",
    "    \"19\": \"Likely pathogenic, low penetrance\",\n",
    "    \"20\": \"Established risk allele\",\n",
    "    \"21\": \"Likely risk allele\",\n",
    "    \"22\": \"Uncertain risk allele\",\n",
    "    \"255\":\"other\",\n",
    "    \".\":  None,\n",
    "    \"\":   None,\n",
    "}\n",
    "CLNSIG_SPLIT = re.compile(r\"[\\/;\\|,]+\")\n",
    "\n",
    "def clnsig_codes_to_labels(clnsig_val: str) -> list[str]:\n",
    "    if not isinstance(clnsig_val, str):\n",
    "        return []\n",
    "    codes = [t.strip() for t in CLNSIG_SPLIT.split(clnsig_val) if t.strip() != \"\"]\n",
    "    labels = []\n",
    "    for code in codes:\n",
    "        if code in CLNSIG_MAP:\n",
    "            lab = CLNSIG_MAP[code]\n",
    "            if lab:\n",
    "                labels.append(lab)\n",
    "    return labels\n",
    "\n",
    "# ----------------- Load phenotype + annotated variants -----------------\n",
    "pheno = pd.read_csv(PHENO_PATH)\n",
    "\n",
    "# Resolve trait columns (handles things like \"asthma \" vs \"asthma\")\n",
    "trait_col_map = resolve_trait_columns(pheno, TRAITS)\n",
    "print(\"[info] Resolved trait columns:\", trait_col_map)\n",
    "\n",
    "# Normalize phenotype IDs to string index\n",
    "pheno_ids = pheno[PHENO_ID_COL].astype(str)\n",
    "pheno = pheno.set_index(pheno_ids).copy()\n",
    "pheno.index.name = \"sample_id\"\n",
    "\n",
    "# Load annotated tables\n",
    "ann_paths = sorted(glob.glob(os.path.join(OUTDIR_ANN, \"chr*.annotated.csv\")))\n",
    "if not ann_paths:\n",
    "    raise FileNotFoundError(f\"No annotated files found in {OUTDIR_ANN}. Expected chr*.annotated.csv\")\n",
    "\n",
    "annot = pd.concat(\n",
    "    [pd.read_csv(p, dtype=str, low_memory=False) for p in ann_paths],\n",
    "    ignore_index=True\n",
    ")\n",
    "\n",
    "# Clinical filter (keep variants with at least one CLNSIG label)\n",
    "annot[\"CLNSIG\"] = annot[\"CLNSIG\"].fillna(\"\")\n",
    "annot[\"CLNSIG_labels_list\"] = annot[\"CLNSIG\"].apply(clnsig_codes_to_labels)\n",
    "annot[\"CLNSIG_labels_str\"]  = annot[\"CLNSIG_labels_list\"].apply(lambda xs: \"|\".join(xs))\n",
    "clin_mask = annot[\"CLNSIG_labels_list\"].apply(lambda xs: len(xs) > 0)\n",
    "annot_clin = annot[clin_mask].copy()\n",
    "\n",
    "print(f\"[info] Variants total: {annot.shape[0]:,}  |  with valid clinical labels: {annot_clin.shape[0]:,}\")\n",
    "\n",
    "# Build variant_id\n",
    "annot_clin[\"variant_id\"] = (\n",
    "    annot_clin[\"#CHROM\"].astype(str) + \":\" +\n",
    "    annot_clin[\"POS\"].astype(str)    + \":\" +\n",
    "    annot_clin[\"REF\"].astype(str)    + \":\" +\n",
    "    annot_clin[\"ALT\"].astype(str)\n",
    ")\n",
    "\n",
    "# Sample columns are those that match phenotype IDs\n",
    "sample_cols_clean = [c for c in annot_clin.columns if c in pheno.index]\n",
    "if not sample_cols_clean:\n",
    "    raise RuntimeError(\"No genotype sample columns matched phenotype IDs.\")\n",
    "overlap = sample_cols_clean\n",
    "print(f\"[info] Overlapping samples (all subjects): {len(overlap)}\")\n",
    "\n",
    "# ---- Build asthma case / non-case subsets for stratified analyses ----\n",
    "asthma_col = trait_col_map[\"asthma\"]\n",
    "asthma_raw = pheno.loc[overlap, asthma_col]\n",
    "asthma_y, asthma_binary = decide_trait(asthma_raw)\n",
    "if not asthma_binary:\n",
    "    print(\"[warn] Asthma column did not resolve cleanly to binary; stratified analyses may be unreliable.\")\n",
    "\n",
    "mask_asthma     = (asthma_y == 1)\n",
    "mask_nonasthma  = (asthma_y == 0)\n",
    "\n",
    "overlap_asthma = [sid for sid, m in zip(overlap, mask_asthma.values) if pd.notna(m) and m]\n",
    "overlap_nonasthma = [sid for sid, m in zip(overlap, mask_nonasthma.values) if pd.notna(m) and m]\n",
    "\n",
    "print(f\"[info] Asthma-only samples: {len(overlap_asthma)}\")\n",
    "print(f\"[info] Non-asthma-only samples: {len(overlap_nonasthma)}\")\n",
    "\n",
    "# ----------------- Association core -----------------\n",
    "def run_assoc_for_trait(trait_name: str, sample_ids, label: str = \"\") -> pd.DataFrame:\n",
    "    \"\"\"\n",
    "    Run association for a given trait in the specified sample subset.\n",
    "\n",
    "    trait_name : one of TRAITS\n",
    "    sample_ids : list of sample IDs to include\n",
    "    label      : string used only for logging (e.g. 'asthmaOnly')\n",
    "    \"\"\"\n",
    "    if len(sample_ids) < 10:\n",
    "        print(f\"[trait {trait_name} | {label}] Too few samples ({len(sample_ids)}) – skipping.\")\n",
    "        return pd.DataFrame(columns=[\n",
    "            \"variant_id\",\"CHROM\",\"POS\",\"REF\",\"ALT\",\"MAF\",\"N\",\n",
    "            \"beta\",\"se\",\"pval\",\"r2\",\"qval_fdr\",\"CLNSIG_labels\"\n",
    "        ])\n",
    "\n",
    "    ph_col = trait_col_map[trait_name]\n",
    "    y_raw = pheno.loc[sample_ids, ph_col]\n",
    "\n",
    "    # SPECIAL CASE: IgE_Value is analyzed on log10 scale\n",
    "    if trait_name == \"IgE_Value\":\n",
    "        y_raw = np.log10(pd.to_numeric(y_raw, errors=\"coerce\"))\n",
    "        print(f\"[trait {trait_name} | {label}] Using log10(IgE_Value) as the quantitative trait.\")\n",
    "\n",
    "    # Decide binary vs numeric for this trait within this subset\n",
    "    y, binary = decide_trait(y_raw)\n",
    "    print(f\"[trait {trait_name} | {label}] Interpreted as {'BINARY' if binary else 'NUMERIC'}; \"\n",
    "          f\"non-missing={y.notna().sum()}\")\n",
    "\n",
    "    results = []\n",
    "    for _, row in annot_clin.iterrows():\n",
    "        # Dosage for the selected samples\n",
    "        g = np.array([gt_to_dosage(row[s]) for s in sample_ids], dtype=float)\n",
    "\n",
    "        # Build df with y and genotype dosage\n",
    "        df = pd.DataFrame({\"y\": y.values, \"g\": g}, index=sample_ids).dropna(subset=[\"y\", \"g\"])\n",
    "        if df.shape[0] < 10:\n",
    "            continue\n",
    "\n",
    "        # Design matrix: intercept + genotype (additive model)\n",
    "        X = sm.add_constant(df[\"g\"].values, has_constant=\"add\")\n",
    "\n",
    "        try:\n",
    "            if binary:\n",
    "                fit = sm.Logit(df[\"y\"].values, X).fit(disp=False)\n",
    "                r2  = float(fit.prsquared)  # McFadden pseudo-R^2\n",
    "            else:\n",
    "                fit = sm.OLS(df[\"y\"].values, X).fit()\n",
    "                r2  = float(fit.rsquared)   # standard R^2\n",
    "            beta = float(fit.params[1])\n",
    "            se   = float(fit.bse[1])\n",
    "            pval = float(fit.pvalues[1])\n",
    "        except Exception:\n",
    "            # If a rare variant causes perfect separation or other issues, skip it\n",
    "            continue\n",
    "\n",
    "        maf = maf_from_dosage(df[\"g\"].values)\n",
    "        results.append({\n",
    "            \"variant_id\": row[\"variant_id\"],\n",
    "            \"CHROM\": row[\"#CHROM\"],\n",
    "            \"POS\": int(row[\"POS\"]),\n",
    "            \"REF\": row[\"REF\"],\n",
    "            \"ALT\": row[\"ALT\"],\n",
    "            \"MAF\": maf,\n",
    "            \"N\": int(df.shape[0]),\n",
    "            \"beta\": beta,\n",
    "            \"se\": se,\n",
    "            \"pval\": pval,\n",
    "            \"r2\": r2,\n",
    "            \"CLNSIG_labels\": row[\"CLNSIG_labels_str\"]\n",
    "        })\n",
    "\n",
    "    if not results:\n",
    "        print(f\"[trait {trait_name} | {label}] No analyzable variants produced results.\")\n",
    "        return pd.DataFrame(columns=[\n",
    "            \"variant_id\",\"CHROM\",\"POS\",\"REF\",\"ALT\",\"MAF\",\"N\",\n",
    "            \"beta\",\"se\",\"pval\",\"r2\",\"qval_fdr\",\"CLNSIG_labels\"\n",
    "        ])\n",
    "\n",
    "    out = pd.DataFrame(results)\n",
    "    out[\"qval_fdr\"] = multipletests(out[\"pval\"].values, method=\"fdr_bh\")[1]\n",
    "    out = out.sort_values(\"pval\", ascending=True).reset_index(drop=True)\n",
    "    return out\n",
    "\n",
    "# ----------------- Run and save -----------------\n",
    "all_outputs = {}\n",
    "\n",
    "# 1) Full-cohort analyses (same as before)\n",
    "for trait in TRAITS:\n",
    "    res = run_assoc_for_trait(trait, overlap, label=\"all\")\n",
    "    out_path = f\"{OUT_PREFIX}_{trait}.csv\"\n",
    "    res.to_csv(out_path, index=False)\n",
    "    all_outputs[f\"{trait}_all\"] = out_path\n",
    "    print(f\"[save] {trait} (all) -> {out_path} (variants tested: {res.shape[0]})\")\n",
    "\n",
    "# 2) Stratified analyses: asthma-only and non-asthma-only\n",
    "#    We only run these for quantitative traits (no point running 'asthma' within asthma-only).\n",
    "STRAT_TRAITS = [t for t in TRAITS if t != \"asthma\"]\n",
    "\n",
    "for trait in STRAT_TRAITS:\n",
    "    # asthma-only\n",
    "    res_a = run_assoc_for_trait(trait, overlap_asthma, label=\"asthmaOnly\")\n",
    "    out_path_a = f\"{OUT_PREFIX}_asthmaOnly_{trait}.csv\"\n",
    "    res_a.to_csv(out_path_a, index=False)\n",
    "    all_outputs[f\"{trait}_asthmaOnly\"] = out_path_a\n",
    "    print(f\"[save] {trait} (asthmaOnly) -> {out_path_a} (variants tested: {res_a.shape[0]})\")\n",
    "\n",
    "    # non-asthma-only\n",
    "    res_n = run_assoc_for_trait(trait, overlap_nonasthma, label=\"nonAsthmaOnly\")\n",
    "    out_path_n = f\"{OUT_PREFIX}_nonAsthmaOnly_{trait}.csv\"\n",
    "    res_n.to_csv(out_path_n, index=False)\n",
    "    all_outputs[f\"{trait}_nonAsthmaOnly\"] = out_path_n\n",
    "    print(f\"[save] {trait} (nonAsthmaOnly) -> {out_path_n} (variants tested: {res_n.shape[0]})\")\n",
    "\n",
    "print(\"\\nDone. Top rows from each output:\")\n",
    "for key, path in all_outputs.items():\n",
    "    try:\n",
    "        print(f\"\\n### {key} ###\")\n",
    "        print(pd.read_csv(path, nrows=5).to_string(index=False))\n",
    "    except Exception:\n",
    "        pass\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "810c521a-1897-4d2a-8d49-052c049a5646",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[plot] Overlapping samples for quicklook: 119\n",
      "[fig] saved -> assoc_clinical_asthma_trait_quicklook.png\n",
      "[fig] saved -> assoc_clinical_asthma_manhattan.png\n",
      "[fig] saved -> assoc_clinical_asthma_qq.png\n",
      "[fig] saved -> assoc_clinical_asthma_volcano.png\n",
      "[fig] saved -> assoc_clinical_asthma_maf_vs_mlog10p.png\n",
      "[fig] saved -> assoc_clinical_max_pets_trait_quicklook.png\n",
      "[fig] saved -> assoc_clinical_max_pets_manhattan.png\n",
      "[fig] saved -> assoc_clinical_max_pets_qq.png\n",
      "[fig] saved -> assoc_clinical_max_pets_volcano.png\n",
      "[fig] saved -> assoc_clinical_max_pets_maf_vs_mlog10p.png\n",
      "[fig] saved -> assoc_clinical_asthmaOnly_max_pets_asthmaOnly_trait_quicklook.png\n",
      "[fig] saved -> assoc_clinical_asthmaOnly_max_pets_asthmaOnly_manhattan.png\n",
      "[fig] saved -> assoc_clinical_asthmaOnly_max_pets_asthmaOnly_qq.png\n",
      "[fig] saved -> assoc_clinical_asthmaOnly_max_pets_asthmaOnly_volcano.png\n",
      "[fig] saved -> assoc_clinical_asthmaOnly_max_pets_asthmaOnly_maf_vs_mlog10p.png\n",
      "[fig] saved -> assoc_clinical_nonAsthmaOnly_max_pets_nonAsthmaOnly_trait_quicklook.png\n",
      "[max_pets | nonAsthmaOnly] No results to plot (Manhattan).\n",
      "[max_pets | nonAsthmaOnly] No results to plot (QQ).\n",
      "[max_pets | nonAsthmaOnly] No results to plot (volcano).\n",
      "[max_pets | nonAsthmaOnly] No results to plot (MAF vs sig).\n",
      "[fig] saved -> assoc_clinical_num_smokers_trait_quicklook.png\n",
      "[fig] saved -> assoc_clinical_num_smokers_manhattan.png\n",
      "[fig] saved -> assoc_clinical_num_smokers_qq.png\n",
      "[fig] saved -> assoc_clinical_num_smokers_volcano.png\n",
      "[fig] saved -> assoc_clinical_num_smokers_maf_vs_mlog10p.png\n",
      "[fig] saved -> assoc_clinical_asthmaOnly_num_smokers_asthmaOnly_trait_quicklook.png\n",
      "[fig] saved -> assoc_clinical_asthmaOnly_num_smokers_asthmaOnly_manhattan.png\n",
      "[fig] saved -> assoc_clinical_asthmaOnly_num_smokers_asthmaOnly_qq.png\n",
      "[fig] saved -> assoc_clinical_asthmaOnly_num_smokers_asthmaOnly_volcano.png\n",
      "[fig] saved -> assoc_clinical_asthmaOnly_num_smokers_asthmaOnly_maf_vs_mlog10p.png\n",
      "[fig] saved -> assoc_clinical_nonAsthmaOnly_num_smokers_nonAsthmaOnly_trait_quicklook.png\n",
      "[num_smokers | nonAsthmaOnly] No results to plot (Manhattan).\n",
      "[num_smokers | nonAsthmaOnly] No results to plot (QQ).\n",
      "[num_smokers | nonAsthmaOnly] No results to plot (volcano).\n",
      "[num_smokers | nonAsthmaOnly] No results to plot (MAF vs sig).\n",
      "[fig] saved -> assoc_clinical_IgE_Value_trait_quicklook.png\n",
      "[fig] saved -> assoc_clinical_IgE_Value_manhattan.png\n",
      "[fig] saved -> assoc_clinical_IgE_Value_qq.png\n",
      "[fig] saved -> assoc_clinical_IgE_Value_volcano.png\n",
      "[fig] saved -> assoc_clinical_IgE_Value_maf_vs_mlog10p.png\n",
      "[fig] saved -> assoc_clinical_asthmaOnly_IgE_Value_asthmaOnly_trait_quicklook.png\n",
      "[fig] saved -> assoc_clinical_asthmaOnly_IgE_Value_asthmaOnly_manhattan.png\n",
      "[fig] saved -> assoc_clinical_asthmaOnly_IgE_Value_asthmaOnly_qq.png\n",
      "[fig] saved -> assoc_clinical_asthmaOnly_IgE_Value_asthmaOnly_volcano.png\n",
      "[fig] saved -> assoc_clinical_asthmaOnly_IgE_Value_asthmaOnly_maf_vs_mlog10p.png\n",
      "[fig] saved -> assoc_clinical_nonAsthmaOnly_IgE_Value_nonAsthmaOnly_trait_quicklook.png\n",
      "[IgE_Value | nonAsthmaOnly] No results to plot (Manhattan).\n",
      "[IgE_Value | nonAsthmaOnly] No results to plot (QQ).\n",
      "[IgE_Value | nonAsthmaOnly] No results to plot (volcano).\n",
      "[IgE_Value | nonAsthmaOnly] No results to plot (MAF vs sig).\n",
      "[fig] saved -> assoc_clinical_BMI_trait_quicklook.png\n",
      "[fig] saved -> assoc_clinical_BMI_manhattan.png\n",
      "[fig] saved -> assoc_clinical_BMI_qq.png\n",
      "[fig] saved -> assoc_clinical_BMI_volcano.png\n",
      "[fig] saved -> assoc_clinical_BMI_maf_vs_mlog10p.png\n",
      "[fig] saved -> assoc_clinical_asthmaOnly_BMI_asthmaOnly_trait_quicklook.png\n",
      "[fig] saved -> assoc_clinical_asthmaOnly_BMI_asthmaOnly_manhattan.png\n",
      "[fig] saved -> assoc_clinical_asthmaOnly_BMI_asthmaOnly_qq.png\n",
      "[fig] saved -> assoc_clinical_asthmaOnly_BMI_asthmaOnly_volcano.png\n",
      "[fig] saved -> assoc_clinical_asthmaOnly_BMI_asthmaOnly_maf_vs_mlog10p.png\n",
      "[fig] saved -> assoc_clinical_nonAsthmaOnly_BMI_nonAsthmaOnly_trait_quicklook.png\n",
      "[BMI | nonAsthmaOnly] No results to plot (Manhattan).\n",
      "[BMI | nonAsthmaOnly] No results to plot (QQ).\n",
      "[BMI | nonAsthmaOnly] No results to plot (volcano).\n",
      "[BMI | nonAsthmaOnly] No results to plot (MAF vs sig).\n",
      "[table] Significant associations (all subjects, FDR<0.05) -> significant_hits_all_FDR_lt_0p05.csv\n",
      "subset     trait         variant_id CHROM      POS REF ALT     MAF   N      beta       se          pval      qval_fdr            r2  bonf_sig        rsID CLNSIG_labels    CSQ_terms_str\n",
      "   all       BMI chr17:21298925:G:C chr17 21298925   G   C 0.50000 119  6.423962 0.070371 2.633672e-111 1.997640e-108  0.000000e+00      True  rs36047035        Benign missense_variant\n",
      "   all       BMI chr17:21415657:C:T chr17 21415657   C   T 0.50000 119  6.423962 0.070371 2.633672e-111 1.997640e-108  0.000000e+00      True  rs79666990        Benign              NaN\n",
      "   all       BMI chr17:21415639:C:T chr17 21415639   C   T 0.50000 119  6.423962 0.070371 2.633672e-111 1.997640e-108  0.000000e+00      True  rs75757803        Benign              NaN\n",
      "   all       BMI chr17:21302148:C:T chr17 21302148   C   T 0.50000 119  6.423962 0.070371 2.633672e-111 1.997640e-108  0.000000e+00      True  rs62057721 Likely benign missense_variant\n",
      "   all       BMI chr17:21415636:C:T chr17 21415636   C   T 0.50000 119  6.423962 0.070371 2.633672e-111 1.997640e-108  0.000000e+00      True rs112314728        Benign missense_variant\n",
      "   all       BMI chr17:21415351:G:A chr17 21415351   G   A 0.50000 119  6.423962 0.070371 2.633672e-111 1.997640e-108  0.000000e+00      True  rs72846666        Benign              NaN\n",
      "   all       BMI chr17:21415669:C:T chr17 21415669   C   T 0.50000 119  6.423962 0.070371 2.633672e-111 1.997640e-108  0.000000e+00      True   rs9909970        Benign missense_variant\n",
      "   all       BMI chr17:21415696:G:C chr17 21415696   G   C 0.50000 119  6.423962 0.070371 2.633672e-111 1.997640e-108  0.000000e+00      True   rs1657741        Benign              NaN\n",
      "   all       BMI chr17:21298407:T:C chr17 21298407   T   C 0.50000 119  6.423962 0.070371 2.633672e-111 1.997640e-108  0.000000e+00      True   rs8076154        Benign              NaN\n",
      "   all       BMI chr17:21415757:G:A chr17 21415757   G   A 0.50000 119  6.423962 0.070371 2.633672e-111 1.997640e-108  0.000000e+00      True  rs76265595        Benign missense_variant\n",
      "   all       BMI chr17:21415555:G:A chr17 21415555   G   A 0.50000 119  6.423962 0.070371 2.633672e-111 1.997640e-108  0.000000e+00      True  rs73979893        Benign missense_variant\n",
      "   all       BMI chr17:21415775:G:A chr17 21415775   G   A 0.50000 119  6.423962 0.070371 2.633672e-111 1.997640e-108  0.000000e+00      True  rs75029097        Benign missense_variant\n",
      "   all       BMI chr17:21415798:G:A chr17 21415798   G   A 0.50000 119  6.423962 0.070371 2.633672e-111 1.997640e-108  0.000000e+00      True  rs77283567        Benign              NaN\n",
      "   all       BMI chr17:21415859:G:A chr17 21415859   G   A 0.50000 119  6.423962 0.070371 2.633672e-111 1.997640e-108  0.000000e+00      True  rs73313922        Benign missense_variant\n",
      "   all       BMI chr17:21415896:C:T chr17 21415896   C   T 0.50000 119  6.423962 0.070371 2.633672e-111 1.997640e-108  0.000000e+00      True  rs73979896        Benign missense_variant\n",
      "   all       BMI chr17:21416080:G:A chr17 21416080   G   A 0.50000 119  6.423962 0.070371 2.633672e-111 1.997640e-108  0.000000e+00      True  rs73979899        Benign              NaN\n",
      "   all       BMI chr17:21300637:C:T chr17 21300637   C   T 0.50000 119  6.423962 0.070371 2.633672e-111 1.997640e-108  0.000000e+00      True  rs62057673 Likely benign              NaN\n",
      "   all       BMI chr17:21298879:C:A chr17 21298879   C   A 0.50000 119  6.423962 0.070371 2.633672e-111 1.997640e-108  0.000000e+00      True  rs33911218        Benign missense_variant\n",
      "   all       BMI chr17:21300581:T:C chr17 21300581   T   C 0.50000 119  6.423962 0.070371 2.633672e-111 1.997640e-108  0.000000e+00      True  rs34105301        Benign missense_variant\n",
      "   all       BMI chr17:21415606:G:A chr17 21415606   G   A 0.50000 119  6.423962 0.070371 2.633672e-111 1.997640e-108  0.000000e+00      True  rs79626424        Benign              NaN\n",
      "   all       BMI chr17:21415640:A:G chr17 21415640   A   G 0.50000 119  6.423962 0.070371 2.633672e-111 1.997640e-108  0.000000e+00      True   rs8076599        Benign missense_variant\n",
      "   all       BMI chr17:21300622:G:A chr17 21300622   G   A 0.50000 119  6.423962 0.070371 2.633672e-111 1.997640e-108  0.000000e+00      True  rs62057672 Likely benign              NaN\n",
      "   all       BMI chr12:47077521:C:T chr12 47077521   C   T 0.05042 119  2.015714 0.430816  7.811975e-06  5.667757e-03  1.576155e-01     False  rs11836128        Benign              NaN\n",
      "   all       BMI chr12:57029413:T:C chr12 57029413   T   C 0.10084 119  1.410744 0.327216  3.402714e-05  2.365879e-02  1.370906e-01     False  rs56128678        Benign              NaN\n",
      "   all IgE_Value chr17:21415639:C:T chr17 21415639   C   T 0.50000 119 -0.391872 0.028700  4.832890e-26  3.665747e-23 -2.220446e-16      True  rs75757803        Benign              NaN\n",
      "   all IgE_Value chr17:21416080:G:A chr17 21416080   G   A 0.50000 119 -0.391872 0.028700  4.832890e-26  3.665747e-23 -2.220446e-16      True  rs73979899        Benign              NaN\n",
      "   all IgE_Value chr17:21415636:C:T chr17 21415636   C   T 0.50000 119 -0.391872 0.028700  4.832890e-26  3.665747e-23 -2.220446e-16      True rs112314728        Benign missense_variant\n",
      "   all IgE_Value chr17:21415669:C:T chr17 21415669   C   T 0.50000 119 -0.391872 0.028700  4.832890e-26  3.665747e-23 -2.220446e-16      True   rs9909970        Benign missense_variant\n",
      "   all IgE_Value chr17:21415896:C:T chr17 21415896   C   T 0.50000 119 -0.391872 0.028700  4.832890e-26  3.665747e-23 -2.220446e-16      True  rs73979896        Benign missense_variant\n",
      "   all IgE_Value chr17:21298407:T:C chr17 21298407   T   C 0.50000 119 -0.391872 0.028700  4.832890e-26  3.665747e-23 -2.220446e-16      True   rs8076154        Benign              NaN\n",
      "[table] Significant associations (asthma-only, FDR<0.05) -> significant_hits_asthmaOnly_FDR_lt_0p05.csv\n",
      "    subset     trait         variant_id CHROM      POS REF ALT      MAF   N      beta       se          pval     qval_fdr            r2  bonf_sig        rsID     CLNSIG_labels    CSQ_terms_str\n",
      "asthmaOnly       BMI chr17:21415896:C:T chr17 21415896   C   T 0.500000 110  6.417711 0.075335 1.446709e-101 1.049619e-98 -2.220446e-16      True  rs73979896            Benign missense_variant\n",
      "asthmaOnly       BMI chr17:21415798:G:A chr17 21415798   G   A 0.500000 110  6.417711 0.075335 1.446709e-101 1.049619e-98 -2.220446e-16      True  rs77283567            Benign              NaN\n",
      "asthmaOnly       BMI chr17:21415775:G:A chr17 21415775   G   A 0.500000 110  6.417711 0.075335 1.446709e-101 1.049619e-98 -2.220446e-16      True  rs75029097            Benign missense_variant\n",
      "asthmaOnly       BMI chr17:21415757:G:A chr17 21415757   G   A 0.500000 110  6.417711 0.075335 1.446709e-101 1.049619e-98 -2.220446e-16      True  rs76265595            Benign missense_variant\n",
      "asthmaOnly       BMI chr17:21415696:G:C chr17 21415696   G   C 0.500000 110  6.417711 0.075335 1.446709e-101 1.049619e-98 -2.220446e-16      True   rs1657741            Benign              NaN\n",
      "asthmaOnly       BMI chr17:21415669:C:T chr17 21415669   C   T 0.500000 110  6.417711 0.075335 1.446709e-101 1.049619e-98 -2.220446e-16      True   rs9909970            Benign missense_variant\n",
      "asthmaOnly       BMI chr17:21415657:C:T chr17 21415657   C   T 0.500000 110  6.417711 0.075335 1.446709e-101 1.049619e-98 -2.220446e-16      True  rs79666990            Benign              NaN\n",
      "asthmaOnly       BMI chr17:21415640:A:G chr17 21415640   A   G 0.500000 110  6.417711 0.075335 1.446709e-101 1.049619e-98 -2.220446e-16      True   rs8076599            Benign missense_variant\n",
      "asthmaOnly       BMI chr17:21415639:C:T chr17 21415639   C   T 0.500000 110  6.417711 0.075335 1.446709e-101 1.049619e-98 -2.220446e-16      True  rs75757803            Benign              NaN\n",
      "asthmaOnly       BMI chr17:21415636:C:T chr17 21415636   C   T 0.500000 110  6.417711 0.075335 1.446709e-101 1.049619e-98 -2.220446e-16      True rs112314728            Benign missense_variant\n",
      "asthmaOnly       BMI chr17:21415606:G:A chr17 21415606   G   A 0.500000 110  6.417711 0.075335 1.446709e-101 1.049619e-98 -2.220446e-16      True  rs79626424            Benign              NaN\n",
      "asthmaOnly       BMI chr17:21415555:G:A chr17 21415555   G   A 0.500000 110  6.417711 0.075335 1.446709e-101 1.049619e-98 -2.220446e-16      True  rs73979893            Benign missense_variant\n",
      "asthmaOnly       BMI chr17:21415351:G:A chr17 21415351   G   A 0.500000 110  6.417711 0.075335 1.446709e-101 1.049619e-98 -2.220446e-16      True  rs72846666            Benign              NaN\n",
      "asthmaOnly       BMI chr17:21302148:C:T chr17 21302148   C   T 0.500000 110  6.417711 0.075335 1.446709e-101 1.049619e-98 -2.220446e-16      True  rs62057721     Likely benign missense_variant\n",
      "asthmaOnly       BMI chr17:21300637:C:T chr17 21300637   C   T 0.500000 110  6.417711 0.075335 1.446709e-101 1.049619e-98 -2.220446e-16      True  rs62057673     Likely benign              NaN\n",
      "asthmaOnly       BMI chr17:21300622:G:A chr17 21300622   G   A 0.500000 110  6.417711 0.075335 1.446709e-101 1.049619e-98 -2.220446e-16      True  rs62057672     Likely benign              NaN\n",
      "asthmaOnly       BMI chr17:21298407:T:C chr17 21298407   T   C 0.500000 110  6.417711 0.075335 1.446709e-101 1.049619e-98 -2.220446e-16      True   rs8076154            Benign              NaN\n",
      "asthmaOnly       BMI chr17:21298879:C:A chr17 21298879   C   A 0.500000 110  6.417711 0.075335 1.446709e-101 1.049619e-98 -2.220446e-16      True  rs33911218            Benign missense_variant\n",
      "asthmaOnly       BMI chr17:21415859:G:A chr17 21415859   G   A 0.500000 110  6.417711 0.075335 1.446709e-101 1.049619e-98 -2.220446e-16      True  rs73313922            Benign missense_variant\n",
      "asthmaOnly       BMI chr17:21298925:G:C chr17 21298925   G   C 0.500000 110  6.417711 0.075335 1.446709e-101 1.049619e-98 -2.220446e-16      True  rs36047035            Benign missense_variant\n",
      "asthmaOnly       BMI chr17:21300581:T:C chr17 21300581   T   C 0.500000 110  6.417711 0.075335 1.446709e-101 1.049619e-98 -2.220446e-16      True  rs34105301            Benign missense_variant\n",
      "asthmaOnly       BMI chr17:21416080:G:A chr17 21416080   G   A 0.500000 110  6.417711 0.075335 1.446709e-101 1.049619e-98 -2.220446e-16      True  rs73979899            Benign              NaN\n",
      "asthmaOnly       BMI    chr6:348906:G:A  chr6   348906   G   A 0.500000 110  6.417711 0.075335 1.446709e-101 1.049619e-98 -2.220446e-16      True  rs11242812            Benign              NaN\n",
      "asthmaOnly       BMI chr12:47077521:C:T chr12 47077521   C   T 0.054545 110  2.048414 0.443720  1.081551e-05 7.519937e-03  1.648085e-01     False  rs11836128            Benign              NaN\n",
      "asthmaOnly       BMI chr12:57029413:T:C chr12 57029413   T   C 0.100000 110  1.511350 0.349353  3.397852e-05 2.267998e-02  1.476971e-01     False  rs56128678            Benign              NaN\n",
      "asthmaOnly       BMI   chr5:1882015:T:G  chr5  1882015   T   G 0.213636 110  1.138285 0.274143  6.599872e-05 4.235848e-02  1.376583e-01     False   rs2232374 Benign|Pathogenic              NaN\n",
      "asthmaOnly IgE_Value chr17:21415555:G:A chr17 21415555   G   A 0.500000 110 -0.384585 0.030784  9.130262e-23 6.624203e-20  1.110223e-16      True  rs73979893            Benign missense_variant\n",
      "asthmaOnly IgE_Value chr17:21415757:G:A chr17 21415757   G   A 0.500000 110 -0.384585 0.030784  9.130262e-23 6.624203e-20  1.110223e-16      True  rs76265595            Benign missense_variant\n",
      "asthmaOnly IgE_Value chr17:21415696:G:C chr17 21415696   G   C 0.500000 110 -0.384585 0.030784  9.130262e-23 6.624203e-20  1.110223e-16      True   rs1657741            Benign              NaN\n",
      "asthmaOnly IgE_Value chr17:21415669:C:T chr17 21415669   C   T 0.500000 110 -0.384585 0.030784  9.130262e-23 6.624203e-20  1.110223e-16      True   rs9909970            Benign missense_variant\n"
     ]
    }
   ],
   "source": [
    "# ===== Plotting + Significant-Hits Table (auto-detects file prefix) =====\n",
    "import os, re\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy.stats import chi2\n",
    "\n",
    "# ---- CONFIG ----\n",
    "PHENO_PATH   = \"ImmPort_URECA_Cross_Black.csv\"\n",
    "PHENO_ID_COL = \"id\"\n",
    "\n",
    "# BMI + IgE as traits\n",
    "TRAITS       = [\"asthma\", \"max_pets\", \"num_smokers\", \"IgE_Value\", \"BMI\"]\n",
    "\n",
    "FDR_ALPHA    = 0.05\n",
    "P_MIN        = 1e-300\n",
    "\n",
    "# Preferred prefixes to search (first that exists per-trait will be used)\n",
    "PREFERRED_PREFIXES = [\"assoc_clinical\", \"assoc_chr5\", \"assoc_chr5_fast\"]\n",
    "\n",
    "# Significant-table outputs\n",
    "SIG_TABLE_ALL          = \"significant_hits_all_FDR_lt_0p05.csv\"\n",
    "SIG_TABLE_ASTHMA_ONLY  = \"significant_hits_asthmaOnly_FDR_lt_0p05.csv\"\n",
    "SIG_TABLE_NONASTH_ONLY = \"significant_hits_nonAsthmaOnly_FDR_lt_0p05.csv\"\n",
    "\n",
    "ANNOTATED_DIR = \"per_chrom_outputs_new\"\n",
    "\n",
    "# ======================================================================\n",
    "# helpers\n",
    "# ======================================================================\n",
    "def _norm(s: str) -> str:\n",
    "    return re.sub(r\"\\W+\", \"_\", str(s).strip().lower())\n",
    "\n",
    "def _resolve_trait_cols(pheno_df: pd.DataFrame, names: list[str]) -> dict:\n",
    "    colmap = { _norm(c): c for c in pheno_df.columns }\n",
    "    resolved = {}\n",
    "    for t in names:\n",
    "        key = _norm(t)\n",
    "        if key not in colmap:\n",
    "            raise KeyError(f\"Trait '{t}' not found (case/space-insensitive).\")\n",
    "        resolved[t] = colmap[key]\n",
    "    return resolved\n",
    "\n",
    "def _ensure_cols(df: pd.DataFrame) -> pd.DataFrame:\n",
    "    if df.empty:\n",
    "        return df\n",
    "    df = df.copy()\n",
    "    df[\"mlog10p\"] = -np.log10(np.clip(df[\"pval\"].values.astype(float), P_MIN, 1.0))\n",
    "    df[\"hit\"] = (df[\"qval_fdr\"].values.astype(float) <= FDR_ALPHA)\n",
    "    df[\"POS\"] = pd.to_numeric(df[\"POS\"], errors=\"coerce\")\n",
    "    df[\"CHROM\"] = df[\"CHROM\"].astype(str)\n",
    "    return df\n",
    "\n",
    "def _savefig(path: str):\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(path, dpi=220, bbox_inches=\"tight\")\n",
    "    plt.close()\n",
    "    print(f\"[fig] saved -> {path}\")\n",
    "\n",
    "def decide_trait(y_raw: pd.Series):\n",
    "    y_num = pd.to_numeric(y_raw, errors=\"coerce\")\n",
    "    nuniq = y_num.dropna().nunique()\n",
    "    sl = y_raw.astype(str).str.strip().str.lower()\n",
    "    bin_map = {\n",
    "        \"1\": 1, \"0\": 0,\n",
    "        \"yes\": 1, \"no\": 0,\n",
    "        \"true\": 1, \"false\": 0,\n",
    "        \"y\": 1, \"n\": 0,\n",
    "        \"asthma\": 1, \"no asthma\": 0,\n",
    "        \"present\": 1, \"absent\": 0,\n",
    "        \"case\": 1, \"control\": 0\n",
    "    }\n",
    "    y_bin = sl.map(bin_map).astype(float)\n",
    "    coverage = y_bin.notna().mean()\n",
    "    if nuniq > 2:\n",
    "        return y_num.astype(float), False\n",
    "    if coverage >= 0.9:\n",
    "        return y_bin, True\n",
    "    if set(pd.unique(y_num.dropna())).issubset({0,1}):\n",
    "        return y_num.astype(float), True\n",
    "    return y_num.astype(float), False\n",
    "\n",
    "def plot_trait_quicklook(trait_name, series, out_prefix, subset_label=\"all\"):\n",
    "    y, is_binary = decide_trait(series)\n",
    "    label_suffix = \"\" if subset_label == \"all\" else f\"_{subset_label}\"\n",
    "    fig_path = f\"{out_prefix}_{trait_name}{label_suffix}_trait_quicklook.png\"\n",
    "    plt.figure(figsize=(6,4))\n",
    "    if is_binary:\n",
    "        vals = y.dropna().astype(int)\n",
    "        plt.bar([\"0\",\"1\"], [(vals==0).sum(), (vals==1).sum()])\n",
    "        plt.ylabel(\"Count\"); plt.title(f\"{trait_name} (binary) counts – {subset_label}\")\n",
    "    else:\n",
    "        yn = y.dropna()\n",
    "        plt.hist(yn.values, bins=30)\n",
    "        plt.xlabel(trait_name); plt.ylabel(\"Frequency\"); plt.title(f\"{trait_name} distribution – {subset_label}\")\n",
    "    _savefig(fig_path)\n",
    "\n",
    "# ======================================================================\n",
    "# MANHATTAN\n",
    "# ======================================================================\n",
    "def manhattan(df, trait_name, out_prefix, subset_label=\"all\", fdr_alpha=FDR_ALPHA):\n",
    "    df = _ensure_cols(df)\n",
    "    if df.empty or df[\"POS\"].isna().all():\n",
    "        print(f\"[{trait_name} | {subset_label}] No results to plot (Manhattan).\")\n",
    "        return\n",
    "\n",
    "    def chr_key(c):\n",
    "        c = str(c).replace(\"chr\",\"\")\n",
    "        try:\n",
    "            return (0, int(c))\n",
    "        except:\n",
    "            order = {\"X\": 23, \"Y\": 24, \"MT\": 25, \"M\":25}\n",
    "            return (1, order.get(c, 99))\n",
    "\n",
    "    chroms = sorted(df[\"CHROM\"].unique(), key=chr_key)\n",
    "\n",
    "    offsets = {}\n",
    "    current = 0\n",
    "    tick_positions = []\n",
    "    tick_labels = []\n",
    "    for c in chroms:\n",
    "        subset = df.loc[df[\"CHROM\"]==c, \"POS\"].dropna()\n",
    "        if subset.empty:\n",
    "            offsets[c] = current\n",
    "            continue\n",
    "        minpos, maxpos = subset.min(), subset.max()\n",
    "        offsets[c] = current\n",
    "        tick_positions.append(current + (minpos + maxpos)/2.0)\n",
    "        tick_labels.append(str(c))\n",
    "        current += maxpos + 3e5\n",
    "\n",
    "    df[\"cum_pos\"] = df.apply(lambda r: r[\"POS\"] + offsets.get(str(r[\"CHROM\"]), 0), axis=1)\n",
    "\n",
    "    m = max(len(df), 1)\n",
    "    bonf_p = 0.05 / m\n",
    "    bonf_y = -np.log10(max(bonf_p, P_MIN))\n",
    "\n",
    "    hits = df[\"hit\"].values\n",
    "    has_fdr_hits = hits.any()\n",
    "    if has_fdr_hits:\n",
    "        p_fdr_cut = float(df.loc[hits, \"pval\"].max())\n",
    "        fdr_y = -np.log10(max(p_fdr_cut, P_MIN))\n",
    "    else:\n",
    "        fdr_y = None\n",
    "\n",
    "    chrom_colors = [\n",
    "        \"#a23b3b\", \"#5f5f5f\", \"#2f6f3e\", \"#4f3b3b\", \"#6f7f7f\",\n",
    "        \"#ac8740\", \"#7ad0f5\", \"#111111\", \"#a65065\", \"#b06840\",\n",
    "        \"#5a4fd0\", \"#c1b741\"\n",
    "    ]\n",
    "\n",
    "    label_suffix = \"\" if subset_label == \"all\" else f\"_{subset_label}\"\n",
    "    fig_path = f\"{out_prefix}_{trait_name}{label_suffix}_manhattan.png\"\n",
    "    plt.figure(figsize=(11.5,4.2))\n",
    "\n",
    "    for i, c in enumerate(chroms):\n",
    "        mask = (df[\"CHROM\"] == c)\n",
    "        col = chrom_colors[i % len(chrom_colors)]\n",
    "        plt.scatter(\n",
    "            df.loc[mask, \"cum_pos\"],\n",
    "            df.loc[mask, \"mlog10p\"],\n",
    "            s=10,\n",
    "            alpha=0.8,\n",
    "            c=col,\n",
    "            edgecolor=\"none\",\n",
    "            zorder=1\n",
    "        )\n",
    "\n",
    "    plt.axhline(bonf_y, linestyle=\"--\", linewidth=1.1, color=\"black\",\n",
    "                label=f\"Bonferroni 0.05 (p<{bonf_p:.1e})\")\n",
    "    if fdr_y is not None:\n",
    "        plt.axhline(fdr_y, linestyle=\":\", linewidth=1.2, color=\"purple\",\n",
    "                    label=f\"FDR<{fdr_alpha}\")\n",
    "\n",
    "    # --- jittered labels for FDR-significant hits ---\n",
    "    if has_fdr_hits:\n",
    "        sig_df = df.loc[hits].sort_values(\"pval\")\n",
    "        top_hits = sig_df.head(15)  # up to 15 labels\n",
    "        # Sort by chromosome position so neighboring peaks are offset nicely\n",
    "        top_hits = top_hits.sort_values([\"CHROM\", \"cum_pos\"]).reset_index(drop=True)\n",
    "\n",
    "        for i, (_, r) in enumerate(top_hits.iterrows()):\n",
    "            label = (\n",
    "                r.get(\"rsID\")\n",
    "                or r.get(\"rsid\")\n",
    "                or r.get(\"ID\")\n",
    "                or r.get(\"SYMBOL\")\n",
    "                or r.get(\"gene\")\n",
    "                or r.get(\"Gene.refGene\")\n",
    "                or r[\"variant_id\"]\n",
    "            )\n",
    "            # deterministic jitter: alternate x and stack vertically\n",
    "            x_jitter = (i % 3 - 1) * 5e5      # -0.5Mb, 0, +0.5Mb\n",
    "            y_jitter = 0.35 + (i // 3) * 0.4  # stack rows\n",
    "            plt.text(\n",
    "                r[\"cum_pos\"] + x_jitter,\n",
    "                r[\"mlog10p\"] + y_jitter,\n",
    "                str(label),\n",
    "                fontsize=7,\n",
    "                rotation=20,\n",
    "                ha=\"left\",\n",
    "                va=\"bottom\",\n",
    "                color=\"black\",\n",
    "                zorder=2\n",
    "            )\n",
    "\n",
    "    plt.xticks(tick_positions, tick_labels, fontsize=8)\n",
    "    plt.xlabel(\"Chromosomes\")\n",
    "    plt.ylabel(r\"$-\\log_{10}(P)$\")\n",
    "    title_subset = \"all subjects\" if subset_label == \"all\" else subset_label\n",
    "    plt.title(f\"Manhattan – {trait_name} ({title_subset})\")\n",
    "    plt.legend(frameon=False, fontsize=7, loc=\"upper right\")\n",
    "    _savefig(fig_path)\n",
    "\n",
    "# ======================================================================\n",
    "# VOLCANO — standard GWAS-style\n",
    "# ======================================================================\n",
    "def volcano(df, trait_name, out_prefix, subset_label=\"all\"):\n",
    "    df = _ensure_cols(df)\n",
    "    if df.empty:\n",
    "        print(f\"[{trait_name} | {subset_label}] No results to plot (volcano).\")\n",
    "        return\n",
    "\n",
    "    x = df[\"beta\"].values.astype(float)\n",
    "    y = df[\"mlog10p\"].values\n",
    "\n",
    "    # Significance by FDR\n",
    "    sig = df[\"qval_fdr\"].astype(float) <= FDR_ALPHA\n",
    "    ns  = ~sig\n",
    "\n",
    "    # Bonferroni line\n",
    "    m = max(len(df), 1)\n",
    "    bonf_p = 0.05 / m\n",
    "    bonf_y = -np.log10(max(bonf_p, P_MIN))\n",
    "\n",
    "    # (optional) FDR cutoff line at the max p among FDR-significant hits\n",
    "    if sig.any():\n",
    "        p_fdr_cut = float(df.loc[sig, \"pval\"].max())\n",
    "        fdr_y = -np.log10(max(p_fdr_cut, P_MIN))\n",
    "    else:\n",
    "        fdr_y = None\n",
    "\n",
    "    label_suffix = \"\" if subset_label == \"all\" else f\"_{subset_label}\"\n",
    "    fig_path = f\"{out_prefix}_{trait_name}{label_suffix}_volcano.png\"\n",
    "    plt.figure(figsize=(6,5))\n",
    "\n",
    "    # NS grey\n",
    "    plt.scatter(x[ns], y[ns], s=12, alpha=0.35, c=\"lightgrey\", edgecolor=\"none\", label=\"Not sig\", zorder=1)\n",
    "\n",
    "    # FDR-significant points in color\n",
    "    plt.scatter(x[sig], y[sig], s=18, alpha=0.9, c=\"#d95f02\", edgecolor=\"none\", label=f\"FDR<{FDR_ALPHA}\", zorder=2)\n",
    "\n",
    "    # Horizontal lines\n",
    "    plt.axhline(bonf_y, linestyle=\"--\", linewidth=1, color=\"black\",\n",
    "                label=f\"Bonf 0.05 (p<{bonf_p:.1e})\")\n",
    "    if fdr_y is not None:\n",
    "        plt.axhline(fdr_y, linestyle=\":\", linewidth=1, color=\"purple\",\n",
    "                    label=\"FDR cutoff\")\n",
    "\n",
    "    # Label top hits (by p-value) with jitter to avoid overlap\n",
    "    if sig.any():\n",
    "        top = df.loc[sig].sort_values(\"pval\").head(12)\n",
    "    else:\n",
    "        top = df.sort_values(\"pval\").head(12)\n",
    "\n",
    "    for i, (_, r) in enumerate(top.iterrows()):\n",
    "        label = (\n",
    "            r.get(\"rsID\")\n",
    "            or r.get(\"rsid\")\n",
    "            or r.get(\"ID\")\n",
    "            or r.get(\"SYMBOL\")\n",
    "            or r.get(\"gene\")\n",
    "            or r.get(\"Gene.refGene\")\n",
    "            or r[\"variant_id\"]\n",
    "        )\n",
    "        x_jitter = (i % 3 - 1) * 0.2       # -0.2, 0, +0.2\n",
    "        y_jitter = 0.2 + (i // 3) * 0.25   # stack rows upwards\n",
    "        plt.text(\n",
    "            r[\"beta\"] + x_jitter,\n",
    "            r[\"mlog10p\"] + y_jitter,\n",
    "            str(label),\n",
    "            fontsize=7,\n",
    "            color=\"red\",\n",
    "            ha=\"center\",\n",
    "            va=\"bottom\",\n",
    "            zorder=3\n",
    "        )\n",
    "\n",
    "    plt.xlabel(\"Effect size (β per ALT allele)\")\n",
    "    plt.ylabel(r\"$-\\log_{10}(\\mathrm{Pvalue})$\")\n",
    "    title_subset = \"all subjects\" if subset_label == \"all\" else subset_label\n",
    "    plt.title(f\"Volcano – {trait_name} ({title_subset})\")\n",
    "    plt.legend(frameon=False, fontsize=7, loc=\"upper right\")\n",
    "    _savefig(fig_path)\n",
    "\n",
    "# ======================================================================\n",
    "# QQ + MAF\n",
    "# ======================================================================\n",
    "def qq_plot(df, trait_name, out_prefix, subset_label=\"all\"):\n",
    "    df = _ensure_cols(df)\n",
    "    if df.empty:\n",
    "        print(f\"[{trait_name} | {subset_label}] No results to plot (QQ).\")\n",
    "        return\n",
    "    p = np.clip(np.sort(df[\"pval\"].values.astype(float)), P_MIN, 1.0)\n",
    "    n = len(p)\n",
    "    exp = (np.arange(1, n+1) - 0.5) / n\n",
    "    o = -np.log10(p); e = -np.log10(exp)\n",
    "\n",
    "    chisq = chi2.isf(p, df=1)\n",
    "    lam_gc = np.median(chisq) / chi2.ppf(0.5, 1)\n",
    "\n",
    "    label_suffix = \"\" if subset_label == \"all\" else f\"_{subset_label}\"\n",
    "    fig_path = f\"{out_prefix}_{trait_name}{label_suffix}_qq.png\"\n",
    "    plt.figure(figsize=(5,5))\n",
    "    plt.scatter(e, o, s=6, alpha=0.6)\n",
    "    lim = max(e.max(), o.max()) * 1.02\n",
    "    plt.plot([0, lim], [0, lim], color=\"black\", linewidth=1)\n",
    "    plt.xlabel(\"Expected -log10(p)\"); plt.ylabel(\"Observed -log10(p)\")\n",
    "    title_subset = \"all subjects\" if subset_label == \"all\" else subset_label\n",
    "    plt.title(f\"QQ plot – {trait_name} ({title_subset}) (λGC={lam_gc:.3f})\")\n",
    "    _savefig(fig_path)\n",
    "\n",
    "def maf_vs_significance(df, trait_name, out_prefix, subset_label=\"all\"):\n",
    "    df = _ensure_cols(df)\n",
    "    if df.empty:\n",
    "        print(f\"[{trait_name} | {subset_label}] No results to plot (MAF vs sig).\")\n",
    "        return\n",
    "    maf = pd.to_numeric(df[\"MAF\"], errors=\"coerce\").values\n",
    "    y   = df[\"mlog10p\"].values\n",
    "    hit = df[\"hit\"].values\n",
    "    label_suffix = \"\" if subset_label == \"all\" else f\"_{subset_label}\"\n",
    "    fig_path = f\"{out_prefix}_{trait_name}{label_suffix}_maf_vs_mlog10p.png\"\n",
    "    plt.figure(figsize=(6,5))\n",
    "    plt.scatter(maf[~hit], y[~hit], s=8, alpha=0.6)\n",
    "    if hit.any(): plt.scatter(maf[hit], y[hit], s=12, alpha=0.9)\n",
    "    plt.xlabel(\"Minor Allele Frequency (MAF)\"); plt.ylabel(\"-log10(p)\")\n",
    "    title_subset = \"all subjects\" if subset_label == \"all\" else subset_label\n",
    "    plt.title(f\"MAF vs significance – {trait_name} ({title_subset})\")\n",
    "    _savefig(fig_path)\n",
    "\n",
    "# ======================================================================\n",
    "# load phenotype & build annotation table (rsID/gene/functional)\n",
    "# ======================================================================\n",
    "pheno = pd.read_csv(PHENO_PATH)\n",
    "trait_col_map = _resolve_trait_cols(pheno, TRAITS)\n",
    "pheno = pheno.set_index(pheno[PHENO_ID_COL].astype(str)).copy()\n",
    "pheno.index.name = \"sample_id\"\n",
    "\n",
    "annot_merge = None\n",
    "if ANNOTATED_DIR and os.path.isdir(ANNOTATED_DIR):\n",
    "    ann_files = [os.path.join(ANNOTATED_DIR, f) for f in os.listdir(ANNOTATED_DIR) if f.endswith(\".annotated.csv\")]\n",
    "    if ann_files:\n",
    "        ann_list = []\n",
    "        for f in sorted(ann_files):\n",
    "            try:\n",
    "                tmp = pd.read_csv(f, dtype=str, low_memory=False)\n",
    "                ann_list.append(tmp)\n",
    "            except Exception:\n",
    "                pass\n",
    "        if ann_list:\n",
    "            annot_merge = pd.concat(ann_list, ignore_index=True)\n",
    "            if \"variant_id\" not in annot_merge.columns and {\"#CHROM\",\"POS\",\"REF\",\"ALT\"}.issubset(annot_merge.columns):\n",
    "                annot_merge[\"variant_id\"] = (\n",
    "                    annot_merge[\"#CHROM\"].astype(str) + \":\" +\n",
    "                    annot_merge[\"POS\"].astype(str) + \":\" +\n",
    "                    annot_merge[\"REF\"].astype(str) + \":\" +\n",
    "                    annot_merge[\"ALT\"].astype(str)\n",
    "                )\n",
    "            keep_cols = [\"variant_id\", \"#CHROM\", \"POS\", \"REF\", \"ALT\"]\n",
    "            rs_like = [c for c in annot_merge.columns if c.lower() in {\"id\",\"rsid\",\"rs_id\",\"rs\"}]\n",
    "            gene_like = [c for c in annot_merge.columns if c.lower() in {\"gene\",\"gene.refgene\",\"symbol\",\"gene_name\"}]\n",
    "            # NEW: functional annotation column (CSQ_terms_str) if present\n",
    "            func_cols = [c for c in annot_merge.columns if c == \"CSQ_terms_str\"]\n",
    "            keep_cols = keep_cols + rs_like + gene_like + func_cols\n",
    "            annot_merge = annot_merge[keep_cols].drop_duplicates()\n",
    "\n",
    "# overlap for trait quicklook (just to get a sample set; plots don't depend on genotypes)\n",
    "overlap = list(pheno.index)\n",
    "if ANNOTATED_DIR and os.path.isdir(ANNOTATED_DIR):\n",
    "    any_file = None\n",
    "    for name in sorted(os.listdir(ANNOTATED_DIR)):\n",
    "        if name.endswith(\".annotated.csv\"):\n",
    "            any_file = os.path.join(ANNOTATED_DIR, name)\n",
    "            break\n",
    "    if any_file:\n",
    "        cols = list(pd.read_csv(any_file, nrows=0).columns)\n",
    "        overlap = [c for c in cols if c in pheno.index]\n",
    "print(f\"[plot] Overlapping samples for quicklook: {len(overlap)}\")\n",
    "\n",
    "# ======================================================================\n",
    "# main loop (all / asthma-only / non-asthma-only)\n",
    "# ======================================================================\n",
    "# contexts: (middle_tag_in_filename, subset_label, human_readable_suffix)\n",
    "CONTEXTS = [\n",
    "    (\"\",              \"all\",            \"\"),\n",
    "    (\"asthmaOnly_\",   \"asthmaOnly\",     \" (asthma only)\"),\n",
    "    (\"nonAsthmaOnly_\",\"nonAsthmaOnly\",  \" (non-asthma only)\")\n",
    "]\n",
    "\n",
    "signif_frames = {c_label: [] for _, c_label, _ in CONTEXTS}\n",
    "\n",
    "for trait in TRAITS:\n",
    "    for mid_tag, subset_label, _suffix in CONTEXTS:\n",
    "        # We do NOT expect stratified outputs for the asthma case/control trait itself\n",
    "        if trait == \"asthma\" and subset_label != \"all\":\n",
    "            continue\n",
    "\n",
    "        assoc_path = None\n",
    "        out_prefix_used = None\n",
    "\n",
    "        for pref in PREFERRED_PREFIXES:\n",
    "            if mid_tag == \"\":\n",
    "                candidate = f\"{pref}_{trait}.csv\"\n",
    "            else:\n",
    "                candidate = f\"{pref}_{mid_tag}{trait}.csv\"\n",
    "            if os.path.exists(candidate):\n",
    "                assoc_path = candidate\n",
    "                # prefix used in figure filenames\n",
    "                out_prefix_used = pref if subset_label == \"all\" else f\"{pref}_{subset_label}\"\n",
    "                break\n",
    "\n",
    "        if assoc_path is None:\n",
    "            print(f\"[skip] No association file found for {trait} ({subset_label}).\")\n",
    "            continue\n",
    "\n",
    "        res = pd.read_csv(assoc_path)\n",
    "\n",
    "        # merge rsID/gene/functional info from annotation\n",
    "        if annot_merge is not None:\n",
    "            if \"variant_id\" in res.columns and \"variant_id\" in annot_merge.columns:\n",
    "                res = res.merge(annot_merge, on=\"variant_id\", how=\"left\", suffixes=(\"\", \"_ann\"))\n",
    "            else:\n",
    "                res = res.merge(\n",
    "                    annot_merge,\n",
    "                    left_on=[\"CHROM\",\"POS\",\"REF\",\"ALT\"],\n",
    "                    right_on=[\"#CHROM\",\"POS\",\"REF\",\"ALT\"],\n",
    "                    how=\"left\",\n",
    "                    suffixes=(\"\", \"_ann\")\n",
    "                )\n",
    "\n",
    "        # trait distribution quicklook (uses all samples; subset doesn't change phenotype distribution much)\n",
    "        ph_col = trait_col_map[trait]\n",
    "        series = pheno.loc[overlap, ph_col]\n",
    "        plot_trait_quicklook(trait, series, out_prefix=out_prefix_used, subset_label=subset_label)\n",
    "\n",
    "        # GWAS-style plots for this subset\n",
    "        manhattan(res, trait, out_prefix=out_prefix_used, subset_label=subset_label)\n",
    "        qq_plot(res, trait, out_prefix=out_prefix_used, subset_label=subset_label)\n",
    "        volcano(res, trait, out_prefix=out_prefix_used, subset_label=subset_label)\n",
    "        maf_vs_significance(res, trait, out_prefix=out_prefix_used, subset_label=subset_label)\n",
    "\n",
    "        # collect significant hits (FDR < 0.05)\n",
    "        if \"qval_fdr\" not in res.columns:\n",
    "            print(f\"[warn] qval_fdr missing for {trait} ({subset_label}); skipping sig-table entry.\")\n",
    "            continue\n",
    "\n",
    "        res_sig = res.loc[(res[\"qval_fdr\"].astype(float) <= FDR_ALPHA)].copy()\n",
    "        if not res_sig.empty:\n",
    "            m = len(res)\n",
    "            res_sig[\"bonf_sig\"] = res_sig[\"pval\"].astype(float) < (0.05 / max(m,1))\n",
    "            res_sig[\"trait\"] = trait\n",
    "            res_sig[\"subset\"] = subset_label\n",
    "\n",
    "            # normalize rsID col name (if present)\n",
    "            if \"ID\" in res_sig.columns:\n",
    "                res_sig.rename(columns={\"ID\":\"rsID\"}, inplace=True)\n",
    "            elif \"rsid\" in res_sig.columns:\n",
    "                res_sig.rename(columns={\"rsid\":\"rsID\"}, inplace=True)\n",
    "            elif \"rs_id\" in res_sig.columns:\n",
    "                res_sig.rename(columns={\"rs_id\":\"rsID\"}, inplace=True)\n",
    "\n",
    "            # gene annotations\n",
    "            extra_anns = []\n",
    "            for cand in [\"gene\",\"GENE\",\"Gene.refGene\",\"SYMBOL\",\"gene_name\"]:\n",
    "                if cand in res_sig.columns:\n",
    "                    extra_anns.append(cand)\n",
    "\n",
    "            # functional annotation column\n",
    "            func_cols = []\n",
    "            if \"CSQ_terms_str\" in res_sig.columns:\n",
    "                func_cols.append(\"CSQ_terms_str\")\n",
    "\n",
    "            base_cols = [\n",
    "                \"subset\",\"trait\",\"variant_id\",\"CHROM\",\"POS\",\"REF\",\"ALT\",\n",
    "                \"MAF\",\"N\",\"beta\",\"se\",\"pval\",\"qval_fdr\",\"r2\",\"bonf_sig\",\n",
    "                \"rsID\",\"CLNSIG_labels\"\n",
    "            ]\n",
    "            base_cols = [c for c in base_cols if c in res_sig.columns]\n",
    "            res_sig = res_sig[base_cols + func_cols + extra_anns]\n",
    "\n",
    "            signif_frames[subset_label].append(res_sig)\n",
    "\n",
    "# ---- Write separate significant tables ----\n",
    "if signif_frames[\"all\"]:\n",
    "    sig_all = pd.concat(signif_frames[\"all\"], ignore_index=True)\n",
    "    sig_all = sig_all.sort_values([\"trait\",\"pval\"]).reset_index(drop=True)\n",
    "    sig_all.to_csv(SIG_TABLE_ALL, index=False)\n",
    "    print(f\"[table] Significant associations (all subjects, FDR<{FDR_ALPHA}) -> {SIG_TABLE_ALL}\")\n",
    "    try:\n",
    "        print(sig_all.head(30).to_string(index=False))\n",
    "    except Exception:\n",
    "        print(sig_all.head(30))\n",
    "\n",
    "if signif_frames[\"asthmaOnly\"]:\n",
    "    sig_asth = pd.concat(signif_frames[\"asthmaOnly\"], ignore_index=True)\n",
    "    sig_asth = sig_asth.sort_values([\"trait\",\"pval\"]).reset_index(drop=True)\n",
    "    sig_asth.to_csv(SIG_TABLE_ASTHMA_ONLY, index=False)\n",
    "    print(f\"[table] Significant associations (asthma-only, FDR<{FDR_ALPHA}) -> {SIG_TABLE_ASTHMA_ONLY}\")\n",
    "    try:\n",
    "        print(sig_asth.head(30).to_string(index=False))\n",
    "    except Exception:\n",
    "        print(sig_asth.head(30))\n",
    "\n",
    "if signif_frames[\"nonAsthmaOnly\"]:\n",
    "    sig_non = pd.concat(signif_frames[\"nonAsthmaOnly\"], ignore_index=True)\n",
    "    sig_non = sig_non.sort_values([\"trait\",\"pval\"]).reset_index(drop=True)\n",
    "    sig_non.to_csv(SIG_TABLE_NONASTH_ONLY, index=False)\n",
    "    print(f\"[table] Significant associations (non-asthma-only, FDR<{FDR_ALPHA}) -> {SIG_TABLE_NONASTH_ONLY}\")\n",
    "    try:\n",
    "        print(sig_non.head(30).to_string(index=False))\n",
    "    except Exception:\n",
    "        print(sig_non.head(30))\n",
    "\n",
    "if not (signif_frames[\"all\"] or signif_frames[\"asthmaOnly\"] or signif_frames[\"nonAsthmaOnly\"]):\n",
    "    print(f\"[table] No variants met FDR<{FDR_ALPHA} in any subset.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c5796d7-bbed-4c70-8522-b1acfdd5fafc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[info] Resolved trait columns: {'asthma': 'asthma ', 'max_pets': 'max_pets', 'num_smokers': 'num_smokers', 'BMI': 'BMI', 'IgE_Value': 'IgE_Value'}\n",
      "[info] Variants total: 1,629,382  |  with valid clinical labels: 16,688\n",
      "[info] Overlapping samples: 119\n",
      "[trait asthma] Using Fisher (trait BINARY; rule: as-is). N non-missing=119\n",
      "[trait asthma] No analyzable variants produced results.\n",
      "[save] asthma -> assoc_exact_fisher_asthma.csv (variants tested: 0)\n",
      "[trait max_pets] Using Fisher (trait NUMERIC→BINARY; rule: >= median (0)). N non-missing=119\n",
      "[trait max_pets] No analyzable variants produced results.\n",
      "[save] max_pets -> assoc_exact_fisher_max_pets.csv (variants tested: 0)\n",
      "[trait num_smokers] Using Fisher (trait NUMERIC→BINARY; rule: >= median (1)). N non-missing=119\n",
      "[trait num_smokers] No analyzable variants produced results.\n",
      "[save] num_smokers -> assoc_exact_fisher_num_smokers.csv (variants tested: 0)\n",
      "[trait BMI] Using Fisher (trait NUMERIC→BINARY; rule: >= median (12.8)). N non-missing=119\n",
      "[trait BMI] No analyzable variants produced results.\n",
      "[save] BMI -> assoc_exact_fisher_bmi.csv (variants tested: 0)\n",
      "[trait IgE_Value] Using Fisher (trait NUMERIC→BINARY; rule: >= median (0.09)). N non-missing=119\n"
     ]
    }
   ],
   "source": [
    "# =============================================================\n",
    "# Association analysis (EXACT) on clinically annotated SNPs\n",
    "# - Loads annotated files from per_chrom_outputs_new (same as before)\n",
    "# - Maps CLNSIG numeric codes → labels (same)\n",
    "# - Filters out missing / unknown codes (same)\n",
    "# - NEW: For each trait:\n",
    "#       * If binary → use as-is\n",
    "#       * If numeric → binarize (default: median split, configurable)\n",
    "# - Runs Fisher's exact tests for:\n",
    "#       * Dominant model: carriers (1/2) vs non-carriers (0)\n",
    "#       * Recessive model: hom-alt (2) vs others (0/1), if any 2s exist\n",
    "# - BH-FDR correction within trait×model\n",
    "# - Outputs per-trait CSVs with OR, pval, qval and model label\n",
    "# =============================================================\n",
    "import os, glob, re, warnings\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from scipy.stats import fisher_exact\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "from statsmodels.tools.sm_exceptions import ConvergenceWarning, PerfectSeparationWarning\n",
    "\n",
    "# ---------- Silence noisy warnings ----------\n",
    "warnings.filterwarnings(\"ignore\", category=ConvergenceWarning)\n",
    "warnings.filterwarnings(\"ignore\", category=PerfectSeparationWarning)\n",
    "warnings.filterwarnings(\"ignore\", category=RuntimeWarning)\n",
    "np.seterr(divide=\"ignore\", invalid=\"ignore\", over=\"ignore\")\n",
    "\n",
    "# ========= User inputs =========\n",
    "PHENO_PATH   = \"ImmPort_URECA_Cross_Black.csv\"\n",
    "PHENO_ID_COL = \"id\"\n",
    "\n",
    "# Include numeric (BMI, IgE) and binary/count traits here. We'll auto-detect numeric vs binary.\n",
    "TRAITS       = [\"asthma\", \"max_pets\", \"num_smokers\", \"BMI\", \"IgE_Value\"]\n",
    "\n",
    "# Optional: provide clinical thresholds to binarize specific numeric traits instead of median\n",
    "# Format: {\"TraitNameInTRAITS\": cutoff_value}  → outcome = 1 if value >= cutoff else 0\n",
    "CUSTOM_CUTOFFS = {\n",
    "    # \"BMI\": 25.0,   # example: overweight+\n",
    "    # \"IgE\": 100.0,  # example threshold (kU/L). Replace with your clinical cutoffs if desired.\n",
    "}\n",
    "\n",
    "# Where your pipeline wrote the annotated tables:\n",
    "OUTDIR_ANN   = \"per_chrom_outputs_new\"\n",
    "# Output prefix for association results:\n",
    "OUT_PREFIX   = \"assoc_exact_fisher\"\n",
    "# ==============================\n",
    "\n",
    "# ---------- Helpers (yours, unchanged except tiny note on typing) ----------\n",
    "def norm(s: str) -> str:\n",
    "    return re.sub(r\"\\W+\", \"_\", str(s).strip().lower())\n",
    "\n",
    "def resolve_trait_columns(pheno_df: pd.DataFrame, wanted_traits: list[str]) -> dict:\n",
    "    colmap = {norm(c): c for c in pheno_df.columns}\n",
    "    resolved = {}\n",
    "    for t in wanted_traits:\n",
    "        key = norm(t)\n",
    "        if key in colmap:\n",
    "            resolved[t] = colmap[key]\n",
    "        else:\n",
    "            cands = [c for k, c in colmap.items() if key == k or key in k]\n",
    "            if len(cands) == 1:\n",
    "                resolved[t] = cands[0]\n",
    "            else:\n",
    "                raise KeyError(f\"Could not find trait column for '{t}'. \"\n",
    "                               f\"Some normalized columns: {list(sorted(colmap.keys()))[:15]} ...\")\n",
    "    return resolved\n",
    "\n",
    "def coerce_binary_series(s: pd.Series) -> pd.Series:\n",
    "    sl = s.astype(str).str.strip().str.lower()\n",
    "    mapping = {\n",
    "        \"1\": 1, \"0\": 0,\n",
    "        \"yes\": 1, \"no\": 0,\n",
    "        \"true\": 1, \"false\": 0,\n",
    "        \"y\": 1, \"n\": 0,\n",
    "        \"asthma\": 1, \"no asthma\": 0,\n",
    "        \"present\": 1, \"absent\": 0,\n",
    "        \"case\": 1, \"control\": 0\n",
    "    }\n",
    "    out = sl.map(mapping)\n",
    "    if out.notna().sum() == 0:\n",
    "        out = pd.to_numeric(s, errors=\"coerce\")\n",
    "    return out.astype(float)\n",
    "\n",
    "def decide_trait(y_raw: pd.Series):\n",
    "    \"\"\"\n",
    "    Decide if a trait is binary or numeric:\n",
    "    - If numeric coercion has >2 unique non-NaN values => numeric\n",
    "    - Else, try string→binary mapping; if coverage >= 90% => binary\n",
    "    - Else, if numeric values are truly {0,1} => binary\n",
    "    - Else => numeric\n",
    "    \"\"\"\n",
    "    y_num = pd.to_numeric(y_raw, errors=\"coerce\")\n",
    "    nuniq = y_num.dropna().nunique()\n",
    "\n",
    "    sl = y_raw.astype(str).str.strip().str.lower()\n",
    "    bin_map = {\n",
    "        \"1\": 1, \"0\": 0,\n",
    "        \"yes\": 1, \"no\": 0,\n",
    "        \"true\": 1, \"false\": 0,\n",
    "        \"y\": 1, \"n\": 0,\n",
    "        \"asthma\": 1, \"no asthma\": 0,\n",
    "        \"present\": 1, \"absent\": 0,\n",
    "        \"case\": 1, \"control\": 0\n",
    "    }\n",
    "    y_bin = sl.map(bin_map).astype(float)\n",
    "    coverage = y_bin.notna().mean()\n",
    "\n",
    "    if nuniq > 2:\n",
    "        return y_num.astype(float), False\n",
    "    if coverage >= 0.9:\n",
    "        return y_bin, True\n",
    "    if set(pd.unique(y_num.dropna())).issubset({0,1}):\n",
    "        return y_num.astype(float), True\n",
    "    return y_num.astype(float), False\n",
    "\n",
    "def gt_to_dosage(field: str) -> float:\n",
    "    # returns 0.0, 1.0, 2.0 or np.nan\n",
    "    if not isinstance(field, str) or field == \"\" or field == \".\":\n",
    "        return np.nan\n",
    "    gt = field.split(\":\", 1)[0]\n",
    "    if gt in {\"./.\", \".|.\", \".\"}:\n",
    "        return np.nan\n",
    "    if \"/\" in gt:\n",
    "        a1, a2 = gt.split(\"/\")\n",
    "    elif \"|\" in gt:\n",
    "        a1, a2 = gt.split(\"|\")\n",
    "    else:\n",
    "        return np.nan\n",
    "    if a1 not in {\"0\",\"1\"} or a2 not in {\"0\",\"1\"}:\n",
    "        return np.nan\n",
    "    return float((a1 == \"1\") + (a2 == \"1\"))\n",
    "\n",
    "def maf_from_dosage(dosages: np.ndarray) -> float | None:\n",
    "    x = dosages[~np.isnan(dosages)]\n",
    "    n = x.size\n",
    "    if n == 0: return None\n",
    "    af = x.sum() / (2.0 * n)\n",
    "    return float(min(af, 1 - af))\n",
    "\n",
    "# Map CLNSIG numeric codes → labels (yours)\n",
    "CLNSIG_MAP = {\n",
    "    \"0\":  \"Uncertain significance\",\n",
    "    \"1\":  \"not provided\",\n",
    "    \"2\":  \"Benign\",\n",
    "    \"3\":  \"Likely benign\",\n",
    "    \"4\":  \"Likely pathogenic\",\n",
    "    \"5\":  \"Pathogenic\",\n",
    "    \"6\":  \"Drug response\",\n",
    "    \"8\":  \"Confers sensitivity\",\n",
    "    \"9\":  \"Risk factor\",\n",
    "    \"10\": \"Association\",\n",
    "    \"11\": \"Protective\",\n",
    "    \"12\": \"Conflicting interpretations of pathogenicity\",\n",
    "    \"13\": \"Affects\",\n",
    "    \"14\": \"Association not found\",\n",
    "    \"15\": \"Benign/Likely benign\",\n",
    "    \"16\": \"Pathogenic/Likely pathogenic\",\n",
    "    \"17\": \"Conflicting data from submitters\",\n",
    "    \"18\": \"Pathogenic, low penetrance\",\n",
    "    \"19\": \"Likely pathogenic, low penetrance\",\n",
    "    \"20\": \"Established risk allele\",\n",
    "    \"21\": \"Likely risk allele\",\n",
    "    \"22\": \"Uncertain risk allele\",\n",
    "    \"255\":\"other\",\n",
    "    \".\":  None,\n",
    "    \"\":   None,\n",
    "}\n",
    "CLNSIG_SPLIT = re.compile(r\"[\\/;\\|,]+\")\n",
    "def clnsig_codes_to_labels(clnsig_val: str) -> list[str]:\n",
    "    if not isinstance(clnsig_val, str):\n",
    "        return []\n",
    "    codes = [t.strip() for t in CLNSIG_SPLIT.split(clnsig_val) if t.strip() != \"\"]\n",
    "    labels = []\n",
    "    for code in codes:\n",
    "        if code in CLNSIG_MAP:\n",
    "            lab = CLNSIG_MAP[code]\n",
    "            if lab:\n",
    "                labels.append(lab)\n",
    "    return labels\n",
    "\n",
    "# ----------------- Load phenotype + annotated variants -----------------\n",
    "pheno = pd.read_csv(PHENO_PATH)\n",
    "\n",
    "# Resolve trait columns (robust to spacing/case)\n",
    "trait_col_map = resolve_trait_columns(pheno, TRAITS)\n",
    "print(\"[info] Resolved trait columns:\", trait_col_map)\n",
    "\n",
    "# Normalize phenotype IDs to string index\n",
    "pheno_ids = pheno[PHENO_ID_COL].astype(str)\n",
    "pheno = pheno.set_index(pheno_ids).copy()\n",
    "pheno.index.name = \"sample_id\"\n",
    "\n",
    "# Load annotated tables\n",
    "ann_paths = sorted(glob.glob(os.path.join(OUTDIR_ANN, \"chr*.annotated.csv\")))\n",
    "if not ann_paths:\n",
    "    raise FileNotFoundError(f\"No annotated files found in {OUTDIR_ANN}. Expected chr*.annotated.csv\")\n",
    "\n",
    "annot = pd.concat([pd.read_csv(p, dtype=str, low_memory=False) for p in ann_paths],\n",
    "                  ignore_index=True)\n",
    "\n",
    "# Keep only variants with at least one valid CLNSIG label\n",
    "annot[\"CLNSIG\"] = annot[\"CLNSIG\"].fillna(\"\")\n",
    "annot[\"CLNSIG_labels_list\"] = annot[\"CLNSIG\"].apply(clnsig_codes_to_labels)\n",
    "annot[\"CLNSIG_labels_str\"]  = annot[\"CLNSIG_labels_list\"].apply(lambda xs: \"|\".join(xs))\n",
    "clin_mask = annot[\"CLNSIG_labels_list\"].apply(lambda xs: len(xs) > 0)\n",
    "annot_clin = annot[clin_mask].copy()\n",
    "\n",
    "print(f\"[info] Variants total: {annot.shape[0]:,}  |  with valid clinical labels: {annot_clin.shape[0]:,}\")\n",
    "\n",
    "# Build variant_id for outputs\n",
    "annot_clin[\"variant_id\"] = (annot_clin[\"#CHROM\"].astype(str) + \":\" +\n",
    "                            annot_clin[\"POS\"].astype(str) + \":\" +\n",
    "                            annot_clin[\"REF\"].astype(str) + \":\" +\n",
    "                            annot_clin[\"ALT\"].astype(str))\n",
    "\n",
    "# ----- Robust sample column detection (yours) -----\n",
    "sample_cols_clean = [c for c in annot_clin.columns if c in pheno.index]\n",
    "if not sample_cols_clean:\n",
    "    raise RuntimeError(\"No genotype sample columns matched phenotype IDs.\")\n",
    "overlap = sample_cols_clean\n",
    "print(f\"[info] Overlapping samples: {len(overlap)}\")\n",
    "\n",
    "# ------------- Trait binarization helpers for Fisher -------------------\n",
    "def binarize_numeric(y: pd.Series, trait_name: str) -> tuple[pd.Series, str]:\n",
    "    \"\"\"Return binary series (0/1) and a label describing the rule.\"\"\"\n",
    "    y_num = pd.to_numeric(y, errors=\"coerce\")\n",
    "    y_bin = None\n",
    "    rule = \"\"\n",
    "    if trait_name in CUSTOM_CUTOFFS:\n",
    "        cut = float(CUSTOM_CUTOFFS[trait_name])\n",
    "        y_bin = (y_num >= cut).astype(float)\n",
    "        rule = f\">= {cut}\"\n",
    "    else:\n",
    "        med = float(np.nanmedian(y_num.values))\n",
    "        y_bin = (y_num >= med).astype(float)\n",
    "        rule = f\">= median ({med:.3g})\"\n",
    "    return y_bin, rule\n",
    "\n",
    "def to_int_no_nan(x):\n",
    "    \"\"\"Coerce genotype dosage to integers {0,1,2} with NaN preserved.\"\"\"\n",
    "    return pd.Series(x).where(~np.isnan(x), np.nan).astype(\"float\")\n",
    "\n",
    "# ----------------- Fisher exact core -----------------\n",
    "def fisher_2x2(a, b, c, d):\n",
    "    \"\"\"Return OR, pval for table [[a,b],[c,d]] using two-sided Fisher.\"\"\"\n",
    "    table = np.array([[a, b],\n",
    "                      [c, d]], dtype=int)\n",
    "    # SciPy returns (oddsratio, pvalue); oddsratio may be inf if zeros present\n",
    "    or_val, p = fisher_exact(table, alternative='two-sided')\n",
    "    return float(or_val), float(p)\n",
    "\n",
    "def run_fisher_for_trait(trait_name: str) -> pd.DataFrame:\n",
    "    ph_col = trait_col_map[trait_name]\n",
    "    y_raw = pheno.loc[overlap, ph_col]\n",
    "\n",
    "    y_num_or_bin, is_bin = decide_trait(y_raw)\n",
    "    if is_bin:\n",
    "        y_bin = y_num_or_bin.astype(float)\n",
    "        trait_kind = \"BINARY\"\n",
    "        rule = \"as-is\"\n",
    "    else:\n",
    "        y_bin, rule = binarize_numeric(y_raw, trait_name)\n",
    "        trait_kind = \"NUMERIC→BINARY\"\n",
    "\n",
    "    print(f\"[trait {trait_name}] Using Fisher (trait {trait_kind}; rule: {rule}). \"\n",
    "          f\"N non-missing={int(y_bin.notna().sum())}\")\n",
    "\n",
    "    results = []\n",
    "    # Iterate over variants\n",
    "    for _, row in annot_clin.iterrows():\n",
    "        # genotype dosage per overlapping sample\n",
    "        g = np.array([gt_to_dosage(row[s]) for s in overlap], dtype=float)\n",
    "        df = pd.DataFrame({\"y\": y_bin.values, \"g\": g}, index=overlap).dropna()\n",
    "\n",
    "        # Need enough data to form both groups\n",
    "        if df.shape[0] < 6:\n",
    "            continue\n",
    "\n",
    "        # integerize genotype dosage (0/1/2)\n",
    "        g_int = to_int_no_nan(df[\"g\"].values)\n",
    "\n",
    "        # --- Dominant model: carriers (1/2) vs non-carriers (0) ---\n",
    "        dom = (g_int >= 1).astype(int)\n",
    "        # 2x2 table: rows = genotype group (non-carrier, carrier), cols = trait (0,1)\n",
    "        a = int(((dom == 0) & (df[\"y\"] == 0)).sum())\n",
    "        b = int(((dom == 0) & (df[\"y\"] == 1)).sum())\n",
    "        c = int(((dom == 1) & (df[\"y\"] == 0)).sum())\n",
    "        d = int(((dom == 1) & (df[\"y\"] == 1)).sum())\n",
    "\n",
    "        # Only run if both rows and both cols present\n",
    "        if (a + b) > 0 and (c + d) > 0 and (a + c) > 0 and (b + d) > 0:\n",
    "            or_dom, p_dom = fisher_2x2(a, b, c, d)\n",
    "            maf = maf_from_dosage(df[\"g\"].values)\n",
    "            results.append({\n",
    "                \"variant_id\": row[\"variant_id\"],\n",
    "                \"CHROM\": row[\"#CHROM\"],\n",
    "                \"POS\": int(row[\"POS\"]),\n",
    "                \"REF\": row[\"REF\"],\n",
    "                \"ALT\": row[\"ALT\"],\n",
    "                \"MAF\": maf,\n",
    "                \"N\": int(df.shape[0]),\n",
    "                \"model\": \"dominant\",\n",
    "                \"trait_binarization\": rule,\n",
    "                \"odds_ratio\": or_dom,\n",
    "                \"pval\": p_dom,\n",
    "                \"CLNSIG_labels\": row[\"CLNSIG_labels_str\"]\n",
    "            })\n",
    "\n",
    "        # --- Recessive model: hom-alt (2) vs others (0/1) ---\n",
    "        if (g_int == 2).any():\n",
    "            rec = (g_int == 2).astype(int)\n",
    "            a = int(((rec == 0) & (df[\"y\"] == 0)).sum())\n",
    "            b = int(((rec == 0) & (df[\"y\"] == 1)).sum())\n",
    "            c = int(((rec == 1) & (df[\"y\"] == 0)).sum())\n",
    "            d = int(((rec == 1) & (df[\"y\"] == 1)).sum())\n",
    "            # Require both genotypes to be represented and both outcomes to occur at least once\n",
    "            if (a + b) > 0 and (c + d) > 0 and (a + c) > 0 and (b + d) > 0:\n",
    "                or_rec, p_rec = fisher_2x2(a, b, c, d)\n",
    "                maf = maf_from_dosage(df[\"g\"].values)\n",
    "                results.append({\n",
    "                    \"variant_id\": row[\"variant_id\"],\n",
    "                    \"CHROM\": row[\"#CHROM\"],\n",
    "                    \"POS\": int(row[\"POS\"]),\n",
    "                    \"REF\": row[\"REF\"],\n",
    "                    \"ALT\": row[\"ALT\"],\n",
    "                    \"MAF\": maf,\n",
    "                    \"N\": int(df.shape[0]),\n",
    "                    \"model\": \"recessive\",\n",
    "                    \"trait_binarization\": rule,\n",
    "                    \"odds_ratio\": or_rec,\n",
    "                    \"pval\": p_rec,\n",
    "                    \"CLNSIG_labels\": row[\"CLNSIG_labels_str\"]\n",
    "                })\n",
    "\n",
    "    if not results:\n",
    "        print(f\"[trait {trait_name}] No analyzable variants produced results.\")\n",
    "        return pd.DataFrame(columns=[\n",
    "            \"variant_id\",\"CHROM\",\"POS\",\"REF\",\"ALT\",\"MAF\",\"N\",\"model\",\n",
    "            \"trait_binarization\",\"odds_ratio\",\"pval\",\"qval_fdr\",\"CLNSIG_labels\"\n",
    "        ])\n",
    "\n",
    "    out = pd.DataFrame(results)\n",
    "    # FDR by model within trait\n",
    "    out[\"qval_fdr\"] = np.nan\n",
    "    for m in [\"dominant\",\"recessive\"]:\n",
    "        mask = out[\"model\"] == m\n",
    "        if mask.any():\n",
    "            out.loc[mask, \"qval_fdr\"] = multipletests(out.loc[mask, \"pval\"].values,\n",
    "                                                      method=\"fdr_bh\")[1]\n",
    "    out = out.sort_values([\"model\",\"pval\"], ascending=[True, True]).reset_index(drop=True)\n",
    "    return out\n",
    "\n",
    "# ----------------- Run and save -----------------\n",
    "all_outputs = {}\n",
    "for trait in TRAITS:\n",
    "    try:\n",
    "        res = run_fisher_for_trait(trait)\n",
    "        out_path = f\"{OUT_PREFIX}_{norm(trait)}.csv\"\n",
    "        res.to_csv(out_path, index=False)\n",
    "        all_outputs[trait] = out_path\n",
    "        print(f\"[save] {trait} -> {out_path} (variants tested: {res.shape[0]})\")\n",
    "    except KeyError as e:\n",
    "        print(f\"[warn] Skipping '{trait}': {e}\")\n",
    "\n",
    "print(\"\\nDone. Top rows from each output:\")\n",
    "for trait, path in all_outputs.items():\n",
    "    try:\n",
    "        print(f\"\\n### {trait} ###\")\n",
    "        print(pd.read_csv(path, nrows=5).to_string(index=False))\n",
    "    except Exception as e:\n",
    "        print(f\"[warn] Could not preview {trait}: {e}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8600105d-179f-4770-9a70-5655fed63141",
   "metadata": {},
   "outputs": [],
   "source": [
    "from statsmodels.stats.contingency_tables import Table2x2\n",
    "\n",
    "# after you add a,b,c,d columns (see patch below)\n",
    "df = pd.read_csv(\"assoc_exact_fisher_bmi.csv\")\n",
    "top = df.sort_values(\"pval\").head(15).copy()\n",
    "\n",
    "cis = []\n",
    "for _, r in top.iterrows():\n",
    "    tab = Table2x2([[r[\"a\"], r[\"b\"]],[r[\"c\"], r[\"d\"]]])\n",
    "    cis.append(tab.oddsratio_confint())  # default alpha=0.05\n",
    "top[\"ci_low\"] = [c[0] for c in cis]\n",
    "top[\"ci_high\"] = [c[1] for c in cis]\n",
    "\n",
    "y = np.arange(len(top))\n",
    "plt.figure(figsize=(6,6))\n",
    "plt.hlines(y, np.log2(top[\"ci_low\"]), np.log2(top[\"ci_high\"]))\n",
    "plt.plot(np.log2(top[\"odds_ratio\"]), y, \"o\")\n",
    "plt.axvline(0, linestyle=\"--\")  # log2(OR)=0 → OR=1\n",
    "plt.yticks(y, top[\"variant_id\"])\n",
    "plt.xlabel(\"log2(OR) [95% CI]\")\n",
    "plt.title(\"Forest plot — top Fisher hits (BMI)\")\n",
    "plt.tight_layout(); plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f0d79331-98cc-4daf-a173-88dbcc0a905b",
   "metadata": {},
   "outputs": [],
   "source": [
    "variants = []  # build a union of interesting variants across trait files\n",
    "# read multiple CSVs, merge on variant_id, pivot to trait_model columns\n",
    "\n",
    "# Example with one file for brevity:\n",
    "df = pd.read_csv(\"assoc_exact_fisher_bmi.csv\")\n",
    "df[\"col\"] = df[\"model\"] + \" (BMI)\"\n",
    "keep = df[df[\"qval_fdr\"]<=0.1].copy()\n",
    "mat = keep.pivot_table(index=\"variant_id\", columns=\"col\", values=\"odds_ratio\")\n",
    "mat = np.log2(mat)\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "plt.figure(figsize=(6, max(3, 0.22*mat.shape[0])))\n",
    "plt.imshow(mat, aspect=\"auto\", interpolation=\"nearest\")\n",
    "plt.colorbar(label=\"log2(OR)\")\n",
    "plt.yticks(range(mat.shape[0]), mat.index)\n",
    "plt.xticks(range(mat.shape[1]), mat.columns, rotation=45, ha=\"right\")\n",
    "plt.title(\"Effect heatmap (significant cells)\")\n",
    "plt.tight_layout(); plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1298a761-fb63-47b8-af6c-179b6dc0947e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[plot] Overlapping samples for quicklook: 119\n",
      "[fig] saved -> assoc_clinical_asthma_trait_quicklook.png\n",
      "[fig] saved -> assoc_clinical_asthma_manhattan.png\n",
      "[fig] saved -> assoc_clinical_asthma_manhattan_by_clinical.png\n",
      "[fig] saved -> assoc_clinical_asthma_qq.png\n",
      "[fig] saved -> assoc_clinical_asthma_volcano.png\n",
      "[fig] saved -> assoc_clinical_asthma_maf_vs_mlog10p.png\n",
      "[fig] saved -> assoc_clinical_max_pets_trait_quicklook.png\n",
      "[fig] saved -> assoc_clinical_max_pets_manhattan.png\n",
      "[fig] saved -> assoc_clinical_max_pets_manhattan_by_clinical.png\n",
      "[fig] saved -> assoc_clinical_max_pets_qq.png\n",
      "[fig] saved -> assoc_clinical_max_pets_volcano.png\n",
      "[fig] saved -> assoc_clinical_max_pets_maf_vs_mlog10p.png\n",
      "[fig] saved -> assoc_clinical_num_smokers_trait_quicklook.png\n",
      "[fig] saved -> assoc_clinical_num_smokers_manhattan.png\n",
      "[fig] saved -> assoc_clinical_num_smokers_manhattan_by_clinical.png\n",
      "[fig] saved -> assoc_clinical_num_smokers_qq.png\n",
      "[fig] saved -> assoc_clinical_num_smokers_volcano.png\n",
      "[fig] saved -> assoc_clinical_num_smokers_maf_vs_mlog10p.png\n",
      "[table] Significant associations (FDR<0.05) -> significant_hits_FDR_lt_0p05.csv\n",
      "   trait         variant_id CHROM      POS REF ALT  MAF   N     beta       se         pval  qval_fdr CLNSIG_bucket        CLNSIG_labels  bonf_sig\n",
      "max_pets  chr5:41158761:G:A  chr5 41158761   G   A  0.0 119 0.292437 0.049999 4.513289e-08  0.000007  Benign (any) Benign|Benign|Benign      True\n",
      "max_pets chr17:78141117:T:G chr17 78141117   T   G  0.0 119 0.292437 0.049999 4.513289e-08  0.000007  Benign (any) Benign|Benign|Benign      True\n",
      "max_pets chr17:78460270:T:C chr17 78460270   T   C  0.0 119 0.292437 0.049999 4.513289e-08  0.000007  Benign (any)               Benign      True\n",
      "max_pets chr17:78527232:T:A chr17 78527232   T   A  0.0 119 0.292437 0.049999 4.513289e-08  0.000007  Benign (any)               Benign      True\n",
      "max_pets chr17:78527233:T:A chr17 78527233   T   A  0.0 119 0.292437 0.049999 4.513289e-08  0.000007  Benign (any)               Benign      True\n",
      "max_pets chr17:78527235:T:A chr17 78527235   T   A  0.0 119 0.292437 0.049999 4.513289e-08  0.000007  Benign (any)               Benign      True\n",
      "max_pets chr17:78562077:T:C chr17 78562077   T   C  0.0 119 0.292437 0.049999 4.513289e-08  0.000007  Benign (any)               Benign      True\n",
      "max_pets    chr5:254327:C:T  chr5   254327   C   T  0.0 119 0.292437 0.049999 4.513289e-08  0.000007  Benign (any) Benign|Benign|Benign      True\n",
      "max_pets chr17:74285420:A:G chr17 74285420   A   G  0.0 119 0.292437 0.049999 4.513289e-08  0.000007  Benign (any)               Benign      True\n",
      "max_pets  chr5:10261430:C:T  chr5 10261430   C   T  0.0 119 0.292437 0.049999 4.513289e-08  0.000007  Benign (any)        Benign|Benign      True\n",
      "max_pets  chr5:13923012:A:T  chr5 13923012   A   T  0.0 119 0.292437 0.049999 4.513289e-08  0.000007  Benign (any)               Benign      True\n",
      "max_pets  chr12:6121122:A:G chr12  6121122   A   G  0.0 119 0.292437 0.049999 4.513289e-08  0.000007  Benign (any)               Benign      True\n",
      "max_pets  chr5:14709926:T:G  chr5 14709926   T   G  0.0 119 0.292437 0.049999 4.513289e-08  0.000007  Benign (any)        Benign|Benign      True\n",
      "max_pets  chr5:14711550:A:G  chr5 14711550   A   G  0.0 119 0.292437 0.049999 4.513289e-08  0.000007  Benign (any)               Benign      True\n",
      "max_pets  chr5:59193558:G:A  chr5 59193558   G   A  0.0 119 0.292437 0.049999 4.513289e-08  0.000007  Benign (any)        Benign|Benign      True\n",
      "max_pets  chr5:80678827:T:C  chr5 80678827   T   C  0.0 119 0.292437 0.049999 4.513289e-08  0.000007  Benign (any)               Benign      True\n",
      "max_pets  chr5:90647953:T:G  chr5 90647953   T   G  0.0 119 0.292437 0.049999 4.513289e-08  0.000007  Benign (any)               Benign      True\n",
      "max_pets  chr5:10265955:G:T  chr5 10265955   G   T  0.0 119 0.292437 0.049999 4.513289e-08  0.000007  Benign (any)               Benign      True\n",
      "max_pets  chr12:5967775:A:G chr12  5967775   A   G  0.0 119 0.292437 0.049999 4.513289e-08  0.000007  Benign (any)               Benign      True\n",
      "max_pets chr17:63971875:A:G chr17 63971875   A   G  0.0 119 0.292437 0.049999 4.513289e-08  0.000007  Benign (any)        Benign|Benign      True\n"
     ]
    }
   ],
   "source": [
    "# ===== Plotting + Significant-Hits Table (with clinical-colored Manhattan) =====\n",
    "import os, re\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy.stats import chi2\n",
    "\n",
    "# ---- CONFIG ----\n",
    "PHENO_PATH   = \"ImmPort_URECA_Cross_Black.csv\"\n",
    "PHENO_ID_COL = \"id\"\n",
    "TRAITS       = [\"asthma\", \"max_pets\", \"num_smokers\"]\n",
    "\n",
    "FDR_ALPHA    = 0.05\n",
    "P_MIN        = 1e-300\n",
    "\n",
    "# Preferred prefixes to search (first that exists per-trait will be used)\n",
    "PREFERRED_PREFIXES = [\"assoc_clinical\", \"assoc_chr5\", \"assoc_chr5_fast\"]\n",
    "\n",
    "SIG_TABLE_OUT = \"significant_hits_FDR_lt_0p05.csv\"   # combined table\n",
    "\n",
    "# Use any annotated file to infer sample columns for trait quicklook\n",
    "ANNOTATED_DIR = \"per_chrom_outputs_new\"\n",
    "\n",
    "# ---- helpers ----\n",
    "def _norm(s: str) -> str:\n",
    "    return re.sub(r\"\\W+\", \"_\", str(s).strip().lower())\n",
    "\n",
    "def _resolve_trait_cols(pheno_df: pd.DataFrame, names: list[str]) -> dict:\n",
    "    colmap = { _norm(c): c for c in pheno_df.columns }\n",
    "    resolved = {}\n",
    "    for t in names:\n",
    "        key = _norm(t)\n",
    "        if key not in colmap:\n",
    "            raise KeyError(f\"Trait '{t}' not found (case/space-insensitive). \"\n",
    "                           f\"Available (normalized) include: {list(colmap.keys())[:15]} ...\")\n",
    "        resolved[t] = colmap[key]\n",
    "    return resolved\n",
    "\n",
    "def _ensure_cols(df: pd.DataFrame) -> pd.DataFrame:\n",
    "    if df.empty:\n",
    "        return df\n",
    "    df = df.copy()\n",
    "    df[\"mlog10p\"] = -np.log10(np.clip(df[\"pval\"].values.astype(float), P_MIN, 1.0))\n",
    "    df[\"hit\"] = (df[\"qval_fdr\"].values.astype(float) <= FDR_ALPHA)\n",
    "    df[\"POS\"] = pd.to_numeric(df[\"POS\"], errors=\"coerce\")\n",
    "    df[\"CHROM\"] = df[\"CHROM\"].astype(str)\n",
    "    # keep CLNSIG fields if present\n",
    "    if \"CLNSIG_labels\" not in df.columns:\n",
    "        df[\"CLNSIG_labels\"] = \"\"\n",
    "    return df\n",
    "\n",
    "def _savefig(path: str):\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(path, dpi=180, bbox_inches=\"tight\")\n",
    "    plt.close()\n",
    "    print(f\"[fig] saved -> {path}\")\n",
    "\n",
    "def decide_trait(y_raw: pd.Series):\n",
    "    y_num = pd.to_numeric(y_raw, errors=\"coerce\")\n",
    "    nuniq = y_num.dropna().nunique()\n",
    "    sl = y_raw.astype(str).str.strip().str.lower()\n",
    "    bin_map = {\n",
    "        \"1\": 1, \"0\": 0, \"yes\": 1, \"no\": 0, \"true\": 1, \"false\": 0, \"y\": 1, \"n\": 0,\n",
    "        \"asthma\": 1, \"no asthma\": 0, \"present\": 1, \"absent\": 0, \"case\": 1, \"control\": 0\n",
    "    }\n",
    "    y_bin = sl.map(bin_map).astype(float); coverage = y_bin.notna().mean()\n",
    "    if nuniq > 2: return y_num.astype(float), False\n",
    "    if coverage >= 0.9: return y_bin, True\n",
    "    if set(pd.unique(y_num.dropna())).issubset({0,1}): return y_num.astype(float), True\n",
    "    return y_num.astype(float), False\n",
    "\n",
    "def plot_trait_quicklook(trait_name, series, out_prefix):\n",
    "    y, is_binary = decide_trait(series)\n",
    "    fig_path = f\"{out_prefix}_{trait_name}_trait_quicklook.png\"\n",
    "    plt.figure(figsize=(6,4))\n",
    "    if is_binary:\n",
    "        vals = y.dropna().astype(int)\n",
    "        plt.bar([\"0\",\"1\"], [(vals==0).sum(), (vals==1).sum()])\n",
    "        plt.ylabel(\"Count\"); plt.title(f\"{trait_name} (binary) counts\")\n",
    "    else:\n",
    "        yn = y.dropna()\n",
    "        plt.hist(yn.values, bins=30)\n",
    "        plt.xlabel(trait_name); plt.ylabel(\"Frequency\"); plt.title(f\"{trait_name} distribution\")\n",
    "    _savefig(fig_path)\n",
    "\n",
    "def _chrom_sort_key(c):\n",
    "    c = str(c).replace(\"chr\",\"\")\n",
    "    try:\n",
    "        return (0, int(c))\n",
    "    except:\n",
    "        order = {\"X\": 23, \"Y\": 24, \"MT\": 25, \"M\":25}\n",
    "        return (1, order.get(c, 99))\n",
    "\n",
    "def _build_cum_pos(df):\n",
    "    # cumulative positions and chromosome tick marks\n",
    "    chroms = sorted(df[\"CHROM\"].dropna().unique(), key=_chrom_sort_key)\n",
    "    offsets, current = {}, 0\n",
    "    tick_positions, tick_labels = [], []\n",
    "    for c in chroms:\n",
    "        subset = df.loc[df[\"CHROM\"]==c, \"POS\"].dropna()\n",
    "        if subset.empty:\n",
    "            offsets[c] = current\n",
    "            continue\n",
    "        minpos, maxpos = subset.min(), subset.max()\n",
    "        offsets[c] = current\n",
    "        tick_positions.append(current + (minpos + maxpos)/2.0)\n",
    "        tick_labels.append(str(c))\n",
    "        current += maxpos + 1e6  # gap\n",
    "    df[\"cum_pos\"] = df.apply(lambda r: (r[\"POS\"] if pd.notna(r[\"POS\"]) else 0) + offsets.get(str(r[\"CHROM\"]), 0), axis=1)\n",
    "    return df, tick_positions, tick_labels\n",
    "\n",
    "# --- Clinical bucketing (from CLNSIG_labels string) ---\n",
    "def labels_to_bucket(labels_str: str) -> str:\n",
    "    \"\"\"\n",
    "    Bucket rules:\n",
    "    - Pathogenic (any): if *any* of these appear: Pathogenic, Likely pathogenic,\n",
    "      Pathogenic/Likely pathogenic, Pathogenic, low penetrance, Likely pathogenic, low penetrance\n",
    "    - Benign (any): Benign, Likely benign, Benign/Likely benign\n",
    "    - Uncertain: Uncertain significance or Uncertain risk allele\n",
    "    - Conflicting: Conflicting interpretations of pathogenicity or Conflicting data from submitters\n",
    "    - Other: anything else / empty\n",
    "    \"\"\"\n",
    "    if not isinstance(labels_str, str) or labels_str.strip()==\"\":\n",
    "        return \"Other\"\n",
    "    labs = {t.strip() for t in re.split(r\"[|,;/]+\", labels_str) if t.strip()}\n",
    "    patho = {\"Pathogenic\",\"Likely pathogenic\",\"Pathogenic/Likely pathogenic\",\n",
    "             \"Pathogenic, low penetrance\",\"Likely pathogenic, low penetrance\"}\n",
    "    benign = {\"Benign\",\"Likely benign\",\"Benign/Likely benign\"}\n",
    "    uncertain = {\"Uncertain significance\",\"Uncertain risk allele\"}\n",
    "    conflicting = {\"Conflicting interpretations of pathogenicity\",\"Conflicting data from submitters\"}\n",
    "    if labs & patho: return \"Pathogenic (any)\"\n",
    "    if labs & benign: return \"Benign (any)\"\n",
    "    if labs & uncertain: return \"Uncertain\"\n",
    "    if labs & conflicting: return \"Conflicting\"\n",
    "    return \"Other\"\n",
    "\n",
    "def manhattan(df, trait_name, out_prefix, fdr_alpha=FDR_ALPHA):\n",
    "    df = _ensure_cols(df)\n",
    "    if df.empty or df[\"POS\"].isna().all():\n",
    "        print(f\"[{trait_name}] No results to plot (Manhattan).\")\n",
    "        return\n",
    "    df, ticks, labels = _build_cum_pos(df)\n",
    "\n",
    "    # thresholds\n",
    "    m = max(len(df), 1)\n",
    "    bonf_p = 0.05 / m\n",
    "    bonf_y = -np.log10(max(bonf_p, P_MIN))\n",
    "    hits = df[\"hit\"].values\n",
    "    has_fdr_hits = hits.any()\n",
    "    if has_fdr_hits:\n",
    "        p_fdr_cut = float(df.loc[hits, \"pval\"].max())\n",
    "        fdr_y = -np.log10(max(p_fdr_cut, P_MIN))\n",
    "\n",
    "    # plot (default coloring)\n",
    "    fig_path = f\"{out_prefix}_{trait_name}_manhattan.png\"\n",
    "    plt.figure(figsize=(11,4))\n",
    "    for i, c in enumerate(sorted(df[\"CHROM\"].unique(), key=_chrom_sort_key)):\n",
    "        mask = (df[\"CHROM\"] == c)\n",
    "        plt.scatter(df.loc[mask & ~hits, \"cum_pos\"], df.loc[mask & ~hits, \"mlog10p\"], s=6, alpha=0.6)\n",
    "        if has_fdr_hits:\n",
    "            plt.scatter(df.loc[mask & hits, \"cum_pos\"], df.loc[mask & hits, \"mlog10p\"], s=10, alpha=0.9)\n",
    "    plt.axhline(bonf_y, linestyle=\"-\", linewidth=1.2, label=f\"Bonferroni 0.05 (p<{bonf_p:.2e})\")\n",
    "    if has_fdr_hits:\n",
    "        plt.axhline(fdr_y, linestyle=\"--\", linewidth=1.2, label=f\"FDR<{fdr_alpha} (p≈{p_fdr_cut:.1e})\")\n",
    "    plt.xticks(ticks, labels)\n",
    "    plt.xlabel(\"Chromosome\"); plt.ylabel(\"-log10(p)\")\n",
    "    plt.title(f\"Manhattan – {trait_name}\")\n",
    "    plt.legend(frameon=False, loc=\"upper right\")\n",
    "    _savefig(fig_path)\n",
    "\n",
    "def manhattan_by_clinical(df, trait_name, out_prefix, fdr_alpha=FDR_ALPHA):\n",
    "    \"\"\"\n",
    "    Manhattan colored by clinical bucket derived from CLNSIG_labels.\n",
    "    \"\"\"\n",
    "    df = _ensure_cols(df)\n",
    "    if df.empty or df[\"POS\"].isna().all():\n",
    "        print(f\"[{trait_name}] No results to plot (Clin-colored Manhattan).\")\n",
    "        return\n",
    "\n",
    "    # Add bucket\n",
    "    df[\"CLNSIG_bucket\"] = df[\"CLNSIG_labels\"].apply(labels_to_bucket)\n",
    "\n",
    "    # Build cumulative positions\n",
    "    df, ticks, labels = _build_cum_pos(df)\n",
    "\n",
    "    # thresholds\n",
    "    m = max(len(df), 1)\n",
    "    bonf_p = 0.05 / m\n",
    "    bonf_y = -np.log10(max(bonf_p, P_MIN))\n",
    "\n",
    "    # colors per bucket (feel free to tweak)\n",
    "    bucket_order = [\"Pathogenic (any)\", \"Benign (any)\", \"Uncertain\", \"Conflicting\", \"Other\"]\n",
    "    bucket_colors = {\n",
    "        \"Pathogenic (any)\": \"#d62728\",\n",
    "        \"Benign (any)\":     \"#2ca02c\",\n",
    "        \"Uncertain\":        \"#1f77b4\",\n",
    "        \"Conflicting\":      \"#ff7f0e\",\n",
    "        \"Other\":            \"#7f7f7f\",\n",
    "    }\n",
    "\n",
    "    fig_path = f\"{out_prefix}_{trait_name}_manhattan_by_clinical.png\"\n",
    "    plt.figure(figsize=(11,4))\n",
    "    for b in bucket_order:\n",
    "        mask = (df[\"CLNSIG_bucket\"] == b)\n",
    "        if mask.any():\n",
    "            plt.scatter(df.loc[mask, \"cum_pos\"], df.loc[mask, \"mlog10p\"],\n",
    "                        s=8, alpha=0.8, label=f\"{b} (n={mask.sum():,})\",\n",
    "                        c=bucket_colors.get(b, \"#7f7f7f\"))\n",
    "\n",
    "    plt.axhline(bonf_y, linestyle=\"-\", linewidth=1.2, color=\"#333333\",\n",
    "                label=f\"Bonferroni 0.05 (p<{bonf_p:.2e})\")\n",
    "    plt.xticks(ticks, labels)\n",
    "    plt.xlabel(\"Chromosome\"); plt.ylabel(\"-log10(p)\")\n",
    "    plt.title(f\"Manhattan by clinical category – {trait_name}\")\n",
    "    plt.legend(frameon=False, loc=\"upper right\", ncol=2)\n",
    "    _savefig(fig_path)\n",
    "\n",
    "def qq_plot(df, trait_name, out_prefix):\n",
    "    df = _ensure_cols(df)\n",
    "    if df.empty:\n",
    "        print(f\"[{trait_name}] No results to plot (QQ).\")\n",
    "        return\n",
    "    p = np.clip(np.sort(df[\"pval\"].values.astype(float)), P_MIN, 1.0)\n",
    "    n = len(p); exp = (np.arange(1, n+1) - 0.5) / n\n",
    "    o = -np.log10(p); e = -np.log10(exp)\n",
    "    chisq = chi2.isf(p, df=1)\n",
    "    lam_gc = np.median(chisq) / chi2.ppf(0.5, 1)\n",
    "    fig_path = f\"{out_prefix}_{trait_name}_qq.png\"\n",
    "    plt.figure(figsize=(5,5))\n",
    "    plt.scatter(e, o, s=6, alpha=0.6)\n",
    "    lim = max(e.max(), o.max()) * 1.02\n",
    "    plt.plot([0, lim], [0, lim])\n",
    "    plt.xlabel(\"Expected -log10(p)\"); plt.ylabel(\"Observed -log10(p)\")\n",
    "    plt.title(f\"QQ plot – {trait_name} (λGC={lam_gc:.3f})\")\n",
    "    _savefig(fig_path)\n",
    "\n",
    "def volcano(df, trait_name, out_prefix):\n",
    "    df = _ensure_cols(df)\n",
    "    if df.empty:\n",
    "        print(f\"[{trait_name}] No results to plot (volcano).\")\n",
    "        return\n",
    "    x, y, hit = df[\"beta\"].values.astype(float), df[\"mlog10p\"].values, df[\"hit\"].values\n",
    "    fig_path = f\"{out_prefix}_{trait_name}_volcano.png\"\n",
    "    plt.figure(figsize=(6,5))\n",
    "    plt.scatter(x[~hit], y[~hit], s=8, alpha=0.6)\n",
    "    if hit.any(): plt.scatter(x[hit], y[hit], s=12, alpha=0.9)\n",
    "    if len(df) > 0: plt.axhline(-np.log10(0.05/len(df)), linestyle=\"--\", linewidth=1)\n",
    "    plt.xlabel(\"Effect size (β per ALT allele)\"); plt.ylabel(\"-log10(p)\")\n",
    "    plt.title(f\"Volcano – {trait_name} (FDR<{FDR_ALPHA})\")\n",
    "    _savefig(fig_path)\n",
    "\n",
    "def maf_vs_significance(df, trait_name, out_prefix):\n",
    "    df = _ensure_cols(df)\n",
    "    if df.empty:\n",
    "        print(f\"[{trait_name}] No results to plot (MAF vs sig).\")\n",
    "        return\n",
    "    maf = pd.to_numeric(df[\"MAF\"], errors=\"coerce\").values\n",
    "    y   = df[\"mlog10p\"].values\n",
    "    hit = df[\"hit\"].values\n",
    "    fig_path = f\"{out_prefix}_{trait_name}_maf_vs_mlog10p.png\"\n",
    "    plt.figure(figsize=(6,5))\n",
    "    plt.scatter(maf[~hit], y[~hit], s=8, alpha=0.6)\n",
    "    if hit.any(): plt.scatter(maf[hit], y[hit], s=12, alpha=0.9)\n",
    "    plt.xlabel(\"Minor Allele Frequency (MAF)\"); plt.ylabel(\"-log10(p)\")\n",
    "    plt.title(f\"MAF vs significance – {trait_name}\")\n",
    "    _savefig(fig_path)\n",
    "\n",
    "# ---- Reload phenotype and infer sample columns for quicklook ----\n",
    "pheno = pd.read_csv(PHENO_PATH)\n",
    "trait_col_map = _resolve_trait_cols(pheno, TRAITS)\n",
    "pheno = pheno.set_index(pheno[PHENO_ID_COL].astype(str)).copy()\n",
    "pheno.index.name = \"sample_id\"\n",
    "\n",
    "# Discover sample columns from any annotated file (robust sample detection)\n",
    "overlap = list(pheno.index)\n",
    "if ANNOTATED_DIR and os.path.isdir(ANNOTATED_DIR):\n",
    "    any_file = None\n",
    "    for name in sorted(os.listdir(ANNOTATED_DIR)):\n",
    "        if name.endswith(\".annotated.csv\"):\n",
    "            any_file = os.path.join(ANNOTATED_DIR, name)\n",
    "            break\n",
    "    if any_file:\n",
    "        cols = list(pd.read_csv(any_file, nrows=0).columns)\n",
    "        overlap = [c for c in cols if c in pheno.index]\n",
    "print(f\"[plot] Overlapping samples for quicklook: {len(overlap)}\")\n",
    "\n",
    "# ---- For each trait, find the existing assoc file, plot, and collect significant hits ----\n",
    "signif_frames = []\n",
    "for trait in TRAITS:\n",
    "    assoc_path = None\n",
    "    out_prefix_used = None\n",
    "    for pref in PREFERRED_PREFIXES:\n",
    "        candidate = f\"{pref}_{trait}.csv\"\n",
    "        if os.path.exists(candidate):\n",
    "            assoc_path = candidate\n",
    "            out_prefix_used = pref\n",
    "            break\n",
    "    if assoc_path is None:\n",
    "        print(f(\"[skip] No association file found for {trait} with prefixes {PREFERRED_PREFIXES}.\"))\n",
    "        continue\n",
    "\n",
    "    res = pd.read_csv(assoc_path)\n",
    "    ph_col = trait_col_map[trait]\n",
    "    series = pheno.loc[overlap, ph_col]\n",
    "\n",
    "    # Plots\n",
    "    plot_trait_quicklook(trait, series, out_prefix=out_prefix_used)\n",
    "    manhattan(res, trait, out_prefix=out_prefix_used)\n",
    "    manhattan_by_clinical(res, trait, out_prefix=out_prefix_used)  # NEW\n",
    "    qq_plot(res, trait, out_prefix=out_prefix_used)\n",
    "    volcano(res, trait, out_prefix=out_prefix_used)\n",
    "    maf_vs_significance(res, trait, out_prefix=out_prefix_used)\n",
    "\n",
    "    # Collect significant hits (FDR < 0.05) + add clinical columns\n",
    "    res = _ensure_cols(res)\n",
    "    res[\"CLNSIG_bucket\"] = res[\"CLNSIG_labels\"].apply(labels_to_bucket) if \"CLNSIG_labels\" in res.columns else \"Other\"\n",
    "    res_sig = res.loc[(res[\"qval_fdr\"].astype(float) <= FDR_ALPHA)].copy()\n",
    "    if not res_sig.empty:\n",
    "        m = len(res)\n",
    "        res_sig[\"bonf_sig\"] = res_sig[\"pval\"].astype(float) < (0.05 / max(m,1))\n",
    "        res_sig[\"trait\"] = trait\n",
    "        cols = [\"trait\",\"variant_id\",\"CHROM\",\"POS\",\"REF\",\"ALT\",\"MAF\",\"N\",\"beta\",\"se\",\"pval\",\"qval_fdr\",\n",
    "                \"CLNSIG_labels\",\"CLNSIG_bucket\",\"bonf_sig\"]\n",
    "        keep = [c for c in cols if c in res_sig.columns]\n",
    "        signif_frames.append(res_sig[keep])\n",
    "\n",
    "# ---- Write combined significant table ----\n",
    "if signif_frames:\n",
    "    sig_table = pd.concat(signif_frames, ignore_index=True)\n",
    "    sig_table = sig_table.sort_values([\"trait\",\"pval\"]).reset_index(drop=True)\n",
    "    sig_table.to_csv(SIG_TABLE_OUT, index=False)\n",
    "    print(f\"[table] Significant associations (FDR<{FDR_ALPHA}) -> {SIG_TABLE_OUT}\")\n",
    "    display_cols = [\"trait\",\"variant_id\",\"CHROM\",\"POS\",\"REF\",\"ALT\",\"MAF\",\"N\",\"beta\",\"se\",\n",
    "                    \"pval\",\"qval_fdr\",\"CLNSIG_bucket\",\"CLNSIG_labels\",\"bonf_sig\"]\n",
    "    try:\n",
    "        print(sig_table.head(20)[display_cols].to_string(index=False))\n",
    "    except Exception:\n",
    "        print(sig_table.head(20).to_string(index=False))\n",
    "else:\n",
    "    print(f\"[table] No variants met FDR<{FDR_ALPHA} across the provided traits.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5c1bf408-9a93-405b-ac1b-853d302c8130",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[info] Resolved trait columns: {'asthma': 'asthma ', 'max_pets': 'max_pets', 'num_smokers': 'num_smokers'}\n",
      "[info] Overlapping samples: 119 / 123\n",
      "[trait asthma] Interpreted as BINARY (after coercion). N non-missing=119\n",
      "[save] asthma -> assoc_chr5_asthma.csv (variants tested: 7997)\n",
      "[trait max_pets] Interpreted as BINARY (after coercion). N non-missing=101\n",
      "[save] max_pets -> assoc_chr5_max_pets.csv (variants tested: 8144)\n",
      "[trait num_smokers] Interpreted as BINARY (after coercion). N non-missing=83\n",
      "[save] num_smokers -> assoc_chr5_num_smokers.csv (variants tested: 8139)\n",
      "\n",
      "Done. Top rows from each output:\n",
      "\n",
      "### asthma ###\n",
      "        variant_id CHROM       POS REF ALT      MAF   N      beta       se     pval  qval_fdr\n",
      " chr5:76403426:T:C  chr5  76403426   T   C 0.071429 119 -2.890372 0.775248 0.000193       1.0\n",
      "  chr5:7829700:G:C  chr5   7829700   G   C 0.264706 119  2.000459 0.593678 0.000753       1.0\n",
      "chr5:149627077:C:T  chr5 149627077   C   T 0.441176 119 -2.632392 0.802169 0.001032       1.0\n",
      "chr5:100391001:T:C  chr5 100391001   T   C 0.092920 113 -1.436135 0.454708 0.001587       1.0\n",
      " chr5:55935542:C:T  chr5  55935542   C   T 0.071429 119 -2.323204 0.737250 0.001626       1.0\n",
      "\n",
      "### max_pets ###\n",
      "        variant_id CHROM       POS REF ALT      MAF   N      beta       se     pval  qval_fdr\n",
      "chr5:103212017:C:T  chr5 103212017   C   T 0.069307 101  2.289232 0.697439 0.001029  0.982862\n",
      "chr5:132694975:C:G  chr5 132694975   C   G 0.356436 101  1.040566 0.329012 0.001563  0.982862\n",
      "chr5:140813058:C:T  chr5 140813058   C   T 0.336634 101 -1.293998 0.409386 0.001573  0.982862\n",
      "chr5:126868928:C:G  chr5 126868928   C   G 0.193069 101  1.220329 0.387852 0.001653  0.982862\n",
      "chr5:103205675:A:G  chr5 103205675   A   G 0.064356 101  2.134821 0.700609 0.002311  0.982862\n",
      "\n",
      "### num_smokers ###\n",
      "        variant_id CHROM       POS REF ALT      MAF  N      beta       se     pval  qval_fdr\n",
      "chr5:115830216:C:A  chr5 115830216   C   A 0.150602 83  2.045534 0.569745 0.000330  0.957029\n",
      "chr5:141103793:T:C  chr5 141103793   T   C 0.174699 83  1.628305 0.499074 0.001104  0.957029\n",
      "chr5:141103287:T:G  chr5 141103287   T   G 0.174699 83  1.628305 0.499074 0.001104  0.957029\n",
      " chr5:10973615:G:A  chr5  10973615   G   A 0.168675 83 -1.704559 0.552011 0.002016  0.957029\n",
      " chr5:17491160:T:C  chr5  17491160   T   C 0.156627 83 -1.566269 0.514879 0.002350  0.957029\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import statsmodels.api as sm\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "\n",
    "import warnings\n",
    "import numpy as np\n",
    "from statsmodels.tools.sm_exceptions import ConvergenceWarning, PerfectSeparationWarning\n",
    "\n",
    "# Silence spammy warnings from per-variant regressions\n",
    "warnings.filterwarnings(\"ignore\", category=ConvergenceWarning)\n",
    "warnings.filterwarnings(\"ignore\", category=PerfectSeparationWarning)\n",
    "warnings.filterwarnings(\"ignore\", category=RuntimeWarning)  # from numpy/statsmodels\n",
    "np.seterr(divide=\"ignore\", invalid=\"ignore\", over=\"ignore\")  # silence NumPy FP warnings\n",
    "\n",
    "\n",
    "# ========= User inputs (edit paths if needed) =========\n",
    "PHENO_PATH = \"ImmPort_URECA_Cross_Black.csv\"   # phenotype file\n",
    "PHENO_ID_COL = \"id\"                            # sample ID column in phenotype file\n",
    "# Trait names are given in a friendly form; we'll resolve to actual columns robustly\n",
    "TRAITS = [\"asthma\", \"max_pets\", \"num_smokers\"]\n",
    "\n",
    "GENO_CSV = \"chr5.vep.selected.csv\"      # processed chr22 genotype matrix (from earlier step)\n",
    "OUT_PREFIX = \"assoc_chr5\"                     # prefix for outputs\n",
    "# ======================================================\n",
    "\n",
    "# ---------- helpers ----------\n",
    "def norm(s: str) -> str:\n",
    "    \"\"\"Normalize a column name for matching: lowercase, strip, collapse non-word to underscore.\"\"\"\n",
    "    return re.sub(r\"\\W+\", \"_\", str(s).strip().lower())\n",
    "\n",
    "def resolve_trait_columns(pheno_df: pd.DataFrame, wanted_traits: list[str]) -> dict:\n",
    "    \"\"\"\n",
    "    Resolve wanted trait names to actual pheno columns, case/space-insensitively.\n",
    "    Returns dict: wanted_name -> actual_column_name (in pheno_df).\n",
    "    Raises if a trait cannot be resolved.\n",
    "    \"\"\"\n",
    "    colmap = {norm(c): c for c in pheno_df.columns}\n",
    "    resolved = {}\n",
    "    for t in wanted_traits:\n",
    "        key = norm(t)\n",
    "        if key in colmap:\n",
    "            resolved[t] = colmap[key]\n",
    "        else:\n",
    "            # try fuzzy: remove trailing/plural underscores, etc.\n",
    "            candidates = [c for k, c in colmap.items() if key == k or key in k]\n",
    "            if len(candidates) == 1:\n",
    "                resolved[t] = candidates[0]\n",
    "            else:\n",
    "                raise KeyError(f\"Could not find trait column for '{t}'. \"\n",
    "                               f\"Available columns (normalized) include: {list(colmap.keys())[:15]} ...\")\n",
    "    return resolved\n",
    "\n",
    "def coerce_binary_series(s: pd.Series) -> pd.Series:\n",
    "    \"\"\"\n",
    "    Convert common binary encodings (including 'Asthma'/'No Asthma') to 0/1 floats.\n",
    "    Unmapped values become NaN.\n",
    "    \"\"\"\n",
    "    # Lowercase/strip for matching\n",
    "    sl = s.astype(str).str.strip().str.lower()\n",
    "\n",
    "    # Common mappings\n",
    "    mapping = {\n",
    "        \"1\": 1, \"0\": 0,\n",
    "        \"yes\": 1, \"no\": 0,\n",
    "        \"true\": 1, \"false\": 0,\n",
    "        \"y\": 1, \"n\": 0,\n",
    "        \"asthma\": 1, \"no asthma\": 0,\n",
    "        \"present\": 1, \"absent\": 0,\n",
    "        \"case\": 1, \"control\": 0\n",
    "    }\n",
    "    out = sl.map(mapping)\n",
    "\n",
    "    # If still all NaN, try numeric coercion (handles 0/1 stored numerically)\n",
    "    if out.notna().sum() == 0:\n",
    "        out = pd.to_numeric(s, errors=\"coerce\")\n",
    "\n",
    "    return out.astype(float)\n",
    "\n",
    "def is_binary(series: pd.Series) -> bool:\n",
    "    vals = pd.unique(series.dropna())\n",
    "    return set(vals).issubset({0,1}) and len(vals) <= 2\n",
    "\n",
    "def gt_to_dosage(field: str) -> float:\n",
    "    \"\"\"\n",
    "    Convert per-sample VCF field to ALT-allele dosage (0/1/2), handling 0/1 and 0|1.\n",
    "    Returns NaN if missing or malformed.\n",
    "    \"\"\"\n",
    "    if not isinstance(field, str) or field == \"\" or field == \".\":\n",
    "        return np.nan\n",
    "    gt = field.split(\":\", 1)[0]\n",
    "    if gt in {\"./.\", \".|.\", \".\"}:\n",
    "        return np.nan\n",
    "    if \"/\" in gt:\n",
    "        a1, a2 = gt.split(\"/\")\n",
    "    elif \"|\" in gt:\n",
    "        a1, a2 = gt.split(\"|\")\n",
    "    else:\n",
    "        return np.nan\n",
    "    if a1 not in {\"0\",\"1\"} or a2 not in {\"0\",\"1\"}:\n",
    "        return np.nan\n",
    "    return float((a1 == \"1\") + (a2 == \"1\"))\n",
    "\n",
    "def maf_from_dosage(dosages: np.ndarray) -> float | None:\n",
    "    x = dosages[~np.isnan(dosages)]\n",
    "    n = x.size\n",
    "    if n == 0: return None\n",
    "    af = x.sum() / (2.0 * n)\n",
    "    return float(min(af, 1 - af))\n",
    "\n",
    "# ----------------- Load data -----------------\n",
    "pheno = pd.read_csv(PHENO_PATH)\n",
    "geno  = pd.read_csv(GENO_CSV, low_memory=False)\n",
    "\n",
    "# Resolve trait columns robustly\n",
    "trait_col_map = resolve_trait_columns(pheno, TRAITS)\n",
    "print(\"[info] Resolved trait columns:\", trait_col_map)\n",
    "\n",
    "# Normalize phenotype IDs to string index\n",
    "pheno_ids = pheno[PHENO_ID_COL].astype(str)\n",
    "pheno = pheno.set_index(pheno_ids).copy()\n",
    "pheno.index.name = \"sample_id\"\n",
    "\n",
    "# Genotype sample columns\n",
    "fixed_cols   = list(geno.columns[:9])   # ['#CHROM','POS','ID','REF','ALT','QUAL','FILTER','INFO','FORMAT']\n",
    "sample_cols  = list(geno.columns[9:])\n",
    "sample_cols_str = [str(s) for s in sample_cols]\n",
    "\n",
    "# Keep only overlapping samples (order = genotype order)\n",
    "overlap = [s for s in sample_cols_str if s in pheno.index]\n",
    "if len(overlap) == 0:\n",
    "    raise RuntimeError(\"No overlapping sample IDs between genotype (columns after FORMAT) and phenotype 'id'.\")\n",
    "print(f\"[info] Overlapping samples: {len(overlap)} / {len(sample_cols_str)}\")\n",
    "\n",
    "# Build variant_id (useful for outputs)\n",
    "geno[\"variant_id\"] = (geno[\"#CHROM\"].astype(str) + \":\" +\n",
    "                      geno[\"POS\"].astype(str) + \":\" +\n",
    "                      geno[\"REF\"].astype(str) + \":\" +\n",
    "                      geno[\"ALT\"].astype(str))\n",
    "\n",
    "# ----------------- Association core -----------------\n",
    "def run_assoc_for_trait(trait_name: str) -> pd.DataFrame:\n",
    "    # Get the actual phenotype column name\n",
    "    ph_col = trait_col_map[trait_name]\n",
    "    y_raw = pheno.loc[overlap, ph_col]\n",
    "\n",
    "    # Try binary coercion first (handles 'Asthma'/'No Asthma'); fallback to numeric\n",
    "    y_bin = coerce_binary_series(y_raw)\n",
    "    if is_binary(y_bin.dropna()):\n",
    "        y = y_bin\n",
    "        binary = True\n",
    "        print(f\"[trait {trait_name}] Interpreted as BINARY (after coercion). N non-missing={y.notna().sum()}\")\n",
    "    else:\n",
    "        y = pd.to_numeric(y_raw, errors=\"coerce\")\n",
    "        binary = False\n",
    "        print(f\"[trait {trait_name}] Interpreted as NUMERIC. N non-missing={y.notna().sum()}\")\n",
    "\n",
    "    results = []\n",
    "\n",
    "    for _, row in geno.iterrows():\n",
    "        # dosage vector for overlapping samples\n",
    "        g = np.array([gt_to_dosage(row[s]) for s in overlap], dtype=float)\n",
    "        df = pd.DataFrame({\"y\": y.values, \"g\": g}, index=overlap).dropna()\n",
    "        if df.shape[0] < 10:\n",
    "            continue\n",
    "\n",
    "        X = sm.add_constant(df[\"g\"].values, has_constant=\"add\")\n",
    "        try:\n",
    "            if binary:\n",
    "                mod = sm.Logit(df[\"y\"].values, X)\n",
    "                fit = mod.fit(disp=False)\n",
    "            else:\n",
    "                mod = sm.OLS(df[\"y\"].values, X)\n",
    "                fit = mod.fit()\n",
    "            beta = float(fit.params[1])\n",
    "            se   = float(fit.bse[1])\n",
    "            pval = float(fit.pvalues[1])\n",
    "        except Exception:\n",
    "            # perfect separation, singular matrix, etc.\n",
    "            continue\n",
    "\n",
    "        maf = maf_from_dosage(df[\"g\"].values)\n",
    "        results.append({\n",
    "            \"variant_id\": row[\"variant_id\"],\n",
    "            \"CHROM\": row[\"#CHROM\"],\n",
    "            \"POS\": int(row[\"POS\"]),\n",
    "            \"REF\": row[\"REF\"],\n",
    "            \"ALT\": row[\"ALT\"],\n",
    "            \"MAF\": maf,\n",
    "            \"N\": int(df.shape[0]),\n",
    "            \"beta\": beta,\n",
    "            \"se\": se,\n",
    "            \"pval\": pval\n",
    "        })\n",
    "\n",
    "    if not results:\n",
    "        print(f\"[trait {trait_name}] No analyzable variants produced results.\")\n",
    "        return pd.DataFrame(columns=[\"variant_id\",\"CHROM\",\"POS\",\"REF\",\"ALT\",\"MAF\",\"N\",\"beta\",\"se\",\"pval\",\"qval_fdr\"])\n",
    "\n",
    "    out = pd.DataFrame(results)\n",
    "    out[\"qval_fdr\"] = multipletests(out[\"pval\"].values, method=\"fdr_bh\")[1]\n",
    "    out = out.sort_values(\"pval\", ascending=True).reset_index(drop=True)\n",
    "    return out\n",
    "\n",
    "# ----------------- Run and save -----------------\n",
    "all_outputs = {}\n",
    "for trait in TRAITS:\n",
    "    res = run_assoc_for_trait(trait)\n",
    "    out_path = f\"{OUT_PREFIX}_{trait}.csv\"\n",
    "    res.to_csv(out_path, index=False)\n",
    "    all_outputs[trait] = out_path\n",
    "    print(f\"[save] {trait} -> {out_path} (variants tested: {res.shape[0]})\")\n",
    "\n",
    "print(\"\\nDone. Top rows from each output:\")\n",
    "for trait, path in all_outputs.items():\n",
    "    try:\n",
    "        print(f\"\\n### {trait} ###\")\n",
    "        print(pd.read_csv(path, nrows=5).to_string(index=False))\n",
    "    except Exception:\n",
    "        pass\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4981678f-4ab0-4c36-b2ea-12a3f2ec8de4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[plot] Overlapping samples for quicklook: 119\n",
      "[fig] saved -> assoc_chr5_asthma_trait_quicklook.png\n",
      "[fig] saved -> assoc_chr5_asthma_manhattan_chr22.png\n",
      "[fig] saved -> assoc_chr5_asthma_qq.png\n",
      "[fig] saved -> assoc_chr5_asthma_volcano.png\n",
      "[fig] saved -> assoc_chr5_asthma_maf_vs_mlog10p.png\n",
      "[fig] saved -> assoc_chr5_max_pets_trait_quicklook.png\n",
      "[fig] saved -> assoc_chr5_max_pets_manhattan_chr22.png\n",
      "[fig] saved -> assoc_chr5_max_pets_qq.png\n",
      "[fig] saved -> assoc_chr5_max_pets_volcano.png\n",
      "[fig] saved -> assoc_chr5_max_pets_maf_vs_mlog10p.png\n",
      "[fig] saved -> assoc_chr5_num_smokers_trait_quicklook.png\n",
      "[fig] saved -> assoc_chr5_num_smokers_manhattan_chr22.png\n",
      "[fig] saved -> assoc_chr5_num_smokers_qq.png\n",
      "[fig] saved -> assoc_chr5_num_smokers_volcano.png\n",
      "[fig] saved -> assoc_chr5_num_smokers_maf_vs_mlog10p.png\n",
      "[table] No variants met FDR<0.05 across the provided traits.\n"
     ]
    }
   ],
   "source": [
    "# ===== Plotting + Significant-Hits Table (auto-detects file prefix) =====\n",
    "import os, re\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy.stats import chi2\n",
    "\n",
    "# ---- CONFIG ----\n",
    "PHENO_PATH   = \"ImmPort_URECA_Cross_Black.csv\"\n",
    "PHENO_ID_COL = \"id\"\n",
    "TRAITS       = [\"asthma\", \"max_pets\", \"num_smokers\"]\n",
    "GENO_CSV     = \"chr5.vep.selected.csv\"\n",
    "FDR_ALPHA    = 0.05\n",
    "P_MIN        = 1e-300\n",
    "# Preferred prefixes to search (first that exists per-trait will be used)\n",
    "PREFERRED_PREFIXES = [\"assoc_chr5\", \"assoc_chr5_fast\"]\n",
    "SIG_TABLE_OUT = \"assoc_chr5_significant_hits.csv\"   # combined FDR<0.05 table\n",
    "\n",
    "# ---- helpers ----\n",
    "def _norm(s: str) -> str:\n",
    "    return re.sub(r\"\\W+\", \"_\", str(s).strip().lower())\n",
    "\n",
    "def _resolve_trait_cols(pheno_df: pd.DataFrame, names: list[str]) -> dict:\n",
    "    colmap = { _norm(c): c for c in pheno_df.columns }\n",
    "    resolved = {}\n",
    "    for t in names:\n",
    "        key = _norm(t)\n",
    "        if key not in colmap:\n",
    "            raise KeyError(f\"Trait '{t}' not found (case/space-insensitive). \"\n",
    "                           f\"Available (normalized) include: {list(colmap.keys())[:15]} ...\")\n",
    "        resolved[t] = colmap[key]\n",
    "    return resolved\n",
    "\n",
    "def _ensure_cols(df: pd.DataFrame) -> pd.DataFrame:\n",
    "    if df.empty:\n",
    "        return df\n",
    "    df = df.copy()\n",
    "    df[\"mlog10p\"] = -np.log10(np.clip(df[\"pval\"].values.astype(float), P_MIN, 1.0))\n",
    "    df[\"hit\"] = (df[\"qval_fdr\"].values.astype(float) <= FDR_ALPHA)\n",
    "    df[\"POS\"] = pd.to_numeric(df[\"POS\"], errors=\"coerce\")\n",
    "    return df\n",
    "\n",
    "def _savefig(path: str):\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(path, dpi=180, bbox_inches=\"tight\")\n",
    "    plt.close()\n",
    "    print(f\"[fig] saved -> {path}\")\n",
    "\n",
    "def plot_trait_quicklook(trait_name, series, out_prefix):\n",
    "    y = series.copy()\n",
    "    fig_path = f\"{out_prefix}_{trait_name}_trait_quicklook.png\"\n",
    "    plt.figure(figsize=(6,4))\n",
    "    uniq = set(pd.unique(y.dropna()))\n",
    "    if uniq <= {0,1}:   # binary\n",
    "        vals = y.dropna().astype(int)\n",
    "        plt.bar([\"0\",\"1\"], [(vals==0).sum(), (vals==1).sum()])\n",
    "        plt.ylabel(\"Count\"); plt.title(f\"{trait_name} (binary) counts\")\n",
    "    else:\n",
    "        yn = pd.to_numeric(y, errors=\"coerce\").dropna()\n",
    "        plt.hist(yn.values, bins=30)\n",
    "        plt.xlabel(trait_name); plt.ylabel(\"Frequency\"); plt.title(f\"{trait_name} distribution\")\n",
    "    _savefig(fig_path)\n",
    "\n",
    "def manhattan_chr22(df, trait_name, out_prefix, fdr_alpha=FDR_ALPHA):\n",
    "    df = _ensure_cols(df)\n",
    "    if df.empty:\n",
    "        print(f\"[{trait_name}] No results to plot (Manhattan).\")\n",
    "        return\n",
    "\n",
    "    x = df[\"POS\"].values\n",
    "    y = df[\"mlog10p\"].values\n",
    "    hit = df[\"hit\"].values  # qval_fdr <= alpha\n",
    "\n",
    "    # Thresholds\n",
    "    m = max(len(df), 1)\n",
    "    bonf_p = 0.05 / m\n",
    "    bonf_y = -np.log10(max(bonf_p, P_MIN))\n",
    "\n",
    "    # FDR cutoff line (largest p among hits)\n",
    "    has_fdr_hits = hit.any()\n",
    "    if has_fdr_hits:\n",
    "        p_fdr_cut = float(df.loc[hit, \"pval\"].max())\n",
    "        fdr_y = -np.log10(max(p_fdr_cut, P_MIN))\n",
    "\n",
    "    fig_path = f\"{out_prefix}_{trait_name}_manhattan_chr22.png\"\n",
    "    plt.figure(figsize=(10,4))\n",
    "    plt.scatter(x[~hit], y[~hit], s=6, alpha=0.6, label=\"tested\")\n",
    "    if has_fdr_hits:\n",
    "        plt.scatter(x[hit], y[hit], s=10, alpha=0.9, label=f\"FDR<{fdr_alpha}\")\n",
    "\n",
    "    # Add threshold lines\n",
    "    plt.axhline(bonf_y, color=\"red\", linestyle=\"-\", linewidth=1.2,\n",
    "                label=f\"Bonferroni 0.05 (p<{bonf_p:.2e})\")\n",
    "    if has_fdr_hits:\n",
    "        plt.axhline(fdr_y, color=\"orange\", linestyle=\"--\", linewidth=1.2,\n",
    "                    label=f\"FDR<{fdr_alpha} (p≈{p_fdr_cut:.1e})\")\n",
    "\n",
    "    plt.xlabel(\"Position on chr22 (bp)\")\n",
    "    plt.ylabel(\"-log10(p)\")\n",
    "    plt.title(f\"Manhattan (chr22) – {trait_name}\")\n",
    "    plt.legend(frameon=False, loc=\"upper right\")\n",
    "    _savefig(fig_path)\n",
    "\n",
    "def qq_plot(df, trait_name, out_prefix):\n",
    "    df = _ensure_cols(df)\n",
    "    if df.empty:\n",
    "        print(f\"[{trait_name}] No results to plot (QQ).\")\n",
    "        return\n",
    "    p = np.clip(np.sort(df[\"pval\"].values.astype(float)), P_MIN, 1.0)\n",
    "    n = len(p)\n",
    "    exp = (np.arange(1, n+1) - 0.5) / n\n",
    "    o = -np.log10(p); e = -np.log10(exp)\n",
    "\n",
    "    # Lambda GC\n",
    "    chisq = chi2.isf(p, df=1)\n",
    "    lam_gc = np.median(chisq) / chi2.ppf(0.5, 1)\n",
    "\n",
    "    fig_path = f\"{out_prefix}_{trait_name}_qq.png\"\n",
    "    plt.figure(figsize=(5,5))\n",
    "    plt.scatter(e, o, s=6, alpha=0.6)\n",
    "    lim = max(e.max(), o.max()) * 1.02\n",
    "    plt.plot([0, lim], [0, lim])\n",
    "    plt.xlabel(\"Expected -log10(p)\"); plt.ylabel(\"Observed -log10(p)\")\n",
    "    plt.title(f\"QQ plot – {trait_name} (λGC={lam_gc:.3f})\")\n",
    "    _savefig(fig_path)\n",
    "\n",
    "def volcano(df, trait_name, out_prefix):\n",
    "    df = _ensure_cols(df)\n",
    "    if df.empty:\n",
    "        print(f\"[{trait_name}] No results to plot (volcano).\")\n",
    "        return\n",
    "    x, y, hit = df[\"beta\"].values.astype(float), df[\"mlog10p\"].values, df[\"hit\"].values\n",
    "    fig_path = f\"{out_prefix}_{trait_name}_volcano.png\"\n",
    "    plt.figure(figsize=(6,5))\n",
    "    plt.scatter(x[~hit], y[~hit], s=8, alpha=0.6)\n",
    "    if hit.any(): plt.scatter(x[hit], y[hit], s=12, alpha=0.9)\n",
    "    if len(df) > 0: plt.axhline(-np.log10(0.05/len(df)), linestyle=\"--\", linewidth=1)\n",
    "    plt.xlabel(\"Effect size (β per ALT allele)\"); plt.ylabel(\"-log10(p)\")\n",
    "    plt.title(f\"Volcano – {trait_name} (FDR<{FDR_ALPHA})\")\n",
    "    _savefig(fig_path)\n",
    "\n",
    "def maf_vs_significance(df, trait_name, out_prefix):\n",
    "    df = _ensure_cols(df)\n",
    "    if df.empty:\n",
    "        print(f\"[{trait_name}] No results to plot (MAF vs sig).\")\n",
    "        return\n",
    "    maf, y, hit = pd.to_numeric(df[\"MAF\"], errors=\"coerce\").values, df[\"mlog10p\"].values, df[\"hit\"].values\n",
    "    fig_path = f\"{out_prefix}_{trait_name}_maf_vs_mlog10p.png\"\n",
    "    plt.figure(figsize=(6,5))\n",
    "    plt.scatter(maf[~hit], y[~hit], s=8, alpha=0.6)\n",
    "    if hit.any(): plt.scatter(maf[hit], y[hit], s=12, alpha=0.9)\n",
    "    plt.xlabel(\"Minor Allele Frequency (MAF)\"); plt.ylabel(\"-log10(p)\")\n",
    "    plt.title(f\"MAF vs significance – {trait_name}\")\n",
    "    _savefig(fig_path)\n",
    "\n",
    "# ---- Reload phenotype/genotype to align quicklook samples ----\n",
    "pheno = pd.read_csv(PHENO_PATH)\n",
    "geno  = pd.read_csv(GENO_CSV, low_memory=False)\n",
    "\n",
    "trait_col_map = _resolve_trait_cols(pheno, TRAITS)  # case/space-insensitive\n",
    "pheno = pheno.set_index(pheno[PHENO_ID_COL].astype(str)).copy()\n",
    "pheno.index.name = \"sample_id\"\n",
    "\n",
    "sample_cols = [str(c) for c in geno.columns[9:]]\n",
    "overlap = [s for s in sample_cols if s in pheno.index]\n",
    "print(f\"[plot] Overlapping samples for quicklook: {len(overlap)}\")\n",
    "\n",
    "# ---- For each trait, find the existing assoc file, plot, and collect significant hits ----\n",
    "signif_frames = []\n",
    "for trait in TRAITS:\n",
    "    assoc_path = None\n",
    "    out_prefix_used = None\n",
    "    for pref in PREFERRED_PREFIXES:\n",
    "        candidate = f\"{pref}_{trait}.csv\"\n",
    "        if os.path.exists(candidate):\n",
    "            assoc_path = candidate\n",
    "            out_prefix_used = pref\n",
    "            break\n",
    "    if assoc_path is None:\n",
    "        print(f\"[skip] No association file found for {trait} with prefixes {PREFERRED_PREFIXES}.\")\n",
    "        continue\n",
    "\n",
    "    res = pd.read_csv(assoc_path)\n",
    "    ph_col = trait_col_map[trait]\n",
    "    series = pheno.loc[overlap, ph_col]\n",
    "\n",
    "    # map 'Asthma'/'No Asthma' etc. to 1/0 for nicer binary bars\n",
    "    s_low = series.astype(str).str.strip().str.lower()\n",
    "    if set(pd.unique(s_low.dropna())) <= {\"asthma\",\"no asthma\",\"yes\",\"no\",\"true\",\"false\",\"1\",\"0\"}:\n",
    "        mapping = {\"asthma\":1,\"no asthma\":0,\"yes\":1,\"no\":0,\"true\":1,\"false\":0,\"1\":1,\"0\":0}\n",
    "        series = s_low.map(mapping).astype(float)\n",
    "\n",
    "    # Make plots\n",
    "    plot_trait_quicklook(trait, series, out_prefix=out_prefix_used)\n",
    "    manhattan_chr22(res, trait, out_prefix=out_prefix_used)\n",
    "    qq_plot(res, trait, out_prefix=out_prefix_used)\n",
    "    volcano(res, trait, out_prefix=out_prefix_used)\n",
    "    maf_vs_significance(res, trait, out_prefix=out_prefix_used)\n",
    "\n",
    "    # Collect significant hits (FDR < 0.05)\n",
    "    res_sig = res.loc[(res[\"qval_fdr\"].astype(float) <= FDR_ALPHA)].copy()\n",
    "    if not res_sig.empty:\n",
    "        # Add Bonferroni flag at study-wise level for this trait\n",
    "        m = len(res)\n",
    "        res_sig[\"bonf_sig\"] = res_sig[\"pval\"].astype(float) < (0.05 / max(m,1))\n",
    "        res_sig[\"trait\"] = trait\n",
    "        # Standard column order\n",
    "        cols = [\"trait\",\"variant_id\",\"CHROM\",\"POS\",\"REF\",\"ALT\",\"MAF\",\"N\",\"beta\",\"se\",\"pval\",\"qval_fdr\",\"bonf_sig\"]\n",
    "        signif_frames.append(res_sig[cols])\n",
    "\n",
    "# ---- Write combined significant table ----\n",
    "if signif_frames:\n",
    "    sig_table = pd.concat(signif_frames, ignore_index=True)\n",
    "    sig_table = sig_table.sort_values([\"trait\",\"pval\"]).reset_index(drop=True)\n",
    "    sig_table.to_csv(SIG_TABLE_OUT, index=False)\n",
    "    print(f\"[table] Significant associations (FDR<{FDR_ALPHA}) -> {SIG_TABLE_OUT}\")\n",
    "    # Show a quick preview\n",
    "    display_cols = [\"trait\",\"variant_id\",\"CHROM\",\"POS\",\"REF\",\"ALT\",\"MAF\",\"N\",\"beta\",\"se\",\"pval\",\"qval_fdr\",\"bonf_sig\"]\n",
    "    print(sig_table.head(20)[display_cols].to_string(index=False))\n",
    "else:\n",
    "    print(f\"[table] No variants met FDR<{FDR_ALPHA} across the provided traits.\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a96f1c88-a8fb-41e9-9b81-6a189009bb34",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:base] *",
   "language": "python",
   "name": "conda-base-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
